<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Per Unneberg">
  <title>DDLS Population genomics in practice - Simulation</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-pointer/pointer.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/logos/nbis-scilifelab.svg" data-background-position="top left" data-background-size="500px" class="center">
  <h1 class="title">Simulation</h1>
  <h3 class="subtitle">Primer on the coalescent and forward simulation</h3>
  <h4 class="author">Per Unneberg</h4>
  <h4 class="institute">NBIS</h4>
  <h4 class="date">16-Oct-2023</h4>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Table of contents</h2>
<ul>
<li><a href="#/the-coalescent" id="/toc-the-coalescent">The coalescent</a></li>
<li><a href="#/the-coalescent-with-recombination" id="/toc-the-coalescent-with-recombination">The coalescent with recombination</a></li>
<li><a href="#/forward-simulation-in-slim" id="/toc-forward-simulation-in-slim">Forward simulation in SLiM</a></li>
<li><a href="#/applications" id="/toc-applications">Applications</a></li>
</ul>
</nav>
</section>

<section id="intended-learning-outcomes" class="slide level2 unnumbered unlisted">
<h2>Intended learning outcomes</h2>
<p>Introduction to the coalescent and simulation in msprime, with two exercises on the coalescent and msprime. The students should be able to:</p>
<ul>
<li>Describe need for simulations</li>
<li>Demonstrate the coalescent step by step</li>
<li>Detail some properties of coalescent trees</li>
<li>Define the site-frequency spectrum</li>
<li>Describe the coalescent with recombination</li>
<li>Perform simple simulations in msprime</li>
</ul>
</section>
<section id="the-wright-fisher-model-and-simulations" class="slide level2">
<h2>The Wright-Fisher model and simulations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-reproductive-success-1_719d7328077e001baa1bad0c014956d1">
<div class="cell-output-display">
<div id="fig-wf-model-reproductive-success-1" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;1: Wright-Fisher model for 50 generations, 30 individuals</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-reproductive-success-1-1.svg" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<aside class="notes">

<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<h3 id="recap">Recap</h3>
<p>Model of a population describing <strong>genealogies</strong> under the following assumptions</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<div class="fragment">
<p><strong>Forward</strong> simulation</p>
</div>
</div>
</div>
<aside class="notes">
<p><span class="citation" data-cites="hein2005gene">(p.&nbsp;15 <a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span> shows the fraction of genes without descendants. Focus on the number of descendants for <em>one</em> gene <em>i</em>, which is X ~ Bin(2N, 1/2N). Since E(X)=1, for large 2N, X is almos Po(1), such that P(no descendants) = P(X=0) = e^{-1}</p>
<p>Low reproductive success: forward simulations costly!</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-wright-fisher-model-and-simulations-1" class="slide level2">
<h2>The Wright-Fisher model and simulations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-reproductive-success-2_88567cc1eed8bac5f03ade9a892b3b70">
<div class="cell-output-display">
<div id="fig-wf-model-reproductive-success-2" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;2: Wright-Fisher model for 50 generations, 30 individuals</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-reproductive-success-2-1.svg" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-wf-model-reproductive-success-graph" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;3: Reproductive success in percent per generation.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-reproductive-success-graph-1.svg" style="width:55.0%"></p>
</figure>
</div>
</div>
</div>
<p>Mean reproductive success = 63.4%. Can show for large populations P(no descendants)=<span class="math inline">\(1 - e^{-1} \approx 0.632\)</span></p>
</div>
</div>
<aside class="notes">
<p><span class="citation" data-cites="hein2005gene">(p.&nbsp;15 <a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span> shows the fraction of genes without descendants. Focus on the number of descendants for <em>one</em> gene <em>i</em>, which is X ~ Bin(2N, 1/2N). Since E(X)=1, for large 2N, X is almos Po(1), such that P(no descendants) = P(X=0) = e^{-1}</p>
<p>Low reproductive success: forward simulations costly!</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="forward-and-backward-simulation" class="slide level2">
<h2>Forward and backward simulation</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-genealogy-forward_ee0044e7a42c77d84996e8674d5ed0cf">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy-forward" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;4: Forward simulation.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-genealogy-forward-1.svg" style="width:60.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-genealogy-backward_56ab3fb486ce7355cb3dcd323593c874">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy-backward" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;5: Backward simulation.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-genealogy-backward-1.svg" style="width:60.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<p>Simulated nodes are filled with black. Genealogy of interest is highlighted in thick black lines.</p>
<aside class="notes">
<p>Forward simulations: require that we keep track of the entire population -&gt; computationally challenging.</p>
<p>Backward simulations: require only tracking the sample -&gt; quicker. However, cannot model selection.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="why-do-we-want-simulations-anyway" class="slide level2">
<h2>Why do we want simulations anyway?</h2>
<div>
<dl>
<dt class="fragment">Null model</dt>
<dd class="fragment">
Generate neutral null distributions to compare with observed data
</dd>
<dt class="fragment">Exploration</dt>
<dd class="fragment">
Use to gain understanding and improve interpretation of mutational processes
</dd>
<dt class="fragment">Mathematical complexity</dt>
<dd class="fragment">
No analytical solutions for linked selection and the like <span class="math inline">\(\rightarrow\)</span> must use simulations
</dd>
<dt class="fragment">Benchmarking</dt>
<dd class="fragment">
Use to generate test data with known properties on which to test and evaluate new methods
</dd>
<dt class="fragment">Model training</dt>
<dd class="fragment">
Generate training data for machine learning, e.g., Approximate Bayesian Computation (ABC) or Neural Networks (NNs)
</dd>
</dl>
</div>
</section>
<section>
<section id="the-coalescent" class="title-slide slide level1 center">
<h1>The coalescent</h1>

</section>
<section id="coalescent-simulations" class="slide level2">
<h2>Coalescent simulations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>The coalescent simulates the genealogy of a <strong>sample</strong> of individuals on which mutations are “sprinkled” according to a Poisson process.</p>
<ol type="1">
<li>Simulate ancestry (genealogy)</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/wf-model-genealogy-backward-1_13233446bd1a27534caa2308794d874b">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-model-genealogy-backward-1-1.svg" style="width:75.0%"></p>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>NB: only the samples on the genealogy are needed, but the entire population is shown as a means to compare with forward simulations.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="coalescent-simulations-1" class="slide level2">
<h2>Coalescent simulations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>The coalescent simulates the genealogy of a <strong>sample</strong> of individuals on which mutations are “sprinkled” according to a Poisson process.</p>
<ol type="1">
<li>Simulate ancestry (genealogy)</li>
<li>Simulate mutations</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/wf-model-genealogy-backward-mutations_9487c515969963460fe5317264fab204">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-model-genealogy-backward-mutations-1.svg" style="width:75.0%"></p>
</div>
</div>
</div>
</div>
</section>
<section id="coalescent-simulations-2" class="slide level2">
<h2>Coalescent simulations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>The coalescent simulates the genealogy of a <strong>sample</strong> of individuals on which mutations are “sprinkled” according to a Poisson process.</p>
<ol type="1">
<li>Simulate ancestry (genealogy)</li>
<li>Simulate mutations</li>
</ol>
<h4>
Exercise
</h4>
<p>How many mutations are common to all samples? How many mutations does sample 1 have? Sample 2?</p>
<p>Assuming the ancestral state is denoted <code>0</code> (prior to the <em>first</em> generation) and the derived state <code>1</code>, what are the sequences of the samples?</p>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/wf-model-genealogy-backward-mutations-1_4c9c887c231fb58d84af52aa07102b87">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-model-genealogy-backward-mutations-1-1.svg" style="width:75.0%"></p>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Answers to exercise: 2 mutations are common to all samples. Sample 1 has 3 mutations, sample 2 has 4.</p>
<p>(arbitrarily) ordering the sites from top to bottom (oldest mutation comes first), the sequences are:</p>
<p>1: 11100 2: 11101 3: 11010</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-1_e0e6f0c579a3fbe75afecd96acb712cf">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-1-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-1" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Start with <span class="math inline">\(i=n\)</span> chromosomes</p></li>
<li><p>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></p></li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-2_b848affc08941a5b18548ec1fa72e791">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-2-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-2" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Start with <span class="math inline">\(i=n\)</span> chromosomes</p></li>
<li><p>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></p></li>
<li><p>Choose two chromosomes at random to coalesce</p></li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-3_d5af43b937f636a454eeb2e278836f27">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-3-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-3" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>Start with <span class="math inline">\(i=n\)</span> chromosomes</p></li>
<li><p>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></p></li>
<li><p>Choose two chromosomes at random to coalesce</p></li>
<li><p>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></p></li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-4_ac9667844f2107107349a0410f2ed4f5">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-4-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-4" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-5_ed8e008d523f85ff7c4ef6e5a31cd9ab">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-5-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-5" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-6_fcb3f5a0e7335c8b83fbee79c62d6610">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-6-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-6" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-7_c9a9e21999993bca1cf4e9261e4211fc">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-7-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-7" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-8_eb0b374189eecf0e4ad39fad9fd0215f">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-8-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-8" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-9_8af6d61e6a1c1942137f4e011b02efc0">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-9-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-9" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-10_bf948f1adf3eab3f0a476ada87048c88">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-10-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-10" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span><span class="math inline">\(^1\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
<div class="smallr">
<p><span class="math inline">\(^1\)</span>: The exponential can be parametrized in two different ways, so that the parameter to the function is either <span class="math inline">\(\lambda\)</span> or <span class="math inline">\(\beta=1/\lambda\)</span>.</p>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/simulate-genealogy-11_5451f464821a7d1a2adf2457dbc0d016">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/simulate-genealogy-11-1.svg" width="500" height="600"></p>
</div>
</div>
</div>
</div>
</section>
<section id="simulating-genealogies-hahn_molecularpopulationgenetics_2019-p.-115-11" class="slide level2">
<h2>Simulating genealogies <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 115</a>)</span></h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Start with <span class="math inline">\(i=n\)</span> chromosomes</li>
<li>Choose time to next coalescent event from an exponential distribution with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span><span class="math inline">\(^1\)</span></li>
<li>Choose two chromosomes at random to coalesce</li>
<li>Merge the two lineages and set <span class="math inline">\(i \rightarrow i - 1\)</span></li>
<li>If <span class="math inline">\(i&gt;1\)</span>, go to step 2; if not, stop.</li>
</ol>
<div class="smallr">
<p><span class="math inline">\(^1\)</span>: The exponential can be parametrized in two different ways, so that the parameter to the function is either <span class="math inline">\(\lambda\)</span> or <span class="math inline">\(\beta=1/\lambda\)</span>.</p>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-simulate-genealogy-end_114e947d1d00724e0bfdcea2ddcf1757">
<div class="cell-output-display">
<div id="fig-simulate-genealogy-end" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;6: A simulated genealogy. The <span class="math inline">\(T_i\)</span> represent the waiting time when the state has <span class="math inline">\(i\)</span> chromosomes.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-simulate-genealogy-end-1.svg" width="500" height="600"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="some-properties-of-the-tree" class="slide level2">
<h2>Some properties of the tree</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-genealogy-properties_29c3784e68df71942a87c73bb66f7fc5">
<div class="cell-output-display">
<div id="fig-genealogy-properties" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;7: A coalescent tree with numbered nodes. Nodes 1-5 correspond to the samples and are leaves. The internal nodes 6-8 (and the unlabelled root) are ancestral chromosomes (unsampled). The <span class="math inline">\(T_i\)</span> represent the waiting time when the state has <span class="math inline">\(i\)</span> chromosomes, whereas <span class="math inline">\(\tau_i\)</span> correspond to the branch length from node <span class="math inline">\(i\)</span> to its parent.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-genealogy-properties-1.svg" width="400" height="500"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>Expected waiting time to coalesce when <span class="math inline">\(i\)</span> lineages: <span class="math inline">\(E(T_i) = \frac{2}{i(i-1)}\)</span></p>
<p>Branch lengths can be derived from waiting times. For instance, <span class="math inline">\(\tau_1=\tau_2=T_5+T_4\)</span> and <span class="math inline">\(\tau_4=\tau_5=T_5\)</span></p>
<p>Time to the <strong>most recent common ancestor (MRCA)</strong>: <span class="math inline">\(T_{MRCA} = \sum_{i=2}^n T_i\)</span></p>
<p>with expected value <span class="math inline">\(E(T_{MRCA}) = \sum_{i=2}^nE(T_i) = 2\left(1 - \frac{1}{n}\right)\)</span></p>
<p>The expected <strong>total tree height</strong> is <span class="math inline">\(E(T_{total}) = \sum_{i=2}^n iE(T_i) = 2\sum_{i=2}^n\frac{1}{i-1}\)</span></p>
</div>
</div>
<aside class="notes">
<p>Note that the parametrization is in 2N generations (continuous time approximation); some use N, others 4N, and so on.</p>
<p>Expected waiting time: the first expression follows from the fact that the waiting time, given <span class="math inline">\(i\)</span> chromosomes, is exponentially distributed with parameter <span class="math inline">\(\lambda=i(i-1)/2\)</span>. For an exponentially distributed variable <span class="math inline">\(X\)</span>, the expected value is simply <span class="math inline">\(E(X)=1/\lambda\)</span>.</p>
<p>Other relations are most easily derived by studying the tree and summing up relevant branches or time intervals.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="coalescent-simulations-vary-in-topology-and-height" class="slide level2">
<h2>Coalescent simulations vary in topology and height</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="im">import</span> msprime</span>
<span id="cb1-2"><a href="#cb1-2"></a>seeds <span class="op">=</span> [<span class="dv">12</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">34</span>, <span class="dv">63</span>, <span class="dv">30</span>]</span>
<span id="cb1-3"><a href="#cb1-3"></a>trees <span class="op">=</span> [msprime.sim_ancestry(<span class="dv">3</span>, random_seed<span class="op">=</span>x) <span class="cf">for</span> x <span class="kw">in</span> seeds]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="index_cache/revealjs/fig-coalescent-trees-with-msprime_b270e914bd4645972f218bb1bbee940c">
<svg baseprofile="full" height="262.5576521441852" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(95.625 26.8)"><g class="a10 i1 leaf node n2 p0 sample" transform="translate(34.375 193.958)"><path class="edge" d="M 0 0 V -193.958 H -34.375"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a10 c2 node n9 p0" transform="translate(-34.375 160.079)"><g class="a9 i0 leaf node n0 p0 sample" transform="translate(-31.25 33.8787)"><path class="edge" d="M 0 0 V -33.8787 H 31.25"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a9 c2 node n8 p0" transform="translate(31.25 9.96887)"><g class="a8 i2 leaf node n4 p0 sample" transform="translate(17.5 23.9099)"><path class="edge" d="M 0 0 V -23.9099 H -17.5"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 c2 node n7 p0" transform="translate(-17.5 8.81347)"><g class="a7 i2 leaf node n5 p0 sample" transform="translate(15 15.0964)"><path class="edge" d="M 0 0 V -15.0964 H -15"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 c2 node n6 p0" transform="translate(-15 14.3228)"><g class="a6 i0 leaf node n1 p0 sample" transform="translate(-10 0.77358)"><path class="edge" d="M 0 0 V -0.77358 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i1 leaf node n3 p0 sample" transform="translate(10 0.77358)"><path class="edge" d="M 0 0 V -0.77358 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -14.3228 H 15"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -8.81347 H 17.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -9.96887 H -31.25"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><path class="edge" d="M 0 0 V -160.079 H 34.375"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
<svg baseprofile="full" height="143.18431650051062" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(105.625 26.8)"><g class="a10 i1 leaf node n2 p0 sample" transform="translate(24.375 74.5843)"><path class="edge" d="M 0 0 V -74.5843 H -24.375"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a10 c2 node n9 p0" transform="translate(-24.375 17.8455)"><g class="a9 i1 leaf node n3 p0 sample" transform="translate(28.75 56.7388)"><path class="edge" d="M 0 0 V -56.7388 H -28.75"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a9 c2 node n8 p0" transform="translate(-28.75 42.5557)"><g class="a8 i0 leaf node n0 p0 sample" transform="translate(-22.5 14.1831)"><path class="edge" d="M 0 0 V -14.1831 H 22.5"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 c2 node n7 p0" transform="translate(22.5 3.65457)"><g class="a7 i2 leaf node n4 p0 sample" transform="translate(15 10.5285)"><path class="edge" d="M 0 0 V -10.5285 H -15"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 c2 node n6 p0" transform="translate(-15 4.37337)"><g class="a6 i0 leaf node n1 p0 sample" transform="translate(-10 6.15517)"><path class="edge" d="M 0 0 V -6.15517 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i2 leaf node n5 p0 sample" transform="translate(10 6.15517)"><path class="edge" d="M 0 0 V -6.15517 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -4.37337 H 15"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -3.65457 H -22.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><path class="edge" d="M 0 0 V -42.5557 H 28.75"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -17.8455 H 24.375"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
<svg baseprofile="full" height="150.46109506901013" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(61.25 26.8)"><g class="a10 i0 leaf node n0 p0 sample" transform="translate(-31.25 81.8611)"><path class="edge" d="M 0 0 V -81.8611 H 31.25"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a10 c2 node n9 p0" transform="translate(31.25 59.5717)"><g class="a9 c2 node n7 p0" transform="translate(27.5 21.1371)"><g class="a7 i1 leaf node n3 p0 sample" transform="translate(-10 1.15234)"><path class="edge" d="M 0 0 V -1.15234 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 i2 leaf node n5 p0 sample" transform="translate(10 1.15234)"><path class="edge" d="M 0 0 V -1.15234 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -21.1371 H -27.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><g class="a9 c2 node n8 p0" transform="translate(-27.5 20.74)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-15 1.54941)"><path class="edge" d="M 0 0 V -1.54941 H 15"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 c2 node n6 p0" transform="translate(15 0.689536)"><g class="a6 i1 leaf node n2 p0 sample" transform="translate(-10 0.859874)"><path class="edge" d="M 0 0 V -0.859874 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i2 leaf node n4 p0 sample" transform="translate(10 0.859874)"><path class="edge" d="M 0 0 V -0.859874 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -0.689536 H -15"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><path class="edge" d="M 0 0 V -20.74 H 27.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -59.5717 H -31.25"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
<svg baseprofile="full" height="76.60317603443052" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(86.25 26.8)"><g class="a10 c2 node n8 p0" transform="translate(33.75 5.80079)"><g class="a8 i1 leaf node n2 p0 sample" transform="translate(-10 2.20239)"><path class="edge" d="M 0 0 V -2.20239 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(10 2.20239)"><path class="edge" d="M 0 0 V -2.20239 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -5.80079 H -33.75"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><g class="a10 c2 node n9 p0" transform="translate(-33.75 5.45098)"><g class="a9 i0 leaf node n0 p0 sample" transform="translate(-22.5 2.55219)"><path class="edge" d="M 0 0 V -2.55219 H 22.5"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a9 c2 node n7 p0" transform="translate(22.5 1.98685)"><g class="a7 i1 leaf node n3 p0 sample" transform="translate(15 0.565342)"><path class="edge" d="M 0 0 V -0.565342 H -15"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 c2 node n6 p0" transform="translate(-15 0.539646)"><g class="a6 i0 leaf node n1 p0 sample" transform="translate(-10 0.0256967)"><path class="edge" d="M 0 0 V -0.0256967 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i2 leaf node n4 p0 sample" transform="translate(10 0.0256967)"><path class="edge" d="M 0 0 V -0.0256967 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -0.539646 H 15"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -1.98685 H -22.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><path class="edge" d="M 0 0 V -5.45098 H 33.75"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
<svg baseprofile="full" height="87.63214958144492" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(86.25 26.8)"><g class="a10 c2 node n8 p0" transform="translate(-33.75 15.3658)"><g class="a8 i0 leaf node n0 p0 sample" transform="translate(-22.5 3.66634)"><path class="edge" d="M 0 0 V -3.66634 H 22.5"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 c2 node n7 p0" transform="translate(22.5 2.53288)"><g class="a7 i2 leaf node n5 p0 sample" transform="translate(15 1.13346)"><path class="edge" d="M 0 0 V -1.13346 H -15"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 c2 node n6 p0" transform="translate(-15 0.0368438)"><g class="a6 i1 leaf node n3 p0 sample" transform="translate(-10 1.09662)"><path class="edge" d="M 0 0 V -1.09662 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i2 leaf node n4 p0 sample" transform="translate(10 1.09662)"><path class="edge" d="M 0 0 V -1.09662 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -0.0368438 H 15"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -2.53288 H -22.5"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><path class="edge" d="M 0 0 V -15.3658 H 33.75"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><g class="a10 c2 node n9 p0" transform="translate(33.75 6.02772)"><g class="a9 i0 leaf node n1 p0 sample" transform="translate(-10 13.0044)"><path class="edge" d="M 0 0 V -13.0044 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a9 i1 leaf node n2 p0 sample" transform="translate(10 13.0044)"><path class="edge" d="M 0 0 V -13.0044 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -6.02772 H -33.75"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
<svg baseprofile="full" height="500.0" id="msprime-simulation" version="1.1" width="200" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}.edge {stroke-width: 2px}</style>
</defs><g class="tree-sequence"><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n10 p0 root" transform="translate(70 26.8)"><g class="a10 c2 node n6 p0" transform="translate(-30 425.824)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-10 5.57571)"><path class="edge" d="M 0 0 V -5.57571 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a6 i2 leaf node n4 p0 sample" transform="translate(10 5.57571)"><path class="edge" d="M 0 0 V -5.57571 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -425.824 H 30"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><g class="a10 c2 node n9 p0" transform="translate(30 388.351)"><g class="a9 c2 node n7 p0" transform="translate(20 35.4797)"><g class="a7 i1 leaf node n2 p0 sample" transform="translate(-10 7.56884)"><path class="edge" d="M 0 0 V -7.56884 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a7 i2 leaf node n5 p0 sample" transform="translate(10 7.56884)"><path class="edge" d="M 0 0 V -7.56884 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -35.4797 H -20"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><g class="a9 c2 node n8 p0" transform="translate(-20 33.0688)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-10 9.9798)"><path class="edge" d="M 0 0 V -9.9798 H 10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><g class="a8 i1 leaf node n3 p0 sample" transform="translate(10 9.9798)"><path class="edge" d="M 0 0 V -9.9798 H -10"></path><rect class="sym" height="0" width="0" x="-0" y="-0"></rect><text class="lab" transform="translate(0 11)"></text></g><path class="edge" d="M 0 0 V -33.0688 H 20"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab lft" transform="translate(-3 -7.0)"></text></g><path class="edge" d="M 0 0 V -388.351 H -30"></path><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab rgt" transform="translate(3 -7.0)"></text></g><circle class="sym" cx="0" cy="0" r="0"></circle><text class="lab" transform="translate(0 -11)"></text></g></g></g></g></g>
</svg>
</div>
<p>Examples of coalescent simulations. Note the variation in tree topology and height.</p>
</section>
<section id="diminishing-returns-of-adding-more-samples" class="slide level2">
<h2>Diminishing returns of adding more samples</h2>

<img data-src="index_files/figure-revealjs/fig-coalescent-diminishing-returns-1.svg" width="600" class="r-stretch quarto-figure-left"><p class="caption">Figure&nbsp;8: Dependence of <span class="math inline">\(E[T_{MRCA}]\)</span> and <span class="math inline">\(E[T_{total}]\)</span> on the sample size n.&nbsp;Already at low values of n the value of <span class="math inline">\(E[T_{MRCA}]\)</span> is close to its asymptotic value, which has practical consequences for the measurement of DNA variation. Adapted from <span class="citation" data-cites="wakeley_CoalescentTheoryIntroduction_2008">Wakeley (<a href="#/bibliography" role="doc-biblioref" onclick="">2008</a>)</span>, Fig. 3.3.</p><aside class="notes">
<p>Adding a sample only adds <span class="math inline">\(2/n\)</span> to the total tree length. Also, adding a sample will add little to <span class="math inline">\(T_{MRCA}\)</span>, or equivalently, will most likely add a very recent coalescent event to the tree for large enough <span class="math inline">\(n\)</span>. Note that <span class="math inline">\(n=20\)</span> here is equivalent to 10 diploid individuals.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="adding-mutations" class="slide level2">
<h2>Adding mutations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-add-mutations_ba13666372e59133e7b9bcd4a43b0d0e">
<div class="cell-output-display">
<div id="fig-add-mutations" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;9: Adding mutations on a coalescent genealogy.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-add-mutations-1.svg" width="500" height="500"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>Mutations are added by placing them on branches, where the probability of ending up on a branch <span class="math inline">\(\tau_i\)</span> is equal to its normalized length, where normalization is by the total tree branch length:</p>
<p><span class="math display">\[
P(\text{mutation on branch }i) = \frac{\tau_i}{\sum_j\tau_j} =
\frac{\tau_i}{T_{total}}
\]</span></p>
<p>The total number of segregating sites <span class="math inline">\(S\)</span> to be thrown on the tree is modelled as a <a href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson random variable</a> which expresses the probability of a given number of events in time <span class="math inline">\(t\)</span>:</p>
<p><span class="math display">\[
S = Po(\theta/2T_{total})
\]</span></p>
</div>
</div>
<aside class="notes">
<p>Sofar we have looked at genealogies; now add mutations.</p>
<p>In words: sample a number of segregating sites from the Poisson distribution, and “sprinkle” them on the tree.</p>
<p>The parametrization <span class="math inline">\(\lambda = \theta / 2\)</span> is defined with reference to the average heterozygosity between two sequences being <span class="math inline">\(\theta\)</span> (by Tajima; see <span class="citation" data-cites="wakeley_CoalescentTheoryIntroduction_2008">Wakeley (<a href="#/bibliography" role="doc-biblioref" onclick="">2008</a>)</span>, p.&nbsp;92)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-coalescent-and-diversity" class="slide level2 smaller">
<h2>The coalescent and diversity</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/coalescent-tree-dna-variation_ef2cbc1f993cda652f4cefd42f969b0c">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/coalescent-tree-dna-variation-1.svg" width="960" height="600"></p>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>Many statistical quantities can be related to the <strong>site frequency spectrum</strong> (<strong>SFS</strong>), which is a summary of the frequencies of the segregating sites. Let <span class="math inline">\(\xi_i\)</span> be the number of chromosomes in the sample with <span class="math inline">\(i\)</span> minor alleles. In the figure we have 6 mutations on <span class="math inline">\(n=4\)</span> chromosomes.</p>
<div class="columns">
<div class="column" style="width:40%;">
<table>
<thead>
<tr class="header">
<th></th>
<th>Name</th>
<th style="text-align: right;">Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="math inline">\(\xi_1\)</span></td>
<td>singleton</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td><span class="math inline">\(\xi_2\)</span></td>
<td>doubleton</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(\xi_3\)</span></td>
<td></td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:40%;">
<div class="translatey50">
<div class="cell" data-hash="index_cache/revealjs/site-frequency-spectrum-graph_22b59fece5523371f4e7486f6dcf2102">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/site-frequency-spectrum-graph-1.svg" width="960"></p>
</div>
</div>
</div>
</div>
</div>
<div class="translatey100">
<p>Note that the number of segregating sites is <span class="math inline">\(S=\sum_{i=1}^{n-1}\xi_i\)</span></p>
<p>In this notation one can show that <span class="math inline">\(\pi\)</span>, the nucleotide diversity, is</p>
<p><span class="math display">\[
\pi = \frac{\sum_{i=1}^{n-1}i(n-i)\xi_i}{n(n-1)/2}
\]</span></p>
</div>
<div class="fragment translatey100">
<!-- -->
<h4 id="exercise">Exercise</h4>
<p>Calculate the nucleotide diversity and compare the results from the foundation lecture</p>
</div>
<div class="fragment translatey100">
<p><span class="math display">\[
\pi = \frac{1(4-1)\xi_1 + 2(4-2)\xi_2 + 3(4-3)\xi_3}{4(4-1)/2} =
\frac{1}{6}\left(12 + 8 + 0\right) = 3\frac{1}{3}
\]</span></p>
</div>
</div>
</div>
<aside class="notes">
<p>Exercise:</p>
<p><span class="math display">\[
\pi = \frac{1(4-1)\xi_1 + 2(4-2)\xi_2 + 3(4-3)\xi_3}{4(4-1)/2}
= \frac{3*4 + 4*2 + 3*0}{6} = \frac{20}{6} = \frac{8}{3}
\]</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-impact-of-topology-on-the-sfs" class="slide level2">
<h2>The impact of topology on the SFS</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div id="fig-neutrality-interpretation" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;10: Interpretation of neutrality tests <span class="citation" data-cites="ferretti_DecomposingSiteFrequency_2017">(<a href="#/bibliography" role="doc-biblioref" onclick="">Ferretti et al., 2017</a>)</span></figcaption>
<p><img data-src="assets/images/ferretti-table1.jpg" width="400"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div id="fig-sfs-neutrality-selection" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;11: The SFS under neutrality and selection</figcaption>
<p><img data-src="assets/images/nielsen-fig2-2005.jpeg" width="480"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<p>Many tests for selection are based on the SFS which in turn is influenced by the topology of the tree.</p>
</div>
</div>
<aside class="notes">
<p>Take home: topologies influence shape of SFS</p>
<h3 id="example-tajimas-d">Example: Tajima’s D</h3>
<p>Tests common versus rare alleles. Numerator compares nucleotide diversity <span class="math inline">\(\pi\)</span> to Watterson’s theta, <span class="math inline">\(\theta_W\)</span>.</p>
<dl>
<dt>D &lt; 0</dt>
<dd>
recent population increase <em>or</em> positive selection
</dd>
<dt>D &gt; 0</dt>
<dd>
population contraction <em>or</em> balancing selection
</dd>
</dl>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section id="exercise-on-the-coalescent" class="title-slide slide level1 unnumbered unlisted center">
<h1>Exercise on the coalescent</h1>

</section>

<section>
<section id="the-coalescent-with-recombination" class="title-slide slide level1 center">
<h1>The coalescent with recombination</h1>
<aside class="notes">
<p>Introduce the coalescent with recombination which in practice is a very brief inroduction to the concept of tree sequences.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="on-non-recombining-chromosomes-and-assortment" class="slide level2">
<h2>On non-recombining chromosomes and assortment</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/non-recombining-chromosomes-assortment-1_42ec96a90faa86175562f0212d7868a5">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/non-recombining-chromosomes-assortment-1-1.svg" width="300"></p>
</div>
</div>
<p>Both siblings inherit chromosome from paternal grandfather</p>
<p>Chromosomes coalesce at father</p>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="cell" data-hash="index_cache/revealjs/non-recombining-chromosomes-assortment-2_f6a91d912bb2cd7d1091c503050a60a9">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/non-recombining-chromosomes-assortment-2-1.svg" width="300"></p>
</div>
</div>
<p>Siblings inherit different grandparental chromosomes <span class="math inline">\(\Rightarrow\)</span> chromosomes coalesce God knows when in the past</p>
<p>Genealogies <strong>differ</strong></p>
</div>
</div>
</div>
<aside class="notes">
<p>NB: the right genealogy is so small the sequences don’t coalesce. The following slide will show a larger genealogy where all sequence coalesce.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-ancestral-recombination-graph" class="slide level2">
<h2>The ancestral recombination graph</h2>
<p><img data-src="assets/images/brandt-genetics-2022-fig1a.jpeg" data-fig-align="center," height="300"> <img data-src="assets/images/brandt-genetics-2022-fig1b.jpeg" data-fig-align="center," height="300"> <img data-src="assets/images/brandt-genetics-2022-fig1c.jpeg" data-fig-align="center," height="300"></p>
<div class="flushright">
<p><span class="citation" data-cites="y.c.brandt_EvaluationMethodsEstimating_2022">Y. C. Brandt et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2022, fig. 1a</a>)</span></p>
</div>
<div>
<p>Properties:</p>
<ul>
<li>marginal trees constitute a sequence of trees (<strong>tree sequence</strong>) along a chromosome</li>
<li>each tree represents the genealogy of a non-recombining part of the chromosome</li>
<li>neighbouring trees are <strong>correlated</strong></li>
</ul>
<p>Interpretation: chromosomes are mosaics of non-recombining units</p>
</div>
<aside class="notes">
<p>Going backwards in time, we now have two events to track. We have already encountered coalescence, that is, the merging of lineages. We now also need to monitor recombination events, at which point a genealogy <em>splits</em> into two separate paths. A recombination event takes place on a sequence, such that the genealogical history of the recombined sequence differs on either side of the recombination breakpoint. In the ARG, by convention the sequence to the left of the breakpoint takes the left path, the sequence to the right the right path.</p>
<p>The end consequence of recombination is that the chromosome/sequence will consist of non-recombining units with different genealogical histories. However, neighbouring genealogies will be similar - they are <em>correlated</em> - a fact that is of importance for the tree sequence data structure used in msprime.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-ancestral-recombination-graph-as-sequence-of-trees" class="slide level2">
<h2>The ancestral recombination graph as sequence of trees</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<figcaption>The ancestral recombination graph as a tree sequence <span class="citation" data-cites="baumdicker_EfficientAncestryMutation_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Baumdicker et al., 2022, fig. 5</a>)</span></figcaption>
<p><img data-src="assets/images/baumdicker-2022-fig5.jpeg" height="500"></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p>Properties:</p>
<ul>
<li>marginal trees constitute a sequence of trees (<strong>tree sequence</strong>) along a chromosome</li>
<li>each tree represents the genealogy of a non-recombining part of the chromosome</li>
<li>neighbouring trees are <strong>correlated</strong></li>
</ul>
<p>Interpretation: chromosomes are mosaics of non-recombining units</p>
</div>
</div>
</section>
<section id="msprime-is-a-fast-coalescent-simulator" class="slide level2">
<h2>msprime is a fast coalescent simulator</h2>
<p>Original coalescent simulator implementions assumed <em>small</em> sample sizes and could only simulate short recombining sequences. <code>msprime</code> was developed to address these shortcomings as today we have</p>
<ol type="1">
<li>chromosome-level assemblies that require full treatment of recombination (<em>very</em> complex problem)</li>
<li>biobank datasets consisting of 100’s of thousands of samples</li>
</ol>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="features">Features</h3>
<ul>
<li>can simulate millions of whole chromosomes</li>
<li>well-designed mature API</li>
<li>succinct data structure (tree sequences) has led to advances in forwards-time simulation</li>
<li>can be used together with forwards-time simulators</li>
</ul>
</div><div class="column" style="width:50%;">
<h3 id="limitations">Limitations</h3>
<ul>
<li>assumes neutrality - can’t simulate selection</li>
</ul>
</div>
</div>
</section>
<section id="msprime-stores-variation-data-as-tree-sequences" class="slide level2">
<h2>msprime stores variation data as tree sequences</h2>

<img data-src="assets/images/baumdicker-2022-fig2.jpg" width="700" class="r-stretch quarto-figure-left"><p class="caption">Tree sequences <span class="citation" data-cites="baumdicker_EfficientAncestryMutation_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Baumdicker et al., 2022, fig. 2</a>)</span></p></section>
<section id="tree-sequences-compress-data" class="slide level2">
<h2>Tree sequences compress data</h2>

<img data-src="assets/images/kelleher-nature-2019-fig1c.png" width="700" class="r-stretch quarto-figure-left"><p class="caption">Data compression <span class="citation" data-cites="kelleher_InferringWholegenomeHistories_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kelleher et al., 2019, fig. 1c</a>)</span></p></section>
<section id="simulating-ancestry-with-msprime" class="slide level2">
<h2>Simulating ancestry with msprime</h2>
<!-- markdownlint-disable MD013 -->
<p>From <a href="https://tskit.dev/msprime/docs/stable/quickstart.html">msprime quickstart</a></p>
<!-- markdownlint-enable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/sim_ancestry_1935924b18bb915673606e603e734fb6">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1"></a><span class="im">import</span> msprime</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co"># Simulate an ancestral history for 3 diploid samples under the coalescent</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co"># with recombination on a 5kb region with human-like parameters.</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>ts <span class="op">=</span> msprime.sim_ancestry(</span>
<span id="cb2-5"><a href="#cb2-5"></a>    samples<span class="op">=</span><span class="dv">3</span>,</span>
<span id="cb2-6"><a href="#cb2-6"></a>    recombination_rate<span class="op">=</span><span class="fl">1e-8</span>,</span>
<span id="cb2-7"><a href="#cb2-7"></a>    sequence_length<span class="op">=</span><span class="dv">5_000</span>,</span>
<span id="cb2-8"><a href="#cb2-8"></a>    population_size<span class="op">=</span><span class="dv">10_000</span>,</span>
<span id="cb2-9"><a href="#cb2-9"></a>    random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="bu">print</span>(ts.draw_svg())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<svg baseprofile="full" height="200" version="1.1" width="800" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}</style>
</defs><g class="tree-sequence"><g class="background"><path d="M20,0 l190,0 l0,138.2 l-168.568,25 l0,5 l-21.432,0 l0,-5 l0,-25 l0,-138.2z"></path><path d="M210,0 l190,0 l0,138.2 l-261.136,25 l0,5 l-97.432,0 l0,-5 l168.568,-25 l0,-138.2z"></path><path d="M400,0 l190,0 l0,138.2 l38.152,25 l0,5 l-489.288,0 l0,-5 l261.136,-25 l0,-138.2z"></path><path d="M590,0 l190,0 l0,138.2 l0,25 l0,5 l-151.848,0 l0,-5 l-38.152,-25 l0,-138.2z"></path></g><g class="axes"><g class="x-axis"><g class="title" transform="translate(400 200)"><text class="lab" text-anchor="middle" transform="translate(0 -11)">Genome position</text></g><line class="ax-line" x1="20" x2="780" y1="163.2" y2="163.2"></line><g class="ticks"><g class="tick" transform="translate(20 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">0</text></g></g><g class="tick" transform="translate(41.432 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">141</text></g></g><g class="tick" transform="translate(138.864 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">782</text></g></g><g class="tick" transform="translate(628.152 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">4001</text></g></g><g class="tick" transform="translate(780 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">5000</text></g></g></g></g></g><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n13 p0 root" transform="translate(124.688 26.8)"><g class="a13 i1 leaf node n3 p0 sample" transform="translate(32.8125 94.6)"><path class="edge" d="M 0 0 V -94.6 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a13 c2 node n10 p0" transform="translate(-32.8125 79.037)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -79.037 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">13</text></g></g></g><g class="tree t1" transform="translate(210 0)"><g class="plotbox"><g class="c2 node n12 p0 root" transform="translate(124.688 50.9994)"><g class="a12 i1 leaf node n3 p0 sample" transform="translate(32.8125 70.4006)"><path class="edge" d="M 0 0 V -70.4006 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a12 c2 node n10 p0" transform="translate(-32.8125 54.8376)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -54.8376 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">12</text></g></g></g><g class="tree t2" transform="translate(400 0)"><g class="plotbox"><g class="c2 node n11 p0 root" transform="translate(124.688 90.3248)"><g class="a11 i1 leaf node n3 p0 sample" transform="translate(32.8125 31.0752)"><path class="edge" d="M 0 0 V -31.0752 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a11 c2 node n10 p0" transform="translate(-32.8125 15.5122)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -15.5122 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">11</text></g></g></g><g class="tree t3" transform="translate(590 0)"><g class="plotbox"><g class="c2 node n11 p0 root" transform="translate(124.688 90.3248)"><g class="a11 i1 leaf node n3 p0 sample" transform="translate(32.8125 31.0752)"><path class="edge" d="M 0 0 V -31.0752 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a11 c2 node n10 p0" transform="translate(-32.8125 15.5122)"><g class="a10 c2 node n8 p0" transform="translate(-28.125 12.0777)"><g class="a8 i2 leaf node n5 p0 sample" transform="translate(18.75 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><g class="a8 c2 node n7 p0" transform="translate(-18.75 1.62753)"><g class="a7 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.85771)"><path class="edge" d="M 0 0 V -1.85771 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a7 i0 leaf node n1 p0 sample" transform="translate(12.5 1.85771)"><path class="edge" d="M 0 0 V -1.85771 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><path class="edge" d="M 0 0 V -1.62753 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">7</text></g><path class="edge" d="M 0 0 V -12.0777 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(12.5 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 i1 leaf node n2 p0 sample" transform="translate(-12.5 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -11.4842 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -15.5122 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">11</text></g></g></g></g></g>
</svg>
</div>
</section>
<section id="simulating-mutations-with-msprime" class="slide level2">
<h2>Simulating mutations with msprime</h2>
<div class="cell" data-hash="index_cache/revealjs/sim_mutations_1cb5a04b871d95bbb0fb051e2ed3a1fd">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1"></a><span class="im">import</span> msprime</span>
<span id="cb3-2"><a href="#cb3-2"></a>ts <span class="op">=</span> msprime.sim_ancestry(</span>
<span id="cb3-3"><a href="#cb3-3"></a>    samples<span class="op">=</span><span class="dv">3</span>,</span>
<span id="cb3-4"><a href="#cb3-4"></a>    recombination_rate<span class="op">=</span><span class="fl">1e-8</span>,</span>
<span id="cb3-5"><a href="#cb3-5"></a>    sequence_length<span class="op">=</span><span class="dv">5_000</span>,</span>
<span id="cb3-6"><a href="#cb3-6"></a>    population_size<span class="op">=</span><span class="dv">10_000</span>,</span>
<span id="cb3-7"><a href="#cb3-7"></a>    random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb3-8"><a href="#cb3-8"></a>mutated_ts <span class="op">=</span> msprime.sim_mutations(ts, rate<span class="op">=</span><span class="fl">1e-8</span>, random_seed<span class="op">=</span><span class="dv">54321</span>)</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="bu">print</span>(mutated_ts.draw_svg())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<svg baseprofile="full" height="200" version="1.1" width="800" xmlns="http://www.w3.org/2000/svg" xmlns:ev="http://www.w3.org/2001/xml-events" xlink="http://www.w3.org/1999/xlink">
<defs>
<style type="text/css">.background path {fill: #808080; fill-opacity: 0}.background path:nth-child(odd) {fill-opacity: .1}.axes {font-size: 14px}.x-axis .tick .lab {font-weight: bold; dominant-baseline: hanging}.axes, .tree {font-size: 14px; text-anchor: middle}.axes line, .edge {stroke: black; fill: none}.axes .ax-skip {stroke-dasharray: 4}.y-axis .grid {stroke: #FAFAFA}.node > .sym {fill: black; stroke: none}.site > .sym {stroke: black}.mut text {fill: red; font-style: italic}.mut.extra text {fill: hotpink}.mut line {fill: none; stroke: none}.mut .sym {fill: none; stroke: red}.mut.extra .sym {stroke: hotpink}.node .mut .sym {stroke-width: 1.5px}.tree text, .tree-sequence text {dominant-baseline: central}.plotbox .lab.lft {text-anchor: end}.plotbox .lab.rgt {text-anchor: start}</style>
</defs><g class="tree-sequence"><g class="background"><path d="M20,0 l190,0 l0,138.2 l-168.568,25 l0,5 l-21.432,0 l0,-5 l0,-25 l0,-138.2z"></path><path d="M210,0 l190,0 l0,138.2 l-261.136,25 l0,5 l-97.432,0 l0,-5 l168.568,-25 l0,-138.2z"></path><path d="M400,0 l190,0 l0,138.2 l38.152,25 l0,5 l-489.288,0 l0,-5 l261.136,-25 l0,-138.2z"></path><path d="M590,0 l190,0 l0,138.2 l0,25 l0,5 l-151.848,0 l0,-5 l-38.152,-25 l0,-138.2z"></path></g><g class="axes"><g class="x-axis"><g class="title" transform="translate(400 200)"><text class="lab" text-anchor="middle" transform="translate(0 -11)">Genome position</text></g><line class="ax-line" x1="20" x2="780" y1="163.2" y2="163.2"></line><g class="ticks"><g class="tick" transform="translate(20 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">0</text></g></g><g class="tick" transform="translate(41.432 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">141</text></g></g><g class="tick" transform="translate(138.864 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">782</text></g></g><g class="tick" transform="translate(628.152 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">4001</text></g></g><g class="tick" transform="translate(780 163.2)"><line x1="0" x2="0" y1="0" y2="5"></line><g transform="translate(0 6)"><text class="lab">5000</text></g></g></g><g class="site s0" transform="translate(33.68 163.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m0"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s1" transform="translate(70.616 163.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m1"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s2" transform="translate(144.488 163.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m2"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g><g class="site s3" transform="translate(507.008 163.2)"><line class="sym" x1="0" x2="0" y1="0" y2="-10"></line><g class="mut m3"><polyline class="sym" points="2.5,-6.5 0,-1.5 -2.5,-6.5"></polyline></g></g></g></g><g class="plotbox trees"><g class="tree t0" transform="translate(20 0)"><g class="plotbox"><g class="c2 node n13 p0 root" transform="translate(124.688 26.8)"><g class="a13 i1 leaf node n3 p0 sample" transform="translate(32.8125 94.6)"><path class="edge" d="M 0 0 V -94.6 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a13 c2 m0 node n10 p0 s0" transform="translate(-32.8125 79.037)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -79.037 H 32.8125"></path><g class="mut m0 s0" transform="translate(0 -0.962491)"><line x1="0" x2="0" y1="0" y2="0.962491"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">0</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">13</text></g></g></g><g class="tree t1" transform="translate(210 0)"><g class="plotbox"><g class="c2 node n12 p0 root" transform="translate(124.688 50.9994)"><g class="a12 i1 leaf node n3 p0 sample" transform="translate(32.8125 70.4006)"><path class="edge" d="M 0 0 V -70.4006 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a12 c2 m1 node n10 p0 s1" transform="translate(-32.8125 54.8376)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -54.8376 H 32.8125"></path><g class="mut m1 s1" transform="translate(0 -44.3139)"><line x1="0" x2="0" y1="0" y2="44.3139"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">1</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">12</text></g></g></g><g class="tree t2" transform="translate(400 0)"><g class="plotbox"><g class="c2 node n11 p0 root" transform="translate(124.688 90.3248)"><g class="a11 i1 leaf m3 node n3 p0 s3 sample" transform="translate(32.8125 31.0752)"><path class="edge" d="M 0 0 V -31.0752 H -32.8125"></path><g class="mut m3 s3" transform="translate(0 -1.57025)"><line x1="0" x2="0" y1="0" y2="1.57025"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab rgt" transform="translate(5 0)">3</text></g><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a11 c2 node n10 p0" transform="translate(-32.8125 15.5122)"><g class="a10 c2 node n8 p0" transform="translate(28.125 12.0777)"><g class="a8 i0 leaf node n1 p0 sample" transform="translate(-12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><g class="a8 i2 leaf node n5 p0 sample" transform="translate(12.5 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><path class="edge" d="M 0 0 V -12.0777 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">8</text></g><g class="a10 c2 m2 node n9 p0 s2" transform="translate(-28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(18.75 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 c2 node n6 p0" transform="translate(-18.75 3.0756)"><g class="a6 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a6 i1 leaf node n2 p0 sample" transform="translate(12.5 1.00319)"><path class="edge" d="M 0 0 V -1.00319 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -3.0756 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">6</text></g><path class="edge" d="M 0 0 V -11.4842 H 28.125"></path><g class="mut m2 s2" transform="translate(0 -8.93025)"><line x1="0" x2="0" y1="0" y2="8.93025"></line><path class="sym" d="M -3,-3 l 6,6 M -3,3 l 6,-6"></path><text class="lab lft" transform="translate(-5 0)">2</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -15.5122 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">11</text></g></g></g><g class="tree t3" transform="translate(590 0)"><g class="plotbox"><g class="c2 node n11 p0 root" transform="translate(124.688 90.3248)"><g class="a11 i1 leaf node n3 p0 sample" transform="translate(32.8125 31.0752)"><path class="edge" d="M 0 0 V -31.0752 H -32.8125"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">3</text></g><g class="a11 c2 node n10 p0" transform="translate(-32.8125 15.5122)"><g class="a10 c2 node n8 p0" transform="translate(-28.125 12.0777)"><g class="a8 i2 leaf node n5 p0 sample" transform="translate(18.75 3.48524)"><path class="edge" d="M 0 0 V -3.48524 H -18.75"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">5</text></g><g class="a8 c2 node n7 p0" transform="translate(-18.75 1.62753)"><g class="a7 i0 leaf node n0 p0 sample" transform="translate(-12.5 1.85771)"><path class="edge" d="M 0 0 V -1.85771 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">0</text></g><g class="a7 i0 leaf node n1 p0 sample" transform="translate(12.5 1.85771)"><path class="edge" d="M 0 0 V -1.85771 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">1</text></g><path class="edge" d="M 0 0 V -1.62753 H 18.75"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">7</text></g><path class="edge" d="M 0 0 V -12.0777 H 28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">8</text></g><g class="a10 c2 node n9 p0" transform="translate(28.125 11.4842)"><g class="a9 i2 leaf node n4 p0 sample" transform="translate(12.5 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H -12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">4</text></g><g class="a9 i1 leaf node n2 p0 sample" transform="translate(-12.5 4.07879)"><path class="edge" d="M 0 0 V -4.07879 H 12.5"></path><rect class="sym" height="6" width="6" x="-3" y="-3"></rect><text class="lab" transform="translate(0 11)">2</text></g><path class="edge" d="M 0 0 V -11.4842 H -28.125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab rgt" transform="translate(3 -7.0)">9</text></g><path class="edge" d="M 0 0 V -15.5122 H 32.8125"></path><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab lft" transform="translate(-3 -7.0)">10</text></g><circle class="sym" cx="0" cy="0" r="3"></circle><text class="lab" transform="translate(0 -11)">11</text></g></g></g></g></g>
</svg>
</div>
</section></section>
<section id="msprime-exercise" class="title-slide slide level1 unnumbered unlisted center">
<h1>msprime exercise</h1>

</section>

<section>
<section id="forward-simulation-in-slim" class="title-slide slide level1 center">
<h1>Forward simulation in SLiM</h1>
<aside class="notes">
<p>This is currently a <em>very</em> brief introduction to SLiM. Future versions of the lecture will add some more detail.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="slim" class="slide level2">
<h2>SLiM</h2>
<p>SLiM (<strong>S</strong>election on <strong>Li</strong>nked <strong>M</strong>utations) <span class="citation" data-cites="haller_SLiMForwardGenetic_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Haller &amp; Messer, 2019</a>)</span> is a forwards-time simulator. As its name implies, it models selection and thus is a complement to coalescent-based simulators.</p>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="why-slim">Why SLiM?</h3>
<ol type="1">
<li><strong>flexibility</strong> - scripting language <em>Eidos</em> allows for modelling complex scenarios with little code</li>
<li><strong>performance</strong> - optimized code base</li>
<li><strong>GUI</strong> - interactive execution and graphical debugging</li>
</ol>
<div class="flushright">
<p><span class="citation" data-cites="haller_SLiMEvolutionarySimulation_2022">Haller &amp; Messer (<a href="#/bibliography" role="doc-biblioref" onclick="">2022</a>)</span></p>
</div>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/images/slimgui.jpeg" width="600"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="forward-simulation-in-slim-1" class="slide level2 smaller">
<h2>Forward simulation in SLiM</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol start="0" type="1">
<li>Execution of <code>first()</code> events</li>
<li>Execution of <code>early()</code> events</li>
<li>Generation of offspring; for each offspring:
<ol type="1">
<li>Choose source subpop for parental individuals, based on migration rates</li>
<li>Choose parent 1, based on cached fitness values</li>
<li>Choose parent 2, based on fitness and any defined <code>mateChoice()</code> callbacks</li>
<li>Generate the candidate offspring, with mutation and recombination (including <code>mutation()</code> and <code>recombination()</code> callbacks)</li>
<li>Suppress/modify the candidate, using defined <code>modifyChild()</code> callbacks</li>
</ol></li>
<li>Removal of fixed mutations unless <code>convertToSubstitution==F</code></li>
<li>Offspring become parents</li>
<li>Execution of <code>late()</code> events</li>
<li>Fitness recalculation using <code>mutationEffect()</code> and <code>fitnessEffect()</code> callbacks</li>
<li>Tick/cycle count increment</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-genealogy-forward-slim_7dc3af4645f698d877c937120426c981">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy-forward-slim" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;12: Forward simulation.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-genealogy-forward-slim-1.svg" style="width:60.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>SLiM works forward in time according to a <em>tick cycle</em>, which basically is a counter of time steps.</p>
<p><span class="citation" data-cites="haller_SLiMForwardGenetic_2019">Haller &amp; Messer (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span>, pp.&nbsp;603–610 describes the steps in more details; some clarifying comments are added belowe wrt to the slide flowchart. There is <strong>a lot</strong> to unpack here, and the summary flowchart captures most of the essentials.</p>
<ol start="0" type="1">
<li>first events: prior to generation of offspring; don’t affect fitness. Rare use.</li>
<li>early: catch-all event to do things before generation of offspring</li>
<li>generation of offspring
<ol type="1">
<li>Choose source subpop for parental individuals, based on migration rates</li>
<li>Choose parent 1, based on cached fitness values</li>
<li>Choose parent 2, based on fitness and any defined <code>mateChoice()</code> callbacks</li>
<li>Generate the candidate offspring, with mutation and recombination (including <code>mutation()</code> and <code>recombination()</code> callbacks)</li>
<li>Suppress/modify the candidate, using defined <code>modifyChild()</code> callbacks</li>
</ol></li>
<li>Removal of fixed mutations unless <code>convertToSubstitution==F</code> - for simulation efficiency reasons</li>
<li>Offspring become parents</li>
<li>Execution of <code>late()</code> events - e.g., changing of selection or dominance coefficients</li>
<li>Fitness recalculation using <code>mutationEffect()</code> and <code>fitnessEffect()</code> callbacks</li>
<li>Tick/cycle count increment</li>
</ol>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="simple-neutral-simulation-in-slim" class="slide level2 smaller">
<h2>Simple neutral simulation in SLiM</h2>
<div class="cell" data-hash="index_cache/revealjs/slim-example_6a9fb82c8410a2fdba87b806d53c7017">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource c number-lines code-with-copy"><code class="sourceCode c"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">// set up a simple neutral simulation</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>initialize<span class="op">()</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="op">{</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">// set the overall mutation rate</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>initializeMutationRate<span class="op">(</span><span class="fl">1e-7</span><span class="op">);</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">// m1 mutation type: neutral</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>initializeMutationType<span class="op">(</span><span class="st">"m1"</span><span class="op">,</span> <span class="dv">0</span><span class="er">.5</span><span class="op">,</span> <span class="st">"f"</span><span class="op">,</span> <span class="dv">0</span><span class="er">.0</span><span class="op">);</span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">// g1 genomic element type: uses m1 for all mutations</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>initializeGenomicElementType<span class="op">(</span><span class="st">"g1"</span><span class="op">,</span> m1<span class="op">,</span> <span class="fl">1.0</span><span class="op">);</span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">// uniform chromosome of length 100 kb</span></span>
<span id="cb4-11"><a href="#cb4-11"></a>initializeGenomicElement<span class="op">(</span>g1<span class="op">,</span> <span class="dv">0</span><span class="op">,</span> <span class="dv">99999</span><span class="op">);</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">// uniform recombination along the chromosome</span></span>
<span id="cb4-13"><a href="#cb4-13"></a>initializeRecombinationRate<span class="op">(</span><span class="fl">1e-8</span><span class="op">);</span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="op">}</span></span>
<span id="cb4-15"><a href="#cb4-15"></a><span class="co">// create a population of 500 individuals</span></span>
<span id="cb4-16"><a href="#cb4-16"></a><span class="dv">1</span> early<span class="op">()</span></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="op">{</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>sim<span class="op">.</span>addSubpop<span class="op">(</span><span class="st">"p1"</span><span class="op">,</span> <span class="dv">500</span><span class="op">);</span></span>
<span id="cb4-19"><a href="#cb4-19"></a><span class="op">}</span></span>
<span id="cb4-20"><a href="#cb4-20"></a><span class="co">// run to tick 10000</span></span>
<span id="cb4-21"><a href="#cb4-21"></a><span class="dv">10000</span> early<span class="op">()</span></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="op">{</span></span>
<span id="cb4-23"><a href="#cb4-23"></a>sim<span class="op">.</span>simulationFinished<span class="op">();</span></span>
<span id="cb4-24"><a href="#cb4-24"></a><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Comes with a huge manual <span class="citation" data-cites="haller_SLiMEvolutionarySimulation_2022">(<a href="#/bibliography" role="doc-biblioref" onclick="">Haller &amp; Messer, 2022</a>)</span> (700+ pages!)</p>
</section>
<section id="recapitation---combining-the-best-of-two-worlds" class="slide level2">
<h2>Recapitation - combining the best of two worlds</h2>
<div style="display: grid; grid-template-columns: 4fr 3fr 3fr; grid-row-gap: 0px; grid-column-gap: 20px;">
<div>
<p>A recent addition to SLiM is that it records data in <em>tree sequence format</em> (as in msprime) <i class="fa-solid fa-arrow-right" aria-label="arrow-right"></i> we can combine backward and forward simulations!</p>
</div>
<div>
<div class="fragment" data-fragment-index="2">
<p>Backward simulation that adds coalescent (neutral) history</p>
</div>
</div>
<div>
<div class="fragment" data-fragment-index="2">
<div style="transform: translate(0, 60px);">
<div class="cell" data-hash="index_cache/revealjs/backward-simulation-recapitation_4fdd9db660f3ad5fcf784ea9d80f07c1">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/backward-simulation-recapitation-1.svg" style="width:70.0%"></p>
</div>
</div>
</div>
</div>
</div>
<div>

</div>
<div>
<div class="fragment" data-fragment-index="1">
<p>Forward simulation with selection or some other process that isn’t supported by the coalescent</p>
</div>
</div>
<div>
<div class="fragment" data-fragment-index="1">
<div class="translatey50">
<div class="cell" data-hash="index_cache/revealjs/forward-simulation-recapitation_673a9a6d74280bd68122d94903fa0ccc">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/forward-simulation-recapitation-1.svg" style="width:70.0%"></p>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="applications" class="title-slide slide level1 center">
<h1>Applications</h1>

</section>
<section id="simulation-of-null-distributions" class="slide level2">
<h2>Simulation of null distributions</h2>
<div class="compact">
<div id="fig-pope-2023-clr" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;13: Empirical distribution function of SweepFinder2 composite likelihood ratio (CLR) scores from observed data versus null simulations for three populations of squash bee (<em>E. pruinosa</em>). msprime was used to simulate data (without selection) using the inferred demographic model and recombination map. The eastern population in particular shows signs of selective sweeps. <span class="citation" data-cites="pope_ExpansionAgricultureHas_2023">Pope et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2023</a>)</span>, Fig S8.</figcaption>
<p><img data-src="assets/images/pope_2023_figs8.png" style="width:90.0%"></p>
</figure>
</div>
</div>
<div>
<p>Authors inferred demographic history and recombination map. The inferred values were used to simulate data with msprime. SweepFinder2 <span class="citation" data-cites="degiorgio_SweepFinder2IncreasedSensitivity_2016">(<a href="#/bibliography" role="doc-biblioref" onclick="">DeGiorgio et al., 2016</a>)</span> was used to calculate composite likelihood ratio (CLR) scores on both observed and simulated data to assess significance quantiles.</p>
</div>
<aside class="notes">
<p>The details of the sweep finding / paper are not the main point here; rather, we want to highlight the use of simulations to generate null distribution scores (here, for CLR values) such that significance quantiles can be derived and applied to the observed data.</p>
<p>Original figure legend:</p>
<blockquote>
<p>Empiricial distribution functions of SWEEPFINDER 2 composite likelihood ratio (CLR) scores from the observed data and null simulations for each major E. pruinosa lineage (log-transformed for visualization). The null simulations were generated using the inferred demographic model and recombination maps (see Methods in main text). The observed and null distributions match each other quite closely in the Mexican lineage (MX). In the Western lineage (CO), the tail of the CLR score distribution deviates slightly from the null simulations. In the Eastern lineage (PA), the tail of the observed CLR score distribution is extremely heavy relative to the null. This suggests the existence of selective sweeps in both CO and PA, but especially in PA</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="simulations-of-genomic-landscapes-in-monkeyflower" class="slide level2">
<h2>Simulations of genomic landscapes in monkeyflower</h2>
<div class="columns">
<div class="column" style="width:30%;">
<!-- markdownlint-disable MD013 -->
<div id="fig-stankowski-figs2" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;14: Sampling locations</figcaption>
<p><img data-src="assets/images/stankowski-2019-figs2.jpg" style="width:80.0%"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:70%;">
<!-- markdownlint-disable MD013 -->
<div id="fig-stankowski-fig7" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;15: Genomic landscapes simulated under different divergence histories.</figcaption>
<p><img data-src="assets/images/stankowski-2019-fig7.png" style="width:70.0%"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
</div>
<p>Monkeyflower (<em>Mimulus aurantiacus</em>) radiation in California shows wide adaptive range and phenotypic diversity. <span class="citation" data-cites="stankowski_WidespreadSelectionGene_2019">Stankowski et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span> <em>et al</em> use simulations (SLiM and msprime) to shed light on processes that shape genomic landscapes, which are correlated between species.</p>
<div class="fragment">
<p>We will look more closely at this system tomorrow.</p>
</div>
<aside class="notes">
<p>Simulations are based on variations of a neutral base model that starts out with an ancestral population (N=10,000) that after 10N generation splits to two daughter populations (N=10,000 each) and are simulated another 10N generations. The chromosome is 21Mbp (similar in size to a monkeyflower chromosome) consisting of a central neutrally evolving region flanked by two regions where non-neutral processes are “allowed”. Models are:</p>
<ol type="1">
<li>neutral</li>
<li>BGS (non-neutral mutations are deleterious)</li>
<li>Bateson-Dobzhansky-Muller incompatibility (BDMI); after split, fraction variants deleterious in one population, neutral in other</li>
<li>positive selection</li>
<li>BGS and positive selection</li>
<li>local adaptation; 4 but also after split some variants beneficial in one population, neutral in other</li>
</ol>
<p>Figure caption:</p>
<blockquote>
<p>Fig 7. Genomic landscapes simulated under different divergence histories. Each row of plots shows patterns of within- and between-population variation (π, dxy, and FST) across the chromosome (500-kb windows) at 5 time points (N generations, where N = 10,000) during one of the scenarios The selection parameter (Ns, where s = Ns/N), proportion of deleterious (−) and positive mutations (+), and number of migrants per generation (Nm; 0 unless stated) for these simulations are as follows: (i) neutral divergence (no selection), (ii) BGS (−Ns = 100; −prop = 0.1), (iii) BDMI (−Ns = 100, −prop = 0.05, Nm = 0.1), (iv) positive selection (+Ns = 100, +prop = 0.001), (v) BGS and positive selection (−Ns = 100, −prop = 0.1; +Ns = 100, +prop = 0.005), and (vi) local adaptation (+Ns = 100, +prop = 0.001, Nm = 0.1). The gray boxes in the first column show the areas of the chromosome that are experiencing selection, while the white central area evolves neutrally. Note that π (in populations a and b) and dxy have been mean centered so they can be viewed on the same scale. Uncentered values and additional simulations with different parameter combinations and more time points can be found in S13 Fig. BDMI, Bateson-Dobzhansky-Muller incompatibility; BGS, background selection.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="neural-network-recombination-landscape-prediction" class="slide level2">
<h2>Neural network recombination landscape prediction</h2>
<!-- markdownlint-disable MD013 -->
<div>
<div id="fig-adrion-2020-fig1" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;16: Cartoon of Recombination Landscape Estimation using Recurrent Neural Networks (ReLERNN) workflow <span class="citation" data-cites="adrion_PredictingLandscapeRecombination_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Adrion et al., 2020</a>)</span></figcaption>
<p><img data-src="assets/images/adrion-2020-fig1.jpg" style="width:60.0%"></p>
</figure>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<p>Use msprime simulations to generate training data for neural network.</p>
<aside class="notes">
<p>Methods paper; one of the first examples of using msprime simulations to generate training data for a neural network.</p>
<p>Figure caption:</p>
<blockquote>
<p>A cartoon depicting a typical workflow using ReLERNN’s four modules (shaded boxes) for (A) individually sequenced genomes or (B) pooled sequences. ReLERNN can optionally (dotted lines) utilize output from stairwayplot, SMC++, and MSMC to simulate under a demographic history with msprime. Training inlays show the network architectures used, with the GRU inlay in (B) depicting the gated connections within each hidden unit. Here, r, z, ht, and ht˜ are the reset gate, update gate, activation, and candidate activation, respectively (Cho et al.&nbsp;2014). The genotype matrix encodes alleles as reference (−1), alternative (1), or padded/missing data (0; not shown). Variant positions are encoded along the real number line (0–1).</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-adrion_PredictingLandscapeRecombination_2020" class="csl-entry" role="listitem">
Adrion, J. R., Galloway, J. G., &amp; Kern, A. D. (2020). Predicting the <span>Landscape</span> of <span>Recombination Using Deep Learning</span>. <em>Molecular Biology and Evolution</em>, <em>37</em>(6), 1790–1808. <a href="https://doi.org/10.1093/molbev/msaa038">https://doi.org/10.1093/molbev/msaa038</a>
</div>
<div id="ref-baumdicker_EfficientAncestryMutation_2022" class="csl-entry" role="listitem">
Baumdicker, F., Bisschop, G., Goldstein, D., Gower, G., Ragsdale, A. P., Tsambos, G., Zhu, S., Eldon, B., Ellerman, E. C., Galloway, J. G., Gladstein, A. L., Gorjanc, G., Guo, B., Jeffery, B., Kretzschumar, W. W., Lohse, K., Matschiner, M., Nelson, D., Pope, N. S., … Kelleher, J. (2022). Efficient ancestry and mutation simulation with msprime 1.0. <em>Genetics</em>, <em>220</em>(3), iyab229. <a href="https://doi.org/10.1093/genetics/iyab229">https://doi.org/10.1093/genetics/iyab229</a>
</div>
<div id="ref-degiorgio_SweepFinder2IncreasedSensitivity_2016" class="csl-entry" role="listitem">
DeGiorgio, M., Huber, C. D., Hubisz, M. J., Hellmann, I., &amp; Nielsen, R. (2016). <span>SweepFinder2</span>: Increased sensitivity, robustness and flexibility. <em>Bioinformatics</em>, <em>32</em>(12), 1895–1897. <a href="https://doi.org/10.1093/bioinformatics/btw051">https://doi.org/10.1093/bioinformatics/btw051</a>
</div>
<div id="ref-ferretti_DecomposingSiteFrequency_2017" class="csl-entry" role="listitem">
Ferretti, L., Ledda, A., Wiehe, T., Achaz, G., &amp; Ramos-Onsins, S. E. (2017). Decomposing the <span>Site Frequency Spectrum</span>: <span>The Impact</span> of <span>Tree Topology</span> on <span>Neutrality Tests</span>. <em>Genetics</em>, <em>207</em>(1), 229–240. <a href="https://doi.org/10.1534/genetics.116.188763">https://doi.org/10.1534/genetics.116.188763</a>
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-haller_SLiMForwardGenetic_2019" class="csl-entry" role="listitem">
Haller, B. C., &amp; Messer, P. W. (2019). <span>SLiM</span> 3: <span>Forward Genetic Simulations Beyond</span> the <span>Wright</span>. <em>Molecular Biology and Evolution</em>, <em>36</em>(3), 632–637. <a href="https://doi.org/10.1093/molbev/msy228">https://doi.org/10.1093/molbev/msy228</a>
</div>
<div id="ref-haller_SLiMEvolutionarySimulation_2022" class="csl-entry" role="listitem">
Haller, B. C., &amp; Messer, P. W. (2022). <em><span>SLiM</span>: <span>An Evolutionary Simulation Framework</span></em>.
</div>
<div id="ref-hein2005gene" class="csl-entry" role="listitem">
Hein, J., Schierup, M. H., &amp; Wiuf, C. (2005). <em>Gene genealogies, variation and evolution: <span>A</span> primer in coalescent theory</em>. <span>Oxford University Press</span>. <a href="https://books.google.se/books?id=CCmLNAEACAAJ">https://books.google.se/books?id=CCmLNAEACAAJ</a>
</div>
<div id="ref-kelleher_InferringWholegenomeHistories_2019" class="csl-entry" role="listitem">
Kelleher, J., Wong, Y., Wohns, A. W., Fadil, C., Albers, P. K., &amp; McVean, G. (2019). Inferring whole-genome histories in large population datasets. <em>Nature Genetics</em>, <em>51</em>(9), 1330–1338. <a href="https://doi.org/10.1038/s41588-019-0483-y">https://doi.org/10.1038/s41588-019-0483-y</a>
</div>
<div id="ref-pope_ExpansionAgricultureHas_2023" class="csl-entry" role="listitem">
Pope, N. S., Singh, A., Childers, A. K., Kapheim, K. M., Evans, J. D., &amp; López-Uribe, M. M. (2023). The expansion of agriculture has shaped the recent evolutionary history of a specialized squash pollinator. <em>Proceedings of the National Academy of Sciences</em>, <em>120</em>(15), e2208116120. <a href="https://doi.org/10.1073/pnas.2208116120">https://doi.org/10.1073/pnas.2208116120</a>
</div>
<div id="ref-stankowski_WidespreadSelectionGene_2019" class="csl-entry" role="listitem">
Stankowski, S., Chase, M. A., Fuiten, A. M., Rodrigues, M. F., Ralph, P. L., &amp; Streisfeld, M. A. (2019). Widespread selection and gene flow shape the genomic landscape during a radiation of monkeyflowers. <em>PLOS Biology</em>, <em>17</em>(7), e3000391. <a href="https://doi.org/10.1371/journal.pbio.3000391">https://doi.org/10.1371/journal.pbio.3000391</a>
</div>
<div id="ref-wakeley_CoalescentTheoryIntroduction_2008" class="csl-entry" role="listitem">
Wakeley, J. (2008). <em>Coalescent <span>Theory</span>: <span>An Introduction</span></em> (1st Edition edition). <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-y.c.brandt_EvaluationMethodsEstimating_2022" class="csl-entry" role="listitem">
Y. C. Brandt, D., Wei, X., Deng, Y., Vaughn, A. H., &amp; Nielsen, R. (2022). Evaluation of methods for estimating coalescence times using ancestral recombination graphs. <em>Genetics</em>, iyac044. <a href="https://doi.org/10.1093/genetics/iyac044">https://doi.org/10.1093/genetics/iyac044</a>
</div>
</div>

<div class="footer footer-default">
<p>Simulation primer</p>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-pointer/pointer.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'pointer': {"key":"q","color":"red","pointerSize":8,"alwaysVisible":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'h.v',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'default',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: true,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, RevealPointer, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script type="text/javascript">
      function hyperlink_logo() {
        var button = document.createElement( 'div' );
        button.className = "logo-button";
        button.id = "logo-button";
        button.style.visibility = "visible";
        button.style.position = "absolute";
        button.style.zIndex = 30;
        button.style.fontSize = "24px";
        button.style.left = "130px";
        button.style.bottom = "14px";
        button.style.top = "auto";
        button.style.right = "auto";
        if (window.location.host.startsWith("localhost")) {
          button.innerHTML = `<a href="http://${window.location.host}/slides" title="Slides listing"><i class="fa fa-home"></i></a>`
        } else {
          button.innerHTML = '<a href="https://nbisweden.github.io/workshop-pgip/slides" title="Slides listing"><i class="fa fa-home"></i></a>';
        }
        document.querySelector( ".reveal" ).appendChild( button );
      }
    window.document.addEventListener("DOMContentLoaded", function (event) {
      hyperlink_logo();
    });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>