---
title: "Data generation and processing"
subtitle: "Sequencing protocols, best practice variant calling and filtering"
author:
    - "Per Unneberg"
format:
  revealjs:
    footer: Data processing
---

## Setup  {visibility="hidden" .unnumbered .unlisted}

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}

```{r libs}
#| echo: false
#| eval: true
#| cache: false
library(readxl)
library(curl)
```

# Data generation

## Textbook data is clean

<!-- markdownlint-disable MD013 -->

```{bash }
#| label: tskit-vcf-data
#| echo: false
#| eval: true
chromhead=$(bcftools view -h ../../data/data/ooa-outgroups/ooa.chr21.gatk.vcf.gz | tail -n 1)
tskit vcf ../../data/data/ooa-outgroups/ooa.chr21.ts | head -4
echo -e "$chromhead"
tskit vcf ../../data/data/ooa-outgroups/ooa.chr21.ts | tail -n +7 | head -n 5
```

<!-- markdownlint-enable MD013 -->

FIXME: Include McDonald-Kreitman ADH alignment

::: {.fragment}

Variant calls need to be generated from sequencing data and need to be
filtered and preprocessed before any subsequent analyses. There are
many possible reasons, including:

- uneven mapping
- incomplete reference

:::

## Real data

![Aligments of three samples to chromosome 21](assets/images/chr21_example.png)

## Population genomics - the data

Since the goal of population genomics is to analyze variation in a set
of individuals, data generation consists of compiling variation data
from individuals. Here the focus is on next-generation sequencing
data.

<!-- markdownlint-disable MD013 -->

```{r }
#| label: fig-sequencing-cost
#| echo: false
#| eval: true
#| fig-cap: Sequencing cost ($) per megabase [@wetterstrandka_DNASequencingCosts_]
url <- "https://www.genome.gov/sites/default/files/media/files/2021-11/Sequencing_Cost_Data_Table_Aug2021.xls"
if (!file.exists(basename(url)))
    download.file(url, destfile=basename(url))
data <- read_excel(basename(url))
colnames(data) <- c("Date", "Mb", "Genome")
ggplot(data, aes(x=as.Date(Date), y=Mb)) + geom_line() + xlab("") + ylab("") + scale_y_continuous(trans="log10", labels=comma, n.breaks=6) + scale_x_date(breaks="year", labels=date_format("%Y"))
```

<!-- markdownlint-enable MD013 -->

## {}

![@lou_BeginnerGuideLowcoverage_2021](assets/images/sequencing_approaches_lou_2021.png)

## RADSeq

## High coverage sequencing data

Pros: call genotypes with confidence

cons: cost

::: {.notes }

[@lou_BeginnerGuideLowcoverage_2021 p.5967] points out that "full
genome coverage is often needed to understand mechanisms of
adaptation" but that high-coverage sequencing still too expensive

:::

## Low coverage sequencing data

lcWGS - alleviates cost issue

# Read mapping and variant calling

## GATK best practice

Point out: not optimal for non-model organims

## Alternative variant callers

Reference bias: plot no. hets vs coverage for real data, e.g., conifer

### freebayes

### bcftools

### ANGSd

## PoolSeq

### Popoolation

# Variant filtering

Show coverage tracks from e.g., krill

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-1
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};

\begin{scope}
  \clip (0, 0) rectangle (0, 0);
  \pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
  \matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};

  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};
\end{scope}

\end{scope}

\begin{scope}[yshift=320, xshift=330]
    \clip (0, 0) rectangle (0, 0);
    \node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
    \matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
    \matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
    \matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
    \matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
    \node at ($(Aindmask.east) + (2, 0)$) {$^*$};
    \node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};

    \begin{scope}
      \clip (0, 0) rectangle (0, 0);
      \pgfplotsset{covaxis/.append style={ymax=45}}
      \pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
      \matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
      \node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };
    \end{scope}

  \draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
  \draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
  \draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);
\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1_axis.outer south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-2
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};

\begin{scope}
  \clip (0,0) rectangle (0,0);
  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};


\end{scope}
\end{scope}

\begin{scope}[yshift=320, xshift=330]
    \clip (0,0) rectangle (0,0);
\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };
\end{scope}
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);
\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1_axis.outer south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-3
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};



  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]
    \clip (0,0) rectangle (0,0);
\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };
\end{scope}
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);
\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1mask.south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-4
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};



  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]

\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};

\begin{scope}
\clip (0,0) rectangle (0,0);
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};
\end{scope}

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);
\end{scope}

\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1mask.south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-5
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};

  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]

\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};

\begin{scope}
\clip (0,0) rectangle (0,0);
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};
\end{scope}

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };

\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);
\end{scope}
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1mask.south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-6
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};

  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]

\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};

\begin{scope}
\clip (0,0) rectangle (0,0);
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};
\end{scope}

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };

\end{scope}
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);

\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1mask.south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-7
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};

  \pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
  \matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]

\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};

\begin{scope}
\clip (0, 0) rectangle (0, 0);
\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{A3}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };

\end{scope}
\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);

\end{scope}

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1mask.south west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Coverage tracks

<!-- markdownlint-disable MD013 -->

```{r, engine='tikz', fig.ext="svg"}
#| label: coverage-filters-final
#| echo: false
#| eval: true
\addcolumnsum{\coveragetable}{A1,A2,A3}{Asum}
\addthresholdmask{\coveragetable}{A1}{A1mask}
\addthresholdmask{\coveragetable}{A2}{A2mask}
\addthresholdmask{\coveragetable}{A3}{A3mask}
\addthresholdmask[min=3, max=11]{\coveragetable}{Asum}{Asummask}
\addindividualmask[min=0, max=1000, nind=2]{\coveragetable}{A1,A2,A3}{Aindmask}

\let\Aonemask\empty
\formatmask{\coveragetable}{\Aonemask}{A1mask}
\let\Atwomask\empty
\formatmask{\coveragetable}{\Atwomask}{A2mask}
\let\Athreemask\empty
\formatmask{\coveragetable}{\Athreemask}{A3mask}
\let\Asummask\empty
\formatmask{\coveragetable}{\Asummask}{Asummask}
\let\Aindmask\empty
\formatmask{\coveragetable}{\Aindmask}{Aindmask}

\begin{tikzpicture}[x=1pt, y=1pt]

\begin{scope}[xshift=0, yshift=0]
\pic[at={(0,300)}] (A1) {coverageplot={\coveragetable}{ref}{A1}{Sample 1}{blue}};
\matrix[mask, anchor=west, at={($(A1_axis.south west)+(3, -10)$)}] (A1mask) {\Aonemask};


\pic[at={(0,150)}] (A2) {coverageplot={\coveragetable}{ref}{A2}{Sample 2}{blue}};a
\matrix[mask, anchor=west, at={($(A2_axis.south west)+(3, -10)$)}] (A2mask) {\Atwomask};


\pic[at={(0, 0)}] (A3) {coverageplot={\coveragetable}{ref}{A3}{Sample 3}{blue}};
\matrix[mask, anchor=west, at={($(A3_axis.south west)+(3, -10)$)}] (A3mask) {\Athreemask};

\end{scope}

\begin{scope}[yshift=320, xshift=330]
\node[anchor=west] at (0, 10) (title) {Individual presence/absence tracks};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -10)$)}] (A1mask1) {\Aonemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -20)$)}] (A2mask1) {\Atwomask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -30)$)}] (A3mask1) {\Athreemask};
\matrix[mask, anchor=west, at={($(0, 0)+(3, -50)$)}] (Aindmask) {\Aindmask};
\node at ($(Aindmask.east) + (2, 0)$) {$^*$};
\node[anchor=east] at ($(Aindmask.south east) + (0, -10)$) {$^* >$50\% filter};

\pgfplotsset{covaxis/.append style={ymax=45}}
\pic[at={(0, -240)}, /pgip/group, showthreshold=true, minthreshold=4, maxthreshold=10] (Asum) {coverageplot={\coveragetable}{ref}{Asum}{Total coverage}{red}};
\matrix[mask, anchor=west, at={($(Asum_axis.south west)+(3, -10)$)}] (Asummask) {\Asummask};
\node[anchor=north west, text width=180] at ($(Asummask.south west) + (0, -10)$) (Asummask_legend) {Mask generated from threshold 4-10 (shaded rectangle) };

\end{scope}


\draw[->] ($(A1mask.east) + (5, 0)$) to[->, out=0, in=180] ($(A1mask1.west) + (-5, 0)$);
\draw[->] ($(A2mask.east) + (5, 0)$) to[->, out=30, in=190] ($(A2mask1.west) + (-5, 0)$);
\draw[->] ($(A3mask.east) + (5, 0)$) to[->, out=30, in=210] ($(A3mask1.west) + (-5, 0)$);

\node at ($(A3mask.south west) + (600, 0)$) (outer) {};
\useasboundingbox (A1_axis.outer north west) rectangle (outer.north east);

%% \draw (current bounding box.north east) -- (current bounding box.north west) -- (current bounding box.south west) -- (current bounding box.south east) -- cycle;


\end{tikzpicture}
```

<!-- markdownlint-enable MD013 -->

## Filters and masks

Implement:

## General filters {.smallest}

<!-- markdownlint-disable MD013 -->

| Category                                       | Filter                                      | Recommendation (examples)                                              |
|------------------------------------------------|---------------------------------------------|------------------------------------------------------------------------|
| General filters                                | Base quality                                | Recalibrate / <Q20                                                     |
|                                                | Mapping quality                             | MAQ < 20 / improper pairs                                              |
|                                                | Minimum depth and/ or number of individuals | Varies; e.g. <50% individuals, <0.8X average depth                     |
|                                                | Maximum depth                               | 1-2 sd above median depth                                              |
|                                                | Duplicate reads                             | Remove                                                                 |
|                                                | Indels                                      | Realign reads / haplotype-based caller / exclude bases flanking indels |
|                                                | Overlapping sections of paired-­end reads   | Soft-clip to avoid double-counting                                     |
| Filters on polymorphic sites                   | $p$-value                                   | $10^{-6}$                                                              |
|                                                | SNPs with more than two alleles             | Filter; methods often assume bi-allelic sites                          |
|                                                | Minimum minor allele frequency (MAF)        | 1%-10% for some analyses (PCA/admixture/LD/$\mathsf{F_{ST}}$           |
| Restricting analysis to a predefined site list | List of global SNPs                         | Use global call set for analyses requiring shared sites                |

: Key data filters [Table 3 @lou_BeginnerGuideLowcoverage_2021, p.
  5974] {#tbl-key-data-filters}

<!-- markdownlint-enable MD013 -->

::: {.notes }

NB: bypassing recommendations often means doing custom analyses. For
instance, @talla_DissectingEffectsSelection_2019 include GATK
tri-allelic sites due to different bi-allelic pairs segregating in
different subpopulations (e.g. A/G in pop 1, A/T in pop 2)

#### Verbose explanations of filters

Source: [@lou_BeginnerGuideLowcoverage_2021]

Base quality scores are factored into the calculation of genotype
likelihoods, so if they accurately reflect the probability of
sequencing error, bases with low scores also carry useful information.
However, base quality scores are sometimes miscalibrated, so noise may
be reduced if bases with scores below a threshold (e.g., 20) are
either trimmed off prior to analysis or ignored. Alternatively, all
base quality scores can be recalibrated based on estimated error
profiles in the data (see Section 3.1).

Mapping quality is not considered in genotype likelihood estimation in
currently available tools, so it is often advisable to remove
low-­confidence and/or nonuniquely mapped reads prior to analysis
(e.g., reads with mapping quality <20). Filtering out reads that do
not map in proper pairs should also further increase confidence in
reads being mapped to the correct location, but could cause biases in
regions with structural variation.

To avoid sites with low or confounding data support in downstream
analysis, minimum depth and/or minimum number of individual filters
can be used to exclude sites with much reduced sequencing coverage
compared to the rest of the genome (e.g., regions with low unique
mapping rates, such as repetitive sequences). Appropriate thresholds
will vary between data sets, but could, for example, exclude sites
with read data for <50% of individuals (globally or within each
population), or with <0.8× average depth across individuals (after
filtering on mapping quality)

Maximum depth filters are used to exclude sites with exceptionally
high coverage (e.g., regions that are susceptible to dubious mapping,
such as copy number variants). Common maximum depth thresholds could
be one or two standard deviations above the median genome-­wide depth.

PCR and optical duplicates can give inflated impressions of how many
unique molecules have been sequenced, which—­particularly in the
presence of preferential amplification of one allele—­ could bias
genotype likelihood estimation. We therefore recommend removing
duplicate reads prior to any analysis.

Reads mapped across indels are frequently misaligned, especially if
the ends of reads span an indel. To avoid false SNP calls, we
recommend either using dedicated tools to realign reads covering
indels, using a haplotype-­based variant caller (e.g., freebayes or
gatk) to estimate genotype likelihoods, or excluding bases flanking
indels.

If the DNA insert in a library fragment is shorter than the combined
length of paired reads, there will be a section of overlap between the
forward and reverse reads. While some variant callers (e.g., gatk)
account for the pseudoreplication in overlapping ends of read pairs,
the current implementation of angsd treats each end of a read pair as
independent (this may change in a future release (T. Korneliussen,
personal communication)). When treated as independent, read support
for overlapping sections will be “double counted,” which may bias
genotype likelihoods. A conservative approach is to soft-­clip one of
the overlapping read ends.

The significance threshold (often in the form of maximum p-­value) can
be adjusted to fine-­tune the sensitivity of polymorphism detection,
with lower p-­values leading to fewer, but higher confidence, SNP
calls. A commonly used cut-­off is 10 −6.

Most software programs for downstream analyses assume that all SNPs
are biallelic, so SNPs with more than two alleles can be filtered out
in the SNP identification step to avoid violation of such assumptions.

For many types of analysis, such as PCA, admixture analysis, detection
of FST outliers and estimation of LD, low-­frequency SNPs are
uninformative and can even bias results (e.g. Linck & Battey, 2019;
Roesti et al., 2012). For those types of analysis, imposing a minimum
MAF filter of 1%–­10% can substantially speed up computation time.
Appropriate thresholds depend on coverage, sample size (how many
copies does an MAF threshold correspond to) and the type of downstream
analysis.

For comparison of parameter estimates for multiple populations, it is
important to ensure that data are obtained for a shared set of sites
and that SNP polarization (which allele we track the frequency of) is
consistent. For programs such as angsd where population-­specific
estimates are obtained by analysing the data from each population
separately, a good strategy is to first conduct a global SNP calling
with all samples and then restrict population-­specific analysis to
those SNPs with consistent major and minor allele designations
(-­doMajorMinor 3 in angsd) no MAF or SNP p-­value filter (because
that would incorrectly generate “missing data” if a site is fixed in a
particular population).

:::

## Gentotype likelihoods

::: {.notes }

NB: samtools and GATK may actually produce different genotypes despite
having identical GLs. Samtools applies prior $10^{-3}$ to het call,
GATK has no prior [@li_BetterUnderstandingArtifacts_2014]

:::

## Bibliography {.unnumbered .unlisted}

::: { #refs }
:::

::: {.notes }
ART: <https://www-niehs-nih-gov.ezproxy.its.uu.se/research/resources/software/biostatistics/art/index.cfm>
:::
