<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Per Unneberg">
<title>DDLS Population genomics in practice - Variant filtering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbislogo-green.svg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">DDLS Population genomics in practice</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../slides/index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Slides</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../exercises/index.html" rel="" target="">
 <span class="menu-text">Exercises</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../recipes/index.html" rel="" target="">
 <span class="menu-text">Code recipes</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/NBISwe" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-pgip" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../exercises/index.html">Exercises</a></li><li class="breadcrumb-item"><a href="../../exercises/variant_filtering/index.html">Variant filtering</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Slides</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Slides</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/pgip/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genomics in practice</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/foundations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genetics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/variant_calling/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/variant_filtering/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/genetic_diversity/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic diversity</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/population_structure/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population structure</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/demography/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Demographic inference</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/selection/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Selection</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Exercises</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Data</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/compute_environment/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compute environment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/datasets/monkeyflowers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Monkeyflowers dataset</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/variant_calling/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/variant_filtering/index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Variant filtering</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/genetic_diversity/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Genetic diversity landscapes</span></a>
  </div>
</li>
          <li class="sidebar-item">
 <span class="menu-text">exercises/population_structure/index.qmd</span>
  </li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/demography/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Demographic inference</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/selection/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Selection</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Code recipes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../recipes/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code recipes</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#background" id="toc-background" class="nav-link active" data-scroll-target="#background">Background</a>
  <ul class="collapse">
<li><a href="#some-recommended-data-filters" id="toc-some-recommended-data-filters" class="nav-link" data-scroll-target="#some-recommended-data-filters">Some recommended data filters</a></li>
  </ul>
</li>
  <li>
<a href="#sec-basic-filtering" id="toc-sec-basic-filtering" class="nav-link" data-scroll-target="#sec-basic-filtering">Basic filtering of a VCF file</a>
  <ul class="collapse">
<li><a href="#generate-random-subset-of-variants" id="toc-generate-random-subset-of-variants" class="nav-link" data-scroll-target="#generate-random-subset-of-variants">Generate random subset of variants</a></li>
  <li><a href="#generate-statistics-for-filters" id="toc-generate-statistics-for-filters" class="nav-link" data-scroll-target="#generate-statistics-for-filters">Generate statistics for filters</a></li>
  <li><a href="#filtering-the-vcf" id="toc-filtering-the-vcf" class="nav-link" data-scroll-target="#filtering-the-vcf">Filtering the VCF</a></li>
  </ul>
</li>
  <li>
<a href="#depth-filtering-of-vcf-with-invariant-sites" id="toc-depth-filtering-of-vcf-with-invariant-sites" class="nav-link" data-scroll-target="#depth-filtering-of-vcf-with-invariant-sites">Depth filtering of VCF with invariant sites</a>
  <ul class="collapse">
<li><a href="#data-summary-and-subset" id="toc-data-summary-and-subset" class="nav-link" data-scroll-target="#data-summary-and-subset">Data summary and subset</a></li>
  <li><a href="#data-for-depth-filters" id="toc-data-for-depth-filters" class="nav-link" data-scroll-target="#data-for-depth-filters">Data for depth filters</a></li>
  <li><a href="#genotype-depth-data-and-bed-output" id="toc-genotype-depth-data-and-bed-output" class="nav-link" data-scroll-target="#genotype-depth-data-and-bed-output">Genotype depth data and BED output</a></li>
  </ul>
</li>
  <li>
<a href="#sec-advanced-filtering" id="toc-sec-advanced-filtering" class="nav-link" data-scroll-target="#sec-advanced-filtering">Advanced depth filtering on BAM files</a>
  <ul class="collapse">
<li><a href="#per-sample-coverage" id="toc-per-sample-coverage" class="nav-link" data-scroll-target="#per-sample-coverage">Per sample coverage</a></li>
  <li><a href="#sample-set-coverages" id="toc-sample-set-coverages" class="nav-link" data-scroll-target="#sample-set-coverages">Sample set coverages</a></li>
  <li><a href="#total-coverage" id="toc-total-coverage" class="nav-link" data-scroll-target="#total-coverage">Total coverage</a></li>
  <li><a href="#filter-on-minimum-number-of-individuals" id="toc-filter-on-minimum-number-of-individuals" class="nav-link" data-scroll-target="#filter-on-minimum-number-of-individuals">Filter on minimum number of individuals</a></li>
  <li><a href="#sequence-masks-1" id="toc-sequence-masks-1" class="nav-link" data-scroll-target="#sequence-masks-1">Sequence masks</a></li>
  </ul>
</li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Variant filtering</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Per Unneberg </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">04-Nov-2023</p>
    </div>
  </div>
  
    
  </div>
  

</header><!-- markdownlint-disable MD041 --><!-- markdownlint-disable MD041 --><!-- markdownlint-enable MD041 --><!-- markdownlint-disable MD041 --><!-- markdownlint-enable MD041 --><!-- markdownlint-enable MD041 --><div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<i class="fa-solid fa-server" aria-label="server"></i> Compute environment setup
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you haven’t already done so, please read <a href="../../exercises/compute_environment/index.html">Compute environment</a> for information on how to prepare your working directory.</p>
</div>
</div>
</div>
<p>In this exercise we will look at ways of filtering variant data. We will begin by applying filters to the variant file containing variant sites only, followed by an approach that filters on sequencing depth in a variant file containing both variant and invariant sites. The latter methodology can then be generalized to generate depth-based filters from BAM files.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The commands of this document have been run on a subset (a subregion) of the data. Therefore, although you will use the same commands, your results will differ from those presented here.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning objectives
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>filter variants by quality, read depth, and other metrics</li>
<li>apply filters to a VCF file</li>
<li>create per sample, per population, and total depth of coverage profiles</li>
<li>generate mask files for downstream processing</li>
</ul>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Data setup
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true"><i class="fa-solid fa-server" aria-label="server"></i> UPPMAX</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false"><i class="fa-solid fa-laptop" aria-label="laptop"></i> Local</a></li>
</ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<!-- markdownlint-disable MD038 -->
<p>Move to your course working directory <code>/proj/naiss2023-22-1084/users/USERNAME</code>, create an exercise directory called <code>monkeyflower</code> and <code>cd</code> to it:</p>
<!-- markdownlint-enable MD038 -->
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /proj/naiss2023-22-1084/users/USERNAME</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> monkeyflower</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> monkeyflower</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Retrieve the data with <code>rsync</code>. You can use the <code>--dry-run</code> option to see what is going to happen; just remove it when you are content:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the dry run option to retrieve data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rsync</span> <span class="at">--dry-run</span> <span class="at">-av</span> <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>      /proj/naiss2023-22-1084/webexport/monkeyflower/LG4/<span class="pp">*</span> .</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>Create an exercise directory called <code>monkeyflower</code> and <code>cd</code> to it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> monkeyflower</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> monkeyflower</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Retrieve the variant files from <code>https://export.uppmax.uu.se/naiss2023-22-1084/monkeyflower/LG4</code> with <code>wget</code><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">--user</span> pgip <span class="at">--password</span> PASSWORD <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">--recursive</span> <span class="at">--accept</span><span class="op">=</span><span class="st">'*.*'</span> <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">--reject</span><span class="op">=</span><span class="st">'*.gif'</span>,<span class="st">'index*'</span> <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">--no-parent</span> <span class="at">--no-directories</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">--no-clobber</span> <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">--directory-prefix</span><span class="op">=</span>. <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>     https://export.uppmax.uu.se/naiss2023-22-1084/monkeyflower/LG4/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tools
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist">
<li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">Listing</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false"><i class="fa-solid fa-server" aria-label="server"></i> UPPMAX modules</a></li>
<li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false"><i class="fa-solid fa-laptop" aria-label="laptop"></i> Conda</a></li>
</ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<ul>
<li>
<a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> <span class="citation" data-cites="danecek_TwelveYearsSAMtools_2021">(<a href="#ref-danecek_TwelveYearsSAMtools_2021" role="doc-biblioref">Danecek et al., 2021</a>)</span>
</li>
<li>
<a href="https://bedtools.readthedocs.io/en/latest/index.html">bedtools</a> <span class="citation" data-cites="quinlan_BEDToolsFlexibleSuite_2010">(<a href="#ref-quinlan_BEDToolsFlexibleSuite_2010" role="doc-biblioref">Quinlan &amp; Hall, 2010</a>)</span>
</li>
<li><a href="https://bioinf.shenwei.me/csvtk/">csvtk</a></li>
<li>
<a href="https://github.com/brentp/mosdepth">mosdepth</a> <span class="citation" data-cites="pedersen_MosdepthQuickCoverage_2018">(<a href="#ref-pedersen_MosdepthQuickCoverage_2018" role="doc-biblioref">Pedersen &amp; Quinlan, 2018</a>)</span>
</li>
<li>
<a href="https://bioinf.shenwei.me/seqkit/">seqkit</a> <span class="citation" data-cites="shen_SeqKitCrossPlatformUltrafast_2016">(<a href="#ref-shen_SeqKitCrossPlatformUltrafast_2016" role="doc-biblioref">Shen et al., 2016</a>)</span>
</li>
<li>
<a href="https://github.com/vcflib/vcflib">vcflib</a> <span class="citation" data-cites="garrison_SpectrumFreeSoftware_2022">(<a href="#ref-garrison_SpectrumFreeSoftware_2022" role="doc-biblioref">Garrison et al., 2022</a>)</span>
</li>
<li>
<a href="https://vcftools.github.io/">vcftools</a> <span class="citation" data-cites="danecek_VariantCallFormat_2011">(<a href="#ref-danecek_VariantCallFormat_2011" role="doc-biblioref">Danecek et al., 2011</a>)</span>
</li>
</ul>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<p>Execute the following command to load modules:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load uppmax bioinfo-tools <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    mosdepth/0.3.3 BEDTools/2.29.2 <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    samtools/1.17 bcftools/1.17 <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    vcftools/0.1.16 vcflib/1.0.1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>csvtk</code> has been manually added to the module system and can be loaded as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> use /proj/naiss2023-22-1084/modules</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load csvtk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<p>Copy the contents to a file <code>environment.yml</code> and install packages with <code>mamba env update -f environment.yml</code>.</p>
<pre lang="text"><code>channels:
  - conda-forge
  - bioconda
  - default
dependencies:
  - bedtools=2.31.0
  - bcftools=1.15.1
  - csvtk=0.28.0
  - mosdepth=0.3.3
  - samtools=1.15.1
  - vcflib=1.0.1
  - vcftools=0.1.16</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<section id="background" class="level2"><h2 class="anchored" data-anchor-id="background">Background</h2>
<p>Regardless of how a raw variant call set has been produced, the calls will be of varying quality for a number of reasons. For high-coverage sequencing, the two most common are incompleteness of the reference sequence and misalignments in repetitive regions <span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">(<a href="#ref-li_BetterUnderstandingArtifacts_2014" role="doc-biblioref">Li, 2014</a>)</span>. Low-coverage sequencing comes with its own biases and issues, with the most important being the difficulty to accurately call genotypes <span class="citation" data-cites="maruki_GenotypeCallingPopulationGenomic_2017">(<a href="#ref-maruki_GenotypeCallingPopulationGenomic_2017" role="doc-biblioref">Maruki &amp; Lynch, 2017</a>)</span>.</p>
<p>In order to improve the accuracy of downstream inference, a number of analysis-dependent quality control filters should be applied to the raw variant call set (for a concise summary, see <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">Lou et al. (<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">2021</a>)</span>). In this exercise, we will begin by applying filters to the variant file containing variant sites only, followed by a more general approach based on depth filtering of a variant file consisting of all sites, variant as well as invariant. If time permits, we will also look at an approach that generates depth profiles directly from read mappings (BAM files) and that works also when it is impractical due to file size to include invariant sites in the variant file.</p>
<p>It is worthwhile to spend time thinking about filtering. As we will see, there are numerous metrics to filter on, and different applications require different filters. This is not as straightforward as it first may seem, and even experts struggle to get filtering settings right.</p>
<section id="some-recommended-data-filters" class="level3"><h3 class="anchored" data-anchor-id="some-recommended-data-filters">Some recommended data filters</h3>
<p>There are many ways to filter data. Choosing the right set of filters is not easy, and choosing appropriate thresholds depends on application, among other things. Below we list some recommended data filters and thresholds that have general applicability and recently have been reviewed <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">Lou et al., 2021</a>, Table 3)</span>:</p>
<ul>
<li>
<strong>depth</strong>: Given the difficulty of accurately genotyping low-coverage sites, it is recommended to set a minimum read depth cutoff to remove false positive calls. It is also recommended to set a maximum depth cutoff as excessive coverage is often due to mappings to repetitive regions. The thresholds will depend on the depth profile over all sites, but is usually chosen as a range around the mean or median depth (e.g.&nbsp;lower threshold 0.8X mean, upper threshold median + 2 standard deviations).</li>
<li>
<strong>minimum number of individuals</strong>: To avoid sites with too much missing data across individuals, a common requirement is that a minimum number (fraction) of individuals, say 75%, have sequence coverage (depth-based filter) or genotype calls.</li>
<li>
<strong>quality</strong> (<strong><em>p</em>-value</strong>): Most variant calling software provide a Phred-scaled probability score that a genotype is a true genotype. Quality values below 20 (i.e., 1%) should not be trusted, but could be set much higher (i.e., lower <em>p</em>-value) depending on application. Note that if a VCF file includes invariant sites, they have quality values set to 0, which renders quality based filtering inappropriate.</li>
<li>
<strong>MAF</strong>: Filter sites based on a minimum minor allele frequency (MAF) threshold. The appropriate choice depends on application. For instance, for PCA or admixture analyses, low-frequency SNPs are uninformative, and a reasonably large cutoff (say, 0.05-0.10) could be set. If an analysis depends on invariant sites, this filter should not be applied.</li>
</ul>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>For applications where invariant sites should be included, such as genetic diversity calculations, neither quality nor MAF filtering should be applied.</p>
</div>
</div>
</section></section><section id="sec-basic-filtering" class="level2"><h2 class="anchored" data-anchor-id="sec-basic-filtering">Basic filtering of a VCF file<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>
</h2>
<p>We will begin by creating filters for a VCF file consisting of variant sites only for <em>red</em> and <em>yellow</em> ecotypes. Before we start, let’s review some statistics for the entire (unfiltered) call set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats variantsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> ^SN</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  128395
SN  0   number of no-ALTs:  0
SN  0   number of SNPs: 105508
SN  0   number of MNPs: 0
SN  0   number of indels:   23086
SN  0   number of others:   0
SN  0   number of multiallelic sites:   10412
SN  0   number of multiallelic SNP sites:   2108</code></pre>
</div>
</div>
<p>Keep track of these numbers as we will use them to evaluate the effects of filtering.</p>
<section id="generate-random-subset-of-variants" class="level3"><h3 class="anchored" data-anchor-id="generate-random-subset-of-variants">Generate random subset of variants</h3>
<p>Depending on the size of a study, both in terms of reference sequence and number of samples, the VCF output can become large; a VCF file may in fact contain millions of variants and be several hundred GB! We want to create filters by examining distributions of VCF quality metrics and setting reasonable cutoffs. In order to decrease run time, we will look at a random sample of sites. We use the <code>vcflib</code> program <code>vcfrandomsample</code> to randomly sample approximately 100,000 sites from our VCF file<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set parameter r = 100000 / total number of variants</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view variantsites.vcf.gz <span class="kw">|</span> <span class="ex">vcfrandomsample</span> <span class="at">-r</span> 0.8 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">bgzip</span> <span class="at">-c</span> <span class="op">&gt;</span> variantsites.subset.vcf.gz</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index variantsites.subset.vcf.gz</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats variantsites.subset.vcf.gz <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grep</span> <span class="st">"number of records:"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of records:  102671</code></pre>
</div>
</div>
<p>The <code>-r</code> parameter sets the <em>rate</em> of sampling which is why we get <em>approximately</em> 100,000 sites. You will need to adjust this parameter accordingly.</p>
<p>We will now use <code>vcftools</code> to compile statistics. By default, <code>vcftools</code> outputs results to files with a prefix <code>out.</code> in the current directory. You can read up on settings and options by consulting the man pages with <code>man vcftools</code><a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>. Therefore, we define a variable <code>OUT</code> where we will output our quality metrics, along with a variable referencing our variant subset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> vcftools</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>vcftools/variantsites.subset</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="va">VCF</span><span class="op">=</span>variantsites.subset.vcf.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="generate-statistics-for-filters" class="level3"><h3 class="anchored" data-anchor-id="generate-statistics-for-filters">Generate statistics for filters</h3>
<p><code>vcftools</code> can compile many different kinds of statistics. Below we will focus on the ones relevant to our data filters. We will generate metrics and plot results as we go along, with the goal of generating a set of filtering thresholds to apply to the data.</p>
<section id="total-depth-per-site" class="level4"><h4 class="anchored" data-anchor-id="total-depth-per-site">Total depth per site</h4>
<p>To get a general overview of depth of coverage, we first generate the average depth per <em>sample</em><a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${OUT}</span>.idepth</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-f</span> MEAN_DEPTH:mean <span class="va">${OUT}</span>.idepth</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INDV    N_SITES MEAN_DEPTH
PUN-R-ELF   102491  8.56185
PUN-R-JMC   102466  9.53679
PUN-R-LH    102501  9.28071
PUN-R-MT    102469  9.47234
PUN-R-UCSD  102546  8.93438
PUN-Y-BCRD  102467  9.95467
PUN-Y-INJ   102443  8.19653
PUN-Y-LO    102429  7.8115
PUN-Y-PCT   102423  9.21892
PUN-Y-POTR  102389  8.4786
MEAN_DEPTH:mean
8.94</code></pre>
</div>
</div>
<p>The average coverage over all samples is 8.9X. This actually is in the low range for a protocol based on explicitly calling genotypes. At 5X coverage, there may be a high probability that only one of the alleles has been sampled <span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">(<a href="#ref-nielsen_GenotypeSNPCalling_2011" role="doc-biblioref">Nielsen et al., 2011</a>)</span>, whereby sequencing errors may be mistaken for true variation.</p>
<p>Then we calculate depth per site to see if we can identify reasonable depth cutoffs:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--site-depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.ldepth</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHROM   POS SUM_DEPTH   SUMSQ_DEPTH
LG4 2105    9   23
LG4 2130    11  21</code></pre>
</div>
</div>
<p>So, for each position, we have a value (column <code>SUM_DEPTH</code>) for the total depth across all samples.</p>
<p>We plot the distribution of total depths by counting how many times each depth is observed. This can be done with <code>csvtk summary</code> where we count positions and group (<code>-g</code>) by the <code>SUM_DEPTH</code> value:<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-g</span> SUM_DEPTH <span class="at">-f</span> POS:count <span class="at">-w</span> 0 <span class="va">${OUT}</span>.ldepth <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> sort <span class="at">-t</span> <span class="at">-k</span> 1:n <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> plot line <span class="at">-t</span> <span class="at">-</span> <span class="at">-x</span> SUM_DEPTH <span class="at">-y</span> POS:count <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ylab</span> <span class="st">"Genome coverage (bp)"</span> <span class="dt">\</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> <span class="va">$OUT</span>.ldepth.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-csvtk-plot-vcftools-depth-per-site" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.ldepth.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="vcftools/variantsites.subset.ldepth.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;1: Distribution of the total depth per site for all samples.</figcaption></figure>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
On csvtk as plotting software
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You are of course perfectly welcome to use <code>R</code> or some other software to make these plots. We choose to generate the plots using <code>csvtk</code> to avoid too much context switching, and also because it emulates much of the functionality in <code>R</code>, albeit much less powerful when it comes to plotting.</p>
</div>
</div>
</div>
<p>As <a href="#fig-csvtk-plot-vcftools-depth-per-site">Figure&nbsp;1</a> shows, most sites fall within a peak, but also that there are sites with very high coverage, up to ten times as high as the depth at the peak maximum. We calculate some range statistics to get an idea of the spread. The following <code>csvtk</code> command will calculate the minimum, first quartile, median, mean, third quartile, and maximum of the third column (<code>SUM_DEPTH</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-f</span> 3:min,3:q1,3:median,3:mean,3:q3,3:max,3:stdev vcftools/variantsites.subset.ldepth</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SUM_DEPTH:min   SUM_DEPTH:q1    SUM_DEPTH:median    SUM_DEPTH:mean  SUM_DEPTH:q3    SUM_DEPTH:max   SUM_DEPTH:stdev
1.00    55.00   72.00   89.27   88.00   33988.00    249.26</code></pre>
</div>
</div>
<p>The range from the first quartile (<code>q1</code>) to the third (<code>q3</code>) is 55-88, showing most sites have a depth between 50-100X. We redraw the plot to zoom in on the peak:</p>
<div id="fig-csvtk-plot-vcftools-depth-per-site-zoom-in" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.ldepth.zoomin.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="vcftools/variantsites.subset.ldepth.zoomin.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;2: Zoomed in version of <a href="#fig-csvtk-plot-vcftools-depth-per-site">Figure&nbsp;1</a> which was achieved by adding the options <code>--x-min 0 --x-max 140 --y-max 2500</code> to the plotting call.</figcaption></figure>
</div>
<p>We could choose a filter based on the quantile statistics above, or by eye-balling the graph. In this example, we could have chosen the range 50-150X, which equates to 5-15X depth <em>per sample</em>; note that your values will probably be different.</p>
<p>As an aside, we mention that there is a command to directly get the per-site mean depth, <code>--site-mean-depth</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--site-mean-depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.ldepth.mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHROM   POS MEAN_DEPTH  VAR_DEPTH
LG4 2105    0.9 1.65556
LG4 2130    1.1 0.988889</code></pre>
</div>
</div>
</section><section id="variant-quality-distribution" class="level4"><h4 class="anchored" data-anchor-id="variant-quality-distribution">Variant quality distribution</h4>
<p>Another quantity of interest is the variant quality. Recall, variant quality scores are Phred-scaled scores, meaning a value of 10 has a 10% chance of being wrong, 20 a 1% chance, and so on; you typically want to at least filter on 20 or even higher. We extract and plot the quality values below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--site-quality</span> <span class="at">--out</span> <span class="va">$OUT</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To improve histogram, filter out extreme quality scores. You</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># may have to fiddle with the exact values</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-f</span> <span class="st">"QUAL&gt;0"</span> <span class="at">-f</span> <span class="st">"QUAL&lt;1000"</span> <span class="va">${OUT}</span>.lqual  <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-g</span> QUAL <span class="at">-f</span> POS:count <span class="at">-w</span> 0 <span class="at">-</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> sort <span class="at">-t</span> <span class="at">-k</span> 1:n <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> plot hist <span class="at">-t</span> <span class="at">--bins</span> 100 <span class="at">-</span> <span class="dt">\</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">--xlab</span> <span class="st">"Quality value"</span> <span class="dt">\</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ylab</span> <span class="st">"Count"</span> <span class="dt">\</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> <span class="va">$OUT</span>.lqual.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-csvtk-vcftools-site-quality" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.lqual.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="vcftools/variantsites.subset.lqual.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;3: Distribution of variant quality scores.</figcaption></figure>
</div>
<p>Clearly most quality scores are above 20-30. For many applications, we <strong>recommend setting 30 as the cutoff</strong>.</p>
</section><section id="minor-allele-frequency-distribution" class="level4"><h4 class="anchored" data-anchor-id="minor-allele-frequency-distribution">Minor allele frequency distribution</h4>
<p>Since we are going to calculate nucleotide diversities, we will not filter on the minor allele frequency (MAF) here. Nevertheless, we generate the distribution and plot for discussion purposes. The <code>--freq2</code> will output the frequencies only, adding the option <code>--max-alleles 2</code> to focus only on bi-allelic sites:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--freq2</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="at">--max-alleles</span> 2 <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.frq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHROM   POS N_ALLELES   N_CHR   {FREQ}
LG4 2105    2   4   0   1
LG4 2130    2   14  0.857143    0.142857</code></pre>
</div>
</div>
<p>The last two columns are frequencies ordered according to the reference allele. Therefore, we need to pick the minimum value to get the MAF. We can use <code>csvtk mutate</code> to create a new column</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> fix <span class="at">-t</span> <span class="va">${OUT}</span>.frq <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> mutate2 <span class="at">-t</span> <span class="at">-n</span> maf <span class="at">-e</span> <span class="st">'${5} &gt; ${6} ? "${6}" : "${5}" '</span> <span class="at">-</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> plot hist <span class="at">-t</span> <span class="at">--bins</span> 20 <span class="at">-f</span> maf <span class="at">-</span> <span class="dt">\</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">--xlab</span> <span class="st">"Minor allele frequency"</span> <span class="dt">\</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">--ylab</span> <span class="st">"Count"</span> <span class="dt">\</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> <span class="va">$OUT</span>.frq.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-cvstk-vcftools-minor-allele-frequency" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.frq.png" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="vcftools/variantsites.subset.frq.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;4: Distribution of minor allele frequencies.</figcaption></figure>
</div>
<p>Since our variant file consists of 10 individuals, that is, 20 chromosomes, there are only so many frequencies that we can observe, which is why the histogram looks a bit disconnected<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>. In fact, given 20 chromosomes, MAF=0.05 corresponds to one alternative allele among all individuals (<em>singleton</em>), MAF=0.1 to two, and so on. The maximum value is 0.5, which is to be expected, as it is a <em>minor</em> allele frequency. We note that there are more sites with a low minor allele frequency, which in practice means there are many singleton variants.</p>
<p>This is where filtering on MAF can get tricky. Singletons may correspond to sequencing error, but if too hard a filter is applied, the resulting site frequency spectrum (SFS) will be skewed. For statistics that are based on the SFS, this may lead biased estimates. Since we will be applying such a statistic, we do not filter on the MAF here. Note, however, that for other applications, such as population structure, it may be warranted to more stringently (say, MAF&gt;0.1) filter out low-frequency variants.</p>
</section><section id="missing-data-for-individuals-and-sites" class="level4"><h4 class="anchored" data-anchor-id="missing-data-for-individuals-and-sites">Missing data for individuals and sites</h4>
<p>The proportion missing data per individual can indicate whether the input DNA was of poor quality, and that the individual should be excluded from analysis. Note that in this case, missing data refers to a missing genotype call and not sequencing depth!</p>
<p>We can calculate the proportion of missing data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--missing-indv</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.imiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INDV    N_DATA  N_GENOTYPES_FILTERED    N_MISS  F_MISS
PUN-R-ELF   102671  0   6962    0.0678088
PUN-R-JMC   102671  0   6603    0.0643122</code></pre>
</div>
</div>
<p>and look at the results, focusing on the <code>F_MISS</code> column (proportion missing sites):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot hist <span class="at">-t</span> <span class="at">--x-min</span> 0 <span class="at">-f</span> F_MISS <span class="va">${OUT}</span>.imiss <span class="op">&gt;</span> <span class="va">${OUT}</span>.imiss.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-cvstk-vcftools-proportion-missing-ind" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.imiss.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="vcftools/variantsites.subset.imiss.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;5: Distribution of missingness per sample.</figcaption></figure>
</div>
<p>Here, the proportion lies in the range 0.06-0.10 for all samples, which indicates good coverage of all samples and we refrain from taking any action.</p>
<p>Similarly, we can look at missingness per site. This is related to the filter based on minimum number of individuals suggested by <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">Lou et al. (<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">2021</a>)</span>. We calculate</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--missing-site</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.lmiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHR POS N_DATA  N_GENOTYPE_FILTERED N_MISS  F_MISS
LG4 2105    20  0   16  0.8
LG4 2130    20  0   6   0.3</code></pre>
</div>
</div>
<p>and plot to get an idea of if there are sites that lack data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot hist <span class="at">--bins</span> 20 <span class="at">-t</span> <span class="at">-f</span> F_MISS <span class="va">${OUT}</span>.lmiss <span class="op">&gt;</span> <span class="va">${OUT}</span>.lmiss.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-cvstk-vcftools-proportion-missing-site" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/variantsites.subset.lmiss.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="vcftools/variantsites.subset.lmiss.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;6: Distribution of missingness among sites.</figcaption></figure>
</div>
<p>As <a href="#fig-cvstk-vcftools-proportion-missing-site">Figure&nbsp;6</a> shows, many sites have no or little missing data, but given the low coverage, there is a non-negligible number of sites with higher missingness. We calculate range statistics to get a feeling for a good cutoff:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-f</span> 6:min,6:q1,6:median,6:mean,6:q3,6:max vcftools/variantsites.subset.lmiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>F_MISS:min  F_MISS:q1   F_MISS:median   F_MISS:mean F_MISS:q3   F_MISS:max
0.00    0.00    0.00    0.08    0.10    0.90</code></pre>
</div>
</div>
<p>The mean missingness is 8%, so we can safely use 25% missingness as threshold. Typical values of tolerated missingness lie in the range 5-25%. Note that <code>vcftools</code> interprets this value as 1 - missingness, so it has to be inverted to 75% when filtering!</p>
</section><section id="heterozygosity" class="level4"><h4 class="anchored" data-anchor-id="heterozygosity">Heterozygosity</h4>
<p><code>vcftools</code> can calculate the heterozygosity per individual. More specifically, it estimates the <a href="https://en.wikipedia.org/wiki/F-statistics">inbreeding coefficient F</a> for each individual.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--het</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${OUT}</span>.het</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INDV    O(HOM)  E(HOM)  N_SITES F
PUN-R-ELF   72254   65392.2 87059   0.31670
PUN-R-JMC   68120   65611.6 87296   0.11568
PUN-R-LH    70643   65844.7 87606   0.22050
PUN-R-MT    69677   65566.6 87194   0.19006
PUN-R-UCSD  74406   65343.0 87244   0.41382
PUN-Y-BCRD  69203   64725.9 86040   0.21005
PUN-Y-INJ   70990   63902.1 84735   0.34023
PUN-Y-LO    72585   63151.3 84157   0.44910
PUN-Y-PCT   74288   64265.6 85679   0.46804
PUN-Y-POTR  71452   64335.1 85541   0.33561</code></pre>
</div>
</div>
<p>Here, F is a measure of how much the observed homozygotes <code>O(HOM)</code> differ from the expected (<code>E(HOM)</code>; expected <em>by chance</em> under Hardy-Weinberg equilibrium), and may be negative. <code>vcftools</code> calculates F from the expression <span class="math inline">\(F=(O-E)/(N-E)\)</span><a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>, which you can verify by substituting the variables in the output.</p>
<p>If F is positive (<span class="math inline">\(O(HOM) &gt; E(HOM)\)</span>), i.e., there are more observed homozygotes than expected, then there is a deficit of heterozygotes, which could be a sign of inbreeding or signs of allelic dropout in case of low sequencing coverage.</p>
<p>If F is negative, there are fewer observed homozygotes than expected, or conversely, an excess of heterozygotes. This could be indicative of poor sequence quality (bad mappings) or contamination <span class="citation" data-cites="purcell_PLINKToolSet_2007">(<a href="#ref-purcell_PLINKToolSet_2007" role="doc-biblioref">Purcell et al., 2007</a>)</span>.</p>
<p>The underlying assumption is HWE, which holds for F=0.</p>
<p>In this case, we know that the samples are from two different populations, red and yellow. In such cases, we actually expect a deficit of heterozygotes (and consequently, positive F) simply due to something called the <a href="https://en.wikipedia.org/wiki/Wahlund_effect">Wahlund effect</a>.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>The inbreeding coefficient is a population-level statistic and is not reliable for small sample sizes (<span class="math inline">\(n&lt;10\)</span>, say). Therefore, our sample size is in the lower range and the results should be taken with a grain of salt.</p>
</div>
</div>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Use <code>bcftools view -s SAMPLENAMES | vcftools --vcf - --het --stdout</code> to calculate the heterozygosity for red and yellow samples. Substitute <code>SAMPLENAMES</code> for a comma-separated list of samples.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view <span class="at">-s</span> PUN-R-ELF,PUN-R-JMC,PUN-R-LH,PUN-R-MT,PUN-R-UCSD <span class="va">$VCF</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">vcftools</span> <span class="at">--vcf</span> <span class="at">-</span> <span class="at">--het</span> <span class="at">--stdout</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INDV    O(HOM)  E(HOM)  N_SITES F
PUN-R-ELF   41631   37597.3 56436   0.21412
PUN-R-JMC   37402   37690.3 56578   -0.01527
PUN-R-LH    39725   37787.3 56688   0.10252
PUN-R-MT    38755   37544.5 56272   0.06464
PUN-R-UCSD  43430   37384.7 56268   0.32014</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="filtering-the-vcf" class="level3"><h3 class="anchored" data-anchor-id="filtering-the-vcf">Filtering the VCF</h3>
<p>Now that we have decided on filters we can apply them to the VCF. We first set the filters as variables:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="va">MISS</span><span class="op">=</span>0.75</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="va">QUAL</span><span class="op">=</span>30</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="va">MIN_DEPTH</span><span class="op">=</span>5</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="va">MAX_DEPTH</span><span class="op">=</span>15</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and run <code>vcftools</code> as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="va">OUTVCF</span><span class="op">=</span><span class="va">${VCF</span><span class="op">%</span>.subset.vcf.gz<span class="va">}</span>.filtered.vcf.gz</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="dt">\</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">--remove-indels</span> <span class="at">--max-missing</span> <span class="va">$MISS</span> <span class="dt">\</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">--min-meanDP</span> <span class="va">$MIN_DEPTH</span> <span class="at">--max-meanDP</span> <span class="va">$MAX_DEPTH</span> <span class="dt">\</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">--minDP</span> <span class="va">$MIN_DEPTH</span> <span class="at">--maxDP</span> <span class="va">$MAX_DEPTH</span> <span class="at">--recode</span> <span class="dt">\</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">--stdout</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">gzip</span> <span class="at">-c</span> <span class="op">&gt;</span> <span class="va">$OUTVCF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare the results with the original input:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats <span class="va">$OUTVCF</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"^SN"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  37645
SN  0   number of no-ALTs:  0
SN  0   number of SNPs: 37448
SN  0   number of MNPs: 0
SN  0   number of indels:   0
SN  0   number of others:   0
SN  0   number of multiallelic sites:   1636
SN  0   number of multiallelic SNP sites:   688</code></pre>
</div>
</div>
<p>Quite a substantial portion variants have in fact been removed, which here can most likely be attributed to the low average sequencing coverage.</p>
</section></section><section id="depth-filtering-of-vcf-with-invariant-sites" class="level2"><h2 class="anchored" data-anchor-id="depth-filtering-of-vcf-with-invariant-sites">Depth filtering of VCF with invariant sites</h2>
<p>Now we turn our attention to a VCF file containing variant and invariant sites. We will generate depth-based filters, with the motivation that they represent portions of the genome that are accessible to analysis, regardless of whether they contain variants or not. In so doing, we treat filtered sites as <em>missing data</em> and do not assume that they are invariant, as many software packages do.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-coverage-filters" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><a href="index_files/figure-html/fig-coverage-filters-1.svg" class="lightbox" title="Coverage distributions for three hypothetical samples along with the cumulative coverage for all samples." data-gallery="quarto-lightbox-gallery-7"><img src="index_files/figure-html/fig-coverage-filters-1.svg" class="img-fluid figure-img" width="672"></a></p>
<figcaption class="figure-caption">Figure&nbsp;7: Coverage distributions for three hypothetical samples along with the cumulative coverage for all samples.</figcaption></figure>
</div>
</div>
</div>
<p><a href="#fig-coverage-filters">Figure&nbsp;7</a> illustrates the sequencing coverage of three samples. The important thing to note is that the coverage is uneven. Some regions lack coverage entirely, e.g., due to random sampling or errors in the reference sequence. Other regions have excessive coverage, which could be a sign of repeats that have been collapsed in the reference. A general coverage filter could then seek to mask out sites where a fraction (50%, say) of individuals have too low or excessive coverage.</p>
<p>The right panel illustrates the sum of coverages across all samples. Minimum and maximum depth filters could be applied to the aggregate coverages of all samples, or samples grouped by population, to eliminate sites confounding data support.</p>
<p>As mentioned, the VCF in this exercise contains all sites; that is, both monomorphic and polymorphic sites are present. Every site contains information about depth and other metadata, which makes it possible to apply coverage filters directly to the variant file itself.</p>
<p>However, it may not always be possible to generate a VCF with all sites. Species with large genomes will produce files so large that they prevent efficient downstream processing. Under these circumstances, <em>ad hoc</em> coverage filters can be applied to the BAM files to in turn generate sequence masks that can be used in conjunction with the variant file. This is the topic for the advanced session (<a href="#sec-advanced-filtering">Section&nbsp;4</a>).</p>
<p>Regardless of approach, the end result is a set of regions that are discarded (masked) for a given analysis. They can be given either as a BED file, or a sequence mask, which is a FASTA-like file consisting of integer digits (between <code>0</code> and <code>9</code>) for each position on a chromosome. Then, by setting a cutoff, an application can mask positions higher than that cutoff. We will generate mask files with <code>0</code> and <code>1</code> digits, an example of which is shown below, where the central 10 bases of the reference (top) are masked (bottom).</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>&gt;LG4 LG4:12000001-12100000
GGACAATTACCCCCTCCGTTATGTTTCAGT

&gt;LG4
000000000011111111110000000000</code></pre>
</div>
</div>
<section id="data-summary-and-subset" class="level3"><h3 class="anchored" data-anchor-id="data-summary-and-subset">Data summary and subset</h3>
<p>We start by summarising the raw data, as before.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> ^SN</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  100000
SN  0   number of no-ALTs:  91890
SN  0   number of SNPs: 3980
SN  0   number of MNPs: 0
SN  0   number of indels:   1197
SN  0   number of others:   0
SN  0   number of multiallelic sites:   541
SN  0   number of multiallelic SNP sites:   69</code></pre>
</div>
</div>
</section><section id="data-for-depth-filters" class="level3"><h3 class="anchored" data-anchor-id="data-for-depth-filters">Data for depth filters</h3>
<p>By now you should be familiar with the <code>vcftools</code> commands to generate relevant data for filters. In particular, we used <code>--site-depth</code> to generate depth profiles over all sites, and <code>--missing-site</code> to generate missingness data, based on genotype presence/abscence, for every site. Use these same commands again to generate a set of depth filters.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Use <code>vcflib</code> and <code>vcftools</code> to select a subset of variants from which to generate data. Use <code>vcftools</code> commands <code>--site-depth</code> and <code>--missing-site</code> as before to</p>
<ol type="1">
<li>generate data</li>
<li>(possibly) compute summary statistics with <code>csvtk</code>
</li>
<li>plot depth distributions</li>
<li>select thresholds for depth-based and missingness filters</li>
<li>filter the input VCF</li>
</ol>
<p>Call the final output file <code>allsites.filtered.vcf.gz</code> and compare your output to the input file.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<section id="allsites-subset" class="level4"><h4 class="anchored" data-anchor-id="allsites-subset">allsites subset</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set parameter r = 100000 / total number of variants; input file</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here consists of 100000 entries. Adjust this parameter.</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> view allsites.vcf.gz <span class="kw">|</span> <span class="ex">vcfrandomsample</span> <span class="at">-r</span> 1.0 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">bgzip</span> <span class="at">-c</span> <span class="op">&gt;</span> allsites.subset.vcf.gz</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> index allsites.subset.vcf.gz</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats allsites.subset.vcf.gz <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grep</span> <span class="st">"^SN"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  100000
SN  0   number of no-ALTs:  91890
SN  0   number of SNPs: 3980
SN  0   number of MNPs: 0
SN  0   number of indels:   1197
SN  0   number of others:   0
SN  0   number of multiallelic sites:   541
SN  0   number of multiallelic SNP sites:   69</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> vcftools</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>vcftools/allsites.subset</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="va">VCF</span><span class="op">=</span>allsites.subset.vcf.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="depth-per-site" class="level4"><h4 class="anchored" data-anchor-id="depth-per-site">Depth per site</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--site-depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.ldepth</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-f</span> 3:min,3:q1,3:median,3:mean,3:q3,3:max,3:stdev <span class="va">${OUT}</span>.ldepth</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHROM   POS SUM_DEPTH   SUMSQ_DEPTH
LG4 1   15  37
LG4 2   16  38
SUM_DEPTH:min   SUM_DEPTH:q1    SUM_DEPTH:median    SUM_DEPTH:mean  SUM_DEPTH:q3    SUM_DEPTH:max   SUM_DEPTH:stdev
0.00    58.00   76.00   78.93   91.00   2041.00 92.21</code></pre>
</div>
</div>
<div id="fig-csvtk-plot-vcftools-depth-per-site-zoom-in-all" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/allsites.subset.ldepth.zoomin.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="vcftools/allsites.subset.ldepth.zoomin.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;8: Zoomed in view of depth distribution for all sites.</figcaption></figure>
</div>
</section><section id="missingness" class="level4"><h4 class="anchored" data-anchor-id="missingness">Missingness</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="at">--missing-site</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.lmiss</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-f</span> 6:min,6:q1,6:median,6:mean,6:q3,6:max <span class="va">${OUT}</span>.lmiss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHR POS N_DATA  N_GENOTYPE_FILTERED N_MISS  F_MISS
LG4 1   20  0   6   0.3
LG4 2   20  0   4   0.2
F_MISS:min  F_MISS:q1   F_MISS:median   F_MISS:mean F_MISS:q3   F_MISS:max
0.00    0.00    0.10    0.16    0.20    1.00</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot hist <span class="at">--bins</span> 20 <span class="at">-t</span> <span class="at">-f</span> F_MISS <span class="va">${OUT}</span>.lmiss <span class="op">&gt;</span> <span class="va">${OUT}</span>.lmiss.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-cvstk-vcftools-proportion-missing-site-allsites" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="vcftools/allsites.subset.lmiss.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="vcftools/allsites.subset.lmiss.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;9: Distribution of missingness among all sites.</figcaption></figure>
</div>
</section><section id="filter-input" class="level4"><h4 class="anchored" data-anchor-id="filter-input">Filter input</h4>
<p>Unless there is any strange bias that leads to a difference in coverage between variant and invariant sites, the final values should be similar to those before. We set filters and generate the output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="va">MISS</span><span class="op">=</span>0.75</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="va">MIN_DEPTH</span><span class="op">=</span>5</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="va">MAX_DEPTH</span><span class="op">=</span>15</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="va">OUTVCF</span><span class="op">=</span><span class="va">${VCF</span><span class="op">%</span>.subset.vcf.gz<span class="va">}</span>.filtered.vcf.gz</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> <span class="va">$VCF</span> <span class="dt">\</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">--remove-indels</span> <span class="at">--max-missing</span> <span class="va">$MISS</span> <span class="dt">\</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">--min-meanDP</span> <span class="va">$MIN_DEPTH</span> <span class="at">--max-meanDP</span> <span class="va">$MAX_DEPTH</span> <span class="dt">\</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">--minDP</span> <span class="va">$MIN_DEPTH</span> <span class="at">--maxDP</span> <span class="va">$MAX_DEPTH</span> <span class="at">--recode</span> <span class="dt">\</span></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">--stdout</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">gzip</span> <span class="at">-c</span> <span class="op">&gt;</span> <span class="va">$OUTVCF</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare the results with the original input:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bcftools</span> stats <span class="va">$OUTVCF</span> <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"^SN"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SN  0   number of samples:  10
SN  0   number of records:  46691
SN  0   number of no-ALTs:  43386
SN  0   number of SNPs: 2244
SN  0   number of MNPs: 0
SN  0   number of indels:   0
SN  0   number of others:   0
SN  0   number of multiallelic sites:   129
SN  0   number of multiallelic SNP sites:   43</code></pre>
</div>
</div>
</section>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="genotype-depth-data-and-bed-output" class="level3"><h3 class="anchored" data-anchor-id="genotype-depth-data-and-bed-output">Genotype depth data and BED output</h3>
<p>Instead of calculating missing genotypes per site, we can retrieve the individual depth for each genotype with <code>--geno-depth</code>. Since we know cutoffs for mean depth (5-15), we can run this command on the main input file (<code>allsites.vcf.gz</code>). For reasons that soon will become clear, we also rerun <code>--site-depth-mean</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="va">VCF</span><span class="op">=</span>allsites.vcf.gz</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>vcftools/allsites</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span>  <span class="va">${VCF}</span> <span class="at">--geno-depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span>  <span class="va">${VCF}</span> <span class="at">--site-mean-depth</span> <span class="at">--out</span> <span class="va">$OUT</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${OUT}</span>.gdepth</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CHROM   POS PUN-R-ELF   PUN-R-JMC   PUN-R-LH    PUN-R-MT    PUN-R-UCSD  PUN-Y-BCRD  PUN-Y-INJ   PUN-Y-LO    PUN-Y-PCT   PUN-Y-POTR
LG4 1   1   0   3   1   3   0   3   0   2   2
LG4 2   1   1   3   1   3   0   3   0   2   2</code></pre>
</div>
</div>
<p>We could then combine these two files and perform filtering as follows: for each site check that</p>
<ol type="1">
<li>the mean depth is within the filter range</li>
<li>there is a minimimum number of genotypes with sufficient depth</li>
<li>individual genotype depth does not exceed a maximum depth threshold</li>
</ol>
<p>If any of the points above fail, the site is discarded. We keep track of sites that pass the filters and output positions in <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED format</a> (a <a href="https://www.biostars.org/p/84686/">0-based</a> tab-separated format consisting of columns <code>chrom</code>, <code>chromStart</code>, and <code>chromEnd</code>).</p>
<p>Here is some code to achieve these goals. Unfortunately <code>csvtk</code> doesn’t seem to have support for calculating column margins out of the box, which is why we have to resort to this complicated construct using <code>awk</code> to count the number of individual genotypes that pass the coverage threshold 5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="va">BEDOUT</span><span class="op">=</span><span class="va">${VCF</span><span class="op">%</span>.vcf.gz<span class="va">}</span>.keep.bed</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> join <span class="at">-t</span> <span class="va">${OUT}</span>.ldepth.mean <span class="va">${OUT}</span>.gdepth <span class="at">-f</span> CHROM,POS <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-f</span> <span class="st">"MEAN_DEPTH&gt;=5"</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-f</span> <span class="st">"MEAN_DEPTH&lt;=15"</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="at">-v</span> OFS=<span class="st">"\t"</span> <span class="dt">\</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">'NR &gt; 1 {count=0; for (i=4; i&lt;=NF; i++)\</span></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a><span class="st"> {if ($i&gt;4) count++ }; if (count&gt;=5) print $1, $2 - 1, $2}'</span><span class="kw">|</span><span class="dt">\</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">bedtools</span> merge <span class="op">&gt;</span> <span class="va">${BEDOUT}</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">$BEDOUT</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LG4 76  99
LG4 278 336
LG4 346 409</code></pre>
</div>
</div>
<p>The BED file contains a list of regions that are accessible to analysis.</p>
<section id="sequence-masks" class="level4"><h4 class="anchored" data-anchor-id="sequence-masks">Sequence masks</h4>
<p>In addition to the BED output files, we can generate sequence masks. First, we set a variable to point to the reference sequence and index it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">REF</span><span class="op">=</span>M_aurantiacus_v1.fasta</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx <span class="va">${REF}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we use the command <code>bedtools makefasta</code> to make a sequence mask file in FASTA format consisting solely of <code>1</code>’s:<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'BEGIN {OFS="\t"} {print $1, 0, $2}'</span> <span class="va">${REF}</span>.fai <span class="op">&gt;</span> <span class="va">${REF}</span>.bed</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span> <span class="at">-mc</span> 1 <span class="at">-fo</span> <span class="va">${REF}</span>.mask.fa <span class="at">-bed</span> <span class="va">${REF}</span>.bed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We generate a file where all positions are masked because the BED files contain regions that we want to keep. Therefore, we want to convert corresponding positions to zeros. This file will be used as a template for all mask files.</p>
<p>We then apply <code>bedtools maskfasta</code> again to unmask (set to <code>0</code>) the positions that overlap with the BED coordinates:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span>.mask.fa <span class="at">-mc</span> 0 <span class="at">-fo</span> <span class="va">${REF}</span>.unmask.fa <span class="dt">\</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">-bed</span> allsites.keep.bed</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${REF}</span>.unmask.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;LG4
111111111111111111111111111111111111111111111111111111111111
111111111111111100000000000000000000000111111111111111111111</code></pre>
</div>
</div>
<p>We can convince ourselves that this has worked by counting the number of unmasked positions in both the BED file (with <code>bedtools genomecov</code>) and sequence mask:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> allsites.keep.bed <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="kw">|</span> <span class="fu">grep</span> genome</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="co"># tr: -d deletes all characters not (-c, complement) in the character</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set '0'. wc: -m option counts characters</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${REF}</span>.unmask.fa <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="at">-c</span> <span class="st">'0'</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-m</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genome  0   20935   100000  0.20935
genome  1   79065   100000  0.79065
79065</code></pre>
</div>
</div>
<p>Note that 0 and 1 in the <code>bedtools genomecov</code> output refers to coverage (i.e., absence/presence) and not unmask/mask as in the mask FASTA file.</p>
</section><section id="example-using-a-sequence-mask-with-vcftools" class="level4"><h4 class="anchored" data-anchor-id="example-using-a-sequence-mask-with-vcftools">Example: Using a sequence mask with <code>vcftools</code>
</h4>
<p>The sequence mask file can be used with <code>vcftools</code> with the option <code>--mask</code>. Before we use, however, we need to convert the mask file to one sequence per line<a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a>. <code>seqkit</code> is a neat tool that allows us to do this without hassle. As an example, we then perform a genetic diversity calculation with and without mask file to highlight the difference:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="va">WIDEMASK</span><span class="op">=</span><span class="va">${REF}</span>.unmask.wide.fa</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> seq <span class="at">-w</span> 0 <span class="va">${REF}</span>.unmask.fa <span class="op">&gt;</span> <span class="va">${WIDEMASK}</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> allsites.vcf.gz <span class="at">--mask</span> <span class="va">$WIDEMASK</span> <span class="dt">\</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">--site-pi</span> <span class="at">--stdout</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">--ignore-non-numbers</span> <span class="at">--decimal-width</span> 4 <span class="dt">\</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fields</span> PI:count,PI:mean</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--gzvcf</span> allsites.vcf.gz <span class="at">--site-pi</span> <span class="at">--stdout</span> <span class="dv">2</span><span class="op">&gt;</span>/dev/null <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">--ignore-non-numbers</span> <span class="at">--decimal-width</span> 4 <span class="dt">\</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fields</span> PI:count,PI:mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PI:count    PI:mean
79065   0.0183
PI:count    PI:mean
100000  0.0208</code></pre>
</div>
</div>
<p>Clearly, filtering may have significant impact on the final outcome. You must choose your filters wisely!</p>
</section></section></section><section id="sec-advanced-filtering" class="level2"><h2 class="anchored" data-anchor-id="sec-advanced-filtering">Advanced depth filtering on BAM files</h2>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>This exercise is optional. It was created prior to the previous filtering exercises and may contain overlapping explanations and information. The procedure closely resembles that from the section on genotype depth data above, but depth profiles are now calculated from BAM files, which requires different tools.</p>
</div>
</div>
<p>In this section, we will restrict our attention to coverage-based filters, with the aim of generating <em>sequence masks</em> to denote regions of a reference sequence that contain sufficient information across individuals and populations. Furthermore, the masks will be applied in the context of genetic diversity calculations, in which case specific filters on polymorphic sites (e.g., <em>p</em>-value or minimum minor allele frequency (MAF)) should <strong>not</strong> be applied (all sites contain information).</p>
<p>Mapped reads provide information about how well a given genomic region has been represented during sequencing, and this information is usually summarized as the sequencing coverage. For any locus, this is equivalent to the number of reads mapping to that locus.</p>
<p>Sequencing coverage is typically not uniformly distributed over the reference. Reasons may vary but include uneven mapping coverage due to repeat regions, low coverage due to mis-assemblies, or coverage biases generated in the sequencing process. Importantly, both variable and monomorphic sites must be treated identically in the filtering process to eliminate biases between the two kinds of sites.</p>
<p>In this part, we will use <code>mosdepth</code> and <code>bedtools</code> to quickly generate depth of coverage profiles of mapped data. <code>mosdepth</code> is an ultra-fast command line tool for calculating coverage from a BAM file. By default, it generates a summary of the global distribution, and per-base coverage in <code>bed.gz</code> format. We will be using the per-base coverage for filtering.</p>
<p>Alternatively, <code>mosdepth</code> can also output results in a highly compressed format <code>d4</code>, which has been developed to handle the ever increasing size of resequencing projects. Files in d4 format can be processed with the <a href="https://github.com/38/d4-format">d4-tools</a> tool <span class="citation" data-cites="hou_BalancingEfficientAnalysis_2021">(<a href="#ref-hou_BalancingEfficientAnalysis_2021" role="doc-biblioref">Hou et al., 2021</a>)</span>. For instance, <code>d4tools view</code> will display the coverage in <code>bed</code> format. We mention this in passing as it may be relevant when working with large genomes or sample sizes, but given the size of our sample data, we will be using <code>bedtools</code> from now on.</p>
<section id="per-sample-coverage" class="level3"><h3 class="anchored" data-anchor-id="per-sample-coverage">Per sample coverage</h3>
<p>We start by calculating per-sample coverages with <code>mosdepth</code>. For downstream purposes, we need to save the size of the chromosomes we’re looking at, and for many applications, a fasta index file is sufficient.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">REF</span><span class="op">=</span>M_aurantiacus_v1.fasta</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx <span class="va">${REF}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The syntax to generate coverage information for a BAM file is <code>mosdepth &lt;prefix&gt; &lt;input file&gt;</code>. Here, we add the <code>-Q</code> option to exclude reads with a mapping quality less than 20:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mosdepth</span> <span class="at">-Q</span> 20 PUN-Y-INJ PUN-Y-INJ.sort.dup.recal.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The per-base coverage output file will be named <code>PUN-Y-INJ.per-base.bed.gz</code> and can be viewed with <code>bgzip</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> PUN-Y-INJ.per-base.bed.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LG4 0   9   3
LG4 9   13  4
LG4 13  53  5
LG4 53  56  6
LG4 56  62  7</code></pre>
</div>
</div>
<p>To get an idea of what the coverage looks like over the chromsome, we will make use of two versatile programs for dealing with genomic interval data and delimited text files. Both programs operate on streams, enabling the construction of powerful piped commands at the command line (see below).</p>
<p>The first is <code>bedtools</code>, which consists of a set of tools to perform <em>genome arithmetic</em> on <code>bed</code>-like file formats. The <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1"><code>bed</code></a> format is a tab-delimited format that at its simplest consists of the three columns <code>chrom</code> (the chromosome name), <code>chromStart</code>, and <code>chromEnd</code>, the start and end coordinates of a region.</p>
<p>The second is <code>csvtk</code>, a program that provides tools for dealing with tab- or comma-separated text files. Avid <code>R</code> users will see that many of the subcommands are similar to those in the <a href="https://dplyr.tidyverse.org/"><code>tidyverse dplyr</code></a> package.</p>
<p>We can use these programs in a one-liner to generate a simple coverage plot (<a href="#fig-plot-coverage">Figure&nbsp;10</a>)<a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="op">&lt;(</span><span class="ex">bedtools</span> makewindows <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-w</span> 1000<span class="op">)</span> <span class="dt">\</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">-b</span> PUN-Y-INJ.per-base.bed.gz <span class="at">-wa</span> <span class="at">-wb</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bedtools</span> groupby <span class="at">-i</span> <span class="at">-</span> <span class="at">-g</span> 1,2,3 <span class="at">-c</span> 7 <span class="at">-o</span> mean <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot <span class="at">-t</span> line <span class="at">-x</span> 2 <span class="at">-y</span> 4 <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> Position <span class="dt">\</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">--ylab</span> Coverage <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-coverage.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-plot-coverage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="fig-plot-coverage.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="fig-plot-coverage.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;10: Coverage for sample PUN-Y-INJ in 1kb windows. Experiment changing the window size (<code>-w</code>) parameter to change smoothing.</figcaption></figure>
</div>
<p>Apparently there are some high-coverage regions that could be associated with, e.g., collapsed repeat regions in the assembly. Let’s compile coverage results for all samples, using bash string manipulation to generate file prefix<a href="#fn12" class="footnote-ref" id="fnref12" role="doc-noteref"><sup>12</sup></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [A-Z] matches characters A-Z, * is a wildcard character that matches</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="co"># anything</span></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> f <span class="kw">in</span> <span class="pp">[</span><span class="ss">A</span><span class="pp">-</span><span class="ss">Z</span><span class="pp">]*</span>.sort.dup.recal.bam<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a> <span class="co"># Extract the prefix by removing .sort.dup.recal.bam</span></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a> <span class="va">prefix</span><span class="op">=</span><span class="va">${f</span><span class="op">%</span>.sort.dup.recal.bam<span class="va">}</span></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a> <span class="ex">mosdepth</span> <span class="at">-Q</span> 20 <span class="va">$prefix</span> <span class="va">$f</span></span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a> <span class="co"># Print sample name</span></span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a> <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">$prefix</span><span class="st">\t"</span></span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a> <span class="co"># Get the summary line containing the total coverage</span></span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a> <span class="fu">cat</span> <span class="va">$prefix</span>.mosdepth.summary.txt <span class="kw">|</span> <span class="fu">grep</span> total</span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span> <span class="op">&gt;</span> ALL.mosdepth.summary.txt</span>
<span id="cb75-12"><a href="#cb75-12" aria-hidden="true" tabindex="-1"></a><span class="co"># View the file</span></span>
<span id="cb75-13"><a href="#cb75-13" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> ALL.mosdepth.summary.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PUN-R-ELF   total   100000  845166  8.45    0   187
PUN-R-JMC   total   100000  934032  9.34    0   244
PUN-R-LH    total   100000  940246  9.40    0   225
PUN-R-MT    total   100000  956272  9.56    0   233
PUN-R-UCSD  total   100000  812507  8.13    0   293
PUN-Y-BCRD  total   100000  882549  8.83    0   320
PUN-Y-INJ   total   100000  818921  8.19    0   229
PUN-Y-LO    total   100000  656112  6.56    0   151
PUN-Y-PCT   total   100000  836694  8.37    0   179
PUN-Y-POTR  total   100000  795201  7.95    0   205</code></pre>
</div>
</div>
<p>We can calculate the total coverage by summing the values of the fifth column with <code>csvtk</code> as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-H</span> <span class="at">-t</span> ALL.mosdepth.summary.txt <span class="at">-f</span> 5:sum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>to get the total coverage 84.78, which gives a hint at where the diploid coverage peak should be.</p>
</section><section id="sample-set-coverages" class="level3"><h3 class="anchored" data-anchor-id="sample-set-coverages">Sample set coverages</h3>
<p>In this section, we will summarize coverage information for different sample sets, the motivation being that different filters may be warranted depending on what samples are being analysed. For instance, the coverage cutoffs for all samples will most likely be different from those applied to the subpopulations. In practice this means summing coverage tracks like those in <a href="#fig-plot-coverage">Figure&nbsp;10</a> for all samples.</p>
<p>We can combine the coverage output from different samples with <code>bedtools unionbedg</code>. We begin by generating a coverage file for all samples, where the output columns will correspond to individual samples. To save typing, we collect the sample names and generate matching BED file names to pass as arguments to options <code>-names</code> and <code>-i</code>, respectively. Also, we include positions with no coverage (<code>-empty</code>) which requires the use of a genome file (option <code>-g</code>). The BED output is piped to <code>bgzip</code> which compresses the output, before finally indexing with <code>tabix</code><a href="#fn13" class="footnote-ref" id="fnref13" role="doc-noteref"><sup>13</sup></a>.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span><span class="va">$(</span><span class="ex">csvtk</span> grep <span class="at">-f</span> Taxon <span class="at">-r</span> <span class="at">-p</span> <span class="st">"yellow"</span> <span class="at">-r</span> <span class="at">-p</span> <span class="st">"red"</span> sampleinfo.csv <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-f</span> SampleAlias <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> SampleAlias <span class="kw">|</span> <span class="fu">tr</span> <span class="st">"\n"</span> <span class="st">" "</span><span class="va">)</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="va">BEDGZ</span><span class="op">=</span><span class="va">$(</span><span class="cf">for</span> sm <span class="kw">in</span> <span class="va">$SAMPLES</span><span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${sm}</span><span class="st">.per-base.bed.gz "</span><span class="kw">;</span> <span class="cf">done</span><span class="va">)</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> unionbedg <span class="at">-header</span> <span class="at">-names</span> <span class="va">$SAMPLES</span> <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-empty</span> <span class="at">-i</span> <span class="va">$BEDGZ</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.bg.gz</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="at">-S</span> 1 ALL.bg.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<i class="fa-brands fa-linux" aria-label="linux"></i> Command line magic
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The code above works as follows. We first use <code>csvtk</code> to <code>grep</code> (search) for the population names red and yellow in the <code>Taxon</code> column in <code>sampleinfo.csv</code>, thereby filtering the output to lines where <code>Taxon</code> matches the population names. Then, we cut out the interesting column <code>SampleAlias</code> and remove the header (<code>grep -v SampleAlias</code> matches anything but <code>SampleAlias</code>). Finally, <code>tr</code> translates newline character <code>\n</code> to space. The output is stored in the <code>SAMPLES</code> variable through the command substitution (<code>$()</code>) syntax.</p>
<p>We then iterate through the <code>$SAMPLES</code> to generate the input file names with the <code>echo</code> command, storing the output in <code>$BEDGZ</code>. These variables are passed on to <code>bedtools unionbedg</code> to generate a <a href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedgraph file</a> combining all samples.</p>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<p>As mentioned previously, we also need to combine coverages per populations yellow and red.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Using the previous command as a template, try to generate per population coverage files.</p>
<div class="callout callout-style-default callout-hint no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Hint
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-hint">
<p>You only need to modify the code that generates <code>SAMPLES</code> by grepping for each population separately (<code>csvtk grep -f Taxon -r -p red</code> and so on), or setting them manually. Remember also to modify the output file name (e.g., <code>red.bg.gz</code>).</p>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<!-- markdownlint-disable MD013 -->
<p>An example using a for loop is shown here. You could copy-paste the code above and explicitly write out the population labels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a> <span class="va">SAMPLES</span><span class="op">=</span><span class="va">$(</span><span class="ex">csvtk</span> grep <span class="at">-f</span> Taxon <span class="at">-r</span> <span class="at">-p</span> <span class="va">$pop</span> sampleinfo.csv <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-f</span> SampleAlias <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> SampleAlias <span class="kw">|</span> <span class="fu">tr</span> <span class="st">"\n"</span> <span class="st">" "</span><span class="va">)</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a> <span class="va">BEDGZ</span><span class="op">=</span><span class="va">$(</span><span class="cf">for</span> sm <span class="kw">in</span> <span class="va">$SAMPLES</span><span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${sm}</span><span class="st">.per-base.bed.gz "</span><span class="kw">;</span> <span class="cf">done</span><span class="va">)</span></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a> <span class="ex">bedtools</span> unionbedg <span class="at">-header</span> <span class="at">-names</span> <span class="va">$SAMPLES</span> <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-empty</span> <span class="at">-i</span> <span class="va">$BEDGZ</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.bg.gz</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="at">-S</span> 1 <span class="va">$pop</span>.bg.gz</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="total-coverage" class="level3"><h3 class="anchored" data-anchor-id="total-coverage">Total coverage</h3>
<p>Since we eventually want to filter on total coverage, we sum per sample coverages for each sample set with <code>awk</code>:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> ALL.bg.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a> <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="at">-v</span> OFS=<span class="st">"\t"</span> <span class="st">'NR &gt; 1 {sum=0; for (i=4; i&lt;=NF; i++) sum+=$i; print $1, $2, $3, sum}'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.sum.bed.gz</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed ALL.sum.bed.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we use <code>awk</code> to sum from columns 4 and up (<code>NF</code> is the number of the last column).</p>
<!-- markdownlint-enable MD013 -->
<p>For illustration, we plot the total coverage:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="op">&lt;(</span><span class="ex">bedtools</span> makewindows <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-w</span> 1000<span class="op">)</span> <span class="dt">\</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-b</span> ALL.sum.bed.gz <span class="at">-wa</span> <span class="at">-wb</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bedtools</span> groupby <span class="at">-i</span> <span class="at">-</span> <span class="at">-g</span> 1,2,3 <span class="at">-c</span> 7 <span class="at">-o</span> mean <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot <span class="at">-t</span> line <span class="at">-x</span> 2 <span class="at">-y</span> 4 <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> Position <span class="dt">\</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ylab</span> Coverage <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-total-coverage.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<div id="fig-plot-total-coverage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><a href="fig-plot-total-coverage.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="fig-plot-total-coverage.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;11: Total coverage in 1kb windows.</figcaption></figure>
</div>
<p>In order to define thresholds for subsequent filtering, we need to know the total coverage distribution. Therefore, we plot the proportion of the genome coverage versus depth of coverage (similar to k-mer plots in sequence assembly projects). To do so we must summarize our coverage file such that we count how many bases have a given coverage. This can be achieved by noting that each row in the BED file consists of the columns <code>CHROM</code>, <code>START</code>, <code>END</code>, and <code>COVERAGE</code>. We can generate a histogram table by, for each value of <code>COVERAGE</code>, summing the length of the regions (<code>END</code> - <code>START</code>).</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add column containing length of region (end - start)</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> mutate2 <span class="at">-t</span> <span class="at">-H</span> <span class="at">-w</span> 0 <span class="at">-e</span> <span class="st">'$3-$2'</span> ALL.sum.bed.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a> <span class="co"># Sum the regions and *group* by the coverage (fourth column);</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a> <span class="co"># this gives the total number of bases with a given coverage</span></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-H</span> <span class="at">-g</span> 4 <span class="at">-f</span> 5:sum <span class="at">-w</span> 0 <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> sort <span class="at">-t</span> <span class="at">-k</span> 1:n <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a> <span class="fu">awk</span> <span class="at">-v</span> cumsum=0 <span class="st">'BEGIN {OFS=","; cumsum=0} {cumsum += $2; print $1,$2,cumsum}'</span> <span class="op">&gt;</span> ALL.sum.bed.csv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<p>We plot the coverage distribution below, along with a plot of the cumulative coverage.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot line <span class="at">-H</span> ALL.sum.bed.csv <span class="at">-x</span> 1 <span class="at">-y</span> 2 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Genome coverage (bp)"</span> <span class="dt">\</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-total-coverage-distribution.png</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot line <span class="at">-H</span> ALL.sum.bed.csv <span class="at">-x</span> 1 <span class="at">-y</span> 3 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Cumulative genome coverage (kbp)"</span> <span class="dt">\</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-total-coverage-distribution-cumulative.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-plot-total-coverage-distribution" class="quarto-layout-panel" data-attr-output=".details summary=&quot;Output&quot;">
<figure class="figure"><div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-hist" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure"><p><a href="fig-plot-total-coverage-distribution.png" class="lightbox" title="Genome coverage" data-gallery="quarto-lightbox-gallery-12"><img src="fig-plot-total-coverage-distribution.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(a) Genome coverage</figcaption></figure>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-cumulative" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure"><p><a href="fig-plot-total-coverage-distribution-cumulative.png" class="lightbox" title="Cumulative genome coverage" data-gallery="quarto-lightbox-gallery-13"><img src="fig-plot-total-coverage-distribution-cumulative.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(b) Cumulative genome coverage</figcaption></figure>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;12: Genome coverage vs depth of coverage.</figcaption><p></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<p>In <a href="#fig-plot-total-coverage-distribution">Figure&nbsp;12</a> a, a diploid peak is evident at around coverage X=100; we zoom in on that region to get a better view:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> plot line <span class="at">-H</span> ALL.sum.bed.csv <span class="at">-x</span> 1 <span class="at">-y</span> 2 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Genome coverage (bp)"</span> <span class="dt">\</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="at">--x-min</span> 40 <span class="at">--x-max</span> 140</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-disable MD013 -->
<div id="fig-plot-total-coverage-distribution-zoom-in" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><div id="fig-plot-total-coverage-distribution-hist-zoom-in" class="quarto-figure quarto-figure-center anchored">
<figure class="figure"><p><a href="fig-plot-total-coverage-distribution-hist-zoom-in.png" class="lightbox" title="Genome coverage" data-gallery="quarto-lightbox-gallery-14"><img src="fig-plot-total-coverage-distribution-hist-zoom-in.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(a) Genome coverage</figcaption></figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;13: Genome coverage vs depth of coverage.</figcaption></figure>
</div>
<!-- markdownlint-enable MD013 -->
<p><span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">Lou et al., 2021</a>)</span> point out that appropriate thresholds depend on the data set, but as a general rule recommend a minimum depth threshold at &lt;0.8X average coverage, and a maximum depth threshold at mean coverage plus one or two standard deviations. For the sake of simplicity, you could here infer a cutoff simply by manually inspecting <a href="#fig-plot-total-coverage-distribution-zoom-in">Figure&nbsp;13</a>; here, we will use the range 50-110.</p>
<p>We then use these thresholds to generate a BED file containing regions that are accessible, i.e., have sufficient coverage for downstream analyses. We also calculate the number of bases that pass the filtering criteria.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> ALL.sum.bed.gz <span class="at">-f</span> <span class="st">'4&gt;50'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> <span class="at">-f</span> <span class="st">'4&lt;110'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">bgzip</span> <span class="at">-c</span> <span class="op">&gt;</span> ALL.sum.depth.bed.gz</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> ALL.sum.depth.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai  <span class="kw">|</span> <span class="fu">grep</span> genome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<p>Consequently, 75.2% of the genome is accessible by depth.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Generate coverage sums for the red and yellow sample sets, and from these determine coverage thresholds and apply the thresholds to generate BED files with accessible regions.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<p>We base the answer on the previous code.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> <span class="va">$pop</span>.bg.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="at">-v</span> OFS=<span class="st">"\t"</span> <span class="st">'NR &gt; 1 {sum=0; for (i=4; i&lt;=NF; i++) sum+=$i; print $1, $2, $3, sum}'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.sum.bed.gz</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="va">$pop</span>.sum.bed.gz</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> mutate2 <span class="at">-t</span> <span class="at">-H</span> <span class="at">-w</span> 0 <span class="at">-e</span> <span class="st">'$3-$2'</span> <span class="va">$pop</span>.sum.bed.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> summary <span class="at">-t</span> <span class="at">-H</span> <span class="at">-g</span> 4 <span class="at">-f</span> 5:sum <span class="at">-w</span> 0 <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> sort <span class="at">-t</span> <span class="at">-k</span> 1:n <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a> <span class="fu">awk</span> <span class="at">-v</span> cumsum=0 <span class="st">'BEGIN {OFS=","; cumsum=0} {cumsum += $2; print $1,$2,cumsum}'</span> <span class="op">&gt;</span> <span class="va">$pop</span>.sum.bed.csv</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${pop}</span>.sum.bed.csv <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot line <span class="at">-x</span> 1 <span class="at">-y</span> 2 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Genome coverage (bp)"</span> <span class="dt">\</span></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="at">--x-min</span> 0 <span class="at">--x-max</span> 100 <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>    fig-plot-total-coverage-distribution-hist-zoom-in-<span class="va">$pop</span>.png</span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-plot-population-coverage-distribution" class="quarto-layout-panel" data-attr-output=".details summary=&quot;Output&quot;">
<figure class="figure"><div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-hist-zoom-in-red" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure"><p><a href="fig-plot-total-coverage-distribution-hist-zoom-in-red.png" class="lightbox" title="Zoomed in genome coverage, red" data-gallery="quarto-lightbox-gallery-15"><img src="fig-plot-total-coverage-distribution-hist-zoom-in-red.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(a) Zoomed in genome coverage, red</figcaption></figure>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-hist-zoom-in-yellow" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure"><p><a href="fig-plot-total-coverage-distribution-hist-zoom-in-yellow.png" class="lightbox" title="Zoomed in genome coverage, yellow" data-gallery="quarto-lightbox-gallery-16"><img src="fig-plot-total-coverage-distribution-hist-zoom-in-yellow.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(b) Zoomed in genome coverage, yellow</figcaption></figure>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;14: Zoomed in coverage distribution for red and yellow ecotypes.</figcaption><p></p>
</figure>
</div>
<p>Based on <a href="#fig-plot-population-coverage-distribution">Figure&nbsp;14</a>, we generate bed files with depths passing cutoffs:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># red filter: 20-60</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> red.sum.bed.gz <span class="at">-f</span> <span class="st">'4&gt;20'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> <span class="at">-f</span> <span class="st">'4&lt;60'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a> <span class="ex">bgzip</span> <span class="at">-c</span> <span class="op">&gt;</span> red.sum.depth.bed.gz</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> red.sum.depth.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai  <span class="kw">|</span> <span class="fu">grep</span> genome</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co"># yellow filter: 20-55</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> yellow.sum.bed.gz <span class="at">-f</span> <span class="st">'4&gt;20'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a> <span class="ex">csvtk</span> filter <span class="at">-t</span> <span class="at">-H</span> <span class="at">-f</span> <span class="st">'4&lt;55'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a> <span class="ex">bgzip</span> <span class="at">-c</span> <span class="op">&gt;</span> yellow.sum.depth.bed.gz</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> yellow.sum.depth.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai  <span class="kw">|</span> <span class="fu">grep</span> genome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genome  0   20853   100000  0.20853
genome  1   79147   100000  0.79147
genome  0   23231   100000  0.23231
genome  1   76769   100000  0.76769</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Now we have combined total per sample coverage for ALL samples, and for sample sets red and yellow. The upcoming task will be to generate sequence masks from the total coverage and minimum number of individuals with coverage greater than zero.</p>
</section><section id="filter-on-minimum-number-of-individuals" class="level3"><h3 class="anchored" data-anchor-id="filter-on-minimum-number-of-individuals">Filter on minimum number of individuals</h3>
<p>In addition to filtering on total coverage, we will also filter on the minimum number of individuals with a minimum depth. This is to account for cases where regions that pass the minimum coverage filter originate from just a few samples with unusually high coverage. Here, we will remove sites where more than 50% of individuals have zero coverage.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> ALL.bg.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="st">'BEGIN {OFS="\t"} NR &gt; 1 {count=0; for (i=4; i&lt;=NF; i++) {if ($i&gt;0) count+=1}; if (count&gt;=((NF-3)*0.5)) {print $1, $2, $3}}'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.ind.bed.gz</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed ALL.ind.bed.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Use <code>bedtools genomecov</code> to determine the proportion of bases that are filtered out. Use <code>${REF}.fai</code> as the argument to the required option <code>-g</code>.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> ALL.ind.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LG4 0   7694    100000  0.07694
LG4 1   92306   100000  0.92306
genome  0   7694    100000  0.07694
genome  1   92306   100000  0.92306</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Generate coverage sums for red and yellow sample sets.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> <span class="va">$pop</span>.bed.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="st">'BEGIN {OFS="\t"} NR &gt; 1 {count=0; for (i=4; i&lt;=NF; i++) {if ($i&gt;0) count+=1}; if (count&gt;=((NF-3)*0.5)) {print $1, $2, $3}}'</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.ind.bed.gz</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="va">$pop</span>.ind.bed.gz</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[bgzip] Could not open red.bed.gz: No such file or directory
[bgzip] Could not open yellow.bed.gz: No such file or directory</code></pre>
</div>
</div>
<!-- markdownlint-enable MD013 -->
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="sequence-masks-1" class="level3"><h3 class="anchored" data-anchor-id="sequence-masks-1">Sequence masks</h3>
<p>Finally, as before, we can convert the BED output to sequence mask files in FASTA format. Recall that we first have to make a template genome mask file where all positions are masked:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'BEGIN {OFS="\t"} {print $1, 0, $2}'</span> <span class="va">${REF}</span>.fai <span class="op">&gt;</span> <span class="va">${REF}</span>.bed</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span> <span class="at">-mc</span> 1 <span class="at">-fo</span> <span class="va">${REF}</span>.mask.fa <span class="at">-bed</span> <span class="va">${REF}</span>.bed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, for each sample set, we will use <code>bedtools intersect</code> to intersect the BED files corresponding to the total sum coverage and the filter on number of individuals. <code>bedtools intersect</code> makes it easy to combine multiple BED files, so any other filters, or genomic features such as exons, could be added to make a compound mask file. The resulting BED files is used as input to <code>bedtools maskfasta</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> ALL.sum.depth.bed.gz <span class="at">-b</span> ALL.ind.bed.gz <span class="dt">\</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.unmask.bed.gz</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed ALL.unmask.bed.gz</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span>.mask.fa <span class="at">-mc</span> 0 <span class="at">-fo</span> <span class="va">${REF}</span>.unmask.fa <span class="dt">\</span></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>   <span class="at">-bed</span> ALL.unmask.bed.gz</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-n</span> 3 <span class="va">${REF}</span>.unmask.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;LG4
111111111111111111111111111111111111111111111111111111111111
111111111111111111100000000000000000000000000000000111111111</code></pre>
</div>
</div>
<p>We can once again convince ourselves that this has worked by counting the number of unmasked positions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tr: -d deletes all characters not (-c, complement) in the character</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="co"># set '0'. wc: -m option counts characters</span></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${REF}</span>.unmask.fa <span class="kw">|</span> <span class="fu">tr</span> <span class="at">-d</span> <span class="at">-c</span> <span class="st">'0'</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-m</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> ALL.unmask.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="kw">|</span> <span class="fu">grep</span> genome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>75236
genome  0   24764   100000  0.24764
genome  1   75236   100000  0.75236</code></pre>
</div>
</div>
<p>Note that 0 and 1 in the <code>bedtools genomecov</code> output refers to coverage (i.e., absence/presence) and not unmask/mask as in the mask FASTA file.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Create unmask files for red and yellow populations.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-24-contents" aria-controls="callout-24" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-24" class="callout-24-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="va">$pop</span>.sum.depth.bed.gz <span class="at">-b</span> <span class="va">$pop</span>.ind.bed.gz <span class="dt">\</span></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.unmask.bed.gz</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a> <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="va">$pop</span>.unmask.bed.gz</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span>.mask.fa <span class="at">-mc</span> 0 <span class="at">-fo</span> <span class="va">${REF}</span>.<span class="va">$pop</span>.unmask.fa <span class="dt">\</span></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-bed</span> <span class="va">$pop</span>.unmask.bed.gz</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="conclusion" class="level2"><h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>Congratulations! You have now gone through a set of tedious and complex steps to generate output files that determine what regions in a reference DNA sequence are amenable to analysis. In the next exercise we will use these files as inputs to different programs that calculate diversity statistics from population genomic data.</p>
</section><section id="references" class="level2">



</section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-danecek_VariantCallFormat_2011" class="csl-entry" role="listitem">
Danecek, P., Auton, A., Abecasis, G., Albers, C. A., Banks, E., DePristo, M. A., Handsaker, R. E., Lunter, G., Marth, G. T., Sherry, S. T., McVean, G., &amp; Durbin, R. (2011). The variant call format and <span>VCFtools</span>. <em>Bioinformatics (Oxford, England)</em>, <em>27</em>(15), 2156–2158. <a href="https://doi.org/10.1093/bioinformatics/btr330">https://doi.org/10.1093/bioinformatics/btr330</a>
</div>
<div id="ref-danecek_TwelveYearsSAMtools_2021" class="csl-entry" role="listitem">
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., Davies, R. M., &amp; Li, H. (2021). Twelve years of <span>SAMtools</span> and <span>BCFtools</span>. <em>GigaScience</em>, <em>10</em>(2), giab008. <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>
</div>
<div id="ref-garrison_SpectrumFreeSoftware_2022" class="csl-entry" role="listitem">
Garrison, E., Kronenberg, Z. N., Dawson, E. T., Pedersen, B. S., &amp; Prins, P. (2022). A spectrum of free software tools for processing the <span>VCF</span> variant call format: Vcflib, bio-vcf, Cyvcf2, hts-nim and slivar. <em>PLOS Computational Biology</em>, <em>18</em>(5), e1009123. <a href="https://doi.org/10.1371/journal.pcbi.1009123">https://doi.org/10.1371/journal.pcbi.1009123</a>
</div>
<div id="ref-hou_BalancingEfficientAnalysis_2021" class="csl-entry" role="listitem">
Hou, H., Pedersen, B., &amp; Quinlan, A. (2021). Balancing efficient analysis and storage of quantitative genomics data with the <span>D4</span> format and D4tools. <em>Nature Computational Science</em>, <em>1</em>(6), 441–447. <a href="https://doi.org/10.1038/s43588-021-00085-0">https://doi.org/10.1038/s43588-021-00085-0</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-maruki_GenotypeCallingPopulationGenomic_2017" class="csl-entry" role="listitem">
Maruki, T., &amp; Lynch, M. (2017). Genotype <span>Calling</span> from <span>Population-Genomic Sequencing Data</span>. <em>G3 Genes|Genomes|Genetics</em>, <em>7</em>(5), 1393–1404. <a href="https://doi.org/10.1534/g3.117.039008">https://doi.org/10.1534/g3.117.039008</a>
</div>
<div id="ref-nielsen_GenotypeSNPCalling_2011" class="csl-entry" role="listitem">
Nielsen, R., Paul, J. S., Albrechtsen, A., &amp; Song, Y. S. (2011). Genotype and <span>SNP</span> calling from next-generation sequencing data. <em>Nature Reviews Genetics</em>, <em>12</em>(6), 443–451. <a href="https://doi.org/10.1038/nrg2986">https://doi.org/10.1038/nrg2986</a>
</div>
<div id="ref-pedersen_MosdepthQuickCoverage_2018" class="csl-entry" role="listitem">
Pedersen, B. S., &amp; Quinlan, A. R. (2018). Mosdepth: Quick coverage calculation for genomes and exomes. <em>Bioinformatics</em>, <em>34</em>(5), 867–868. <a href="https://doi.org/10.1093/bioinformatics/btx699">https://doi.org/10.1093/bioinformatics/btx699</a>
</div>
<div id="ref-purcell_PLINKToolSet_2007" class="csl-entry" role="listitem">
Purcell, S., Neale, B., Todd-Brown, K., Thomas, L., Ferreira, M. A. R., Bender, D., Maller, J., Sklar, P., de Bakker, P. I. W., Daly, M. J., &amp; Sham, P. C. (2007). <span>PLINK</span>: <span>A Tool Set</span> for <span>Whole-Genome Association</span> and <span>Population-Based Linkage Analyses</span>. <em>The American Journal of Human Genetics</em>, <em>81</em>(3), 559–575. <a href="https://doi.org/10.1086/519795">https://doi.org/10.1086/519795</a>
</div>
<div id="ref-quinlan_BEDToolsFlexibleSuite_2010" class="csl-entry" role="listitem">
Quinlan, A. R., &amp; Hall, I. M. (2010). <span>BEDTools</span>: A flexible suite of utilities for comparing genomic features. <em>Bioinformatics</em>, <em>26</em>(6), 841–842. <a href="https://doi.org/10.1093/bioinformatics/btq033">https://doi.org/10.1093/bioinformatics/btq033</a>
</div>
<div id="ref-shen_SeqKitCrossPlatformUltrafast_2016" class="csl-entry" role="listitem">
Shen, W., Le, S., Li, Y., &amp; Hu, F. (2016). <span>SeqKit</span>: <span>A Cross-Platform</span> and <span>Ultrafast Toolkit</span> for <span>FASTA</span>/<span>Q File Manipulation</span>. <em>PLOS ONE</em>, <em>11</em>(10), e0163962. <a href="https://doi.org/10.1371/journal.pone.0163962">https://doi.org/10.1371/journal.pone.0163962</a>
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>
<ol>
<li id="fn1"><p>The password is provided by the course instructor<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>This exercise is inspired by and based on <a href="https://speciationgenomics.github.io/filtering_vcfs/" class="uri">https://speciationgenomics.github.io/filtering_vcfs/</a>]<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>We select a fix number that should contain enough information to generate reliable statistics. This number should not change significantly even when files contain vastly different numbers of sites, which is why we need adjust the parameter <em>r</em> to the number of sites in the file.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p><code>vcftools</code> does <strong>not</strong> have a <code>-h</code> or <code>--help</code> option.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>The <code>2&gt;/dev/null</code> outputs messages from <code>vcftools</code> to a special file <code>/dev/null</code> which is a kind of electronic dustbin.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>On viewing <code>csvtk</code> plots: either you can <strong>redirect</strong> (<code>&gt;</code>) the results from <code>csvtk</code> to a png output file, or you can <strong>pipe</strong> (<code>|</code>) it to the command <code>display</code> (replace <code>&gt; $OUT.ldepth.png</code> by <code>| display</code>, which should fire up a window with your plot.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p>You can try different values of the <code>--bins</code> option<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p><span class="citation" data-cites="purcell_PLINKToolSet_2007">Purcell et al. (<a href="#ref-purcell_PLINKToolSet_2007" role="doc-biblioref">2007</a>)</span>, p.&nbsp;565 gives a coherent derivation of this estimator<a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn9"><p>We need to generate a BED file representation of the FASTA index unfortuanely as <code>bedtools makefasta</code> doesn’t handle FASTA indices natively.<a href="#fnref9" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn10"><p>For <code>vcftools</code>; unfortunate, but that’s the way it is<a href="#fnref10" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn11"><p>The one-liner combines the results of several commands in a pipe stream. Also, <a href="https://www.gnu.org/software/bash/manual/html_node/Redirections.html">Bash redirections</a> are used to gather the results from the output of <code>bedtools makewindows</code> to <code>bedtools intersect</code>. The intersection commands collects coverage data in 1kb windows that are then summarized by <code>bedtools groupby</code>.<a href="#fnref11" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn12"><p>The <code>%</code> operator deletes the shortest match of <code>$substring</code> from back of <code>$string</code>: <code>${string%substring}</code>. See <a href="https://tldp.org/LDP/abs/html/string-manipulation.html">Bash string manipulation</a> for more information.<a href="#fnref12" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn13"><p>Instead of using the command substitution, you could look into the sample info file and set the sample names manually: <code>SAMPLES="PUN-Y-BCRD PUN-R-ELF PUN-Y-INJ PUN-R-JMC PUN-R-LH PUN-Y-LO PUN-R-MT PUN-Y-PCT PUN-Y-POTR PUN-R-UCSD"</code><a href="#fnref13" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></div></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">2023 <a href="about.html">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v<!--?quarto.version ?--></div>
  </div>
</footer><script>var lightboxQuarto = GLightbox({"loop":true,"openEffect":"zoom","selector":".lightbox","descPosition":"bottom","closeEffect":"zoom"});</script>


</body></html>