<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DDLS Population genomics in practice – references</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./assets/logos/nbislogo-green.svg" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">DDLS Population genomics in practice</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./slides/index.html" rel="" target="">
 <span class="menu-text">Slides</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./exercises/index.html" rel="" target="">
 <span class="menu-text">Exercises</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./recipes/index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Code recipes</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/NBISwe" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-pgip" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Slides</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides/pgip/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genomics in practice</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides/foundations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genetics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./slides/population_structure/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population structure</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises/compute_environment/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compute environment</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises/datasets/monkeyflowers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Monkeyflowers dataset</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises/variant_calling/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exercises/variant_filtering/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant filtering</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Code recipes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./recipes/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code recipes</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#references" id="toc-references" class="nav-link active" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<section id="references" class="level1">
<h1>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-GenomeAtlasEuropean_2022" class="csl-entry" role="listitem">
<em>A genome atlas of european biodiversity</em>. (n.d.). https://www.erga-biodiversity.eu. Retrieved November 3, 2022, from <a href="https://www.erga-biodiversity.eu">https://www.erga-biodiversity.eu</a>
</div>
<div id="ref-adrion_PredictingLandscapeRecombination_2020" class="csl-entry" role="listitem">
Adrion, J. R., Galloway, J. G., &amp; Kern, A. D. (2020). Predicting the <span>Landscape</span> of <span>Recombination Using Deep Learning</span>. <em>Molecular Biology and Evolution</em>, <em>37</em>(6), 1790–1808. <a href="https://doi.org/10.1093/molbev/msaa038">https://doi.org/10.1093/molbev/msaa038</a>
</div>
<div id="ref-allendorf_PopulationGenomics_2022" class="csl-entry" role="listitem">
Allendorf, F. W., Funk, W. C., Aitken, S. N., Byrne, M., &amp; Luikart, G. (2022). Population <span>Genomics</span>. In F. W. Allendorf, W. C. Funk, S. N. Aitken, M. Byrne, G. Luikart, &amp; A. Antunes (Eds.), <em>Conservation and the <span>Genomics</span> of <span>Populations</span></em> (p. 0). <span>Oxford University Press</span>. <a href="https://doi.org/10.1093/oso/9780198856566.003.0004">https://doi.org/10.1093/oso/9780198856566.003.0004</a>
</div>
<div id="ref-allendorf_GenomicsFutureConservation_2010" class="csl-entry" role="listitem">
Allendorf, F. W., Hohenlohe, P. A., &amp; Luikart, G. (2010). Genomics and the future of conservation genetics. <em>Nature Reviews Genetics</em>, <em>11</em>(10), 697–709. <a href="https://doi.org/10.1038/nrg2844">https://doi.org/10.1038/nrg2844</a>
</div>
<div id="ref-baird_RapidSNPDiscovery_2008" class="csl-entry" role="listitem">
Baird, N. A., Etter, P. D., Atwood, T. S., Currey, M. C., Shiver, A. L., Lewis, Z. A., Selker, E. U., Cresko, W. A., &amp; Johnson, E. A. (2008). Rapid <span>SNP Discovery</span> and <span>Genetic Mapping Using Sequenced RAD Markers</span>. <em>PLOS ONE</em>, <em>3</em>(10), e3376. <a href="https://doi.org/10.1371/journal.pone.0003376">https://doi.org/10.1371/journal.pone.0003376</a>
</div>
<div id="ref-barrera-redondo_GenomicTranscriptomicEpigenomic_2020" class="csl-entry" role="listitem">
Barrera-Redondo, J., Piñero, D., &amp; Eguiarte, L. E. (2020). Genomic, <span>Transcriptomic</span> and <span>Epigenomic Tools</span> to <span>Study</span> the <span>Domestication</span> of <span>Plants</span> and <span>Animals</span>: <span>A Field Guide</span> for <span>Beginners</span>. <em>Frontiers in Genetics</em>, <em>11</em>.
</div>
<div id="ref-barton2007evolution" class="csl-entry" role="listitem">
Barton, N. H., Briggs, D. E. G., Eisen, J. A., Goldstein, D. B., &amp; Patel, N. H. (2007). <em>Evolution</em>. <span>Cold Spring Harbor Laboratory Press</span>.
</div>
<div id="ref-baumdicker_EfficientAncestryMutation_2022" class="csl-entry" role="listitem">
Baumdicker, F., Bisschop, G., Goldstein, D., Gower, G., Ragsdale, A. P., Tsambos, G., Zhu, S., Eldon, B., Ellerman, E. C., Galloway, J. G., Gladstein, A. L., Gorjanc, G., Guo, B., Jeffery, B., Kretzschumar, W. W., Lohse, K., Matschiner, M., Nelson, D., Pope, N. S., … Kelleher, J. (2022). Efficient ancestry and mutation simulation with msprime 1.0. <em>Genetics</em>, <em>220</em>(3), iyab229. <a href="https://doi.org/10.1093/genetics/iyab229">https://doi.org/10.1093/genetics/iyab229</a>
</div>
<div id="ref-begun_PopulationGenomicsWholeGenome_2007" class="csl-entry" role="listitem">
Begun, D. J., Holloway, A. K., Stevens, K., Hillier, L. W., Poh, Y.-P., Hahn, M. W., Nista, P. M., Jones, C. D., Kern, A. D., Dewey, C. N., Pachter, L., Myers, E., &amp; Langley, C. H. (2007). Population <span>Genomics</span>: <span>Whole-Genome Analysis</span> of <span>Polymorphism</span> and <span>Divergence</span> in <span>Drosophila</span> simulans. <em>PLOS Biology</em>, <em>5</em>(11), e310. <a href="https://doi.org/10.1371/journal.pbio.0050310">https://doi.org/10.1371/journal.pbio.0050310</a>
</div>
<div id="ref-bergeron_EvolutionGermlineMutation_2023" class="csl-entry" role="listitem">
Bergeron, L. A., Besenbacher, S., Zheng, J., Li, P., Bertelsen, M. F., Quintard, B., Hoffman, J. I., Li, Z., St. Leger, J., Shao, C., Stiller, J., Gilbert, M. T. P., Schierup, M. H., &amp; Zhang, G. (2023). Evolution of the germline mutation rate across vertebrates. <em>Nature</em>, <em>615</em>(7951), 285–291. <a href="https://doi.org/10.1038/s41586-023-05752-y">https://doi.org/10.1038/s41586-023-05752-y</a>
</div>
<div id="ref-berner_AlleleFrequencyDifference_2019" class="csl-entry" role="listitem">
Berner, D. (2019). Allele <span>Frequency Difference AFD</span> to <span>FST</span> for <span>Quantifying Genetic Population Differentiation</span>. <em>Genes</em>, <em>10</em>(4), 308. <a href="https://doi.org/10.3390/genes10040308">https://doi.org/10.3390/genes10040308</a>
</div>
<div id="ref-bhatia_EstimatingInterpretingFST_2013" class="csl-entry" role="listitem">
Bhatia, G., Patterson, N., Sankararaman, S., &amp; Price, A. L. (2013). Estimating and interpreting <span>FST</span>: <span>The</span> impact of rare variants. <em>Genome Research</em>, <em>23</em>(9), 1514–1521. <a href="https://doi.org/10.1101/gr.154831.113">https://doi.org/10.1101/gr.154831.113</a>
</div>
<div id="ref-buri_GeneFrequencySmall_1956" class="csl-entry" role="listitem">
Buri, P. (1956). Gene <span>Frequency</span> in <span>Small Populations</span> of <span>Mutant Drosophila</span>. <em>Evolution</em>, <em>10</em>(4), 367–402. <a href="https://doi.org/10.1111/j.1558-5646.1956.tb02864.x">https://doi.org/10.1111/j.1558-5646.1956.tb02864.x</a>
</div>
<div id="ref-burri_InterpretingDifferentiationLandscapes_2017" class="csl-entry" role="listitem">
Burri, R. (2017). Interpreting differentiation landscapes in the light of long-term linked selection. <em>Evolution Letters</em>, <em>1</em>(3), 118–131. <a href="https://doi.org/10.1002/evl3.14">https://doi.org/10.1002/evl3.14</a>
</div>
<div id="ref-casillas_MolecularPopulationGenetics_2017" class="csl-entry" role="listitem">
Casillas, S., &amp; Barbadilla, A. (2017). Molecular <span>Population Genetics</span>. <em>Genetics</em>, <em>205</em>(3), 1003–1035. <a href="https://doi.org/10.1534/genetics.116.196493">https://doi.org/10.1534/genetics.116.196493</a>
</div>
<div id="ref-chan_GenomeWideFineScaleRecombination_2012" class="csl-entry" role="listitem">
Chan, A. H., Jenkins, P. A., &amp; Song, Y. S. (2012). Genome-<span>Wide Fine-Scale Recombination Rate Variation</span> in <span>Drosophila</span> melanogaster. <em>PLOS Genetics</em>, <em>8</em>(12), e1003090. <a href="https://doi.org/10.1371/journal.pgen.1003090">https://doi.org/10.1371/journal.pgen.1003090</a>
</div>
<div id="ref-chang_DataManagementSummary_2020" class="csl-entry" role="listitem">
Chang, C. C. (2020). Data <span>Management</span> and <span>Summary Statistics</span> with <span>PLINK</span>. In J. Y. Dutheil (Ed.), <em>Statistical <span>Population Genomics</span></em> (pp. 49–65). <span>Springer US</span>. <a href="https://doi.org/10.1007/978-1-0716-0199-0_3">https://doi.org/10.1007/978-1-0716-0199-0_3</a>
</div>
<div id="ref-charlesworth_ElementsEvolutionaryGenetics_2010" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Charlesworth, D. (2010). <em>Elements of <span>Evolutionary Genetics</span></em>. <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-charlesworth_EffectsOfSelection_2021" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Jensen, J. D. (2021). Effects of selection at linked sites on patterns of genetic variability. <em>Annual Review of Ecology, Evolution, and Systematics</em>, <em>52</em>(1), 177–197. <a href="https://doi.org/10.1146/annurev-ecolsys-010621-044528">https://doi.org/10.1146/annurev-ecolsys-010621-044528</a>
</div>
<div id="ref-charlesworth_EffectDeleteriousMutations_1993" class="csl-entry" role="listitem">
Charlesworth, B., Morgan, M. T., &amp; Charlesworth, D. (1993). <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1205596">The <span>Effect</span> of <span>Deleterious Mutations</span> on <span>Neutral Molecular Variation</span></a>. <em>Genetics</em>, <em>134</em>(4), 1289–1303.
</div>
<div id="ref-cooplab_PopulationGeneticsCourse_2011" class="csl-entry" role="listitem">
cooplab. (2011). Population genetics course resources: <span>Hardy-Weinberg Eq</span>. In <em>gcbias</em>. <a href="https://gcbias.org/2011/10/13/population-genetics-course-resources-hardy-weinberg-eq/">https://gcbias.org/2011/10/13/population-genetics-course-resources-hardy-weinberg-eq/</a>
</div>
<div id="ref-corbett-detig_NaturalSelectionConstrains_2015" class="csl-entry" role="listitem">
Corbett-Detig, R. B., Hartl, D. L., &amp; Sackton, T. B. (2015). Natural <span>Selection Constrains Neutral Diversity</span> across <span>A Wide Range</span> of <span>Species</span>. <em>PLOS Biology</em>, <em>13</em>(4), e1002112. <a href="https://doi.org/10.1371/journal.pbio.1002112">https://doi.org/10.1371/journal.pbio.1002112</a>
</div>
<div id="ref-Lawson_2012" class="csl-entry" role="listitem">
D. Lawson, S. M., G. Hellenthal. (2012). Inference of population structure using dense haplotype data. <em>PLoS Genet. 8(1):e1002453</em>. <a href="https://doi.org/doi: 10.1371/journal.pgen.1002453">https://doi.org/doi: 10.1371/journal.pgen.1002453</a>
</div>
<div id="ref-danecek_VariantCallFormat_2011" class="csl-entry" role="listitem">
Danecek, P., Auton, A., Abecasis, G., Albers, C. A., Banks, E., DePristo, M. A., Handsaker, R. E., Lunter, G., Marth, G. T., Sherry, S. T., McVean, G., &amp; Durbin, R. (2011). The variant call format and <span>VCFtools</span>. <em>Bioinformatics (Oxford, England)</em>, <em>27</em>(15), 2156–2158. <a href="https://doi.org/10.1093/bioinformatics/btr330">https://doi.org/10.1093/bioinformatics/btr330</a>
</div>
<div id="ref-danecek_TwelveYearsSAMtools_2021" class="csl-entry" role="listitem">
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., Davies, R. M., &amp; Li, H. (2021). Twelve years of <span>SAMtools</span> and <span>BCFtools</span>. <em>GigaScience</em>, <em>10</em>(2), giab008. <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>
</div>
<div id="ref-davey_RADSeqNextgenerationPopulation_2010" class="csl-entry" role="listitem">
Davey, J. W., &amp; Blaxter, M. L. (2010). <span>RADSeq</span>: Next-generation population genetics. <em>Briefings in Functional Genomics</em>, <em>9</em>(5-6), 416–423. <a href="https://doi.org/10.1093/bfgp/elq031">https://doi.org/10.1093/bfgp/elq031</a>
</div>
<div id="ref-davydov_IdentifyingHighFraction_2010" class="csl-entry" role="listitem">
Davydov, E. V., Goode, D. L., Sirota, M., Cooper, G. M., Sidow, A., &amp; Batzoglou, S. (2010). Identifying a <span>High Fraction</span> of the <span>Human Genome</span> to be under <span>Selective Constraint Using GERP</span>++. <em>PLOS Computational Biology</em>, <em>6</em>(12), e1001025. <a href="https://doi.org/10.1371/journal.pcbi.1001025">https://doi.org/10.1371/journal.pcbi.1001025</a>
</div>
<div id="ref-degiorgio_SweepFinder2IncreasedSensitivity_2016" class="csl-entry" role="listitem">
DeGiorgio, M., Huber, C. D., Hubisz, M. J., Hellmann, I., &amp; Nielsen, R. (2016). <span>SweepFinder2</span>: Increased sensitivity, robustness and flexibility. <em>Bioinformatics</em>, <em>32</em>(12), 1895–1897. <a href="https://doi.org/10.1093/bioinformatics/btw051">https://doi.org/10.1093/bioinformatics/btw051</a>
</div>
<div id="ref-depristo_FrameworkVariationDiscovery_2011" class="csl-entry" role="listitem">
DePristo, M. A., Banks, E., Poplin, R., Garimella, K. V., Maguire, J. R., Hartl, C., Philippakis, A. A., del Angel, G., Rivas, M. A., Hanna, M., McKenna, A., Fennell, T. J., Kernytsky, A. M., Sivachenko, A. Y., Cibulskis, K., Gabriel, S. B., Altshuler, D., &amp; Daly, M. J. (2011). A framework for variation discovery and genotyping using next-generation <span>DNA</span> sequencing data. <em>Nature Genetics</em>, <em>43</em>(5), 491–498. <a href="https://doi.org/10.1038/ng.806">https://doi.org/10.1038/ng.806</a>
</div>
<div id="ref-desroches_ConservingIntraspecificVariation_2021" class="csl-entry" role="listitem">
Des Roches, S., Pendleton, L. H., Shapiro, B., &amp; Palkovacs, E. P. (2021). Conserving intraspecific variation for nature’s contributions to people. <em>Nature Ecology &amp; Evolution</em>, <em>5</em>(5), 574–582. <a href="https://doi.org/10.1038/s41559-021-01403-5">https://doi.org/10.1038/s41559-021-01403-5</a>
</div>
<div id="ref-donato_NewEvaluationMethods_2021" class="csl-entry" role="listitem">
Donato, L., Scimone, C., Rinaldi, C., D’Angelo, R., &amp; Sidoti, A. (2021). New evaluation methods of read mapping by 17 aligners on simulated and empirical <span>NGS</span> data: An updated comparison of <span>DNA-</span> and <span>RNA-Seq</span> data from <span>Illumina</span> and <span>Ion Torrent</span> technologies. <em>Neural Computing and Applications</em>, <em>33</em>(22), 15669–15692. <a href="https://doi.org/10.1007/s00521-021-06188-z">https://doi.org/10.1007/s00521-021-06188-z</a>
</div>
<div id="ref-dutheiljulien_StatisticalPopulationGenomics_2020" class="csl-entry" role="listitem">
Dutheil, Julien (Ed.). (2020). <em>Statistical <span>Population Genomics</span></em>.
</div>
<div id="ref-ellegren_DeterminantsGeneticDiversity_2016" class="csl-entry" role="listitem">
Ellegren, H., &amp; Galtier, N. (2016). Determinants of genetic diversity. <em>Nature Reviews Genetics</em>, <em>17</em>(7), 422–433. <a href="https://doi.org/10.1038/nrg.2016.58">https://doi.org/10.1038/nrg.2016.58</a>
</div>
<div id="ref-ewels_MultiQCSummarizeAnalysis_2016" class="csl-entry" role="listitem">
Ewels, P., Magnusson, M., Lundin, S., &amp; Käller, M. (2016). <span>MultiQC</span>: Summarize analysis results for multiple tools and samples in a single report. <em>Bioinformatics</em>, <em>32</em>(19), 3047–3048. <a href="https://doi.org/10.1093/bioinformatics/btw354">https://doi.org/10.1093/bioinformatics/btw354</a>
</div>
<div id="ref-ewens_MathematicalPopulationGenetics_2004" class="csl-entry" role="listitem">
Ewens, W. J. (2004). <em>Mathematical <span>Population Genetics</span></em> (S. S. Antman, J. E. Marsden, L. Sirovich, &amp; S. Wiggins, Eds.; Vol. 27). <span>Springer</span>. <a href="https://doi.org/10.1007/978-0-387-21822-9">https://doi.org/10.1007/978-0-387-21822-9</a>
</div>
<div id="ref-excoffier_AnalysisPopulationSubdivision_2004" class="csl-entry" role="listitem">
Excoffier, L. (2004). Analysis of <span>Population Subdivision</span>. In <em>Handbook of <span>Statistical Genetics</span></em>. <span>John Wiley &amp; Sons, Ltd</span>. <a href="https://doi.org/10.1002/0470022620.bbc25">https://doi.org/10.1002/0470022620.bbc25</a>
</div>
<div id="ref-ferretti_DecomposingSiteFrequency_2017" class="csl-entry" role="listitem">
Ferretti, L., Ledda, A., Wiehe, T., Achaz, G., &amp; Ramos-Onsins, S. E. (2017). Decomposing the <span>Site Frequency Spectrum</span>: <span>The Impact</span> of <span>Tree Topology</span> on <span>Neutrality Tests</span>. <em>Genetics</em>, <em>207</em>(1), 229–240. <a href="https://doi.org/10.1534/genetics.116.188763">https://doi.org/10.1534/genetics.116.188763</a>
</div>
<div id="ref-fitzpatrick_EcologicalGenomicsMeets_2015" class="csl-entry" role="listitem">
Fitzpatrick, M. C., &amp; Keller, S. R. (2015). Ecological genomics meets community-level modelling of biodiversity: Mapping the genomic landscape of current and future environmental adaptation. <em>Ecology Letters</em>, <em>18</em>(1), 1–16. <a href="https://doi.org/10.1111/ele.12376">https://doi.org/10.1111/ele.12376</a>
</div>
<div id="ref-foll_GenomeScanMethodIdentify_2008" class="csl-entry" role="listitem">
Foll, M., &amp; Gaggiotti, O. (2008). A <span>Genome-Scan Method</span> to <span>Identify Selected Loci Appropriate</span> for <span>Both Dominant</span> and <span>Codominant Markers</span>: <span>A Bayesian Perspective</span>. <em>Genetics</em>, <em>180</em>(2), 977–993. <a href="https://doi.org/10.1534/genetics.108.092221">https://doi.org/10.1534/genetics.108.092221</a>
</div>
<div id="ref-formenti_EraReferenceGenomes_2022a" class="csl-entry" role="listitem">
Formenti, G., Theissinger, K., Fernandes, C., Bista, I., Bombarely, A., Bleidorn, C., Ciofi, C., Crottini, A., Godoy, J. A., Höglund, J., Malukiewicz, J., Mouton, A., Oomen, R. A., Paez, S., Palsbøll, P. J., Pampoulie, C., Ruiz-López, M. J., Svardal, H., Theofanopoulou, C., … Zammit, G. (2022). The era of reference genomes in conservation genomics. <em>Trends in Ecology &amp; Evolution</em>, <em>37</em>(3), 197–202. <a href="https://doi.org/10.1016/j.tree.2021.11.008">https://doi.org/10.1016/j.tree.2021.11.008</a>
</div>
<div id="ref-fuller_PopulationGeneticsCoral_2020" class="csl-entry" role="listitem">
Fuller, Z. L., Mocellin, V. J. L., Morris, L. A., Cantin, N., Shepherd, J., Sarre, L., Peng, J., Liao, Y., Pickrell, J., Andolfatto, P., Matz, M., Bay, L. K., &amp; Przeworski, M. (2020). Population genetics of the coral <span>Acropora</span> millepora: <span>Toward</span> genomic prediction of bleaching. <em>Science</em>, <em>369</em>(6501), eaba4674. <a href="https://doi.org/10.1126/science.aba4674">https://doi.org/10.1126/science.aba4674</a>
</div>
<div id="ref-garrison_SpectrumFreeSoftware_2022" class="csl-entry" role="listitem">
Garrison, E., Kronenberg, Z. N., Dawson, E. T., Pedersen, B. S., &amp; Prins, P. (2022). A spectrum of free software tools for processing the <span>VCF</span> variant call format: Vcflib, bio-vcf, Cyvcf2, hts-nim and slivar. <em>PLOS Computational Biology</em>, <em>18</em>(5), e1009123. <a href="https://doi.org/10.1371/journal.pcbi.1009123">https://doi.org/10.1371/journal.pcbi.1009123</a>
</div>
<div id="ref-garrison_HaplotypebasedVariantDetection_2012" class="csl-entry" role="listitem">
Garrison, E., &amp; Marth, G. (2012). Haplotype-based variant detection from short-read sequencing. <em>arXiv:1207.3907 [q-Bio]</em>. <a href="http://arxiv.org/abs/1207.3907">http://arxiv.org/abs/1207.3907</a>
</div>
<div id="ref-gillespie_PopulationGeneticsConcise_2004" class="csl-entry" role="listitem">
Gillespie, J. H. (2004). <em>Population <span>Genetics</span>: <span>A Concise Guide</span></em> (2nd edition). <span>Johns Hopkins University Press</span>.
</div>
<div id="ref-grahamcoop_NotesPopulationGenetics_2020" class="csl-entry" role="listitem">
Graham Coop. (2020). <em>Notes on <span>Population Genetics</span></em>. <a href="https://github.com/cooplab/popgen-notes">https://github.com/cooplab/popgen-notes</a>
</div>
<div id="ref-gutenkunst_InferringJointDemographic_2009" class="csl-entry" role="listitem">
Gutenkunst, R. N., Hernandez, R. D., Williamson, S. H., &amp; Bustamante, C. D. (2009). Inferring the joint demographic history of multiple populations from multidimensional <span>SNP</span> frequency data. <em>PLoS Genetics</em>, <em>5</em>(10), e1000695. <a href="https://doi.org/10.1371/journal.pgen.1000695">https://doi.org/10.1371/journal.pgen.1000695</a>
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-haller_SLiMForwardGenetic_2019" class="csl-entry" role="listitem">
Haller, B. C., &amp; Messer, P. W. (2019). <span>SLiM</span> 3: <span>Forward Genetic Simulations Beyond</span> the <span>Wright</span>. <em>Molecular Biology and Evolution</em>, <em>36</em>(3), 632–637. <a href="https://doi.org/10.1093/molbev/msy228">https://doi.org/10.1093/molbev/msy228</a>
</div>
<div id="ref-haller_SLiMEvolutionarySimulation_2022" class="csl-entry" role="listitem">
Haller, B. C., &amp; Messer, P. W. (2022). <em><span>SLiM</span>: <span>An Evolutionary Simulation Framework</span></em>.
</div>
<div id="ref-hallerben_MesserLabSLiM_2016" class="csl-entry" role="listitem">
Haller, Ben. (2016). Messer <span>Lab</span> <span>SLiM</span>. In <em>Messer Lab</em>. <a href="https://messerlab.org/slim/">https://messerlab.org/slim/</a>
</div>
<div id="ref-hansen_VariantCallingNext_2016" class="csl-entry" role="listitem">
Hansen, N. F. (2016). Variant <span>Calling From Next Generation Sequence Data</span>. In E. Mathé &amp; S. Davis (Eds.), <em>Statistical <span>Genomics</span>: <span>Methods</span> and <span>Protocols</span></em> (pp. 209–224). <span>Springer</span>. <a href="https://doi.org/10.1007/978-1-4939-3578-9_11">https://doi.org/10.1007/978-1-4939-3578-9_11</a>
</div>
<div id="ref-hartl1997principles" class="csl-entry" role="listitem">
Hartl, D. L., &amp; Clark, A. G. (1997). <em>Principles of population genetics</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-hein2005gene" class="csl-entry" role="listitem">
Hein, J., Schierup, M. H., &amp; Wiuf, C. (2005). <em>Gene genealogies, variation and evolution: <span>A</span> primer in coalescent theory</em>. <span>Oxford University Press</span>. <a href="https://books.google.se/books?id=CCmLNAEACAAJ">https://books.google.se/books?id=CCmLNAEACAAJ</a>
</div>
<div id="ref-hein_GeneGenealogies_2004" class="csl-entry" role="listitem">
Hein, J., Schierup, M., &amp; Wiuf, C. (2004). Gene genealogies, variation and evolution. A primer in coalescent theory. In <em>Systematic Biology - SYST BIOL</em> (Vol. 54).
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2017" class="csl-entry" role="listitem">
Hermisson, J. (2017). <em>Mathematical population genetics</em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017</a>
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2018" class="csl-entry" role="listitem">
Hermisson, J. (2018). <em>Mathematical population genetics <span>II</span></em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2018">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2018</a>
</div>
<div id="ref-hohenlohe_PopulationGenomicsProvides_2019" class="csl-entry" role="listitem">
Hohenlohe, P. A., Hand, B. K., Andrews, K. R., &amp; Luikart, G. (2019). Population <span>Genomics Provides Key Insights</span> in <span>Ecology</span> and <span>Evolution</span>. In O. P. Rajora (Ed.), <em>Population <span>Genomics</span>: <span>Concepts</span>, <span>Approaches</span> and <span>Applications</span></em> (pp. 483–510). <span>Springer International Publishing</span>. <a href="https://doi.org/10.1007/13836_2018_20">https://doi.org/10.1007/13836_2018_20</a>
</div>
<div id="ref-hou_BalancingEfficientAnalysis_2021" class="csl-entry" role="listitem">
Hou, H., Pedersen, B., &amp; Quinlan, A. (2021). Balancing efficient analysis and storage of quantitative genomics data with the <span>D4</span> format and D4tools. <em>Nature Computational Science</em>, <em>1</em>(6), 441–447. <a href="https://doi.org/10.1038/s43588-021-00085-0">https://doi.org/10.1038/s43588-021-00085-0</a>
</div>
<div id="ref-HTSFormatSpecifications_2023" class="csl-entry" role="listitem">
<em><span>HTS</span> format specifications</em>. (2023). <a href="https://samtools.github.io/hts-specs/">https://samtools.github.io/hts-specs/</a>
</div>
<div id="ref-hubisz_InferenceAncestralRecombination_2020" class="csl-entry" role="listitem">
Hubisz, M., &amp; Siepel, A. (2020). Inference of <span>Ancestral Recombination Graphs Using ARGweaver</span>. In J. Y. Dutheil (Ed.), <em>Statistical <span>Population Genomics</span></em> (pp. 231–266). <span>Springer US</span>. <a href="https://doi.org/10.1007/978-1-0716-0199-0_10">https://doi.org/10.1007/978-1-0716-0199-0_10</a>
</div>
<div id="ref-hurst_KaKsRatio_2002" class="csl-entry" role="listitem">
Hurst, L. D. (2002). The <span>Ka</span>/<span>Ks</span> ratio: Diagnosing the form of sequence evolution. <em>Trends in Genetics</em>, <em>18</em>(9), 486–487. <a href="https://doi.org/10.1016/S0168-9525(02)02722-1">https://doi.org/10.1016/S0168-9525(02)02722-1</a>
</div>
<div id="ref-hurst_GeneticsUnderstandingSelection_2009" class="csl-entry" role="listitem">
Hurst, L. D. (2009). Genetics and the understanding of selection. <em>Nature Reviews Genetics</em>, <em>10</em>(2), 83–93. <a href="https://doi.org/10.1038/nrg2506">https://doi.org/10.1038/nrg2506</a>
</div>
<div id="ref-johri_RecommendationsImprovingStatistical_2022" class="csl-entry" role="listitem">
Johri, P., Aquadro, C. F., Beaumont, M., Charlesworth, B., Excoffier, L., Eyre-Walker, A., Keightley, P. D., Lynch, M., McVean, G., Payseur, B. A., Pfeifer, S. P., Stephan, W., &amp; Jensen, J. D. (2022). Recommendations for improving statistical inference in population genomics. <em>PLOS Biology</em>, <em>20</em>(5), e3001669. <a href="https://doi.org/10.1371/journal.pbio.3001669">https://doi.org/10.1371/journal.pbio.3001669</a>
</div>
<div id="ref-kelleher_InferringWholegenomeHistories_2019" class="csl-entry" role="listitem">
Kelleher, J., Wong, Y., Wohns, A. W., Fadil, C., Albers, P. K., &amp; McVean, G. (2019). Inferring whole-genome histories in large population datasets. <em>Nature Genetics</em>, <em>51</em>(9), 1330–1338. <a href="https://doi.org/10.1038/s41588-019-0483-y">https://doi.org/10.1038/s41588-019-0483-y</a>
</div>
<div id="ref-kimura_1983" class="csl-entry" role="listitem">
Kimura, M. (1983). <em>The neutral theory of molecular evolution</em>. <span>Cambridge University Press</span>. <a href="https://doi.org/10.1017/CBO9780511623486">https://doi.org/10.1017/CBO9780511623486</a>
</div>
<div id="ref-kimura_ProteinPolymorphismPhase_1971" class="csl-entry" role="listitem">
Kimura, M., &amp; Ohta, T. (1971). Protein <span>Polymorphism</span> as a <span>Phase</span> of <span>Molecular Evolution</span>. <em>Nature</em>, <em>229</em>(5285), 467–469. <a href="https://doi.org/10.1038/229467a0">https://doi.org/10.1038/229467a0</a>
</div>
<div id="ref-Korneliussen_2014" class="csl-entry" role="listitem">
Korneliussen T, N. R., Albrechtsen A. (2014). ANGSD: Analysis of next generation sequencing data. <em>BMC Bioinformatics 15(1):356</em>. <a href="https://doi.org/doi: 10.1186/s12859-014-0356-4">https://doi.org/doi: 10.1186/s12859-014-0356-4</a>
</div>
<div id="ref-korneliussen_ANGSDAnalysisNext_2014" class="csl-entry" role="listitem">
Korneliussen, T. S., Albrechtsen, A., &amp; Nielsen, R. (2014). <span>ANGSD</span>: <span>Analysis</span> of <span>Next Generation Sequencing Data</span>. <em>BMC Bioinformatics</em>, <em>15</em>(1), 356. <a href="https://doi.org/10.1186/s12859-014-0356-4">https://doi.org/10.1186/s12859-014-0356-4</a>
</div>
<div id="ref-korunes_PixyUnbiasedEstimation_2021" class="csl-entry" role="listitem">
Korunes, K. L., &amp; Samuk, K. (2021). Pixy: <span>Unbiased</span> estimation of nucleotide diversity and divergence in the presence of missing data. <em>Molecular Ecology Resources</em>, <em>21</em>(4), 1359–1368. <a href="https://doi.org/10.1111/1755-0998.13326">https://doi.org/10.1111/1755-0998.13326</a>
</div>
<div id="ref-kreitman_NucleotidePolymorphismAlcohol_1983" class="csl-entry" role="listitem">
Kreitman, M. (1983). Nucleotide polymorphism at the alcohol dehydrogenase locus of <span>Drosophila</span> melanogaster. <em>Nature</em>, <em>304</em>(5925), 412. <a href="https://doi.org/10.1038/304412a0">https://doi.org/10.1038/304412a0</a>
</div>
<div id="ref-kumar_PredictingEffectsCoding_2009" class="csl-entry" role="listitem">
Kumar, P., Henikoff, S., &amp; Ng, P. C. (2009). Predicting the effects of coding non-synonymous variants on protein function using the <span>SIFT</span> algorithm. <em>Nature Protocols</em>, <em>4</em>(7), 1073–1081. <a href="https://doi.org/10.1038/nprot.2009.86">https://doi.org/10.1038/nprot.2009.86</a>
</div>
<div id="ref-kumar_MutationRatesMammalian_2002" class="csl-entry" role="listitem">
Kumar, S., &amp; Subramanian, S. (2002). Mutation rates in mammalian genomes. <em>Proceedings of the National Academy of Sciences</em>, <em>99</em>(2), 803–808. <a href="https://doi.org/10.1073/pnas.022629899">https://doi.org/10.1073/pnas.022629899</a>
</div>
<div id="ref-lan_PaleogenomicsGenomeScaleAnalysis_2019" class="csl-entry" role="listitem">
Lan, T., &amp; Lindqvist, C. (2019). Paleogenomics: <span>Genome-Scale Analysis</span> of <span>Ancient DNA</span> and <span>Population</span> and <span>Evolutionary Genomic Inferences</span>. In O. P. Rajora (Ed.), <em>Population <span>Genomics</span>: <span>Concepts</span>, <span>Approaches</span> and <span>Applications</span></em> (pp. 323–360). <span>Springer International Publishing</span>. <a href="https://doi.org/10.1007/13836_2017_7">https://doi.org/10.1007/13836_2017_7</a>
</div>
<div id="ref-laurie_LinkageDisequilibriumWild_2007" class="csl-entry" role="listitem">
Laurie, C. C., Nickerson, D. A., Anderson, A. D., Weir, B. S., Livingston, R. J., Dean, M. D., Smith, K. L., Schadt, E. E., &amp; Nachman, M. W. (2007). Linkage <span>Disequilibrium</span> in <span>Wild Mice</span>. <em>PLOS Genetics</em>, <em>3</em>(8), e144. <a href="https://doi.org/10.1371/journal.pgen.0030144">https://doi.org/10.1371/journal.pgen.0030144</a>
</div>
<div id="ref-Lawson_2018" class="csl-entry" role="listitem">
Lawson DJ, F. D., van Dorp L. (2018). A tutorial on how not to over-interpret STRUCTURE and ADMIXTURE bar plots. <em>Nat Commun. 14;9(1):3258</em>. <a href="https://doi.org/doi: 10.1038/s41467-018-05257-7">https://doi.org/doi: 10.1038/s41467-018-05257-7</a>
</div>
<div id="ref-leffler_RevisitingOldRiddle_2012" class="csl-entry" role="listitem">
Leffler, E. M., Bullaughey, K., Matute, D. R., Meyer, W. K., Ségurel, L., Venkat, A., Andolfatto, P., &amp; Przeworski, M. (2012). Revisiting an <span>Old Riddle</span>: <span>What Determines Genetic Diversity Levels</span> within <span>Species</span>? <em>PLOS Biology</em>, <em>10</em>(9), e1001388. <a href="https://doi.org/10.1371/journal.pbio.1001388">https://doi.org/10.1371/journal.pbio.1001388</a>
</div>
<div id="ref-lewis_BiggestEverStudy_2023" class="csl-entry" role="listitem">
Lewis, D. (2023). Biggest ever study of primate genomes has surprises for humanity. <em>Nature</em>. <a href="https://doi.org/10.1038/d41586-023-01776-6">https://doi.org/10.1038/d41586-023-01776-6</a>
</div>
<div id="ref-li_AligningSequenceReads_2013" class="csl-entry" role="listitem">
Li, H. (2013). Aligning sequence reads, clone sequences and assembly contigs with <span>BWA-MEM</span>. <em>arXiv:1303.3997 [q-Bio]</em>. <a href="https://arxiv.org/abs/1303.3997">https://arxiv.org/abs/1303.3997</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-li_Minimap2PairwiseAlignment_2018" class="csl-entry" role="listitem">
Li, H. (2018). Minimap2: Pairwise alignment for nucleotide sequences. <em>Bioinformatics</em>, <em>34</em>(18), 3094–3100. <a href="https://doi.org/10.1093/bioinformatics/bty191">https://doi.org/10.1093/bioinformatics/bty191</a>
</div>
<div id="ref-li_InferenceHumanPopulation_2011" class="csl-entry" role="listitem">
Li, H., &amp; Durbin, R. (2011). Inference of human population history from individual whole-genome sequences. <em>Nature</em>, <em>475</em>(7357), 493–496. <a href="https://doi.org/10.1038/nature10231">https://doi.org/10.1038/nature10231</a>
</div>
<div id="ref-li_SNPDetectionMassively_2009" class="csl-entry" role="listitem">
Li, R., Li, Y., Fang, X., Yang, H., Wang, J., Kristiansen, K., &amp; Wang, J. (2009). <span>SNP</span> detection for massively parallel whole-genome resequencing. <em>Genome Research</em>, <em>19</em>(6), 1124–1132. <a href="https://doi.org/10.1101/gr.088013.108">https://doi.org/10.1101/gr.088013.108</a>
</div>
<div id="ref-lipson_ApplyingF4statisticsAdmixture_2020" class="csl-entry" role="listitem">
Lipson, M. (2020). Applying F4-statistics and admixture graphs: <span>Theory</span> and examples. <em>Molecular Ecology Resources</em>, <em>20</em>(6), 1658–1667. <a href="https://doi.org/10.1111/1755-0998.13230">https://doi.org/10.1111/1755-0998.13230</a>
</div>
<div id="ref-liu_ExploringPopulationStructure_2020" class="csl-entry" role="listitem">
Liu, C.-C., Shringarpure, S., Lange, K., &amp; Novembre, J. (2020). Exploring <span>Population Structure</span> with <span>Admixture Models</span> and <span>Principal Component Analysis</span>. In J. Y. Dutheil (Ed.), <em>Statistical <span>Population Genomics</span></em> (pp. 67–86). <span>Springer US</span>. <a href="https://doi.org/10.1007/978-1-0716-0199-0_4">https://doi.org/10.1007/978-1-0716-0199-0_4</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-luu_PcadaptPackagePerform_2017" class="csl-entry" role="listitem">
Luu, K., Bazin, E., &amp; Blum, M. G. B. (2017). Pcadapt: An <span>R</span> package to perform genome scans for selection based on principal component analysis. <em>Molecular Ecology Resources</em>, <em>17</em>(1), 67–77. <a href="https://doi.org/10.1111/1755-0998.12592">https://doi.org/10.1111/1755-0998.12592</a>
</div>
<div id="ref-lynch_OriginsGenomeArchitecture_2007" class="csl-entry" role="listitem">
Lynch, M. (2007). <em>The origins of genome architecture</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-maier_LimitsFittingComplex_2022" class="csl-entry" role="listitem">
Maier, R., Flegontov, P., Flegontova, O., Changmai, P., &amp; Reich, D. (2022). <em>On the limits of fitting complex models of population history to genetic data</em> (p. 2022.05.08.491072). <span>bioRxiv</span>. <a href="https://doi.org/10.1101/2022.05.08.491072">https://doi.org/10.1101/2022.05.08.491072</a>
</div>
<div id="ref-maierrobert_PaperNotes_2022" class="csl-entry" role="listitem">
Maier, Robert. (2022). <em>Paper notes</em>. <a href="https://uqrmaie1.github.io/admixtools/articles/paper.html">https://uqrmaie1.github.io/admixtools/articles/paper.html</a>
</div>
<div id="ref-maruki_GenotypeCallingPopulationGenomic_2017" class="csl-entry" role="listitem">
Maruki, T., &amp; Lynch, M. (2017). Genotype <span>Calling</span> from <span>Population-Genomic Sequencing Data</span>. <em>G3 Genes|Genomes|Genetics</em>, <em>7</em>(5), 1393–1404. <a href="https://doi.org/10.1534/g3.117.039008">https://doi.org/10.1534/g3.117.039008</a>
</div>
<div id="ref-mary-huard_FastAccurateJoint_2023" class="csl-entry" role="listitem">
Mary-Huard, T., &amp; Balding, D. (2023). Fast and accurate joint inference of coancestry parameters for populations and/or individuals. <em>PLOS Genetics</em>, <em>19</em>(1), e1010054. <a href="https://doi.org/10.1371/journal.pgen.1010054">https://doi.org/10.1371/journal.pgen.1010054</a>
</div>
<div id="ref-mclaren_EnsemblVariantEffect_2016" class="csl-entry" role="listitem">
McLaren, W., Gil, L., Hunt, S. E., Riat, H. S., Ritchie, G. R. S., Thormann, A., Flicek, P., &amp; Cunningham, F. (2016). The <span>Ensembl Variant Effect Predictor</span>. <em>Genome Biology</em>, <em>17</em>(1), 122. <a href="https://doi.org/10.1186/s13059-016-0974-4">https://doi.org/10.1186/s13059-016-0974-4</a>
</div>
<div id="ref-mclaren_DerivingConsequencesGenomic_2010" class="csl-entry" role="listitem">
McLaren, W., Pritchard, B., Rios, D., Chen, Y., Flicek, P., &amp; Cunningham, F. (2010). Deriving the consequences of genomic variants with the <span>Ensembl API</span> and <span>SNP Effect Predictor</span>. <em>Bioinformatics</em>, <em>26</em>(16), 2069–2070. <a href="https://doi.org/10.1093/bioinformatics/btq330">https://doi.org/10.1093/bioinformatics/btq330</a>
</div>
<div id="ref-McVean_2009" class="csl-entry" role="listitem">
McVean. (2009). A genealogical interpretation of principal components analysis. <em>PLoS Genetics Oct;5(10):e1000686</em>. <a href="https://doi.org/doi: 10.1371/journal.pgen.1000686">https://doi.org/doi: 10.1371/journal.pgen.1000686</a>
</div>
<div id="ref-Cavalli_Sforza_1978" class="csl-entry" role="listitem">
Menozzi P, C.-S. L., Piazza A. (1978). Synthetic maps of human gene frequencies in europeans. <em>Science 1;201(4358):786-92</em>. <a href="https://doi.org/doi: 10.1126/science.356262">https://doi.org/doi: 10.1126/science.356262</a>
</div>
<div id="ref-miller_HumanBiology_2020" class="csl-entry" role="listitem">
Miller, C. (2020). <em>Human <span>Biology</span></em>. <span>Thompson Rivers University</span>.
</div>
<div id="ref-nei_MolecularEvolutionPhylogenetics_2000" class="csl-entry" role="listitem">
Nei, M., &amp; Kumar, S. (2000). <em>Molecular <span>Evolution</span> and <span>Phylogenetics</span></em>. <span>Oxford University Press</span>.
</div>
<div id="ref-nielsen_MolecularSignaturesNatural_2005" class="csl-entry" role="listitem">
Nielsen, R. (2005). Molecular <span>Signatures</span> of <span>Natural Selection</span>. <em>Annual Review of Genetics</em>, <em>39</em>(1), 197–218. <a href="https://doi.org/10.1146/annurev.genet.39.073003.112420">https://doi.org/10.1146/annurev.genet.39.073003.112420</a>
</div>
<div id="ref-nielsen_SNPCallingGenotype_2012" class="csl-entry" role="listitem">
Nielsen, R., Korneliussen, T., Albrechtsen, A., Li, Y., &amp; Wang, J. (2012). <span>SNP Calling</span>, <span>Genotype Calling</span>, and <span>Sample Allele Frequency Estimation</span> from <span>New-Generation Sequencing Data</span>. <em>PLOS ONE</em>, <em>7</em>(7), e37558. <a href="https://doi.org/10.1371/journal.pone.0037558">https://doi.org/10.1371/journal.pone.0037558</a>
</div>
<div id="ref-nielsen_GenotypeSNPCalling_2011" class="csl-entry" role="listitem">
Nielsen, R., Paul, J. S., Albrechtsen, A., &amp; Song, Y. S. (2011). Genotype and <span>SNP</span> calling from next-generation sequencing data. <em>Nature Reviews Genetics</em>, <em>12</em>(6), 443–451. <a href="https://doi.org/10.1038/nrg2986">https://doi.org/10.1038/nrg2986</a>
</div>
<div id="ref-Novembre_2008" class="csl-entry" role="listitem">
Novembre J, B. K., Johnson T. (2008). Genes mirror geography within europe. <em>Nature. 2008 Nov 6;456(7218):98-101</em>. <a href="https://doi.org/doi: 10.1038/nature07331">https://doi.org/doi: 10.1038/nature07331</a>
</div>
<div id="ref-ohta_SlightlyDeleteriousMutant_1973" class="csl-entry" role="listitem">
Ohta, T. (1973). Slightly <span>Deleterious Mutant Substitutions</span> in <span>Evolution</span>. <em>Nature</em>, <em>246</em>(5428), 96. <a href="https://doi.org/10.1038/246096a0">https://doi.org/10.1038/246096a0</a>
</div>
<div id="ref-okonechnikov_QualimapAdvancedMultisample_2016" class="csl-entry" role="listitem">
Okonechnikov, K., Conesa, A., &amp; García-Alcalde, F. (2016). Qualimap 2: Advanced multi-sample quality control for high-throughput sequencing data. <em>Bioinformatics</em>, <em>32</em>(2), 292–294. <a href="https://doi.org/10.1093/bioinformatics/btv566">https://doi.org/10.1093/bioinformatics/btv566</a>
</div>
<div id="ref-patterson_PopulationStructureEigenanalysis_2006" class="csl-entry" role="listitem">
Patterson, N., Price, A. L., &amp; Reich, D. (2006). Population <span>Structure</span> and <span>Eigenanalysis</span>. <em>PLOS Genetics</em>, <em>2</em>(12), e190. <a href="https://doi.org/10.1371/journal.pgen.0020190">https://doi.org/10.1371/journal.pgen.0020190</a>
</div>
<div id="ref-pedersen_MosdepthQuickCoverage_2018" class="csl-entry" role="listitem">
Pedersen, B. S., &amp; Quinlan, A. R. (2018). Mosdepth: Quick coverage calculation for genomes and exomes. <em>Bioinformatics</em>, <em>34</em>(5), 867–868. <a href="https://doi.org/10.1093/bioinformatics/btx699">https://doi.org/10.1093/bioinformatics/btx699</a>
</div>
<div id="ref-pennisi_AllureMonkeyflowers_2019" class="csl-entry" role="listitem">
Pennisi, E. (2019). The allure of monkeyflowers. <em>Science</em>, <em>365</em>(6456), 854–857. <a href="https://doi.org/10.1126/science.365.6456.854">https://doi.org/10.1126/science.365.6456.854</a>
</div>
<div id="ref-Peter_2016" class="csl-entry" role="listitem">
Peter, B. (2016). Admixture, population structure, and f-statistics. <em>Genetics 202(4):1485-501</em>. <a href="https://doi.org/doi: 10.1534/genetics.115.183913">https://doi.org/doi: 10.1534/genetics.115.183913</a>
</div>
<div id="ref-Peter_2022" class="csl-entry" role="listitem">
Peter, B. (2022). A geometric relationship of F2, F3 and F4-statistics with principal component analysis. <em>Philos Trans R Soc Lond B Biol Sci. 377(1852):20200413</em>. <a href="https://doi.org/doi: 10.1098/rstb.2020.0413">https://doi.org/doi: 10.1098/rstb.2020.0413</a>
</div>
<div id="ref-peter_AdmixturePopulationStructure_2016" class="csl-entry" role="listitem">
Peter, B. M. (2016). Admixture, <span>Population Structure</span>, and <span>F-Statistics</span>. <em>Genetics</em>, <em>202</em>(4), 1485–1501. <a href="https://doi.org/10.1534/genetics.115.183913">https://doi.org/10.1534/genetics.115.183913</a>
</div>
<div id="ref-Picard2019toolkit" class="csl-entry" role="listitem">
Picard toolkit. (2019). In <em>Broad Institute, GitHub repository</em>. <a href="https://broadinstitute.github.io/picard/" class="uri">https://broadinstitute.github.io/picard/</a>; Broad Institute.
</div>
<div id="ref-pope_ExpansionAgricultureHas_2023" class="csl-entry" role="listitem">
Pope, N. S., Singh, A., Childers, A. K., Kapheim, K. M., Evans, J. D., &amp; López-Uribe, M. M. (2023). The expansion of agriculture has shaped the recent evolutionary history of a specialized squash pollinator. <em>Proceedings of the National Academy of Sciences</em>, <em>120</em>(15), e2208116120. <a href="https://doi.org/10.1073/pnas.2208116120">https://doi.org/10.1073/pnas.2208116120</a>
</div>
<div id="ref-purcell_PLINKToolSet_2007" class="csl-entry" role="listitem">
Purcell, S., Neale, B., Todd-Brown, K., Thomas, L., Ferreira, M. A. R., Bender, D., Maller, J., Sklar, P., de Bakker, P. I. W., Daly, M. J., &amp; Sham, P. C. (2007). <span>PLINK</span>: <span>A Tool Set</span> for <span>Whole-Genome Association</span> and <span>Population-Based Linkage Analyses</span>. <em>The American Journal of Human Genetics</em>, <em>81</em>(3), 559–575. <a href="https://doi.org/10.1086/519795">https://doi.org/10.1086/519795</a>
</div>
<div id="ref-quinlan_BEDToolsFlexibleSuite_2010" class="csl-entry" role="listitem">
Quinlan, A. R., &amp; Hall, I. M. (2010). <span>BEDTools</span>: A flexible suite of utilities for comparing genomic features. <em>Bioinformatics</em>, <em>26</em>(6), 841–842. <a href="https://doi.org/10.1093/bioinformatics/btq033">https://doi.org/10.1093/bioinformatics/btq033</a>
</div>
<div id="ref-ReadGroups_2023" class="csl-entry" role="listitem">
Read groups. (2023). In <em>GATK</em>. <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups">https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups</a>
</div>
<div id="ref-reinert_AlignmentNextGenerationSequencing_2015" class="csl-entry" role="listitem">
Reinert, K., Langmead, B., Weese, D., &amp; Evers, D. J. (2015). Alignment of <span>Next-Generation Sequencing Reads</span>. <em>Annual Review of Genomics and Human Genetics</em>, <em>16</em>(1), 133–151. <a href="https://doi.org/10.1146/annurev-genom-090413-025358">https://doi.org/10.1146/annurev-genom-090413-025358</a>
</div>
<div id="ref-rodrigues_SharedEvolutionaryProcesses_2023" class="csl-entry" role="listitem">
Rodrigues, M. F., Kern, A. D., &amp; Ralph, P. L. (2023). <em>Shared evolutionary processes shape landscapes of genomic variation in the great apes</em> (p. 2023.02.07.527547). <span>bioRxiv</span>. <a href="https://doi.org/10.1101/2023.02.07.527547">https://doi.org/10.1101/2023.02.07.527547</a>
</div>
<div id="ref-romiguier_ComparativePopulationGenomics_2014" class="csl-entry" role="listitem">
Romiguier, J., Gayral, P., Ballenghien, M., Bernard, A., Cahais, V., Chenuil, A., Chiari, Y., Dernat, R., Duret, L., Faivre, N., Loire, E., Lourenco, J. M., Nabholz, B., Roux, C., Tsagkogeorga, G., Weber, A. a.-T., Weinert, L. A., Belkhir, K., Bierne, N., … Galtier, N. (2014). Comparative population genomics in animals uncovers the determinants of genetic diversity. <em>Nature</em>, <em>515</em>(7526), 261–263. <a href="https://doi.org/10.1038/nature13685">https://doi.org/10.1038/nature13685</a>
</div>
<div id="ref-schlotterer_SequencingPoolsIndividuals_2014" class="csl-entry" role="listitem">
Schlötterer, C., Tobler, R., Kofler, R., &amp; Nolte, V. (2014). Sequencing pools of individuals - mining genome-wide polymorphism data without big funding. <em>Nature Reviews. Genetics</em>, <em>15</em>(11), 749–763. <a href="https://doi.org/10.1038/nrg3803">https://doi.org/10.1038/nrg3803</a>
</div>
<div id="ref-Schraiber_2015" class="csl-entry" role="listitem">
Schraiber, J., &amp; Akey, J. (2015). Methods and models for unravelling human evolutionary history. <em>Nat Rev Genet. 16(12):727-40</em>. <a href="https://doi.org/doi: 10.1038/nrg4005">https://doi.org/doi: 10.1038/nrg4005</a>
</div>
<div id="ref-shen_SeqKitCrossPlatformUltrafast_2016" class="csl-entry" role="listitem">
Shen, W., Le, S., Li, Y., &amp; Hu, F. (2016). <span>SeqKit</span>: <span>A Cross-Platform</span> and <span>Ultrafast Toolkit</span> for <span>FASTA</span>/<span>Q File Manipulation</span>. <em>PLOS ONE</em>, <em>11</em>(10), e0163962. <a href="https://doi.org/10.1371/journal.pone.0163962">https://doi.org/10.1371/journal.pone.0163962</a>
</div>
<div id="ref-smith_HitchhikingEffectFavourable_1974" class="csl-entry" role="listitem">
Smith, J. M., &amp; Haigh, J. (1974). The hitch-hiking effect of a favourable gene. <em>Genetics Research</em>, <em>23</em>(1), 23–35. <a href="https://doi.org/10.1017/S0016672300014634">https://doi.org/10.1017/S0016672300014634</a>
</div>
<div id="ref-stankowski_WidespreadSelectionGene_2019" class="csl-entry" role="listitem">
Stankowski, S., Chase, M. A., Fuiten, A. M., Rodrigues, M. F., Ralph, P. L., &amp; Streisfeld, M. A. (2019). Widespread selection and gene flow shape the genomic landscape during a radiation of monkeyflowers. <em>PLOS Biology</em>, <em>17</em>(7), e3000391. <a href="https://doi.org/10.1371/journal.pbio.3000391">https://doi.org/10.1371/journal.pbio.3000391</a>
</div>
<div id="ref-stoler_SequencingErrorProfiles_2021" class="csl-entry" role="listitem">
Stoler, N., &amp; Nekrutenko, A. (2021). Sequencing error profiles of <span>Illumina</span> sequencing instruments. <em>NAR Genomics and Bioinformatics</em>, <em>3</em>(1), lqab019. <a href="https://doi.org/10.1093/nargab/lqab019">https://doi.org/10.1093/nargab/lqab019</a>
</div>
<div id="ref-talla_DissectingEffectsSelection_2019" class="csl-entry" role="listitem">
Talla, V., Soler, L., Kawakami, T., Dincă, V., Vila, R., Friberg, M., Wiklund, C., &amp; Backström, N. (2019). Dissecting the <span>Effects</span> of <span>Selection</span> and <span>Mutation</span> on <span>Genetic Diversity</span> in <span>Three Wood White</span> (<span>Leptidea</span>) <span>Butterfly Species</span>. <em>Genome Biology and Evolution</em>, <em>11</em>(10), 2875–2886. <a href="https://doi.org/10.1093/gbe/evz212">https://doi.org/10.1093/gbe/evz212</a>
</div>
<div id="ref-ToolDocumentationIndex_2023" class="csl-entry" role="listitem">
<span>Tool Documentation Index</span>. (2023). In <em>GATK</em>. <a href="https://gatk.broadinstitute.org/hc/en-us/articles/13832655155099--Tool-Documentation-Index">https://gatk.broadinstitute.org/hc/en-us/articles/13832655155099--Tool-Documentation-Index</a>
</div>
<div id="ref-vandervalk_MillionyearoldDNASheds_2021" class="csl-entry" role="listitem">
van der Valk, T., Pečnerová, P., Díez-del-Molino, D., Bergström, A., Oppenheimer, J., Hartmann, S., Xenikoudakis, G., Thomas, J. A., Dehasque, M., Sağlıcan, E., Fidan, F. R., Barnes, I., Liu, S., Somel, M., Heintzman, P. D., Nikolskiy, P., Shapiro, B., Skoglund, P., Hofreiter, M., … Dalén, L. (2021). Million-year-old <span>DNA</span> sheds light on the genomic history of mammoths. <em>Nature</em>, <em>591</em>(7849), 265–269. <a href="https://doi.org/10.1038/s41586-021-03224-9">https://doi.org/10.1038/s41586-021-03224-9</a>
</div>
<div id="ref-VertebrateGenomesProject_2022" class="csl-entry" role="listitem">
<em>Vertebrate <span>Genomes Project</span></em>. (n.d.). https://vertebrategenomesproject.org. Retrieved November 3, 2022, from <a href="https://vertebrategenomesproject.org">https://vertebrategenomesproject.org</a>
</div>
<div id="ref-wakeley_CoalescentTheoryIntroduction_2008" class="csl-entry" role="listitem">
Wakeley, J. (2008). <em>Coalescent <span>Theory</span>: <span>An Introduction</span></em> (1st Edition edition). <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-waples_WhatNeAnyway_2022" class="csl-entry" role="listitem">
Waples, R. S. (2022). What <span>Is Ne</span>, <span>Anyway</span>? <em>Journal of Heredity</em>, <em>113</em>(4), 371–379. <a href="https://doi.org/10.1093/jhered/esac023">https://doi.org/10.1093/jhered/esac023</a>
</div>
<div id="ref-webster_PopulationGenomicsInsect_2023" class="csl-entry" role="listitem">
Webster, M. T., Beaurepaire, A., Neumann, P., &amp; Stolle, E. (2023). Population <span>Genomics</span> for <span>Insect Conservation</span>. <em>Annual Review of Animal Biosciences</em>, <em>11</em>(1), 115–140. <a href="https://doi.org/10.1146/annurev-animal-122221-075025">https://doi.org/10.1146/annurev-animal-122221-075025</a>
</div>
<div id="ref-webstermatthew_GeneticVariationMountain_2022" class="csl-entry" role="listitem">
Webster, Matthew, &amp; Yuanzhen Liu. (2022). <em>Genetic <span>Variation</span> in <span>Mountain Bumblebees</span></em>.
</div>
<div id="ref-wetterstrandka_DNASequencingCosts_" class="csl-entry" role="listitem">
Wetterstrand, KA. <em><span>DNA Sequencing Costs</span>: <span>Data</span> from the <span>NHGRI Genome Sequencing Program</span> (<span>GSP</span>)</em>. <a href="https://www.genome.gov/sequencingcostsdata">www.genome.gov/sequencingcostsdata</a>
</div>
<div id="ref-willi_ConservationGeneticsManagement_2022" class="csl-entry" role="listitem">
Willi, Y., Kristensen, T. N., Sgrò, C. M., Weeks, A. R., Ørsted, M., &amp; Hoffmann, A. A. (2022). Conservation genetics as a management tool: <span>The</span> five best-supported paradigms to assist the management of threatened species. <em>Proceedings of the National Academy of Sciences</em>, <em>119</em>(1), e2105076119. <a href="https://doi.org/10.1073/pnas.2105076119">https://doi.org/10.1073/pnas.2105076119</a>
</div>
<div id="ref-wisely_GeneticDiversityFitness_2002" class="csl-entry" role="listitem">
Wisely, S. M., Buskirk, S. W., Fleming, M. A., McDonald, D. B., &amp; Ostrander, E. A. (2002). Genetic <span>Diversity</span> and <span>Fitness</span> in <span>Black-Footed Ferrets Before</span> and <span>During</span> a <span>Bottleneck</span>. <em>Journal of Heredity</em>, <em>93</em>(4), 231–237. <a href="https://doi.org/10.1093/jhered/93.4.231">https://doi.org/10.1093/jhered/93.4.231</a>
</div>
<div id="ref-y.c.brandt_EvaluationMethodsEstimating_2022" class="csl-entry" role="listitem">
Y. C. Brandt, D., Wei, X., Deng, Y., Vaughn, A. H., &amp; Nielsen, R. (2022). Evaluation of methods for estimating coalescence times using ancestral recombination graphs. <em>Genetics</em>, iyac044. <a href="https://doi.org/10.1093/genetics/iyac044">https://doi.org/10.1093/genetics/iyac044</a>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2023 <a href="about.html">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v<!--?quarto.version?--></div>
  </div>
</footer>



</body></html>