<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Nikolay Oskolkov">
  <meta name="author" content="(uses material from lectures of Ben Peter, Anders Albrechtsen, Ida Moltke, Graham Coop)">
  <title>DDLS Population genomics in practice - Population structure</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link rel="stylesheet" href="assets/styles.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-pointer/pointer.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/logos/nbis-scilifelab.svg" data-background-position="top left" data-background-size="500px" class="center">
  <h1 class="title">Population structure</h1>
  <h3 class="subtitle">Principal Component Analysis, Admixture and F-statistics</h3>
  <h4 class="author">Nikolay Oskolkov</h4>
  <h4 class="author">(uses material from lectures of Ben Peter, Anders Albrechtsen, Ida Moltke, Graham Coop)</h4>
  <h4 class="institute">NBIS</h4>
  <h4 class="date">23-Oct-2023</h4>
</section>

<section id="pca-in-population-genetics-cavalli-sforza" class="slide level2 center">
<h2>PCA in population genetics: Cavalli-Sforza</h2>

<img data-src="assets/Cavalli_Sforza.png" style="width:90.0%" class="r-stretch"><p><span class="citation" data-cites="Cavalli_Sforza_1978">(<a href="" role="doc-biblioref" onclick="return false;">Menozzi P, 1978</a>)</span> noticed that genetic distances seem to correlate with geographical distances</p>
</section>
<section id="pca-in-population-genetics-john-novembre" class="slide level2 center">
<h2>PCA in population genetics: John Novembre</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p><img data-src="assets/Novembre.png" height="510"></p>
<p><span class="citation" data-cites="Novembre_2008">(<a href="" role="doc-biblioref" onclick="return false;">Novembre J, 2008</a>)</span></p>
</div><div class="column" style="width:40%;">
<ul>
<li>genetics seems to mirrow the map of Europe</li>
<li>caveate: it is not always the case</li>
<li>J. Novembre selected samples</li>
</ul>
</div>
</div>
</section>
<section id="pca-and-mds-are-matrix-factorization-techniques" class="slide level2 center">
<h2>PCA and MDS are matrix factorization techniques</h2>

<img data-src="assets/MatrixFactor.png" class="r-stretch quarto-figure-center"></section>
<section id="toy-example-of-pca-for-population-genetics" class="slide level2 center">
<h2>Toy example of PCA for population genetics</h2>
<div class="columns">
<div class="column" style="width:49%;">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-3_4419d81d8b9f66d4c5a8a98f9ccba484">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>gen <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>,</span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>), <span class="dv">5</span>, <span class="at">by =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">colnames</span>(gen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Ind"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">rownames</span>(gen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"SNP"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">print</span>(gen)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ind1 Ind2 Ind3 Ind4 Ind5
SNP1    1    1    1    0    0
SNP2    0    1    2    1    2
SNP3    2    1    1    0    1
SNP4    0    0    1    2    2
SNP5    2    1    1    0    0
SNP6    0    0    1    1    1
SNP7    2    2    1    1    0</code></pre>
</div>
</div>
<div class="fragment">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-4_c0f7b93fdcf26942f62ae5c8c7477d48">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>gen_centered <span class="ot">&lt;-</span> <span class="fu">scale</span>(gen, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>covariance <span class="ot">&lt;-</span> <span class="fu">t</span>(gen_centered) <span class="sc">%*%</span> gen_centered</span>
<span id="cb3-3"><a href="#cb3-3"></a>eig <span class="ot">&lt;-</span> <span class="fu">eigen</span>(covariance)</span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="fu">barplot</span>(eig<span class="sc">$</span>values<span class="sc">/</span><span class="fu">sum</span>(eig<span class="sc">$</span>values), <span class="at">names =</span> <span class="fu">paste0</span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/unnamed-chunk-4-1.svg" width="960"></p>
</div>
</div>
</div>
</div><div class="column" style="width:51%;">
<div class="fragment">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-5_4ed22ec2b10ae780484196d36fd72870">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">plot</span>(eig<span class="sc">$</span>vectors[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="dv">5</span>, <span class="at">col =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="fu">points</span>(eig<span class="sc">$</span>vectors[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">pch =</span> <span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/unnamed-chunk-5-1.svg" width="960"></p>
</div>
</div>
<p>Idea: project the data into a low dimensional space that explains the largest amount of variance</p>
</div>
</div>
</div>
</section>
<section id="variance-maximization-and-eigen-value-decomposition" class="slide level2 center">
<h2>Variance maximization and eigen value decomposition</h2>

<img data-src="assets/max_variance.png" class="r-stretch quarto-figure-center"></section>
<section id="toy-example-of-mds-for-population-genetics" class="slide level2 center">
<h2>Toy example of MDS for population genetics</h2>
<div class="columns">
<div class="column" style="width:49%;">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-6_f1702d8b8c4a4eb73a323f61349e0179">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>gen <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>,</span>
<span id="cb5-2"><a href="#cb5-2"></a>    <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>), <span class="dv">5</span>, <span class="at">by =</span> <span class="cn">TRUE</span>))</span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="fu">colnames</span>(gen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Ind"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="fu">rownames</span>(gen) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"SNP"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="fu">print</span>(gen)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ind1 Ind2 Ind3 Ind4 Ind5
SNP1    1    1    1    0    0
SNP2    0    1    2    1    2
SNP3    2    1    1    0    1
SNP4    0    0    1    2    2
SNP5    2    1    1    0    0
SNP6    0    0    1    1    1
SNP7    2    2    1    1    0</code></pre>
</div>
</div>
<div class="fragment">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-7_01f6f2c3967e82f94ca91658ee1923f5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>my_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(gen), <span class="at">method =</span> <span class="st">"manhattan"</span>, <span class="at">upper =</span> <span class="cn">TRUE</span>, <span class="at">diag =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a>my_dist</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ind1 Ind2 Ind3 Ind4 Ind5
Ind1    0    3    7   10   11
Ind2    3    0    4    7    8
Ind3    7    4    0    5    4
Ind4   10    7    5    0    3
Ind5   11    8    4    3    0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>my_mds <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(my_dist)</span>
<span id="cb9-2"><a href="#cb9-2"></a>my_mds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]       [,2]
Ind1 -6.114866  0.2977272
Ind2 -3.077094  0.2402932
Ind3  0.621230 -1.7944048
Ind4  3.723118  2.1483083
Ind5  4.847612 -0.8919238</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:51%;">
<div class="fragment">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-8_84323eeee01e66e4b917aee7d86d77d5">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">plot</span>(my_mds, <span class="at">xlab =</span> <span class="st">"MDS1"</span>, <span class="at">ylab =</span> <span class="st">"MDS2"</span>, <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">cex =</span> <span class="dv">5</span>, <span class="at">col =</span> <span class="dv">6</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="fu">points</span>(my_mds, <span class="at">pch =</span> <span class="fu">as.character</span>(<span class="dv">5</span><span class="sc">:</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/unnamed-chunk-8-1.svg" width="960"></p>
</div>
</div>
<p>Idea: project the data into a low dimensional space that preserves distances</p>
</div>
</div>
</div>
</section>
<section id="example-of-pca-plot-from-1000g-project" class="slide level2 center">
<h2>Example of PCA plot from 1000G project</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>We will use ANGSD <span class="citation" data-cites="Korneliussen_2014">(<a href="" role="doc-biblioref" onclick="return false;">Korneliussen T, 2014</a>)</span> and compute genotype likelihoods on 435 bam-files from the 1000G project.</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-9_95447516b70ac77698d53309927a20b6">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="bu">cd</span> /home/nikolay/Documents/Teaching/PopGen_2023/data/1000G_bam_files</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">find</span> <span class="va">$PWD</span> <span class="at">-name</span> <span class="st">'*.bam'</span> <span class="op">&gt;</span> 1000G_bam_list.txt</span>
<span id="cb12-3"><a href="#cb12-3"></a><span class="ex">angsd</span> <span class="at">-bam</span> 1000G_bam_list.txt <span class="at">-GL</span> 2 <span class="at">-doMajorMinor</span> 1 <span class="at">-doMaf</span> 1 <span class="at">-SNP_pval</span> 2e-6 <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4"></a>-minMapQ 30 <span class="at">-minQ</span> 20 <span class="at">-minInd</span> 25 <span class="at">-minMaf</span> 0.05 <span class="at">-doGlf</span> 2 <span class="at">-out</span> 1000G <span class="at">-P</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we will run PCAANGSD to infer admixures proportions:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-10_32baac443cb84b15ce537f11638228b5">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="ex">pcangsd</span> <span class="at">-b</span> 1000G.beagle.gz <span class="at">-o</span> pca_1000G <span class="at">-t</span> 4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and finally we will plots the results using custom R scripts:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-11_e87769179ab4faf6d024491f96365b70">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>C <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">read.table</span>(<span class="st">"pca_1000G.cov"</span>))</span>
<span id="cb14-2"><a href="#cb14-2"></a>e <span class="ot">&lt;-</span> <span class="fu">eigen</span>(C)</span>
<span id="cb14-3"><a href="#cb14-3"></a>pops <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">"1000G_bam_list.txt"</span>)</span>
<span id="cb14-4"><a href="#cb14-4"></a>pops <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(pops, <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>), <span class="cf">function</span>(x) x[<span class="dv">6</span>])</span>
<span id="cb14-5"><a href="#cb14-5"></a>mycolor <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"red"</span>, <span class="fu">length</span>(pops))</span>
<span id="cb14-6"><a href="#cb14-6"></a>mycolor[pops <span class="sc">==</span> <span class="st">"CEU"</span>] <span class="ot">&lt;-</span> <span class="st">"blue"</span></span>
<span id="cb14-7"><a href="#cb14-7"></a>mycolor[pops <span class="sc">==</span> <span class="st">"CHB"</span>] <span class="ot">&lt;-</span> <span class="st">"green"</span></span>
<span id="cb14-8"><a href="#cb14-8"></a>mycolor[pops <span class="sc">==</span> <span class="st">"MXL"</span>] <span class="ot">&lt;-</span> <span class="st">"brown"</span></span>
<span id="cb14-9"><a href="#cb14-9"></a>mycolor[pops <span class="sc">==</span> <span class="st">"ASW"</span>] <span class="ot">&lt;-</span> <span class="st">"magenta"</span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="fu">plot</span>(e<span class="sc">$</span>vectors[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">main =</span> <span class="st">"PCA 1000G Project"</span>,</span>
<span id="cb14-11"><a href="#cb14-11"></a>    <span class="at">col =</span> mycolor, <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">c</span>(<span class="st">"YRI"</span>, <span class="st">"CEU"</span>, <span class="st">"CHB"</span>, <span class="st">"MXL"</span>, <span class="st">"ASW"</span>), <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>,</span>
<span id="cb14-13"><a href="#cb14-13"></a>    <span class="st">"green"</span>, <span class="st">"brown"</span>, <span class="st">"magenta"</span>), <span class="at">cex =</span> <span class="dv">2</span>, <span class="at">inset =</span> <span class="fl">0.02</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="assets/pca_1000G.png" style="width:200.0%"></p>
</div>
</div>
</section>
<section id="angsd-genotype-likelihood-tool-for-low-coverage-data" class="slide level2 center">
<h2>ANGSD: genotype likelihood tool for low-coverage data</h2>

<img data-src="assets/angsd_publ.png" class="r-stretch quarto-figure-center"><p><span class="citation" data-cites="Korneliussen_2014">(<a href="" role="doc-biblioref" onclick="return false;">Korneliussen T, 2014</a>)</span></p>
</section>
<section id="hard-genotype-calls-vs.-genotype-likelihoods" class="slide level2 center">
<h2>Hard genotype calls vs.&nbsp;genotype likelihoods</h2>

<img data-src="assets/likes.png" class="r-stretch quarto-figure-center"><div class="fragment">
<ul>
<li>Minor allele frequency based on hard calls: <span class="math display">\[\rm{MAF_{hc}}=\frac{1+1+2}{2*5}=\frac{4}{10}\]</span></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Minor allele frequency based on genotype likelihoods: <span class="math display">\[\rm{MAF_{gl}}=\frac{0*(0.98+0.20+0.89+0.10+0.15)+1*(0.02+0.75+0.10+0.80+0.20)+2*(0.00+0.05+0.01+0.10+0.65)}{2*5}=\frac{3.49}{10}\]</span></li>
</ul>
</div>
</section>
<section id="a-known-pitfall-of-pca-uneven-sampling" class="slide level2 center">
<h2>A known pitfall of PCA: uneven sampling</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/McVean.png" height="510"></p>
</figure>
</div>
<p><span class="citation" data-cites="McVean_2009">(<a href="" role="doc-biblioref" onclick="return false;">McVean, 2009</a>)</span></p>
</div><div class="column" style="width:60%;">
<p>“The results described here provide an explanation. First, from Equation 10 it can be seen that <strong>the matrix M is influenced by the relative sample size from each population through the components <span class="math inline">\(t_i\)</span></strong>. For instance, even if all populations are equally divergent from each other, those for which there are fewer samples will have larger values of <span class="math inline">\(t_i\)</span> because relatively more pairwise comparisons are between populations.”</p>
<p><span class="math display">\[M=XX^T=\frac{1}{N}\sum_{ij}x_ix_j\]</span> <span class="math display">\[N=N_{pop1}+N_{pop2}+N_{pop3}+...=\sum_k N_k\]</span> <span class="math display">\[M_{uneven}=\sum_{ijk}\frac{1}{N_k}x_{ik}x_{jk}\]</span></p>
<ul>
<li>Potential solution: normalize each sample by its population size before computing the covariance matrix.</li>
<li>Is it still an unsupervised technique?</li>
</ul>
</div>
</div>
</section>
<section id="example-of-uneven-sampling-from-1000g-project" class="slide level2 center">
<h2>Example of uneven sampling from 1000G project</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="assets/Downsampled_Europeans.png"></p>
<p><strong>Downsampled Europeans</strong></p>
</div><div class="column" style="width:50%;">
<p><img data-src="assets/Downsampled_Asians.png"></p>
<p><strong>Downsampled Asians</strong></p>
</div>
</div>
</section>
<section id="admixture-underlying-assumtions" class="slide level2 center">
<h2>Admixture: underlying assumtions</h2>
<div class="columns">
<div class="column" style="width:70%;">
<p><img data-src="assets/IBD.png"></p>
</div><div class="column" style="width:30%;">
<ul>
<li>genetic clustering algorithm</li>
<li>at least two unadmixed populations</li>
<li>at least one admixed population</li>
<li>Maximum Likelihood: ADMIXTURE</li>
<li>Bayesian: STRUCTURE</li>
<li>delivers admixture proportions Q</li>
<li>delivers allele frequencies F for all loci for all K populations</li>
</ul>
</div>
</div>
</section>
<section id="q-and-f-outputs-of-admixture-analysis" class="slide level2 center">
<h2>Q and F outputs of admixture analysis</h2>

<img data-src="assets/QF.png" class="r-stretch quarto-figure-center"></section>
<section id="ngsadmix-vs.-admixture" class="slide level2 center">
<h2>NGSadmix vs.&nbsp;ADMIXTURE</h2>
<div class="columns">
<div class="column" style="width:45%;">
<p><img data-src="assets/NGSadmix_publ.png"></p>
</div><div class="column" style="width:55%;">
<p><img data-src="assets/NGSadmix_vs_Admixture.png"></p>
</div>
</div>
</section>
<section id="example-of-admixture-plot-from-1000g-project" class="slide level2 center">
<h2>Example of admixture plot from 1000G project</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>We will use ANGSD <span class="citation" data-cites="Korneliussen_2014">(<a href="" role="doc-biblioref" onclick="return false;">Korneliussen T, 2014</a>)</span> and compute genotype likelihoods on 435 bam-files from the 1000G project.</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-12_b5cd5a89b17ea6d089248da997878ed1">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="bu">cd</span> /home/nikolay/Documents/Teaching/PopGen_2023/data/1000G_bam_files</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="fu">find</span> <span class="va">$PWD</span> <span class="at">-name</span> <span class="st">'*.bam'</span> <span class="op">&gt;</span> 1000G_bam_list.txt</span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="ex">angsd</span> <span class="at">-bam</span> 1000G_bam_list.txt <span class="at">-GL</span> 2 <span class="at">-doMajorMinor</span> 1 <span class="at">-doMaf</span> 1 <span class="at">-SNP_pval</span> 2e-6 <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4"></a>-minMapQ 30 <span class="at">-minQ</span> 20 <span class="at">-minInd</span> 25 <span class="at">-minMaf</span> 0.05 <span class="at">-doGlf</span> 2 <span class="at">-out</span> 1000G <span class="at">-P</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we will run NGSadmix to infer admixures proportions:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-13_c6823072c8e73154c59aa183aac9e060">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1"></a><span class="ex">NGSadmix</span> <span class="at">-likes</span> 1000G.beagle.gz <span class="at">-K</span> 3 <span class="at">-minMaf</span> 0.05 <span class="at">-seed</span> 1 <span class="at">-o</span> 1000G</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and finally we will plots the results using custom R scripts:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-14_4ac9af9ca4fd84610ed13970648ad0a3">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>pops <span class="ot">&lt;-</span> <span class="fu">readLines</span>(<span class="st">"1000G_bam_list.txt"</span>)</span>
<span id="cb17-2"><a href="#cb17-2"></a>pops <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(pops, <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>), <span class="cf">function</span>(x) x[<span class="dv">6</span>])</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="fu">source</span>(<span class="st">"https://raw.githubusercontent.com/GenisGE/evalAdmix/master/visFuns.R"</span>)</span>
<span id="cb17-4"><a href="#cb17-4"></a>qopts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"1000G.qopt"</span>)</span>
<span id="cb17-5"><a href="#cb17-5"></a>ord <span class="ot">&lt;-</span> <span class="fu">orderInds</span>(<span class="at">pop =</span> pops, <span class="at">q =</span> qopts, <span class="at">popord =</span> <span class="fu">c</span>(<span class="st">"YRI"</span>, <span class="st">"ASW"</span>, <span class="st">"CEU"</span>, <span class="st">"MXL"</span>,</span>
<span id="cb17-6"><a href="#cb17-6"></a>    <span class="st">"CHB"</span>))</span>
<span id="cb17-7"><a href="#cb17-7"></a><span class="fu">barplot</span>(<span class="fu">t</span>(qopts)[, ord], <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>), <span class="at">las =</span> <span class="dv">2</span>, <span class="at">space =</span> <span class="dv">0</span>, <span class="at">border =</span> <span class="cn">NA</span>)</span>
<span id="cb17-8"><a href="#cb17-8"></a><span class="fu">text</span>(<span class="fu">sort</span>(<span class="fu">tapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(pops), pops[ord], mean)), <span class="sc">-</span><span class="fl">0.05</span>, <span class="fu">unique</span>(pops[ord]))</span>
<span id="cb17-9"><a href="#cb17-9"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">cumsum</span>(<span class="fu">sapply</span>(<span class="fu">unique</span>(pops[ord]), <span class="cf">function</span>(x) {</span>
<span id="cb17-10"><a href="#cb17-10"></a>    <span class="fu">sum</span>(pops[ord] <span class="sc">==</span> x)</span>
<span id="cb17-11"><a href="#cb17-11"></a>})), <span class="at">col =</span> <span class="dv">1</span>, <span class="at">lwd =</span> <span class="fl">1.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="assets/admixture_1000G.png" style="width:200.0%"></p>
</div>
</div>
</section>
<section id="admixture-bar-plots-should-not-be-over-interpreted" class="slide level2 center">
<h2>Admixture bar plots should not be over-interpreted</h2>

<img data-src="assets/Falush.png" class="r-stretch quarto-figure-center"><p><span class="citation" data-cites="Lawson_2018">(<a href="" role="doc-biblioref" onclick="return false;">Lawson DJ, 2018</a>)</span></p>
</section>
<section id="haplotype-clustering" class="slide level2 center">
<h2>Haplotype clustering</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/haplo_clustering.png"></p>
</figure>
</div>
</div><div class="column" style="width:30%;">
<ul>
<li><p>Ancestry assignments in STRUCTURE / ADMIXTURE do not identify where admixture has occurred</p></li>
<li><p>Haplotype-based methods explore <strong>local ancestry</strong>, LD, recombination and subsequently fine-scale patterns of population structure at different scales</p></li>
<li><p>Chromosome painting: the genome is a mosaic of LD blocks separated by recombination</p></li>
<li><p>ChromoPainter and fineSTRUCTURE are tools for resolving <strong>subtle differences</strong> between populations.</p></li>
</ul>
<p><span class="citation" data-cites="Lawson_2012">(<a href="" role="doc-biblioref" onclick="return false;">D. Lawson, 2012</a>)</span></p>
<p><span class="citation" data-cites="Schraiber_2015">(<a href="" role="doc-biblioref" onclick="return false;">Schraiber &amp; Akey, 2015</a>)</span></p>
</div>
</div>
</section>
<section id="fixation-index-f_st" class="slide level2 center">
<h2>Fixation index <span class="math inline">\(F_{st}\)</span></h2>
<div class="columns">
<div class="column" style="width:45%;">
<ul>
<li><span class="math inline">\(F_{st}\)</span> is a measure of population differentiation due to population structure</li>
<li>Expressed through nucleotide diversity <span class="math inline">\(\pi\)</span> as relative nucleotide diversity between populations minus average nucleotide diversity within populations. If we have two populations 1 and 2, then:</li>
</ul>
<p><span class="math display">\[
F_{st}=\frac{\pi_{12}-\frac{\pi_{1}+\pi_{2}}{2}}{\pi_{12}} \sim
\frac{\sigma_{subpops}^2}{\sigma_{total}^2}
\]</span></p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-15_6524748472d00d9867e1111e398bb319">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">library</span>(<span class="st">"admixtools"</span>)</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="fu">library</span>(<span class="st">"tidyverse"</span>)</span>
<span id="cb18-3"><a href="#cb18-3"></a></span>
<span id="cb18-4"><a href="#cb18-4"></a>admixtools<span class="sc">::</span><span class="fu">extract_f2</span>(<span class="st">"AADR"</span>, <span class="at">outdir =</span> <span class="st">"f2_AADR"</span>)</span>
<span id="cb18-5"><a href="#cb18-5"></a></span>
<span id="cb18-6"><a href="#cb18-6"></a>f2_aadr <span class="ot">&lt;-</span> <span class="fu">read_f2</span>(<span class="st">"f2_AADR"</span>)</span>
<span id="cb18-7"><a href="#cb18-7"></a>fst_aadr <span class="ot">&lt;-</span> <span class="fu">fst</span>(f2_aadr)</span>
<span id="cb18-8"><a href="#cb18-8"></a>fst_aadr</span>
<span id="cb18-9"><a href="#cb18-9"></a></span>
<span id="cb18-10"><a href="#cb18-10"></a>mat <span class="ot">&lt;-</span> <span class="fu">f2</span>(f2_aadr, <span class="at">unique_only =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb18-11"><a href="#cb18-11"></a>    <span class="fu">select</span>(<span class="sc">-</span>se) <span class="sc">%&gt;%</span></span>
<span id="cb18-12"><a href="#cb18-12"></a>    <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> pop2, <span class="at">values_from =</span> est) <span class="sc">%&gt;%</span></span>
<span id="cb18-13"><a href="#cb18-13"></a>    <span class="fu">column_to_rownames</span>(<span class="st">"pop1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-14"><a href="#cb18-14"></a>    <span class="fu">as.matrix</span>()</span>
<span id="cb18-15"><a href="#cb18-15"></a></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="fu">library</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb18-17"><a href="#cb18-17"></a><span class="fu">pheatmap</span>(mat, <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>, <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:55%;">
<p><img data-src="assets/fst_heatmap.png" style="width:200.0%"></p>
</div>
</div>
</section>
<section id="allen-ancient-dna-resource-aadr-dataset" class="slide level2 center">
<h2>Allen Ancient DNA Resource (AADR) dataset</h2>
<!-- markdownlint-disable MD013 -->
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="assets/AADR.png" data-fig-align="center"> <a href="https://reich.hms.harvard.edu/allen-ancient-dna-resource-aadr-downloadable-genotypes-present-day-and-ancient-dna-data">https://reich.hms.harvard.edu/allen-ancient-dna-resource-aadr-downloadable-genotypes-present-day-and-ancient-dna-data</a></p>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-16_b6011e0b67dfb184003ab3006c92b76b">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">head</span> data/DavidReich/AADR.ind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             ALT-116 M    Tubalar
             ALT-117 M    Tubalar
             ALT-165 M    Tubalar
             ALT-845 M    Tubalar
             ALT-846 M    Tubalar
             GEO-002 M   Georgian
             GEO-005 M   Georgian
             GEO-010 M   Georgian
             GEO-015 M   Georgian
             GEO-020 M   Georgian</code></pre>
</div>
</div>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-17_8ccd9921646a544d8bb8560ce3d003f5">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">head</span> <span class="at">-n</span> 8 data/DavidReich/AADR.snp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           rs3094315     1        0.020130          752566 G A
           rs7419119     1        0.022518          842013 T G
          rs13302957     1        0.024116          891021 G A
           rs6696609     1        0.024457          903426 C T
              rs8997     1        0.025727          949654 A G
           rs9442372     1        0.026288         1018704 A G
         rs147606383     1        0.026665         1045331 G A
           rs4970405     1        0.026674         1048955 A G</code></pre>
</div>
</div>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-18_ac3883211112d9017bb043d81d92e0a0">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">head</span> data/DavidReich/AADR.geno</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>12109000100001100012110220021200101020000110010200001011011000011010000011000111000011011010212912100101100100011102112122221212221110000010100000001002291001111100000011000010000000100000000001100001100000111102222220111211202221121121120000001210110100001000100000001111111021222210000000000110000001011122010100011000101120101211000000010100000000000000101010100111100110011000100000011001100011001111111110110211120012021102222101112011222221122222222221212110011001100011010000000001100101011001000111011009001122112010011121112112102101010000001000000010100001000000000000101100011001000010000000000000002000000001011000000001911001
12122211222121212210121002100111200122211022221121010222112122001012002201122111211111221221222222112211222122111111212202221212212112221222222222212222202122112221222222212212212102222222122222222221122222221222112222222222222222212012222212222111210222221222122221222111111211011122222222222122222201111222222121212210222121112222222122222222212211112222122222212122121122122221222021001112122210210212212122222121111222222222222112122212221212212212222222022202222212122122112121112112211212122212222212212122222222212221222212101021220022021222222121122222222222222222221222222212220122222122202222121222122222222121221212001121102210
22000100100000000010011000001111001120001111001110001010021001111200020010000000100001110000001112000000000010012020211201001012111022221222222112200121000010000100000000000000110001000101001102101121011029011011112000001110010000101101110001000111211022121221111200020001000000000000000000000100001001002110000100001100000001000101201010001100010201111210000110001010100000000000010200210111100010001010101001001101111112020011000110200011010200000000111001000020100000100000000200012000101111201000001000002000100000011011112000201110200100110100101000100010100000010000000101000000010000000111000111011001110001100111011010111021111110
20212111212120212212102212121100202122211121121202122222122221121022022222022112100111210121122221122212222212210201012100221100021120000001011001121122210112121111222222210212102101221121221020121212201121210000210210111002192111010100111211220001210200101001111020202122222221222222222122221221222100210001211121221211222111112101021112212122212121111122121112112022222222122221112221222122922211211100111121112121000000100001219011001100009000110000110011001902211211222222222022211221121010121221222212211012212100000020221112221221101121021222212122122212222222111222212112122222210121222112222111210221122222222111221212011101201111
00000000011111900000000000000000000000000000000100010000000000000000000000000010000010001000000000000010000000011120121122110122212120000000000000010101200000000010200000100020000000000010000010000000090900102100110112101209110010121111220000000212911100000900000000002000000000000000000000000000000101202102000100010000011001001111000001001000000000001000100100000100000000000010000000000000000000000211211001010201021101200111102112121911221121110112102112111101000010010000010000000002000100012000000100010012000112111110100001001001000011000000001000000001000010001000000001000110000000001000000000000000001001000000000000000100000000
00001100010112211012210220121110121000000120020200101001101000111011100011000100021110210010101211210010011110000110010002000122102000000010101001121220011221100122210001100100010110000110000000000000009000011111202211100122211112201121121101012111000000000000000001000110101201010111111100100000001121111200000210011121111101000110010001010001011000001200101121122210100101000121000001011002100101001201021110211111112211112012111121111120112011201112111111121112011011011000110000210001000210210012000210021122012110021211100011012111122211112100011110001010000020100000000000000010001101000000100001001002111000000000010000000100000111
22222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222211222222222222222222222222222222222222222222222222222222292222222222222222222222222222222222222222222222222222222222222222222222221222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222122222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222222
22222212222122112220021200001212221221222101222121221222211121212222222112222222222222121222121122122212222222222222222222222222222221212222221221102221212121222201122222112222222112222122222122222222222222222222222222222222222222222222221122211222222222222222222221222112111020212211222222122212112222222222022122922122222221121222222212212221222222221022221222112122222212122221212221212210221222222222222222222222222222222222222222222222222222222222222222222222222212221222222222212222222222112212222222221212222222222212222211110122121011122222120112221212211212122222211222212222221121122222222221221221112222222222222222222212222222
22222102212120012220021200001212201221222101222121121222111121212221222111221121201112021212111012012201211122121122121101221111112121212222221221102201201000122100022222022122222112222112222122292222222222112000220011221211121129121212221121210122002122222222222221210112111020212210121122122211111101211110022122121111111120121110922211212221211222221022221121111021122111121111212221211210221922222212122221211220222002221119222012110201202122221211202212222201222211211222212222112222222212110211222222211201220011112212212211110121101011120122110012221212211212122222211222112212221121122222122221221220011222229222222222222112222221
22222222222222222222220220222222222222222222221222212222221222222222222222222212122221212222222222222222222222211221101220111211121210212222112221222222222212222222222222222222222222222222222222222222222222111110110012100212012222111202211222222111110022222222222222221012919990910202222122222222222222120100222222221222222222222200222222222222222222229222221222222221122222222222222222122212221222222221212111112211021110111112122002101211201122021211011121121101122222122221222221222221122212221121222222212122222110221202212221222222122212221222222222122222222222222122222222222222222222222212222212222221212222222222222222222222222222</code></pre>
</div>
</div>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
</section>
<section id="example-of-fst-computation-from-1000g-project" class="slide level2 center">
<h2>Example of Fst computation from 1000G project</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>We will use ANGSD <span class="citation" data-cites="Korneliussen_2014">(<a href="" role="doc-biblioref" onclick="return false;">Korneliussen T, 2014</a>)</span> and compute Fst between CEU and YRI populations from the 1000G project.</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-19_8f1c419d72ed0c54b7d5b8ba03e1ad9c">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1"></a><span class="ex">angsd</span> <span class="at">-b</span> 1000G_CEU_bam_list.txt <span class="at">-anc</span> anc.fa <span class="at">-out</span> CEU <span class="at">-dosaf</span> 1 <span class="at">-gl</span> 1</span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="ex">angsd</span> <span class="at">-b</span> 1000G_YRI_bam_list.txt <span class="at">-anc</span> anc.fa <span class="at">-out</span> YRI <span class="at">-dosaf</span> 1 <span class="at">-gl</span> 1</span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="ex">realSFS</span> CEU.saf.idx YRI.saf.idx <span class="op">&gt;</span> CEU.YRI.ml</span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="ex">realSFS</span> fst index CEU.saf.idx YRI.saf.idx <span class="at">-sfs</span> CEU.YRI.ml <span class="at">-fstout</span> CEU_YRI</span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="ex">realSFS</span> fst stats  CEU_YRI.fst.idx</span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="ex">realSFS</span> fst stats2 CEU_YRI.fst.idx <span class="at">-win</span> 50 <span class="at">-step</span> 10 <span class="op">&gt;</span> slidingwindow_win50_step10</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and finally we will plots the results using custom R scripts:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-20_52514b93b885e66bebc724aaeac13a88">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"slidingwindow_win50_step10"</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a>df <span class="ot">&lt;-</span> df[<span class="sc">-</span><span class="dv">1</span>, ]</span>
<span id="cb26-3"><a href="#cb26-3"></a>df<span class="sc">$</span>V2 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>V2)</span>
<span id="cb26-4"><a href="#cb26-4"></a>df<span class="sc">$</span>V3 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>V3)</span>
<span id="cb26-5"><a href="#cb26-5"></a>df<span class="sc">$</span>V5 <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>V5)</span>
<span id="cb26-6"><a href="#cb26-6"></a>df <span class="ot">&lt;-</span> df[<span class="fu">order</span>(df<span class="sc">$</span>V2, df<span class="sc">$</span>V3), ]</span>
<span id="cb26-7"><a href="#cb26-7"></a><span class="fu">rownames</span>(df) <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">dim</span>(df)[<span class="dv">1</span>], <span class="dv">1</span>)</span>
<span id="cb26-8"><a href="#cb26-8"></a><span class="fu">plot</span>(df<span class="sc">$</span>V5, <span class="at">col =</span> df<span class="sc">$</span>V2, <span class="at">xlab =</span> <span class="st">"Chromosomes"</span>, <span class="at">ylab =</span> <span class="st">"Fst"</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>)</span>
<span id="cb26-9"><a href="#cb26-9"></a>myticks <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">rownames</span>(df[<span class="sc">!</span><span class="fu">duplicated</span>(df<span class="sc">$</span>V2), ]))</span>
<span id="cb26-10"><a href="#cb26-10"></a><span class="fu">axis</span>(<span class="at">side =</span> <span class="dv">1</span>, <span class="at">at =</span> myticks, <span class="at">labels =</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">21</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="assets/Fst_1000G.png" style="width:200.0%"></p>
</div>
</div>
</section>
<section id="f-statistics-underlying-assumtions" class="slide level2 center">
<h2>F-statistics: underlying assumtions</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="assets/fstat_tree.png"></p>
</div><div class="column" style="width:50%;">
<ul>
<li><p>Treeness test: are populations related in a tree-like fashion (Reich et al.&nbsp;2009)?</p></li>
<li><p>Admixture test: ss a particular population descended from multiple ancestral populations (Reich et al.&nbsp;2009)?</p></li>
<li><p>Admixture proportions: what are the contributions from different populations to a focal population (Green et al.&nbsp;2010; Haak et al. 2015)?</p></li>
<li><p>Number of founders: how many founder populations are there for a certain region (Reich et al.&nbsp;2012; Lazaridis et al.&nbsp;2014)?</p></li>
<li><p>Complex demography: how can mixtures and splits of population explain demography (Patterson et al.&nbsp;2012; Lipson et al.&nbsp;2013)?</p></li>
<li><p>Closest relative: what is the closest relative to a contemporary or ancient population (Raghavan et al.&nbsp;2014)?</p></li>
</ul>
<p><span class="citation" data-cites="Peter_2016">(<a href="" role="doc-biblioref" onclick="return false;">Peter, 2016</a>)</span></p>
</div>
</div>
</section>
<section id="f2-statistic-and-f3-statistic" class="slide level2 center">
<h2>F2-statistic and F3-statistic</h2>
<div class="columns">
<div class="column" style="width:65%;">
<ul>
<li><span class="math inline">\(F_2\)</span> is a covarinace in contrast to <span class="math inline">\(F_{st}\)</span> which is a correlation coefficeint, therefore <span class="math inline">\(F_2\)</span> is tree-additive by its definition</li>
</ul>
<p><span class="math display">\[F_2=2\pi_{12}-\pi_{1}-\pi_{2}\]</span></p>
<ul>
<li><span class="math inline">\(F_3\)</span> can be expressed through a combination of <span class="math inline">\(F_2\)</span>-statistic</li>
</ul>
<p><span class="math display">\[F_3(P_X; P_1, P_2)=\frac{1}{2}\left(F_2(P_X, P_1)+F_2(P_X, P_2)-F_2(P_1, P_2)\right)\]</span></p>
<ul>
<li>Outgroup <span class="math inline">\(F_3\)</span>-statistic: for an unknown population <span class="math inline">\(P_X\)</span> find the closest population from a panel <span class="math inline">\(P_i\)</span> computing the distances via an outgroup <span class="math inline">\(P_O\)</span>.</li>
</ul>
<p><img data-src="assets/F3_outgroup.png" style="width:30.0%"></p>
</div><div class="column" style="width:35%;">
<p><img data-src="assets/F3_tree.png"></p>
</div>
</div>
</section>
<section id="example-of-f3-outgroup-statistic-for-an-unknown-population" class="slide level2 center">
<h2>Example of F3-outgroup-statistic for an Unknown population</h2>
<div class="columns">
<div class="column" style="width:53%;">
<p>We will use the AADR dataset and compute F3-statistic for a (presumably!) European population, which is marked as <strong>Unknown</strong> in the dataset, with respect to a panel of other World populations using a Mbuti as an outgroup population.</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-21_cd05e6bcc3a0ce054a45391722fa3e1b">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">library</span>(<span class="st">"admixtools"</span>)</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="fu">library</span>(<span class="st">"tidyverse"</span>)</span>
<span id="cb27-3"><a href="#cb27-3"></a></span>
<span id="cb27-4"><a href="#cb27-4"></a>admixtools<span class="sc">::</span><span class="fu">extract_f2</span>(<span class="st">"AADR"</span>, <span class="at">outdir =</span> <span class="st">"f2_AADR"</span>)</span>
<span id="cb27-5"><a href="#cb27-5"></a></span>
<span id="cb27-6"><a href="#cb27-6"></a>f2_aadr <span class="ot">&lt;-</span> <span class="fu">read_f2</span>(<span class="st">"f2_AADR"</span>)</span>
<span id="cb27-7"><a href="#cb27-7"></a>fst_aadr <span class="ot">&lt;-</span> <span class="fu">fst</span>(f2_aadr)</span>
<span id="cb27-8"><a href="#cb27-8"></a>fst_aadr</span>
<span id="cb27-9"><a href="#cb27-9"></a></span>
<span id="cb27-10"><a href="#cb27-10"></a>f3_aadr <span class="ot">&lt;-</span> <span class="fu">f3</span>(f2_aadr, <span class="at">pop1 =</span> <span class="st">"Mbuti"</span>, <span class="at">pop2 =</span> <span class="st">"Unknown"</span>, <span class="at">pop3 =</span> <span class="fu">unique</span>(ind<span class="sc">$</span>pop)) <span class="sc">%&gt;%</span></span>
<span id="cb27-11"><a href="#cb27-11"></a>    <span class="fu">arrange</span>(est)</span>
<span id="cb27-12"><a href="#cb27-12"></a>f3_aadr</span>
<span id="cb27-13"><a href="#cb27-13"></a></span>
<span id="cb27-14"><a href="#cb27-14"></a></span>
<span id="cb27-15"><a href="#cb27-15"></a>viz_outgroup_f3 <span class="ot">&lt;-</span> <span class="cf">function</span>(f3_res) {</span>
<span id="cb27-16"><a href="#cb27-16"></a>    f3_res <span class="sc">%&gt;%</span></span>
<span id="cb27-17"><a href="#cb27-17"></a>        <span class="fu">mutate</span>(<span class="at">pop3 =</span> <span class="fu">fct_reorder</span>(pop3, est)) <span class="sc">%&gt;%</span></span>
<span id="cb27-18"><a href="#cb27-18"></a>        <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> pop3, <span class="at">y =</span> est, <span class="at">ymin =</span> est <span class="sc">-</span> <span class="dv">3</span> <span class="sc">*</span> se, <span class="at">ymax =</span> est <span class="sc">+</span> <span class="dv">3</span> <span class="sc">*</span> se)) <span class="sc">+</span></span>
<span id="cb27-19"><a href="#cb27-19"></a>        <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_errorbar</span>() <span class="sc">+</span> <span class="fu">coord_flip</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>(<span class="dv">25</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb27-20"><a href="#cb27-20"></a>        <span class="fu">ylab</span>(<span class="st">"f3"</span>)</span>
<span id="cb27-21"><a href="#cb27-21"></a>}</span>
<span id="cb27-22"><a href="#cb27-22"></a><span class="fu">viz_outgroup_f3</span>(f3_aadr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Can you guess what was the Unknown population?</strong></p>
</div><div class="column" style="width:47%;">
<p><img data-src="assets/f3.png"></p>
</div>
</div>
</section>
<section id="relation-between-f-statistics-and-pca" class="slide level2 center">
<h2>Relation between F-statistics and PCA</h2>

<img data-src="assets/F_stats_PCA.png" class="r-stretch quarto-figure-center"><p><span class="citation" data-cites="Peter_2022">(<a href="" role="doc-biblioref" onclick="return false;">Peter, 2022</a>)</span></p>
</section>
<section id="f4-statistic-d-statistic-abba-baba-test" class="slide level2 center">
<h2>F4-statistic, D-statistic: ABBA-BABA test</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/D_stat_tree.png"></p>
</figure>
</div>
</div><div class="column" style="width:30%;">
<ul>
<li><p>Four populations considered: Ape is an outgroup, Europeans and Africans are tested for their gene flow with Neanderthals</p></li>
<li><p>A – ancestral allele, B – derived allele</p></li>
<li><p>Count number of sites corresponding to ABBA and BABA situations</p></li>
</ul>
<p><span class="math display">\[D=\frac{N_{ABBA}-N_{BABA}}{N_{ABBA}+N_{BABA}}\]</span></p>
<ul>
<li><p><span class="math inline">\(N_{ABBA}\)</span> the total counts of ABBA patterns, <span class="math inline">\(N_{BABA}\)</span> the total counts of BABA patterns</p></li>
<li><p>Exess of shared derived alleles between Europeans and Neanderthals indicates a gene flow between them</p></li>
<li><p>Positive values of D imply Neanderthals are closer to Europeans</p></li>
<li><p>Negative values of D imply Neanderthals are closer to Africans</p></li>
</ul>
</div>
</div>
</section>
<section id="example-of-d-statistic-computation-from-1000g-project" class="slide level2 center">
<h2>Example of D-statistic computation from 1000G project</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>We will use ANGSD <span class="citation" data-cites="Korneliussen_2014">(<a href="" role="doc-biblioref" onclick="return false;">Korneliussen T, 2014</a>)</span> and compute D-statistic in irder to test potential Neanderthal introgression to European (CEU) and African (YRI) populations from the 1000G project.</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-22_ff8f0514ce4af98514fd3a9cee6dbbde">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource bash my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1"></a><span class="bu">echo</span> Neandertal.bam <span class="op">&gt;</span> bams</span>
<span id="cb28-2"><a href="#cb28-2"></a><span class="fu">find</span> <span class="va">$PWD</span> <span class="at">-name</span> <span class="st">'*YRI*.bam'</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-10</span> <span class="op">&gt;&gt;</span> bams</span>
<span id="cb28-3"><a href="#cb28-3"></a><span class="fu">find</span> <span class="va">$PWD</span> <span class="at">-name</span> <span class="st">'*CEU*.bam'</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-10</span> <span class="op">&gt;&gt;</span> bams</span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="ex">angsd</span> <span class="at">-out</span> out <span class="at">-doAbbababa</span> 1 <span class="at">-bam</span> bams <span class="at">-doCounts</span> 1 <span class="at">-anc</span> anc.fa</span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="ex">Rscript</span> jackKnife.R file=out.abbababa indNames=bams outfile=out_abbababa_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and, finally, we will display the resulting table of D-statistics and plot the ABBA and BABA counts using custom R scripts:</p>
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-24_7f6f1a6787bbc41348714b9ae5d96ee8">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">file.path</span>(path, <span class="st">"out_abbababa_results.txt"</span>), <span class="at">header =</span> T, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb29-2"><a href="#cb29-2"></a>df <span class="ot">&lt;-</span> df[<span class="fu">grepl</span>(<span class="st">"Neandertal"</span>, <span class="fu">as.character</span>(df<span class="sc">$</span>H3)), ]</span>
<span id="cb29-3"><a href="#cb29-3"></a>df <span class="ot">&lt;-</span> df[<span class="fu">grepl</span>(<span class="st">"CEU"</span>, <span class="fu">as.character</span>(df<span class="sc">$</span>H2)), ]</span>
<span id="cb29-4"><a href="#cb29-4"></a>df <span class="ot">&lt;-</span> df[<span class="fu">grepl</span>(<span class="st">"YRI"</span>, <span class="fu">as.character</span>(df<span class="sc">$</span>H1)), ]</span>
<span id="cb29-5"><a href="#cb29-5"></a>YRI <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">unlist</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(df<span class="sc">$</span>H1), <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>)), <span class="at">ncol =</span> <span class="dv">9</span>, <span class="at">byrow =</span> T)</span>
<span id="cb29-6"><a href="#cb29-6"></a>df<span class="sc">$</span>H1 <span class="ot">&lt;-</span> <span class="fu">paste</span>(YRI[, <span class="dv">6</span>], YRI[, <span class="dv">2</span>], <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb29-7"><a href="#cb29-7"></a>CEU <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">unlist</span>(<span class="fu">strsplit</span>(<span class="fu">as.character</span>(df<span class="sc">$</span>H2), <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>)), <span class="at">ncol =</span> <span class="dv">9</span>, <span class="at">byrow =</span> T)</span>
<span id="cb29-8"><a href="#cb29-8"></a>df<span class="sc">$</span>H2 <span class="ot">&lt;-</span> <span class="fu">paste</span>(CEU[, <span class="dv">6</span>], CEU[, <span class="dv">2</span>], <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb29-9"><a href="#cb29-9"></a>df<span class="sc">$</span>H3 <span class="ot">&lt;-</span> <span class="st">"NEAND"</span></span>
<span id="cb29-10"><a href="#cb29-10"></a>df[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            H1          H2    H3 nABBA nBABA       Dstat     jackEst        SE
11 YRI_NA19102 CEU_NA12342 NEAND    51    46  0.05154639  0.05154639 0.1220787
12 YRI_NA19213 CEU_NA12342 NEAND    38    31  0.10144930  0.10144930 0.1286760
13 YRI_NA18504 CEU_NA12342 NEAND    50    53 -0.02912621 -0.02912621 0.1353893
14 YRI_NA19108 CEU_NA12342 NEAND    38    42 -0.05000000 -0.05000000 0.1393287
15 YRI_NA19238 CEU_NA12342 NEAND    44    38  0.07317073  0.07317073 0.1305890
16 YRI_NA19102 CEU_NA12348 NEAND    46    45  0.01098901  0.01098901 0.1076116
17 YRI_NA19213 CEU_NA12348 NEAND    37    34  0.04225352  0.04225352 0.1122396
18 YRI_NA18504 CEU_NA12348 NEAND    42    49 -0.07692308 -0.07692308 0.1068372</code></pre>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell my_code_size_abitsmaller" data-hash="index_cache/revealjs/unnamed-chunk-26_ef0fcc7cc11950ea39e93eca64ee3f80">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="fu">boxplot</span>(df<span class="sc">$</span>nABBA, df<span class="sc">$</span>nBABA, <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"ABBA"</span>, <span class="st">"BABA"</span>), <span class="at">ylab =</span> <span class="st">"COUNTS"</span>, <span class="at">col =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/unnamed-chunk-26-1.svg" width="960"></p>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r my_code_size_abitsmaller number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="fu">wilcox.test</span>(df<span class="sc">$</span>nABBA, df<span class="sc">$</span>nBABA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  df$nABBA and df$nBABA
W = 367, p-value = 0.2933
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="refs" class="slide level2 unnumbered unlisted center">
<h2>Refs</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Lawson_2012" class="csl-entry" role="listitem">
D. Lawson, S. M., G. Hellenthal. (2012). Inference of population structure using dense haplotype data. <em>PLoS Genet. 8(1):e1002453</em>. <a href="https://doi.org/doi: 10.1371/journal.pgen.1002453">https://doi.org/doi: 10.1371/journal.pgen.1002453</a>
</div>
<div id="ref-Korneliussen_2014" class="csl-entry" role="listitem">
Korneliussen T, N. R., Albrechtsen A. (2014). ANGSD: Analysis of next generation sequencing data. <em>BMC Bioinformatics 15(1):356</em>. <a href="https://doi.org/doi: 10.1186/s12859-014-0356-4">https://doi.org/doi: 10.1186/s12859-014-0356-4</a>
</div>
<div id="ref-Lawson_2018" class="csl-entry" role="listitem">
Lawson DJ, F. D., van Dorp L. (2018). A tutorial on how not to over-interpret STRUCTURE and ADMIXTURE bar plots. <em>Nat Commun. 14;9(1):3258</em>. <a href="https://doi.org/doi: 10.1038/s41467-018-05257-7">https://doi.org/doi: 10.1038/s41467-018-05257-7</a>
</div>
<div id="ref-McVean_2009" class="csl-entry" role="listitem">
McVean. (2009). A genealogical interpretation of principal components analysis. <em>PLoS Genetics Oct;5(10):e1000686</em>. <a href="https://doi.org/doi: 10.1371/journal.pgen.1000686">https://doi.org/doi: 10.1371/journal.pgen.1000686</a>
</div>
<div id="ref-Cavalli_Sforza_1978" class="csl-entry" role="listitem">
Menozzi P, C.-S. L., Piazza A. (1978). Synthetic maps of human gene frequencies in europeans. <em>Science 1;201(4358):786-92</em>. <a href="https://doi.org/doi: 10.1126/science.356262">https://doi.org/doi: 10.1126/science.356262</a>
</div>
<div id="ref-Novembre_2008" class="csl-entry" role="listitem">
Novembre J, B. K., Johnson T. (2008). Genes mirror geography within europe. <em>Nature. 2008 Nov 6;456(7218):98-101</em>. <a href="https://doi.org/doi: 10.1038/nature07331">https://doi.org/doi: 10.1038/nature07331</a>
</div>
<div id="ref-Peter_2016" class="csl-entry" role="listitem">
Peter, B. (2016). Admixture, population structure, and f-statistics. <em>Genetics 202(4):1485-501</em>. <a href="https://doi.org/doi: 10.1534/genetics.115.183913">https://doi.org/doi: 10.1534/genetics.115.183913</a>
</div>
<div id="ref-Peter_2022" class="csl-entry" role="listitem">
Peter, B. (2022). A geometric relationship of F2, F3 and F4-statistics with principal component analysis. <em>Philos Trans R Soc Lond B Biol Sci. 377(1852):20200413</em>. <a href="https://doi.org/doi: 10.1098/rstb.2020.0413">https://doi.org/doi: 10.1098/rstb.2020.0413</a>
</div>
<div id="ref-Schraiber_2015" class="csl-entry" role="listitem">
Schraiber, J., &amp; Akey, J. (2015). Methods and models for unravelling human evolutionary history. <em>Nat Rev Genet. 16(12):727-40</em>. <a href="https://doi.org/doi: 10.1038/nrg4005">https://doi.org/doi: 10.1038/nrg4005</a>
</div>
</div>

<div class="footer footer-default">
<p>Population structure</p>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-pointer/pointer.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'pointer': {"key":"q","color":"red","pointerSize":8,"alwaysVisible":false},
'smaller': true,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'h.v',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: true,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'default',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: true,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, RevealPointer, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script type="text/javascript">
      function hyperlink_logo() {
        var button = document.createElement( 'div' );
        button.className = "logo-button";
        button.id = "logo-button";
        button.style.visibility = "visible";
        button.style.position = "absolute";
        button.style.zIndex = 30;
        button.style.fontSize = "24px";
        button.style.left = "130px";
        button.style.bottom = "14px";
        button.style.top = "auto";
        button.style.right = "auto";
        if (window.location.host.startsWith("localhost")) {
          button.innerHTML = `<a href="http://${window.location.host}/slides" title="Slides listing"><i class="fa fa-home"></i></a>`
        } else {
          button.innerHTML = '<a href="https://nbisweden.github.io/workshop-pgip/slides" title="Slides listing"><i class="fa fa-home"></i></a>';
        }
        document.querySelector( ".reveal" ).appendChild( button );
      }
    window.document.addEventListener("DOMContentLoaded", function (event) {
      hyperlink_logo();
    });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>