---
format: html
---

# Data {#sec-exercise-data}

```{python}
#| label: link-data
#| echo: false
#| eval: true
import os
if not os.path.exists("data"):
    os.symlink("../data", "data")
```

::: {.callout-note}

Datasets are presumably of three types:

1. simulated data
2. experimental data
3. precomputed results

Datasets 1 and 2 will be used for analyses. Datasets of type 3 are
results that take a long time to compute (e.g. on a HPC) and will be
needed to enable compilation of jupyter book examples

:::

::: {.callout-note}

This document describes data sets used to generate slides, exercise
results and examples in general.

:::

All data reside in a dedicated data repository at
<https://github.com/NBISweden/pgip-data>. Datasets consist of genome
sequences, bam files, variant call files, and tree sequences, among
other things. Data generation is discussed briefly below. See the
[pgip-data README](https://github.com/NBISweden/pgip-data/README.md)
for more information.

## Datasets used on website

Simulated datasets reside in the `data` subdirectory
(@tbl-simulated-datasets).

| Dataset       | Description                  | Location             |
|---------------|:-----------------------------|:---------------------|
| ooa           | Out of Africa                | `data/ooa`           |
| ooa-outgroups | Out of Africa with outgroups | `data/ooa-outgroups` |
| sweep         | Selective sweep data         | `results/slim`       |

: Simulated dataset {#tbl-simulated-datasets}

## Simulated SLiM recipes

[SLiM](https://messerlab.org/slim/) recipes reside in
`docs/recipes/slim`. The output can be generated with the helper
script `pgip-slim` from the software repo
[pgip-tools](https://github.com/NBISweden/pgip-tools).

## Simulated demographic models {#sec-simulated-demographic-models}

::: {.callout-warning}

## FIXME

Desired improvements:

- base mutations are random; model realistic ti/tv ratios better

:::

Demographic models are defined using the [Demes
specification](https://popsim-consortium.github.io/demes-spec-docs/main/introduction.html).
For neutral models, genealogies and mutations are [simulated with
msprime](https://tskit.dev/msprime/docs/stable/ancestry.html) and
saved as [succinct tree sequences](https://tskit.dev/learn.html#what)
that can be converted to standard variant call format (vcf) files.

The variant data represents the true state of the model. In addition,
we can assign haplotypes to each individual from which sequence reads
can be simulated. One simulated individual is chosen as reference and
a reference sequence is generated, either from experimental data or
randomly, for that individual. Given the reference and the variant
data, haplotype sequences are generated for each individual in the
model. The haplotype sequences are then used as templates to generate
sequencing reads with realistic error profiles. Currently
[InSilicoSeq](https://insilicoseq.readthedocs.io/en/latest/index.html)
is used to simulate reads.

Simulated reads are mapped with bwa to the reference and saved to bam
files. The raw data reads and haplotype sequences are finally
discarded to save space. Variant calling is done with GATK.

### Out of Africa (ooa) {#sec-data-simulated-ooa}

The out of Africa dataset is based on the
@gutenkunst_InferringJointDemographic_2009 demographic model
(@fig-ooa-demesdraw; see also
[Population tree example](https://tskit.dev/msprime/docs/stable/demography.html#population-tree)
and [the stdpopsim out of Africa model
specification](https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec_catalog_homsap_models_outofafrica_3g09)).

<!-- markdownlint-disable MD013 -->

```{python }
#| label: fig-ooa-demesdraw
#| echo: false
#| eval: true
#| fig-cap: "[demesdraw](https://grahamgower.github.io/demesdraw/latest/quickstart.html) illustration of out of Africa model."
import demes
import demesdraw
ooa = demes.load("data/data/ooa/ooa.demes.yaml")
ax = demesdraw.tubes(ooa, num_lines_per_migration=6, seed=32)
```

<!-- markdownlint-enable MD013 -->

Note that this dataset is based on a subset of the data generated in
@sec-data-simulated-ooa-outgroups.

### Out of Africa with outgroups (ooa-outgroups) {#sec-data-simulated-ooa-outgroups}

The out of Africa with outgroups demographic model is an extension of
the previous data set (@fig-ooa-with-outgroups-demesdraw). In addition
to simulating human data, the model has been expanded to include three
outgroups chimpanzee, gorilla and orangutan.

<!-- markdownlint-disable MD013 -->

```{python }
#| label: fig-ooa-with-outgroups-demesdraw
#| echo: false
#| eval: true
#| fig-cap: "[demesdraw](https://grahamgower.github.io/demesdraw/latest/quickstart.html) illustration of out of Africa model, including outgroups chimpanzee, gorilla, and orangutan. Note that the time scale is in log units."
import demes
import demesdraw
ooa = demes.load("data/data/ooa-outgroups/ooa_with_outgroups.demes.yaml")
ax = demesdraw.tubes(ooa, num_lines_per_migration=6, log_time=True, seed=32)
```

<!-- markdownlint-enable MD013 -->

The data generation has been implemented in a
[Snakemake](https://snakemake.readthedocs.io/en/stable/) workflow. See
the rulegraph  @fig-snakemake-rulegraph for an overview of the steps
to generate data.

![Snakemake rulegraph](data/data/ooa-outgroups/rulegraph.png){#fig-snakemake-rulegraph}
