---
format: html
---

<!-- markdownlint-disable MD041 -->

{{< include ../_knitr.qmd >}}

{{< include ../_rlibs.qmd >}}

<!-- markdownlint-enable MD041 -->

# Variant calling {#sec-variantcalling}

In this exercise we will perform the basic steps in variant calling.

## Best practice pipeline

Briefly describe best practice pipeline (e.g. GATK, bcftools,
freebayes), highlighting some of the problems with natural
populations:

- many workflows are human-centric and don't scale to non-model
  organisms (e.g. chromosome numbers in GATK)

Workflows begin with mapping; do this as common step. Then, move on to
variant calling and present three alternatives to highlight differences:

1. GATK (human-centric; highlight problems)
2. freebayes
3. ANGSD - especially for low-coverage, however need to work with
   genotype likelihoods

Make 2 and 3 optional.

## Read mapping

### Duplicate removal

Picard MarkDuplicates or bamUtil dedup (samtools rmdup deprecated)

## Variant calling

### GATK workflow

[@hansen_VariantCallingNext_2016]

### Freebayes (optional)

### ANGSD (optional)

# References
