<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Per Unneberg">
  <title>DDLS Population genomics in practice - Variant calling</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-pointer/pointer.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
<script type="text/x-mathjax-config">MathJax.Hub.Config({tex2jax: {enableAssistiveMml: false}});</script>
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/logos/nbis-scilifelab.svg" data-background-position="top left" data-background-size="500px" class="center">
  <h1 class="title">Variant calling</h1>
  <h3 class="subtitle">From sequence data to variant call set</h3>
  <h4 class="author">Per Unneberg</h4>
  <h4 class="institute">NBIS</h4>
  <h4 class="date">08-Nov-2023</h4>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Table of contents</h2>
<ul>
<li><a href="#/sequencing-dna" id="/toc-sequencing-dna">Sequencing DNA</a></li>
<li><a href="#/read-mapping" id="/toc-read-mapping">Read mapping</a></li>
<li><a href="#/variant-calling-and-genotyping" id="/toc-variant-calling-and-genotyping">Variant calling and genotyping</a></li>
<li><a href="#/exercise-on-variant-calling" id="/toc-exercise-on-variant-calling">Exercise on variant calling</a></li>
</ul>
</nav>
</section>

<section id="yesterday" class="slide level2">
<h2>Yesterday</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="dna smallr">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>A</strong></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
</tr>
</tbody>
</table>
</div>
<div class="fragment smallr">
<p><span class="math display">\[
\begin{align}
\pi &amp; = \sum_{j=1}^S h_j = \sum_{j=1}^{S} \frac{n}{n-1}\left(1 - \sum_i p_i^2 \right) \\
&amp; \stackrel{S=6,\\ n=4}{=} \sum_{j=1}^{6} \frac{4}{3}\left(1 - \sum_i p_i^2\right) \\
&amp; = \frac{4}{3}\left(\mathbf{\color{#a7c947}{4}}\left(1-\frac{1}{16}-
\frac{9}{16}\right) + \mathbf{\color{#a7c947}{2}}\left(1 - \frac{1}{4} -
\frac{1}{4}\right)\right) = \frac{10}{3}
\end{align}
\]</span></p>
</div>
</div><div class="column fragment" style="width:20%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/coalescent-tree-dna-variation-1.svg" style="width:80.0%"></p>
</div>
</div>
</div><div class="column fragment" style="width:30%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/site-frequency-spectrum-graph-1.svg" style="width:80.0%"></p>
</div>
</div>
<div class="fragment smallr">
<p><span class="math display">\[
\begin{align}
\pi &amp; = \frac{\sum_{i=1}^{n-1}i(n-i)\xi_i}{n(n-1)/2} \\
&amp; \stackrel{n=4}{=} \frac{1*(4-1)*4 + 2*(4-2)*2}{6} = \frac{10}{3}
\end{align}
\]</span></p>
</div>
</div>
</div>
<div class="fragment">
<p>This is not how real data looks like from the beginning…</p>
</div>
<aside class="notes">
<p>We have previously looked at how to simulate sequence data and performed some simple calculations on data that are “well-behaved” in some sense; textbook data is usually clean without missing data points or at least high quality in some sense. This lecture will focus on how data is generated through the process of <em>variant calling</em>, starting from a set of reads and a reference sequence (it does <em>not</em> cover sequencing technologies or assembly).</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-real-data" class="slide level2">
<h2>The real data</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>@SRR9309790.10003134
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJFJJJJJJJJJJJJJJJJJ
@SRR9309790.10003222
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJ</code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461           </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTG</span><br><span class="tviewcu4">...............M...............................</span><br><span class="tviewc4">...............A...</span><span>                            </span><br><span class="tviewc4">...............A.................</span><span>              </span><br><span class="tviewc4">..............................................</span><span> </span><br><span class="tviewc4">..............................................</span><span> </span><br><span class="tviewc4">..............................................</span><span> </span><br><span class="tviewc4">...............A...............................</span><br><span class="tviewc4">...............A...............................</span><br><span class="tviewc4">,,,,,,,,,,,,a,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,a,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,a,,,,,aa,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span></pre>
</div>


</div>
</div>
<aside class="notes">
<p>Before getting to variants and genotypes a lot of processing has to be done, from FASTQ input, to mapped data, to variant and genotype calls.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-process" class="slide level2">
<h2>The process</h2>
<div>
<p><img data-src="../foundations/assets/images/popgen.svg"></p>
</div>
<aside class="notes">
<p>In short: going from sample collection through data processing to variant data is a long and winding road. Here we will focus on the data processing part that generates a raw variant call set.</p>
<p>The end result is a table consisting of variant positions, and for each individual, information about genetic variation.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="variant-and-genotype-calling" class="slide level2">
<h2>Variant and genotype calling</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fnrg2986/MediaObjects/41576_2011_Article_BFnrg2986_Fig1_HTML.jpg?as=webp" style="width:60.0%"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<div class="flushright smallest translatey50">
<p><span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">Nielsen et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2011</a>)</span></p>
</div>
</div><div class="column" style="width:50%;">
<dl>
<dt>SNP calling</dt>
<dd>
Identification of polymorphic sites (<span class="math inline">\(&gt;1\)</span>% allele frequency)
</dd>
<dt>Variant calling</dt>
<dd>
Identification of variant sites (sufficient that <strong>any</strong> allele differs)
</dd>
<dt>Genotype calling</dt>
<dd>
Determine the allele combination for each individual (aa, aA, or AA for bi-allelic variants)
</dd>
</dl>
<div class="fragment">
<p>Knowing variant sites informs us of possible genotypes and improves genotyping.</p>
<p>Example: knowing a site has <code>A</code> or <code>C</code> limits possible genotype calls to <code>AA</code>, <code>AC</code>, or <code>CC</code></p>
</div>
</div>
</div>
<aside class="notes">
<p>Figure caption from <span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen et al., 2011</a>)</span>:</p>
<blockquote>
<p>Pre-processing steps (shown in yellow) transform the raw data from next-generation sequencing technology into a set of aligned reads that have a measure of confidence, or quality score, associated with the bases of each read. The per-base quality scores produced by base-calling algorithms may need to be recalibrated to accurately reflect the true error rates. Depending on the number of samples and the depth of coverage, either a multi-sample calling procedure (green) or a single-sample calling procedure (orange) may then be applied to obtain SNP or genotype calls and associated quality scores. Note that the multi-sample procedure may include a linkage-based analysis, which can substantially improve the accuracy of SNP or genotype calls. Finally, post-processing (purple) uses both known data and simple heuristics to filter the set of SNPs and/or improve the associated quality scores. Optional, although recommended, steps are shown in dashed lines.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section>
<section id="sequencing-dna" class="title-slide slide level1 center">
<h1>Sequencing DNA</h1>

</section>
<section id="sequencing-technologies" class="slide level2">
<h2>Sequencing technologies</h2>
<div class="columns">
<div class="column" style="width:90%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://ngisweden.scilifelab.se/wp-content/uploads/2023/06/tech3-1024x171.png" style="width:80.0%"></p>
</figure>
</div>
</div><div class="column" data-fig-align="left" style="width:10%;">
<p><img data-src="https://ngisweden.scilifelab.se/wp-content/themes/ngisweden/img/NGI-logo.svg" style="width:100.0%"></p>
</div>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<h5 id="illumina-novaseq-600">Illumina NovaSeq 600</h5>
<div class="smallr">
<blockquote>
<p>Scale up and down with a tunable output of up to 6 Tb and 20B single reads in &lt; 2 days.</p>
</blockquote>
</div>
<p>Up to 2X250 bp read length. Price example: 8,000 SEK total for resequencing 3Gbp genome to 30X</p>
<div class="flushright smallest">
<p><a href="https://www.illumina.com/systems/sequencing-platforms/novaseq.html" class="uri">https://www.illumina.com/systems/sequencing-platforms/novaseq.html</a></p>
</div>
</div><div class="column" style="width:50%;">
<h5 id="pacbio-revio">PacBio Revio</h5>
<div class="smallr">
<blockquote>
<p>Up to 360 Gb of HiFi reads per day, equivalent to 1,300 human whole genomes per year.</p>
</blockquote>
</div>
<p>Tens of kilobases long HiFi reads. Price example (Sequel II): ~35kSEK per library and SMRT cell</p>
<div class="flushright smallest">
<p><a href="https://www.pacb.com/revio/" class="uri">https://www.pacb.com/revio/</a></p>
</div>
</div>
</div>
</section>
<section id="dna-sequences-in-fastq-format" class="slide level2 smaller">
<h2>DNA sequences in FASTQ format</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">ls</span> <span class="at">--long</span> <span class="at">--human</span> <span class="pp">*</span>.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>-rw-r--r-- 1 runner docker 302K Nov  8 14:14 PUN-Y-INJ_R1.fastq.gz
-rw-r--r-- 1 runner docker 393K Nov  8 14:14 PUN-Y-INJ_R2.fastq.gz</code></pre>
</div>
</div>
<div class="fragment">
<p>Count number of lines:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="fu">zcat</span> PUN-Y-INJ_R1.fastq.gz <span class="kw">|</span> <span class="fu">wc</span> <span class="at">--lines</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>15768</code></pre>
</div>
</div>
</div>
<div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:30%;">
<p>Format:</p>
<ol type="1">
<li>sequence id (prefixed by <code>@</code>)</li>
<li>DNA sequence</li>
<li>separator (<code>+</code>)</li>
<li>Phred base quality scores</li>
</ol>
</div><div class="column" style="width:70%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">zcat</span> PUN-Y-INJ_R1.fastq.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 8 <span class="kw">|</span> <span class="fu">cut</span> <span class="at">--characters</span> <span class="at">-30</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@SRR9309790.10003134
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJFJJJJJJJJJJJJJJJJJ
@SRR9309790.10003222
TAAATCGATTCGTTTTTGCTATCTTCGTCT
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJ</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>First data encounter is usually FASTQ files from sequencing center. Files contain <strong>sequence reads</strong> that are generated by the sequencing machine. They correspond to short stretches of DNA that have been sequentially read by the machine, resulting in a string of <strong>base calls</strong> that constitute the reads. Every DNA base has an associated so-called <strong>Phred-scaled quality score</strong> from 0 to 93 that are encoded using ASCII 33 to 126.</p>
<p>The format consists of 1. @ title with id 2. raw sequence 3. +: a separator with optional description 4. the quality scores</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dna-sequence-quality-control" class="slide level2">
<h2>DNA sequence quality control</h2>
<p>Quality values represent the probability <span class="math inline">\(P\)</span> that the call is <em>incorrect</em>. They are coded as <strong>Phred</strong> quality scores <span class="math inline">\(Q\)</span>. Here, <span class="math inline">\(Q=20\)</span> implies 1% probability of error, <span class="math inline">\(Q=30\)</span> 0.1% and so on. Typically you should not rely on quality values below <span class="math inline">\(20\)</span>.</p>
<p><span class="math display">\[
Q = -10 \log_{10} P
\]</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<p>A common way to do QC is with <code>fastqc</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">fastqc</span> <span class="at">--outdir</span> . <span class="at">--extract</span> <span class="pp">*</span>fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<p><img data-src="assets/PUN-Y-INJ_R1_fastqc/Images/per_base_quality.svg" style="width:80.0%"></p>
</div>
</div>
</section>
<section id="sequencing-approaches" class="slide level2">
<h2>Sequencing approaches</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/sequencing-cost-1.svg" style="width:90.0%"></p>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<p>Despite price drop, still need to make choices regarding depth and breadth of sequencing coverage and number of samples.</p>
</div><div class="column fragment" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/images/sequencing_approaches_lou_2021.png" style="width:70.0%"></p>
</figure>
</div>
<div class="flushright smallest translatey50">
<p><span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">Lou et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span></p>
</div>
<p>Here focus on Whole Genome reSequencing (WGS), mostly high-coverage.</p>
</div>
</div>
<aside class="notes">
<p>Despite the fact that sequencing costs have dropped dramatically (left), there still are choices to be made regarding the distribution of costs along 1) sequencing coverage depth, i.e., the mean depth of sequencing 2) sequencing coverage breadth, i.e., whether or not to do targeted or whole-genome resequencing or 3) sample size; how many individuals to sample. Whole-genome resequencing of individuals from populations to sufficient depth (30X) is still very expensive, but often needed to understand mechanisms of adaptation <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lou et al., 2021, p. 5967</a>)</span>. Various protocols have been developed to meet the challenges that cost imposes:</p>
<ol type="1">
<li>RAD-seq, restriction site-associated DNA sequencing, targets regions flanking given restriction sites. Downside: much of genome is missed</li>
<li>pool-seq, pooled sequencing. Cost-effective, but loses information about individuals</li>
<li>lcWGS, low-coverage whole genome sequencing increasing in popularity. Genotyping low coverage is problematic however.</li>
</ol>
<p>Our focus here is WGS (whole-genome resequencing), primarily high-coverage, despite the cost it may incur.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genome-assembly-and-population-resequencing" class="slide level2">
<h2>Genome assembly and population resequencing</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h5>
Genome assembly
</h5>
<p><img data-src="assets/images/oso-genome-assembly.png" style="width:80.0%"></p>
<div class="flushright smallest translatey50">
<p><span class="citation" data-cites="allendorf_PopulationGenomics_2022">Allendorf et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2022</a>)</span></p>
</div>
</div><div class="column fragment" style="width:50%;">
<h5>
Population resequencing
</h5>
<p><img data-src="assets/images/population-resequencing.png"></p>
</div>
</div>
<aside class="notes">
<p>DNA sequences are generated from DNA fragments, often as paired-end reads, or long reads. For population genomics, it is often necessary to generate a <em>reference sequence</em> through <em>genome assembly</em>, the quality of which will impact the types of downstream analyses that can be done. Reads are assembled into <em>contigs</em> (contiguous sequences) that with the aid of genetic maps are pieced together into <em>scaffolds</em>. In the left figure, note that the scaffolds contain areas of unknown sequence.</p>
<p>Once a reference sequence has been produced, (short) reads from individuals are mapped to the reference (right).</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="read-mapping" class="title-slide slide level1 center">
<h1>Read mapping</h1>

</section>
<section id="sequence-alignment-maps-reads-to-a-reference" class="slide level2">
<h2>Sequence alignment maps reads to a reference</h2>
<div>
<div id="fig-read-mapping" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;1: Screenshot of reference sequence (top) and aligned reads (bottom). Second line with <code>.</code> characters is the consensus sequence. Bases are colored by nucleotide. Letter case indicates forward (upper-case) or reverse (lower-case) alignment. <code>*</code> is placeholder for deleted base.</figcaption>
<p><img data-src="assets/images/read-mapping.png" style="width:100.0%"></p>
</figure>
</div>
</div>
<p>Aim of sequence alignment (read mapping) is to determine source in reference sequence. Some commonly used read mappers for resequencing are</p>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>BWA, BWA-MEM <span class="citation" data-cites="li_AligningSequenceReads_2013">(<a href="#/bibliography" role="doc-biblioref" onclick="">H. Li, 2013</a>)</span></li>
<li>Novoalign (<a href="https://www.novocraft.com/" class="uri">https://www.novocraft.com/</a>)</li>
<li>Minimap2 <span class="citation" data-cites="li_Minimap2PairwiseAlignment_2018">(<a href="#/bibliography" role="doc-biblioref" onclick="">H. Li, 2018</a>)</span></li>
</ul>
</div><div class="column" style="width:50%;">
<p>For a recent comprehensive comparison see <span class="citation" data-cites="donato_NewEvaluationMethods_2021">Donato et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2021</a>)</span></p>
</div>
</div>
</section>
<section id="alignments-are-stored-in-bam-format" class="slide level2 smaller">
<h2>Alignments are stored in BAM format</h2>
<div>
<h5>
Header information
</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">samtools</span> view <span class="at">--header-only</span> PUN-Y-INJ.sort.dup.bam <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@HD VN:1.6  SO:coordinate
@SQ SN:LG4  LN:100000
@RG ID:SRR9309790   SM:PUN-Y-INJ    PL:ILLUMINA
@PG ID:bwa  PN:bwa  VN:0.7.17-r1188 CL:bwa mem -R @RG\tID:SRR9309790\tSM:PUN-Y-INJ\tPL:ILLUMINA -p -t 14 -M tiny/M_aurantiacus_v1_splitline_ordered.fasta -</code></pre>
</div>
</div>
<p>Format: metadata record types prefixed with <code>@</code>, e.g., <code>@RG</code> is the <em>read group</em></p>
</div>
<div class="fragment">
<h5>
Alignments
</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="ex">samtools</span> view PUN-Y-INJ.sort.dup.bam <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SRR9309790.7750070  65  LG4 1   39  45S9M1I96M  =   83782   83781   TGAACTATAGTCGATGGGACGAATACCCCCCTGAACTTGCGAAGGGGACAATTACCCCCCTCTGTTATGTTTCAGTCAATTTCATGTTTGATTTTTAGATTTTTAATTAATTATATATTTTTTGCAATTTGTAACCTCTTTAACCTTTATT AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJFFJFJJJJJJJJJJFJJJJJJJJFJJJJJJJJJAJJJJJJJJJJJJJJJJJFA&lt;FFJJ7FF SA:Z:LG4,83018,+,30M2I28M91S,37,5;  MC:Z:12S10M2I123M4S MD:Z:16C28C59   PG:Z:MarkDuplicates RG:Z:SRR9309790 NM:i:3  MQ:i:50 AS:i:88 XS:i:70 ms:i:5185</code></pre>
</div>
</div>
<p>Some important columns: 1:<code>QUERY</code>, 3:<code>REFERENCE</code>, 4:<code>POSITION</code>, 5:<code>MAPQ</code>, 6:<code>CIGAR</code>. The <code>CIGAR</code> string compiles information on the alignment, such as match (<code>M</code>), soft clipping (<code>S</code>), and insertion to reference (<code>I</code>)</p>
</div>
<div class="flushright smallest">
<p>cf <a href="https://samtools.github.io/hts-specs/SAMv1.pdf" class="uri">https://samtools.github.io/hts-specs/SAMv1.pdf</a></p>
</div>
<aside class="notes">
<p>For a complete description, see the specification. Suffice to say that the alignment format consists of a <strong>header</strong> section, with metadata and provenance data, and an <strong>alignment</strong> section, which is a column-based format with information pertaining to the query sequence being mapped and the reference sequence to which the query is mapped.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="mapped-alignments-can-be-viewed-with-samtools-tview" class="slide level2">
<h2>Mapped alignments can be viewed with <code>samtools tview</code></h2>
<div class="columns smallr">
<div class="column" style="width:50%;">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="ex">samtools</span> tview <span class="at">-p</span> LG4:30430 <span class="at">-d</span> H <span class="at">-w</span> 60 <span class="dt">\</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>   PUN-Y-INJ.sort.dup.bam <span class="dt">\</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>   M_aurantiacus_v1.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............M...............................</span><span class="tviewcu2">.</span><span class="tviewcu4">............</span><br><span class="tviewc4">...............A...</span><span>                               </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">...............A.................</span><span>                 </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>      </span><span class="tviewc4">,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,a,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,a,,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,a,,,,,aa,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span></pre>
</div>


</div><div class="column" style="width:50%;">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1"></a><span class="ex">samtools</span> tview <span class="at">-p</span> LG4:30430 <span class="at">-d</span> H <span class="at">-w</span> 60 <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>   PUN-R-ELF.sort.dup.bam <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>   M_aurantiacus_v1.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............A............................................</span><br><span class="tviewc4">,,,,,,,</span><span>                           </span><span class="tviewc4">..........................</span><br><span class="tviewc4">...............A.....</span><span>                  </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A...A....</span><span>               </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A.........</span><span>                                   </span><br><span class="tviewc4">...............A...........C.............</span><span>                   </span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,g,,,,,,,a,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">.............A............................................</span></pre>
</div>


</div>
</div>
<p>Aka <code>pileup</code> format. Forward (<code>.</code>) and backward (<code>,</code>) mapping reads. Mismatches shown as letters.</p>
<aside class="notes">
<p>After preprocessing, reads are mapped to a reference. Observations:</p>
<ol type="1">
<li>different coverage</li>
<li>sequencing error randomly distributed</li>
</ol>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="potential-error-corrections-and-pitfalls" class="slide level2">
<h2>Potential error corrections and pitfalls</h2>
<h4 id="instrument">Instrument</h4>
<ul>
<li>PCR duplicates -&gt; MarkDuplicates</li>
<li>systematic errors from sequencing machine -&gt; Base Quality Score Recalibration (BQSR)</li>
</ul>
<h4 id="reference">Reference</h4>
<ul>
<li>quality of reference sequence!</li>
<li>repetitive sequence</li>
</ul>
</section></section>
<section>
<section id="variant-calling-and-genotyping" class="title-slide slide level1 center">
<h1>Variant calling and genotyping</h1>

</section>
<section id="variants-show-up-in-pileup-alignments" class="slide level2 smaller">
<h2>Variants show up in pileup alignments</h2>
<div class="columns">
<div class="column smallr" style="width:50%;">
<h5>
Sample PUN-Y-INJ
</h5>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............M...............................</span><span class="tviewcu2">.</span><span class="tviewcu4">............</span><br><span class="tviewc4">...............A...</span><span>                               </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">...............A.................</span><span>                 </span><span class="tviewc4">,,,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>      </span><span class="tviewc4">,,,,,,,,</span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">..............................................</span><span>              </span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,a,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,a,,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,a,,,,,aa,a,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span></pre>
</div>


</div><div class="column smallr" style="width:50%;">
<h5>
Sample PUN-R-ELF
</h5>


<title>
LG4:30430
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
LG4:30430
</div>
<pre class="tviewpre"><span> 30431     30441     30451     30461     30471              </span><br><span>CATTGGCAATGGCATCAGTTGAGCATCTTAGTACGAACTAAAAGCTGCGAAAAAATATTT</span><br><span class="tviewcu4">...............A............................................</span><br><span class="tviewc4">,,,,,,,</span><span>                           </span><span class="tviewc4">..........................</span><br><span class="tviewc4">...............A.....</span><span>                  </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A...A....</span><span>               </span><span class="tviewc4">.....................</span><br><span class="tviewc4">...............A.........</span><span>                                   </span><br><span class="tviewc4">...............A...........C.............</span><span>                   </span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span class="tviewc4">...............A............................................</span><br><span class="tviewc4">,,,g,,,,,,,a,,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</span><br><span>  </span><span class="tviewc4">.............A............................................</span></pre>
</div>


</div>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<p>Potential variants show up as multiple mismatches in a column. Two questions arise:</p>
<ul>
<li>how do we detect variant sites?</li>
<li>how do we distinguish variants from sequencing error?</li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Simple approach: filter bases on quality (e.g., Q20), call heterozygous if 20-80% bases non-reference.</p>
<p>Issues: undercalls heterozygotes, no measure of uncertainty</p>
</div>
<div class="fragment">
<p>Solution: probabilistic methods!</p>
</div>
<div class="flushright smallest">
<p><span class="citation" data-cites="nielsen_GenotypeSNPCalling_2011">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen et al., 2011</a>)</span></p>
</div>
</div>
</div>
<aside class="notes">
<p>Potential variants show up as multiple mismatches in a column. Left sample has three reads that match the reference so is probably heterozygote. For the right sample no read matches reference so most likely call is homozygote alternate.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="we-can-calculate-likelihoods-of-observed-data" class="slide level2">
<h2>We can calculate likelihoods of observed data</h2>
<h3 id="example-exluding-sequencing-error">Example (exluding sequencing error!)</h3>
<div class="columns">
<div class="column" style="width:5%;">


<title>
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">

</div>
<pre class="tviewpre"><span>   </span><br><span>TGC</span><br><span class="tviewcu4">.K.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:5%;">

</div><div class="column" style="width:90%;">
<p>Goal: calculate likelihood of observing <span class="math inline">\(X=\)</span><code>G,g,T,T,G,g,t</code> from a genotype <span class="math inline">\(G\)</span>. We assume each observation <span class="math inline">\(X_i\)</span> can be treated independently. We restrict possible genotypes to the observed alleles (i.e., <code>G, T</code>). Some observations:</p>
<div class="fragment">
<p>Prob(<span class="math inline">\(X_1=\)</span><code>G</code> <em>assuming</em> genotype <code>T,T</code>) = P(<span class="math inline">\(X_1=\)</span><code>G</code>|<code>T,T</code>) = <span class="math inline">\(0\)</span></p>
</div>
<div class="fragment">
<p>Prob(<span class="math inline">\(X_1=\)</span><code>G</code> assuming genotype <code>G,G</code>) = P(<span class="math inline">\(X_1=\)</span><code>G</code>|<code>G,G</code>) = <span class="math inline">\(1\)</span></p>
</div>
<div class="fragment">
<p>Prob(<span class="math inline">\(X_1=\)</span><code>G</code> assuming genotype <code>G,T</code>) = P(<span class="math inline">\(X_1=\)</span><code>G</code>|<code>G,T</code>) = <span class="math inline">\(0.5\)</span></p>
</div>
<div class="fragment">
<p>To get total likelihood <span class="math inline">\(P(X|G)\)</span> assuming a genotype <span class="math inline">\(G\)</span> (here <code>G,T</code>), we can multiply over all observations (reads):</p>
<p><span class="math display">\[
\begin{align}
P(X|G,T) &amp; = P(X_1=G|G,T)P(X_2=g|G,T)P(X_3=T|G,T)P(X_4=T|G,T) \\
    &amp; P(X_5=G|G,T)P(X_6=g|G,T)P(X_7=t|G,T) = 0.5^7
\end{align}
\]</span></p>
</div>
</div>
</div>
<aside class="notes">
<p>We restrict the possible genotypes to the observed alleles at a site (here G and T). If there are more than two observed alleles, a common procedure is to pick the two with highest frequencies, under the assumption that the rarest observation is a sequencing error.</p>
<p>In reality, we also need to take into account sequencing error. There are different ways of doing this (e.g. <span class="citation" data-cites="maruki_GenotypeCallingPopulationGenomic_2017">Maruki &amp; Lynch (<a href="#/bibliography" role="doc-biblioref" onclick="">2017</a>)</span>, <span class="citation" data-cites="depristo_FrameworkVariationDiscovery_2011">DePristo et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2011</a>)</span>), but we leave the details to the interested reader.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="we-can-use-bayes-theorem-to-genotype" class="slide level2 smaller">
<h2>We can use Bayes’ theorem to genotype</h2>
<h3 id="example">Example</h3>
<div class="columns">
<div class="column" style="width:5%;">


<title>
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">

</div>
<pre class="tviewpre"><span>   </span><br><span>TGC</span><br><span class="tviewcu4">.K.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:5%;">

</div><div class="column" style="width:90%;">
<p>For a given site, we have a number of observations <span class="math inline">\(X\)</span>. We have shown we can calculate the likelihood of observing <span class="math inline">\(X\)</span> given a genotype <span class="math inline">\(G\)</span>, <span class="math inline">\(P(X|G)\)</span>.</p>
<div class="fragment">
<p>However; what we really want to know is the most likely genotype <span class="math inline">\(G\)</span> <em>given the data</em> <span class="math inline">\(X\)</span>, or <span class="math inline">\(P(G|X)\)</span>.</p>
</div>
<div class="fragment">
<p>Apply Bayes’ theorem:</p>
<div class="large">
<p><span class="math display">\[
P(G|X) \sim P(X|G)\cdot P(G)
\]</span></p>
</div>
</div>
<div class="fragment large">
<p><span class="math display">\[
\text{posterior} \sim \text{likelihood} \cdot \text{prior}
\]</span></p>
</div>
<div class="fragment">
<p>Consequently we need to set a prior on <span class="math inline">\(G\)</span>. If allele frequencies are known, we can constrain the frequencies; for example, if A is known to be low (<span class="math inline">\(\sim1\)</span>%) AA genotype is <em>very</em> unlikely. Otherwise, could set all equal (flat prior).</p>
</div>
</div>
</div>
<aside class="notes">
<p>cf <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890511" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360035890511</a></p>
<p><span class="citation" data-cites="li_SNPDetectionMassively_2009">R. Li et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2009</a>)</span>, Table 1, shows a nice numerical example of one way of setting priors. The authors assume a specific allele, G, in the <em>reference sequence</em>:</p>
<p>–G–</p>
<p>They start by calculating the frequency of <em>haploid</em> genotypes. They first determine <span class="math inline">\(p_G\)</span> by assuming a <em>heterozygous SNP rate</em> <span class="math inline">\(f=0.001\)</span>, which means 1 in a 1000 sites has G/G genotype mutated to G/X, where X is one of {A,C,T}. They assume a transition to transversion (ts/tv) ratio of 4, meaning X=A four times as often as C or T (there is an error in the text where C is taken to be the transition; the numbers in the table are correct however). This gives the following haploid genotype frequencies:</p>
<p><span class="math display">\[\begin{align}
p_G &amp; = 1-f = 0.999 \\
p_A &amp; = 4f/6 = 6.67\times10^{-4} \\
p_C &amp; = f/6 = 1.67\times10^{-4}\\
p_T &amp; = f/6 =1.67\times10^{-4}
\end{align}\]</span></p>
<p>To get the <em>diploid</em> genotypes, we simply multiply the corresponding entries, e.g., <span class="math inline">\(p_{AC} = p_Ap_C\)</span>. For homozygote ALT, we need to account for the <em>homozygous SNP rate</em> <span class="math inline">\(r = 0.0005\)</span>, where G/G mutates to X/X, for X one of {A,C,T}:</p>
<p><span class="math display">\[\begin{align}
p_{AA} &amp; = p_Ap_A + 4r/6 = 3.33\times10^{-4} \\
p_{AC} &amp; = p_Ap_C = 1.11\times10^{-7}\\
p_{AT} &amp; = p_Ap_T = 1.11\times10^{-7}\\
p_{CC} &amp; = p_Cp_C + r/6 = 8.34\times10^{-5} \\
p_{CG} &amp; = p_Cp_G = 1.67\times10^{-4}\\
p_{CT} &amp; = p_Cp_T = 2.78\times10^{-8}\\
p_{GG} &amp; = 1 - f - r = 0.9985 \\
p_{GT} &amp; = p_Gp_T = 1.67\times10^{-4}\\
p_{TT} &amp; = p_Tp_T + r = 8.34\times10^{-5}\\
\end{align}\]</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genotype-likelihoods" class="slide level2">
<h2>Genotype likelihoods</h2>
<p>We have outlined a probabilistic approach to variant calling where we obtain a posterior probability of observing a genotype <span class="math inline">\(G\)</span> given data <span class="math inline">\(X\)</span>:</p>
<p><span class="math display">\[
P(G|X) \sim P(X|G)P(G)
\]</span></p>
<div class="fragment">
<p>Assuming a bi-allelic site, and letting <span class="math inline">\(H_1, H_2\)</span> denote the two alleles, we have three possible genotype likelihoods <span class="math inline">\(P(H_1H_1|X)\)</span>, <span class="math inline">\(P(H_1H_2|X)\)</span>, and <span class="math inline">\(P(H_2H_2|X)\)</span>.</p>
</div>
<div class="fragment">
<p>The <em>highest</em> posterior probability is typically chosen as the genotype call, with a measure confidence represented by the genotype probability or ratio between the two most probable calls.</p>
</div>
<div class="fragment">
<p>Genotype likelihoods are often represented as Phred-scaled likelihoods (again!):</p>
<p><span class="math display">\[
\text{QUAL} = -10 \log_{10} P(G|X)
\]</span></p>
</div>
</section>
<section id="variant-call-format-vcf---header" class="slide level2 smaller">
<h2>Variant Call Format (VCF) - header</h2>
<div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> allsites.vcf.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">--lines</span> 1</span>
<span id="cb15-2"><a href="#cb15-2"></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##FILTER"</span></span>
<span id="cb15-3"><a href="#cb15-3"></a><span class="ex">bcftools</span> view <span class="at">-h</span> allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##INFO"</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 4</span>
<span id="cb15-4"><a href="#cb15-4"></a><span class="ex">bcftools</span> view <span class="at">-h</span> allsites.vcf.gz <span class="kw">|</span> <span class="fu">grep</span> <span class="st">"##FORMAT"</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>##fileformat=VCFv4.2
##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;
##FILTER=&lt;ID=LowQual,Description="Low quality"&gt;
##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes, for each ALT allele, in the same order as listed"&gt;
##INFO=&lt;ID=AF,Number=A,Type=Float,Description="Allele Frequency, for each ALT allele, in the same order as listed"&gt;
##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;
##INFO=&lt;ID=BaseQRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities"&gt;
##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed"&gt;
##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)"&gt;
##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description="Genotype Quality"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##FORMAT=&lt;ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block"&gt;
##FORMAT=&lt;ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another; will always be heterozygous and is not intended to describe called alleles"&gt;
##FORMAT=&lt;ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group"&gt;
##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification"&gt;</code></pre>
</div>
</div>
<p><code>FILTER</code> defines applied filters , <code>INFO</code> fields provide additional information to genotypes, <code>FORMAT</code> specification fields define genotype entries, and more. NB: <code>PL</code> format definition.</p>
</div>
<div class="flushright smallr">
<p><a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf" class="uri">https://samtools.github.io/hts-specs/VCFv4.4.pdf</a></p>
</div>
</section>
<section id="variant-call-format-vcf---data" class="slide level2">
<h2>Variant Call Format (VCF) - data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1"></a><span class="ex">bcftools</span> view <span class="at">--header-only</span> <span class="at">--samples</span> PUN-R-ELF,PUN-Y-INJ allsites.vcf.gz <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2"></a> <span class="fu">tail</span> <span class="at">--lines</span> 1</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="ex">bcftools</span> view <span class="at">--no-header</span>  <span class="at">--samples</span> PUN-R-ELF,PUN-Y-INJ allsites.vcf.gz LG4:6886</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  PUN-R-ELF   PUN-Y-INJ
LG4 6886    .   C   T   722.43  .   AC=2;AF=0.222;AN=4;BaseQRankSum=0;DP=82;ExcessHet=1.8123;FS=1.309;InbreedingCoeff=-0.155;MLEAC=4;MLEAF=0.222;MQ=60;MQRankSum=0;QD=18.52;ReadPosRankSum=0.577;SOR=0.44   GT:AD:DP:GQ:PGT:PID:PL:PS   0/1:3,8:11:50:.:.:189,0,50:.    0/1:2,2:4:45:.:.:45,0,45:.</code></pre>
</div>
</div>
<div>
<p><code>QUAL</code>: Phred-scaled quality score for Prob(ALT is wrong): <span class="math inline">\(722.43\)</span> (<span class="math inline">\(p=10^{-Q/10}=5.7e-73\)</span>)</p>
<p><code>INFO</code> field summarizes data for all samples. For instance:</p>
<ul>
<li>allele count 2 (<code>AC=2</code>)</li>
<li>allele frequency minor allele 0.222 (<code>AF=0.222</code>)</li>
</ul>
</div>
<div class="flushright smallest">
<p><a href="https://samtools.github.io/hts-specs/VCFv4.4.pdf" class="uri">https://samtools.github.io/hts-specs/VCFv4.4.pdf</a></p>
</div>
</section>
<section id="variant-call-format-vcf---data-1" class="slide level2 smaller">
<h2>Variant Call Format (VCF) - data</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  PUN-R-ELF   PUN-Y-INJ
LG4 6886    .   C   T   722.43  .   AC=2;AF=0.222;AN=4;BaseQRankSum=0;DP=82;ExcessHet=1.8123;FS=1.309;InbreedingCoeff=-0.155;MLEAC=4;MLEAF=0.222;MQ=60;MQRankSum=0;QD=18.52;ReadPosRankSum=0.577;SOR=0.44   GT:AD:DP:GQ:PGT:PID:PL:PS   0/1:3,8:11:50:.:.:189,0,50:.    0/1:2,2:4:45:.:.:45,0,45:.</code></pre>
</div>
</div>
<div class="columns">
<div class="column" style="width:80%;">
<h5>
Genotypes (<code>GT:AD:DP:GQ:PGT:PID:PL:PS</code>)
</h5>
<p>PUN-R-ELF: <code>0/1:3,8:11:50:.:.:189,0,50:.</code></p>
<p><code>GT=0/1</code>, <code>AD=3,8</code> =&gt; 3 REF, 8 ALT, <code>DP=11</code> =&gt; sequence depth = 11, <code>PL=189,0,50</code></p>
<p>PUN-Y-INJ: <code>0/1:2,2:4:45:.:.:45,0,45:.</code></p>
<p><code>GT=0/1</code>, <code>AD=2,2</code> =&gt; 2 REF, 2 ALT, <code>DP=4</code> =&gt; sequence depth = 4, <code>PL=45,0,45</code></p>
<div class="fragment" data-fragment-index="2">
<h5>
Relative genotype probabilities
</h5>
<p>Can convert Phred-scaled quality scores to probabilities as</p>
<p><span class="math display">\[
p = 10^{-Q/10}
\]</span></p>
<p>For PUN-R-ELF the relative probabilities are <span class="math inline">\(10^{-189/10}\approx1.26e-9\)</span>, <span class="math inline">\(10^{0}=1\)</span>, <span class="math inline">\(10^{50}=10^{-5}\)</span>.</p>
<p>Interpretation: <code>0/1</code> 10,000 times more likely than <code>1/1</code> (<span class="math inline">\(1/10^{-5}\)</span>)</p>
</div>
</div><div class="column" style="width:3%;">

</div><div class="column" style="width:4%;">


<title>
ELF
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
ELF
</div>
<pre class="tviewpre"><span>   </span><br><span>TCT</span><br><span class="tviewcu4">.Y.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">...</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span><br><span class="tviewc4">,t,</span></pre>
</div>


</div><div class="column" style="width:4%;">


<title>
INJ
</title>
<style type="text/css">
.tviewbody { margin:5px; background-color:white;text-align:center;}
.tviewtitle {text-align:center;}
.tviewpre { margin:5px; background-color:white;}
.tviewc0 {color:black;}
.tviewcu0 {color:black;text-decoration:underline;}
.tviewc1 {color:blue;}
.tviewcu1 {color:blue;text-decoration:underline;}
.tviewc2 {color:green;}
.tviewcu2 {color:green;text-decoration:underline;}
.tviewc3 {color:yellow;}
.tviewcu3 {color:yellow;text-decoration:underline;}
.tviewc4 {color:black;}
.tviewcu4 {color:black;text-decoration:underline;}
.tviewc5 {color:green;}
.tviewcu5 {color:green;text-decoration:underline;}
.tviewc6 {color:cyan;}
.tviewcu6 {color:cyan;text-decoration:underline;}
.tviewc7 {color:yellow;}
.tviewcu7 {color:yellow;text-decoration:underline;}
.tviewc8 {color:red;}
.tviewcu8 {color:red;text-decoration:underline;}
.tviewc9 {color:blue;}
.tviewcu9 {color:blue;text-decoration:underline;}
</style>


<div class="tviewbody">
<div class="tviewtitle">
INJ
</div>
<pre class="tviewpre"><span>   </span><br><span>TCT</span><br><span class="tviewcu4">.</span><span class="tviewcu3">Y</span><span class="tviewcu4">.</span><br><span class="tviewc4">.T.</span><br><span class="tviewc4">...</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">,,,</span><br><span class="tviewc4">.T.</span></pre>
</div>


</div>
</div>
<p>:::::</p>
<p>::::::</p>
<aside class="notes">
<p>QUAL: Phred-scaled quality score for the assertion made in ALT. i.e. <span class="math inline">\(-10log_{10}\)</span> prob(call in ALT is wrong)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="gatk-best-practice" class="slide level2">
<h2>GATK best practice</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/images/gatk-best-practice.png" style="width:90.0%"></p>
</figure>
</div>
<div class="flushright smallest">
<p><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-" class="uri">https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-</a></p>
</div>
</div><div class="column" style="width:30%;">
<h5>
Pros
</h5>
<ul>
<li>Best practices</li>
<li>Large documentation</li>
<li>Variant quality score recalibration</li>
</ul>
<h5>
Cons
</h5>
<ul>
<li>Human-centric - very slow runtime on genomes with many sequences</li>
<li>Complicated setup</li>
</ul>
</div>
</div>
</section>
<section id="alternative-variant-callers" class="slide level2">
<h2>Alternative variant callers</h2>
<div class="columns">
<div class="column" style="width:33%;">
<h5 id="freebayes"><a href="https://github.com/freebayes/freebayes">freebayes</a></h5>
<p>Bayesian genetic variant detector. Simpler setup.</p>
<p>May struggle in high-coverage regions.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="garrison_HaplotypebasedVariantDetection_2012">(<a href="#/bibliography" role="doc-biblioref" onclick="">Garrison &amp; Marth, 2012</a>)</span></p>
</div>
</div><div class="column" style="width:33%;">
<h5 id="bcftools"><a href="https://github.com/samtools/bcftools">bcftools</a></h5>
<p>Utilities for variant calling and manipulating VCFs and BCFs.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="danecek_TwelveYearsSAMtools_2021">(<a href="#/bibliography" role="doc-biblioref" onclick="">Danecek et al., 2021</a>)</span></p>
</div>
</div><div class="column" style="width:33%;">
<h5 id="angsd"><a href="http://www.popgen.dk/angsd/index.php/ANGSD">ANGSD</a></h5>
<p>For low-coverage sequencing. Doesn’t do explicit genotyping; most methods take genotype uncertainty into account.</p>
<div class="flushright smallr">
<p><span class="citation" data-cites="korneliussen_ANGSDAnalysisNext_2014">(<a href="#/bibliography" role="doc-biblioref" onclick="">Korneliussen et al., 2014</a>)</span></p>
</div>
</div>
</div>
<div class="hidden">
<p>Reference bias: plot no. hets vs coverage for real data, e.g., conifer</p>
</div>
<aside class="notes">
<p>NB: samtools and GATK may actually produce different genotypes despite having identical GLs. Samtools applies prior <span class="math inline">\(10^{-3}\)</span> to het call, GATK has no prior <span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">(<a href="#/bibliography" role="doc-biblioref" onclick="">H. Li, 2014</a>)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="exercise-on-variant-calling" class="title-slide slide level1 center">
<h1>Exercise on variant calling</h1>

</section>
<section id="the-monkeyflower-system" class="slide level2 smaller">
<h2>The monkeyflower system</h2>
<div>
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://www.science.org/cms/10.1126/science.365.6456.854/asset/db271bee-8914-4712-9965-dcf1615b6a43/assets/graphic/365_854_f1.jpeg" style="width:50.0%" alt="The allure of monkeyflowers (DOI: 10.1126/science.365.6456.854)"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
</div>
<p>From <a href="https://jgi.doe.gov/csp-2021-genomic-resources-for-mimulus/" class="uri">https://jgi.doe.gov/csp-2021-genomic-resources-for-mimulus/</a></p>
<blockquote>
<p>Plants in the genus Mimulus inhabit highly variable habitats and are famous for their extraordinary ecological diversity. Mimulus is now a powerful system for ecological genomic studies, thanks to its experimental tractability, rapidly growing research community, and the JGI-generated reference genome for M. guttatus.</p>
</blockquote>
</section>
<section id="learning-outcomes" class="slide level2">
<h2>Learning outcomes</h2>
<p>Follow GATK best practices to</p>
<ul>
<li>Perform qc on sequencing reads and interpret results</li>
<li>Prepare reference for read mapping</li>
<li>Map reads to reference</li>
<li>Mark duplicates</li>
<li>Perform raw variant calling to generate a set of sites to exclude from recalibration</li>
<li>Perform base quality score recalibration</li>
<li>Perform variant calling on base recalibrated data</li>
<li>Do genotyping on all samples and combine results to a raw variant call set</li>
</ul>
</section>
<section id="bibliography" class="slide level2 smaller scrollable">
<h2>Bibliography</h2>

<div class="footer footer-default">
<p>Variant calling</p>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-allendorf_PopulationGenomics_2022" class="csl-entry" role="listitem">
Allendorf, F. W., Funk, W. C., Aitken, S. N., Byrne, M., &amp; Luikart, G. (2022). Population <span>Genomics</span>. In F. W. Allendorf, W. C. Funk, S. N. Aitken, M. Byrne, G. Luikart, &amp; A. Antunes (Eds.), <em>Conservation and the <span>Genomics</span> of <span>Populations</span></em> (p. 0). <span>Oxford University Press</span>. <a href="https://doi.org/10.1093/oso/9780198856566.003.0004">https://doi.org/10.1093/oso/9780198856566.003.0004</a>
</div>
<div id="ref-danecek_TwelveYearsSAMtools_2021" class="csl-entry" role="listitem">
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., Davies, R. M., &amp; Li, H. (2021). Twelve years of <span>SAMtools</span> and <span>BCFtools</span>. <em>GigaScience</em>, <em>10</em>(2), giab008. <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>
</div>
<div id="ref-depristo_FrameworkVariationDiscovery_2011" class="csl-entry" role="listitem">
DePristo, M. A., Banks, E., Poplin, R., Garimella, K. V., Maguire, J. R., Hartl, C., Philippakis, A. A., del Angel, G., Rivas, M. A., Hanna, M., McKenna, A., Fennell, T. J., Kernytsky, A. M., Sivachenko, A. Y., Cibulskis, K., Gabriel, S. B., Altshuler, D., &amp; Daly, M. J. (2011). A framework for variation discovery and genotyping using next-generation <span>DNA</span> sequencing data. <em>Nature Genetics</em>, <em>43</em>(5), 491–498. <a href="https://doi.org/10.1038/ng.806">https://doi.org/10.1038/ng.806</a>
</div>
<div id="ref-donato_NewEvaluationMethods_2021" class="csl-entry" role="listitem">
Donato, L., Scimone, C., Rinaldi, C., D’Angelo, R., &amp; Sidoti, A. (2021). New evaluation methods of read mapping by 17 aligners on simulated and empirical <span>NGS</span> data: An updated comparison of <span>DNA-</span> and <span>RNA-Seq</span> data from <span>Illumina</span> and <span>Ion Torrent</span> technologies. <em>Neural Computing and Applications</em>, <em>33</em>(22), 15669–15692. <a href="https://doi.org/10.1007/s00521-021-06188-z">https://doi.org/10.1007/s00521-021-06188-z</a>
</div>
<div id="ref-garrison_HaplotypebasedVariantDetection_2012" class="csl-entry" role="listitem">
Garrison, E., &amp; Marth, G. (2012). Haplotype-based variant detection from short-read sequencing. <em>arXiv:1207.3907 [q-Bio]</em>. <a href="http://arxiv.org/abs/1207.3907">http://arxiv.org/abs/1207.3907</a>
</div>
<div id="ref-korneliussen_ANGSDAnalysisNext_2014" class="csl-entry" role="listitem">
Korneliussen, T. S., Albrechtsen, A., &amp; Nielsen, R. (2014). <span>ANGSD</span>: <span>Analysis</span> of <span>Next Generation Sequencing Data</span>. <em>BMC Bioinformatics</em>, <em>15</em>(1), 356. <a href="https://doi.org/10.1186/s12859-014-0356-4">https://doi.org/10.1186/s12859-014-0356-4</a>
</div>
<div id="ref-li_AligningSequenceReads_2013" class="csl-entry" role="listitem">
Li, H. (2013). Aligning sequence reads, clone sequences and assembly contigs with <span>BWA-MEM</span>. <em>arXiv:1303.3997 [q-Bio]</em>. <a href="https://arxiv.org/abs/1303.3997">https://arxiv.org/abs/1303.3997</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-li_Minimap2PairwiseAlignment_2018" class="csl-entry" role="listitem">
Li, H. (2018). Minimap2: Pairwise alignment for nucleotide sequences. <em>Bioinformatics</em>, <em>34</em>(18), 3094–3100. <a href="https://doi.org/10.1093/bioinformatics/bty191">https://doi.org/10.1093/bioinformatics/bty191</a>
</div>
<div id="ref-li_SNPDetectionMassively_2009" class="csl-entry" role="listitem">
Li, R., Li, Y., Fang, X., Yang, H., Wang, J., Kristiansen, K., &amp; Wang, J. (2009). <span>SNP</span> detection for massively parallel whole-genome resequencing. <em>Genome Research</em>, <em>19</em>(6), 1124–1132. <a href="https://doi.org/10.1101/gr.088013.108">https://doi.org/10.1101/gr.088013.108</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-maruki_GenotypeCallingPopulationGenomic_2017" class="csl-entry" role="listitem">
Maruki, T., &amp; Lynch, M. (2017). Genotype <span>Calling</span> from <span>Population-Genomic Sequencing Data</span>. <em>G3 Genes|Genomes|Genetics</em>, <em>7</em>(5), 1393–1404. <a href="https://doi.org/10.1534/g3.117.039008">https://doi.org/10.1534/g3.117.039008</a>
</div>
<div id="ref-nielsen_GenotypeSNPCalling_2011" class="csl-entry" role="listitem">
Nielsen, R., Paul, J. S., Albrechtsen, A., &amp; Song, Y. S. (2011). Genotype and <span>SNP</span> calling from next-generation sequencing data. <em>Nature Reviews Genetics</em>, <em>12</em>(6), 443–451. <a href="https://doi.org/10.1038/nrg2986">https://doi.org/10.1038/nrg2986</a>
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-pointer/pointer.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'pointer': {"key":"q","color":"red","pointerSize":8,"alwaysVisible":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'h.v',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'default',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: true,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 8.0e-2,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, RevealPointer, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script type="text/javascript">
      function hyperlink_logo() {
        var button = document.createElement( 'div' );
        button.className = "logo-button";
        button.id = "logo-button";
        button.style.visibility = "visible";
        button.style.position = "absolute";
        button.style.zIndex = 30;
        button.style.fontSize = "24px";
        button.style.left = "130px";
        button.style.bottom = "14px";
        button.style.top = "auto";
        button.style.right = "auto";
        if (window.location.host.startsWith("localhost")) {
          button.innerHTML = `<a href="http://${window.location.host}/slides" title="Slides listing"><i class="fa fa-home"></i></a>`
        } else {
          button.innerHTML = '<a href="https://nbisweden.github.io/workshop-pgip/slides" title="Slides listing"><i class="fa fa-home"></i></a>';
        }
        document.querySelector( ".reveal" ).appendChild( button );
      }
    window.document.addEventListener("DOMContentLoaded", function (event) {
      hyperlink_logo();
    });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>