<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Per Unneberg">

<title>DDLS Population genomics in practice - Filtering of resequencing data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../assets/logos/nbislogo-green.svg" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">DDLS Population genomics in practice</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../syllabus.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../precourse.html" rel="" target="">
 <span class="menu-text">Precourse</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../contents.html" rel="" target="">
 <span class="menu-text">Contents</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../info.html" rel="" target="">
 <span class="menu-text">Info</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../slides/index.html" rel="" target="">
 <span class="menu-text">Slides</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../exercises/index.html" rel="" target="">
 <span class="menu-text">Exercises</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../../recipes/index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Code recipes</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/NBISwe" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/NBISweden/workshop-pgip" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../exercises/simulation/index.html">Exercises</a></li><li class="breadcrumb-item"><a href="../../exercises/data_filtering/index.html">Filtering of resequencing data</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Slides</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/pgip/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genomics in practice</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/foundations/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population genetics</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../slides/population_structure/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Population structure</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Exercises</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/simulation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to simulation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/variant_calling/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../exercises/data_filtering/index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Filtering of resequencing data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Code recipes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../recipes/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code recipes</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#background" id="toc-background" class="nav-link active" data-scroll-target="#background">Background</a></li>
  <li><a href="#sec-basic-filtering" id="toc-sec-basic-filtering" class="nav-link" data-scroll-target="#sec-basic-filtering">Basic filtering</a></li>
  <li><a href="#sec-advanced-filtering" id="toc-sec-advanced-filtering" class="nav-link" data-scroll-target="#sec-advanced-filtering">Advanced filtering based coverage analyses</a>
  <ul class="collapse">
  <li><a href="#per-sample-coverage" id="toc-per-sample-coverage" class="nav-link" data-scroll-target="#per-sample-coverage">Per sample coverage</a></li>
  <li><a href="#sample-set-coverages" id="toc-sample-set-coverages" class="nav-link" data-scroll-target="#sample-set-coverages">Sample set coverages</a></li>
  <li><a href="#total-coverage" id="toc-total-coverage" class="nav-link" data-scroll-target="#total-coverage">Total coverage</a></li>
  <li><a href="#filter-on-minimum-number-of-individuals" id="toc-filter-on-minimum-number-of-individuals" class="nav-link" data-scroll-target="#filter-on-minimum-number-of-individuals">Filter on minimum number of individuals</a></li>
  </ul></li>
  <li><a href="#sequence-masks" id="toc-sequence-masks" class="nav-link" data-scroll-target="#sequence-masks">Sequence masks</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Filtering of resequencing data</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Per Unneberg </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">12-Oct-2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<!-- markdownlint-disable MD041 -->
<!-- markdownlint-disable MD041 -->
<!-- markdownlint-enable MD041 -->
<!-- markdownlint-disable MD041 -->
<!-- markdownlint-enable MD041 -->
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>12-Oct-2023: This exercise is WIP!</p>
</div>
</div>
<!-- markdownlint-enable MD041 -->
<p>In this exercise we will look at ways of filtering data, mainly based on coverage.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning objectives
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>create per sample, per population, and total depth of coverage profiles</li>
<li>generate mask files for downstream processing</li>
</ul>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Data setup
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Create an exercise directory and <code>cd</code> to it:</p>
<div class="cell" data-hash="index_cache/html/create-monkeyflower-directory_bf7fc7552f8c48cf1c72b6deeb380ee5">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> monkeyflower</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> monkeyflower</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then download the zip archives and unzip them.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tools
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In this exercise we will use the tools listed below. See <a href="../../exercises/compute_environment/index.html">compute environment</a> for instructions on how to install.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Listing</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Conda</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">UPPMAX modules</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/">csvtk</a></li>
<li><a href="https://bedtools.readthedocs.io/en/latest/index.html">bedtools</a> <span class="citation" data-cites="quinlan_BEDToolsFlexibleSuite_2010">(<a href="#ref-quinlan_BEDToolsFlexibleSuite_2010" role="doc-biblioref">Quinlan &amp; Hall, 2010</a>)</span></li>
<li><a href="https://bioinf.shenwei.me/seqkit/">seqkit</a> <span class="citation" data-cites="shen_SeqKitCrossPlatformUltrafast_2016">(<a href="#ref-shen_SeqKitCrossPlatformUltrafast_2016" role="doc-biblioref">Shen et al., 2016</a>)</span></li>
<li><a href="https://github.com/brentp/mosdepth">mosdepth</a> <span class="citation" data-cites="pedersen_MosdepthQuickCoverage_2018">(<a href="#ref-pedersen_MosdepthQuickCoverage_2018" role="doc-biblioref">Pedersen &amp; Quinlan, 2018</a>)</span></li>
<li><a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> <span class="citation" data-cites="danecek_TwelveYearsSAMtools_2021">(<a href="#ref-danecek_TwelveYearsSAMtools_2021" role="doc-biblioref">Danecek et al., 2021</a>)</span></li>
<li><a href="https://vcftools.github.io/">vcftools</a> <span class="citation" data-cites="danecek_VariantCallFormat_2011">(<a href="#ref-danecek_VariantCallFormat_2011" role="doc-biblioref">Danecek et al., 2011</a>)</span></li>
</ul>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p>Copy the contents to a file <code>environment.yml</code> and install packages with <code>mamba env update -f environment.yml</code>.</p>
<pre lang="text"><code>channels:
  - conda-forge
  - bioconda
  - default
dependencies:
  - bedtools=2.31.0
  - csvtk=0.25.0
  - mosdepth=0.3.3</code></pre>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p>Execute the following command to load modules:</p>
<div class="cell" data-hash="index_cache/html/uppmax-load-modules_ac5dd1572dba3511e9136ae3e8fbc5bb">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load uppmax bioinfo-tools mosdepth/0.3.3 BEDTools/2.29.2 bcftools/1.17 vcftools/0.1.16</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that <code>csvtk</code> may need to be manually installed.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<!-- markdownlint-disable MD013 -->
<!-- markdownlint-enable MD013 -->
<section id="background" class="level2">
<h2 class="anchored" data-anchor-id="background">Background</h2>
<p>Regardless of how a raw variant call set has been produced, the calls will be of varying quality for a number of reasons. For high-coverage sequencing, the two most common are incompleteness of the reference sequence and misaligmnents in repetitive regions <span class="citation" data-cites="li_BetterUnderstandingArtifacts_2014">(<a href="#ref-li_BetterUnderstandingArtifacts_2014" role="doc-biblioref">Li, 2014</a>)</span>. Low-coverage sequencing comes with its own biases and issues, with the most important being the difficulty to accurately call genotypes do to the low coverage, which in practice leads to an underrepresentation of heterozygote calls.</p>
<p>To improve the accuracy of downstream inference, a number of analysis-dependent quality control filters can be applied to the raw variant call set (for a concise summary, see <span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">Lou et al. (<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">2021</a>)</span>). In this exercise, we will focus on coverage-based filters, with the aim of generating <em>sequence masks</em> to denote regions of a reference sequence that contain sufficient information across individuals and populations. Furthermore, the masks will be applied in the context of genetic diversity calculations, in which case specific filters on polymorphic sites (e.g., <em>p</em>-value or minimum minor allele frequency (MAF)) should <strong>not</strong> be applied (all sites contain information).</p>
<div class="cell" data-hash="index_cache/html/fig-coverage-filters_eadb13177b1841b37b9b5176b776a90c">
<div class="cell-output-display">
<div id="fig-coverage-filters" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><a href="index_files/figure-html/fig-coverage-filters-1.svg" class="lightbox" title="Coverage distributions for three hypothetical samples along with the cumulative coverage for all samples." data-gallery="quarto-lightbox-gallery-1"><img src="index_files/figure-html/fig-coverage-filters-1.svg" class="img-fluid figure-img" width="672"></a></p>
<figcaption class="figure-caption">Figure&nbsp;1: Coverage distributions for three hypothetical samples along with the cumulative coverage for all samples.</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-coverage-filters">Figure&nbsp;1</a> illustrates the sequencing coverage of three samples. The important thing to note is that the coverage is uneven. Some regions lack coverage entirely, e.g., due to random sampling or errors in the reference sequence. Other regions have excessive coverage, which could be a sign of repeats that have been collapsed in the reference. A general coverage filter could then seek to mask out sites where a fraction (50%, say) of individuals have too low or excessive coverage.</p>
<p>The right panel illustrates the sum of coverages across all samples. Minimum and maximum depth filters could be applied to the aggregate coverages of all samples, or samples grouped by population, to eliminate sites confounding data support.</p>
<p>The vcf we have produced contains all sites; that is, both monomorphic and polymorphic sites are present. Every site contains information about depth and other metadata, which makes it possible to apply coverage filters directly to the variant file itself. We will make use of direct filters in the first session below (<a href="#sec-basic-filtering">Section&nbsp;2</a>).</p>
<p>However, it may not always be possible to generate a vcf with all sites. Species with large genomes will produce files so large that they prevent efficient downstream processing. Under these circumstances, <em>ad hoc</em> coverage filters can be applied to the bam files to in turn generate sequence masks that can be used in conjunction with the variant file. This is the topic for the second session (<a href="#sec-advanced-filtering">Section&nbsp;3</a>).</p>
</section>
<section id="sec-basic-filtering" class="level2">
<h2 class="anchored" data-anchor-id="sec-basic-filtering">Basic filtering</h2>
<p>FIXME: add basic filtering commands with <code>vcftools</code></p>
</section>
<section id="sec-advanced-filtering" class="level2">
<h2 class="anchored" data-anchor-id="sec-advanced-filtering">Advanced filtering based coverage analyses</h2>
<p>Mapped reads provide information about how well a given genomic region has been represented during sequencing, and this information is usually summarized as the sequencing coverage. For any locus, this is equivalent to the number of reads mapping to that locus.</p>
<p>Sequencing coverage is typically not uniformly distributed over the reference. Reasons may vary but include uneven mapping coverage due to repeat regions, low coverage due to mis-assemblies, or coverage biases generated in the sequencing process. Importantly, both variable and monomorphic sites must be treated identically in the filtering process to eliminate biases between the two kinds of sites.</p>
<p>In this part, we will use <code>mosdepth</code> and <code>bedtools</code> to quickly generate depth of coverage profiles of mapped data. <code>mosdepth</code> is an ultra-fast command line tool for calculating coverage from a bam file. By default, it generates a summary of the global distribution, and per-base coverage in <code>bed.gz</code> format. We will be using the per-base coverage for filtering.</p>
<p>Alternatively, <code>mosdepth</code> can also output results in a highly compressed format <code>d4</code>, which has been developed to handle the ever increasing size of resequencing projects. Files in d4 format can be processed with the <a href="https://github.com/38/d4-format">d4-tools</a> tool <span class="citation" data-cites="hou_BalancingEfficientAnalysis_2021">(<a href="#ref-hou_BalancingEfficientAnalysis_2021" role="doc-biblioref">Hou et al., 2021</a>)</span>. For instance, <code>d4tools view</code> will display the coverage in <code>bed</code> format. We mention this in passing as it may be relevant when working with large genomes or sample sizes, but given the size of our sample data, we will be using <code>bedtools</code> from now on.</p>
<section id="per-sample-coverage" class="level3">
<h3 class="anchored" data-anchor-id="per-sample-coverage">Per sample coverage</h3>
<p>We start by calculating per-sample coverages with <code>mosdepth</code>. For downstream purposes, we need to save the size of the chromosomes we’re looking at, and for many applications, a fasta index file is sufficient.</p>
<div class="cell" data-hash="index_cache/html/samtools-faidx_639f39a3b52d5ee8342ca61b6ac671c9">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">REF</span><span class="op">=</span>M_aurantiacus_v1.fasta</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> faidx <span class="va">${REF}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The syntax to generate coverage information for a bam file is <code>mosdepth &lt;prefix&gt; &lt;input file&gt;</code>. Here, we add the <code>-Q</code> option to exclude reads with a mapping quality less than 20:</p>
<div class="cell" data-hash="index_cache/html/mosdepth-one-sample_defa86d19cd9f2b012fb084f02aa843b">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mosdepth</span> <span class="at">-Q</span> 20 INJ INJ.sort.dup.recal.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The per-base coverage output file will be named <code>INJ.per-base.bed.gz</code> and can be viewed with <code>bgzip</code>:</p>
<div class="cell" data-hash="index_cache/html/bgzip-view-sample_7a8d3fcf8995ec672c713765590a5c92">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> INJ.per-base.bed.gz <span class="kw">|</span> <span class="fu">head</span> <span class="at">-n</span> 5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>LG4:12000001-12100000   0   9   3
LG4:12000001-12100000   9   13  4
LG4:12000001-12100000   13  53  5
LG4:12000001-12100000   53  56  6
LG4:12000001-12100000   56  62  7</code></pre>
</div>
</div>
<p>To get an idea of what the coverage looks like over the chromsome, we can use <code>bedtools</code> and <code>cvstk</code> in a one-liner to generate a simple coverage plot (<a href="#fig-plot-coverage">Figure&nbsp;2</a>)<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
<div class="cell" data-hash="index_cache/html/bash-plot-coverage_55455b71bd10ecd6f0838e4fd1aecfcd">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="op">&lt;(</span><span class="ex">bedtools</span> makewindows <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-w</span> 1000<span class="op">)</span> <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">-b</span> INJ.per-base.bed.gz <span class="at">-wa</span> <span class="at">-wb</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bedtools</span> groupby <span class="at">-i</span> <span class="at">-</span> <span class="at">-g</span> 1,2,3 <span class="at">-c</span> 7 <span class="at">-o</span> mean <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot <span class="at">-t</span> line <span class="at">-x</span> 2 <span class="at">-y</span> 4 <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> Position <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">--ylab</span> Coverage <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-coverage.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-plot-coverage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="fig-plot-coverage.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="fig-plot-coverage.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;2: Coverage for sample INJ in 1kb windows. Experiment changing the window size (<code>-w</code>) parameter to change smoothing.</figcaption>
</figure>
</div>
<p>Apparently there are some high-coverage regions that could be associated with, e.g., collapsed repeat regions in the assembly. Let’s compile coverage results for all samples, using bash string manipulation to generate file prefix<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></p>
<div class="cell" data-hash="index_cache/html/mosdepth-compile-coverage-data_33ce4a5877e176ab9c80fab7608286f1">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> f <span class="kw">in</span> <span class="pp">[</span><span class="ss">A</span><span class="pp">-</span><span class="ss">Z</span><span class="pp">]*</span>.sort.dup.recal.bam<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="va">prefix</span><span class="op">=</span><span class="va">${f</span><span class="op">%</span>.sort.dup.recal.bam<span class="va">}</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">mosdepth</span> <span class="at">-Q</span> 20 <span class="va">$prefix</span> <span class="va">$f</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">$prefix</span><span class="st">\t"</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span> <span class="va">$prefix</span>.mosdepth.summary.txt <span class="kw">|</span> <span class="fu">grep</span> total</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span> <span class="op">&gt;</span> ALL.mosdepth.summary.txt</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> ALL.mosdepth.summary.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>BCRD    total   100000  882549  8.83    0   320
ELF total   100000  845166  8.45    0   187
INJ total   100000  818921  8.19    0   229
JMC total   100000  934032  9.34    0   244
LH  total   100000  940246  9.40    0   225
LO  total   100000  656112  6.56    0   151
MT  total   100000  956272  9.56    0   233
PCT total   100000  836694  8.37    0   179
POTR    total   100000  795201  7.95    0   205
UCSD    total   100000  812507  8.13    0   293</code></pre>
</div>
</div>
<p>We can calculate the total coverage with <code>csvtk</code> as follows:</p>
<div class="cell" data-hash="index_cache/html/mosdepth-compile-coverage-data-sum_9c720befb69e3f6085ee959e48320122">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">csvtk</span> summary <span class="at">-H</span> <span class="at">-t</span> ALL.mosdepth.summary.txt <span class="at">-f</span> 5:sum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>to get the total coverage 84.78, which gives a hint at where the diploid coverage peak should be.</p>
</section>
<section id="sample-set-coverages" class="level3">
<h3 class="anchored" data-anchor-id="sample-set-coverages">Sample set coverages</h3>
<p>We can combine coverage intervals with <code>bedtools unionbedg</code>. We collect the bed file names and generate matching sample names to pass as arguments to option <code>-names</code>. Also, we include positions with no coverage (<code>-empty</code>) which requires the use of a genome file (option <code>-g</code>). The bed output is piped to <code>bgzip</code> which compresses the output, before finally indexing with <code>tabix</code>:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/bedtools-unionbedg-all_453edadd8d9de44376eb6a8fd7e10341">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span><span class="va">$(</span><span class="fu">cat</span> sampleinfo.csv <span class="kw">|</span> <span class="ex">csvtk</span> grep <span class="at">-r</span> <span class="at">-p</span> .<span class="pp">*</span>red.<span class="pp">*</span> <span class="at">-r</span> <span class="at">-p</span> .<span class="pp">*</span>yellow.<span class="pp">*</span> <span class="at">-f</span> Taxon <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-f</span> Sample <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> Sample <span class="kw">|</span> <span class="fu">tr</span> <span class="st">"\n"</span> <span class="st">" "</span><span class="va">)</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="va">BEDGZ</span><span class="op">=</span><span class="va">$(</span><span class="cf">for</span> sm <span class="kw">in</span> <span class="va">$SAMPLES</span><span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${sm}</span><span class="st">.per-base.bed.gz "</span><span class="kw">;</span> <span class="cf">done</span><span class="va">)</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> unionbedg <span class="at">-header</span> <span class="at">-names</span> <span class="va">$SAMPLES</span> <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-empty</span> <span class="at">-i</span> <span class="va">$BEDGZ</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.bg.gz</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="at">-S</span> 1 ALL.bg.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<p>We also need to combine coverages per populations yellow and red.</p>
<div class="callout callout-style-default callout-exercise no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<div>
<div class="callout-exercise">
<p>Using the previous command as a template, try to generate per population coverage files.</p>
<div class="callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div>
<div class="callout-answer">
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/bedtools-unionbedg-per-population_8c569d92811464efae9d13518cb08cc9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a> <span class="va">SAMPLES</span><span class="op">=</span><span class="va">$(</span><span class="fu">cat</span> sampleinfo.csv <span class="kw">|</span> <span class="ex">csvtk</span> grep <span class="at">-r</span> <span class="at">-p</span> .<span class="pp">*</span><span class="va">${pop}</span>.<span class="pp">*</span> <span class="at">-f</span> Taxon <span class="kw">|</span> <span class="ex">csvtk</span> cut <span class="at">-f</span> Sample <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> Sample <span class="kw">|</span> <span class="fu">tr</span> <span class="st">"\n"</span> <span class="st">" "</span><span class="va">)</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a> <span class="va">BEDGZ</span><span class="op">=</span><span class="va">$(</span><span class="cf">for</span> sm <span class="kw">in</span> <span class="va">$SAMPLES</span><span class="kw">;</span> <span class="cf">do</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="at">-n</span> <span class="st">"</span><span class="va">${sm}</span><span class="st">.per-base.bed.gz "</span><span class="kw">;</span> <span class="cf">done</span><span class="va">)</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a> <span class="ex">bedtools</span> unionbedg <span class="at">-header</span> <span class="at">-names</span> <span class="va">$SAMPLES</span> <span class="at">-g</span> <span class="va">${REF}</span> <span class="at">-empty</span> <span class="at">-i</span> <span class="va">$BEDGZ</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.bed.gz</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a> <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="at">-S</span> 1 <span class="va">$pop</span>.bed.gz</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Less than the req'd two fields were encountered in the genome file (M_aurantiacus_v1.fasta) at line 1.  Exiting.
Less than the req'd two fields were encountered in the genome file (M_aurantiacus_v1.fasta) at line 1.  Exiting.</code></pre>
</div>
</div>
<!-- markdownlint-enable MD013 -->
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="total-coverage" class="level3">
<h3 class="anchored" data-anchor-id="total-coverage">Total coverage</h3>
<p>Since we eventually want to filter on total coverage, we sum per sample coverages for each sample set with <code>awk</code>:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/awk-sum-ALL-coverage_71b73b6155e60f71f51318001bfb30bb">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> ALL.bg.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="at">-v</span> OFS=<span class="st">"\t"</span> <span class="st">'NR &gt; 1 {sum=0; for (i=4; i&lt;=NF; i++) sum+=$i; print $1, $2, $3, sum}'</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.sum.bed.gz</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed ALL.sum.bed.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<p>For illustration, we plot the total coverage:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/total-coverage_625f270236975e82e49387d51da35d39">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="op">&lt;(</span><span class="ex">bedtools</span> makewindows <span class="at">-g</span> <span class="va">${REF}</span>.fai <span class="at">-w</span> 1000<span class="op">)</span> <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-b</span> ALL.sum.bed.gz <span class="at">-wa</span> <span class="at">-wb</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bedtools</span> groupby <span class="at">-i</span> <span class="at">-</span> <span class="at">-g</span> 1,2,3 <span class="at">-c</span> 7 <span class="at">-o</span> mean <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot <span class="at">-t</span> line <span class="at">-x</span> 2 <span class="at">-y</span> 4 <span class="at">--point-size</span> 0.01 <span class="at">--xlab</span> Position <span class="dt">\</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ylab</span> Coverage <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="op">&gt;</span> fig-plot-total-coverage.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<div id="fig-plot-total-coverage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="fig-plot-total-coverage.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="fig-plot-total-coverage.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;3: Total coverage in 1kb windows.</figcaption>
</figure>
</div>
<p>In order to define thresholds for subsequent filtering, we plot the proportion of the genome coverage versus depth of coverage (similar to k-mer plots in sequence assembly projects). In</p>
<!-- markdownlint-disable MD013 -->
<div id="fig-plot-total-coverage-distribution" class="quarto-layout-panel" data-attr-output=".details summary=&quot;Output&quot;">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-hist" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure">
<p><a href="fig-plot-total-coverage-distribution.png" class="lightbox" title="Genome coverage" data-gallery="quarto-lightbox-gallery-4"><img src="fig-plot-total-coverage-distribution.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(a) Genome coverage</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div id="fig-plot-total-coverage-distribution-cumulative" class="quarto-figure quarto-figure-center anchored" style="flex-basis: 100.0%;justify-content: center;">
<figure class="figure">
<p><a href="fig-plot-total-coverage-distribution-cumulative.png" class="lightbox" title="Cumulative genome coverage" data-gallery="quarto-lightbox-gallery-5"><img src="fig-plot-total-coverage-distribution-cumulative.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(b) Cumulative genome coverage</figcaption>
</figure>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;4: Genome coverage vs depth of coverage.</figcaption><p></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<p>In <a href="#fig-plot-total-coverage-distribution">Figure&nbsp;4</a> a, a diploid peak is evident just below coverage X=150; we zoom in on that region to get a better view:</p>
<div class="cell" data-hash="index_cache/html/total-depth-of-coverage-zoom-in_11c4198142d18de8e4be3af408895635">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> ALL.sum.bed.csv <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot line <span class="at">-x</span> 1 <span class="at">-y</span> 2 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Genome coverage (kbp)"</span> <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="at">--x-min</span> 90 <span class="at">--x-max</span> 160</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-disable MD013 -->
<div id="fig-plot-total-coverage-distribution-zoom-in" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="fig-plot-total-coverage-distribution-hist-zoom-in" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><a href="fig-plot-total-coverage-distribution-hist-zoom-in.png" class="lightbox" title="Genome coverage" data-gallery="quarto-lightbox-gallery-6"><img src="fig-plot-total-coverage-distribution-hist-zoom-in.png" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">(a) Genome coverage</figcaption>
</figure>
</div>
<figcaption class="figure-caption">Figure&nbsp;5: Genome coverage vs depth of coverage.</figcaption>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<p><span class="citation" data-cites="lou_BeginnerGuideLowcoverage_2021">(<a href="#ref-lou_BeginnerGuideLowcoverage_2021" role="doc-biblioref">Lou et al., 2021</a>)</span> point out that appropriate thresholds depend on the data set, but as a general rule recommends a minimum depth threshold at &lt;0.8x average coverage, and a maximum depth threshold at mean coverage plus one or two standard deviations. Given the unrealistically homogenous coverage of our simulated data, we set thresholds based on <a href="#fig-plot-total-coverage-distribution-zoom-in">Figure&nbsp;5</a> to e.g., 100-160.</p>
<p>We then use these thresholds to generate a bed file containing regions that are accessible, i.e., have sufficient coverage for downstream analyses. We also calculate the number of bases that pass the filtering criteria.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/filter-all-bed-file-on-coverage_56a77021cdb6b126b1a0711d9a1296d2">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">zcat</span> ALL.sum.bed.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{if (($4 &gt;= 100) &amp;&amp; ($4 &lt;= 160)) print $0}'</span> <span class="op">&gt;</span> ALL.sum.depth.bed.gz</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-i</span> ALL.sum.depth.bed.gz <span class="at">-g</span> <span class="va">${REF}</span>.fai  <span class="kw">|</span> <span class="fu">grep</span> genome</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<p>Consequently, 10% of the genome is accessible by depth.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate coverage sums for sample sets CEU, CHB, and YRI, and from these determine coverage thresholds and apply the thresholds to generate bed files with accessible regions.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/awk-sum-sample-set-coverage_be61b0152159f92f86d04144aae1875c">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> <span class="va">$pop</span>.bed.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="at">-v</span> OFS=<span class="st">"\t"</span> <span class="st">'NR &gt; 1 {sum=0; for (i=4; i&lt;=NF; i++) sum+=$i; print $1, $2, $3, sum}'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.sum.bed.gz</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="va">$pop</span>.sum.bed.gz</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">zcat</span> <span class="va">$pop</span>.sum.bed.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">awk</span> <span class="at">-v</span> h=0 <span class="st">'BEGIN {OFS=","; cumsum=0} {a[$4] += ($3-$2)/1000; if ($4 &gt; h) {h=$4}} END{for (i=0; i&lt;=h; i++) {if (a[i]) {cumsum+=a[i]} else {a[i]=0;} print i, a[i], cumsum}}'</span> <span class="dt">\</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>    <span class="op">&gt;</span> <span class="va">$pop</span>.sum.bed.csv</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- markdownlint-enable MD013 -->
<div class="cell" data-hash="index_cache/html/plot-sum-sample-set-coverage_232985ecd2c901070696422a8eaa738e">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> red yellow<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="va">${pop}</span>.sum.bed.csv <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">csvtk</span> plot line <span class="at">-x</span> 1 <span class="at">-y</span> 2 <span class="at">--point-size</span> 0.01 <span class="dt">\</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--xlab</span> <span class="st">"Depth of coverage (X)"</span> <span class="at">--ylab</span> <span class="st">"Genome coverage (kbp)"</span> <span class="dt">\</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--width</span> 9.0 <span class="at">--height</span> 3.5 <span class="at">--x-min</span> 10 <span class="at">--x-max</span> 100 <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    fig-plot-total-coverage-distribution-hist-zoom-in-<span class="va">$pop</span>.png</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>cat: red.sum.bed.csv: No such file or directory
[31m[ERRO][0m xopen: no content
cat: yellow.sum.bed.csv: No such file or directory
[31m[ERRO][0m xopen: no content</code></pre>
</div>
</div>
</div>
</div>
<p>Now we have combined total per sample coverage for ALL samples, and for sample sets CEU, CHB, and YRI. The upcoming task will be to generate sequence masks from the total coverage and minimum number of individuals with coverage greater than zero.</p>
</section>
<section id="filter-on-minimum-number-of-individuals" class="level3">
<h3 class="anchored" data-anchor-id="filter-on-minimum-number-of-individuals">Filter on minimum number of individuals</h3>
<p>In addition to filtering on coverage, we will also filter on the minimum number of individuals with calls. This is to account for cases where coverages that pass the coverage filter originate from just a few samples. Here, we will remove sites where less than 50% of individuals have a call.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/awk-filter-50pct-ALL_c694dbfffc1b61422c770ccb00d331b3">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> ALL.bg.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="st">'BEGIN {OFS="\t"} NR &gt; 1 {count=0; for (i=4; i&lt;=NF; i++) {if ($i&gt;0) count+=1}; if (count&gt;=((NF-3)*0.5)) {print $1, $2, $3}}'</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="op">&gt;</span> ALL.ind.bed.gz</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed ALL.ind.bed.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->
<section id="exercise-1" class="level4">
<h4 class="anchored" data-anchor-id="exercise-1">Exercise</h4>
<p>Generate coverage sums for sample sets CEU, CHB, and YRI.</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/awk-filter-50pct-sample-set_e35e9140411fe600b5094c7fd2b5b757">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pop <span class="kw">in</span> CEU CHB YRI<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bgzip</span> <span class="at">-c</span> <span class="at">-d</span> <span class="va">$pop</span>.bed.gz <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-v</span> FS=<span class="st">"\t"</span> <span class="st">'BEGIN {OFS="\t"} NR &gt; 1 {count=0; for (i=4; i&lt;=NF; i++) {if ($i&gt;0) count+=1}; if (count&gt;=((NF-3)*0.5)) {print $1, $2, $3}}'</span> <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$pop</span>.ind.bed.gz</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="ex">tabix</span> <span class="at">-f</span> <span class="at">-p</span> bed <span class="va">$pop</span>.ind.bed.gz</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- markdownlint-enable MD013 -->
</section>
</section>
</section>
<section id="sequence-masks" class="level2">
<h2 class="anchored" data-anchor-id="sequence-masks">Sequence masks</h2>
<p>Finally, for each sample set, we will use <code>bedtools intersect</code> to generate combined bed files for total sum coverages and the filter on number of individuals. <code>bedtools intersect</code> makes it easy to combine multiple bed files, so other filters could be added.</p>
<p>FIXME: make coverage filter above and intersect with ROI files</p>
<div class="cell" data-hash="index_cache/html/bedtools-intersect-all_8c9b4ee99cb8a732d507291e0b226053">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bedtools intersect -a ALL.sum.bed.gz -b ALL.ind.bed.gz \</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># -g ${REF}.fai | bgzip &gt; ALL.mask.bed.gz</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#tabix -f -p bed ALL.mask.bed.gz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can use the command <code>bedtools makefasta</code> to make a sequence mask file in fasta format. The file will consist of characters <code>0</code> and <code>1</code>, where the latter are regions that will be masked out in subsequent analyses. First, we make a mask file that consists solely of <code>1</code>’s:</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/html/bedtools-maskfasta-make-genome-mask_33bb602c586009a6ea2c80dae4bc2beb">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'BEGIN {OFS="\t"} {print $1, 0, $2}'</span> <span class="va">${REF}</span>.fai <span class="op">&gt;</span> <span class="va">${REF}</span>.bed</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> maskfasta <span class="at">-fi</span> <span class="va">${REF}</span> <span class="at">-mc</span> 1 <span class="at">-fo</span> <span class="va">${REF}</span>.mask.fa <span class="at">-bed</span> <span class="va">${REF}</span>.bed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- markdownlint-enable MD013 -->



</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-danecek_VariantCallFormat_2011" class="csl-entry" role="listitem">
Danecek, P., Auton, A., Abecasis, G., Albers, C. A., Banks, E., DePristo, M. A., Handsaker, R. E., Lunter, G., Marth, G. T., Sherry, S. T., McVean, G., &amp; Durbin, R. (2011). The variant call format and <span>VCFtools</span>. <em>Bioinformatics (Oxford, England)</em>, <em>27</em>(15), 2156–2158. <a href="https://doi.org/10.1093/bioinformatics/btr330">https://doi.org/10.1093/bioinformatics/btr330</a>
</div>
<div id="ref-danecek_TwelveYearsSAMtools_2021" class="csl-entry" role="listitem">
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., Davies, R. M., &amp; Li, H. (2021). Twelve years of <span>SAMtools</span> and <span>BCFtools</span>. <em>GigaScience</em>, <em>10</em>(2), giab008. <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>
</div>
<div id="ref-hou_BalancingEfficientAnalysis_2021" class="csl-entry" role="listitem">
Hou, H., Pedersen, B., &amp; Quinlan, A. (2021). Balancing efficient analysis and storage of quantitative genomics data with the <span>D4</span> format and D4tools. <em>Nature Computational Science</em>, <em>1</em>(6), 441–447. <a href="https://doi.org/10.1038/s43588-021-00085-0">https://doi.org/10.1038/s43588-021-00085-0</a>
</div>
<div id="ref-li_BetterUnderstandingArtifacts_2014" class="csl-entry" role="listitem">
Li, H. (2014). Toward better understanding of artifacts in variant calling from high-coverage samples. <em>Bioinformatics</em>, <em>30</em>(20), 2843–2851. <a href="https://doi.org/10.1093/bioinformatics/btu356">https://doi.org/10.1093/bioinformatics/btu356</a>
</div>
<div id="ref-lou_BeginnerGuideLowcoverage_2021" class="csl-entry" role="listitem">
Lou, R. N., Jacobs, A., Wilder, A. P., &amp; Therkildsen, N. O. (2021). A beginner’s guide to low-coverage whole genome sequencing for population genomics. <em>Molecular Ecology</em>, <em>30</em>(23), 5966–5993. <a href="https://doi.org/10.1111/mec.16077">https://doi.org/10.1111/mec.16077</a>
</div>
<div id="ref-pedersen_MosdepthQuickCoverage_2018" class="csl-entry" role="listitem">
Pedersen, B. S., &amp; Quinlan, A. R. (2018). Mosdepth: Quick coverage calculation for genomes and exomes. <em>Bioinformatics</em>, <em>34</em>(5), 867–868. <a href="https://doi.org/10.1093/bioinformatics/btx699">https://doi.org/10.1093/bioinformatics/btx699</a>
</div>
<div id="ref-quinlan_BEDToolsFlexibleSuite_2010" class="csl-entry" role="listitem">
Quinlan, A. R., &amp; Hall, I. M. (2010). <span>BEDTools</span>: A flexible suite of utilities for comparing genomic features. <em>Bioinformatics</em>, <em>26</em>(6), 841–842. <a href="https://doi.org/10.1093/bioinformatics/btq033">https://doi.org/10.1093/bioinformatics/btq033</a>
</div>
<div id="ref-shen_SeqKitCrossPlatformUltrafast_2016" class="csl-entry" role="listitem">
Shen, W., Le, S., Li, Y., &amp; Hu, F. (2016). <span>SeqKit</span>: <span>A Cross-Platform</span> and <span>Ultrafast Toolkit</span> for <span>FASTA</span>/<span>Q File Manipulation</span>. <em>PLOS ONE</em>, <em>11</em>(10), e0163962. <a href="https://doi.org/10.1371/journal.pone.0163962">https://doi.org/10.1371/journal.pone.0163962</a>
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>The one-liner combines the results of several commands in a pipe stream. Also, <a href="https://www.gnu.org/software/bash/manual/html_node/Redirections.html">Bash redirections</a> are used to gather the results from the output of <code>bedtools makewindows</code> to <code>bedtools intersect</code>. The intersection commands collects coverage data in 1kb windows that are then summarized by <code>bedtools groupby</code>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>The <code>%</code> operator deletes the shortest match of <code>$substring</code> from back of <code>$string</code>: <code>${string%substring}</code>. See <a href="https://tldp.org/LDP/abs/html/string-manipulation.html">Bash string manipulation</a> for more information.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">2023 <a href="about.html">NBIS</a> | <a href="https://choosealicense.com/licenses/gpl-3.0/">GPL-3 License</a></div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">Published with <a href="https://quarto.org/">Quarto</a> v<!--?quarto.version?--></div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","loop":true,"descPosition":"bottom","openEffect":"zoom","selector":".lightbox"});</script>



</body></html>