<!DOCTYPE html>
<html lang="en"><head>
<link href="../../assets/logos/nbislogo-green.svg" rel="icon" type="image/svg+xml">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.450">

  <meta name="author" content="Per Unneberg">
  <title>DDLS Population genomics in practice - Population genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #3c3836;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #3c3836; } /* Normal */
    code span.al { color: #282828; background-color: #cc241d; font-weight: bold; } /* Alert */
    code span.an { color: #98971a; } /* Annotation */
    code span.at { color: #d79921; } /* Attribute */
    code span.bn { color: #f67400; } /* BaseN */
    code span.bu { color: #d65d0e; } /* BuiltIn */
    code span.cf { color: #cc241d; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #b16286; } /* Char */
    code span.cn { color: #b16286; font-weight: bold; } /* Constant */
    code span.co { color: #928374; } /* Comment */
    code span.cv { color: #928374; } /* CommentVar */
    code span.do { color: #98971a; } /* Documentation */
    code span.dt { color: #d79921; } /* DataType */
    code span.dv { color: #f67400; } /* DecVal */
    code span.er { color: #cc241d; text-decoration: underline; } /* Error */
    code span.ex { color: #689d6a; font-weight: bold; } /* Extension */
    code span.fl { color: #f67400; } /* Float */
    code span.fu { color: #689d6a; } /* Function */
    code span.im { color: #689d6a; } /* Import */
    code span.in { color: #282828; background-color: #83a598; } /* Information */
    code span.kw { color: #3c3836; font-weight: bold; } /* Keyword */
    code span.op { color: #3c3836; } /* Operator */
    code span.ot { color: #689d6a; } /* Other */
    code span.pp { color: #d65d0e; } /* Preprocessor */
    code span.re { color: #928374; background-color: #f9f5d7; } /* RegionMarker */
    code span.sc { color: #b16286; } /* SpecialChar */
    code span.ss { color: #98971a; } /* SpecialString */
    code span.st { color: #98971a; } /* String */
    code span.va { color: #458588; } /* Variable */
    code span.vs { color: #98971a; } /* VerbatimString */
    code span.wa { color: #282828; background-color: #fabd2f; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-pointer/pointer.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="../../assets/logos/nbis-scilifelab.svg" data-background-position="top left" data-background-size="500px" class="center">
  <h1 class="title">Population genetics</h1>
  <h3 class="subtitle">Foundations</h3>
  <h4 class="author">Per Unneberg</h4>
  <h4 class="institute">NBIS</h4>
  <h4 class="date">12-Oct-2023</h4>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Table of contents</h2>
<ul>
<li><a href="#/dna-variation" id="/toc-dna-variation">DNA variation</a></li>
<li><a href="#/models-of-populations" id="/toc-models-of-populations">Models of populations</a></li>
<li><a href="#/origin-and-change-of-variation" id="/toc-origin-and-change-of-variation">Origin and change of variation</a></li>
<li><a href="#/selection" id="/toc-selection">Selection</a></li>
</ul>
</nav>
</section>

<section id="intended-learning-outcomes" class="slide level2 unnumbered unlisted">
<h2>Intended learning outcomes</h2>
<p>Introduction to foundations of population genetics with an emphasis on genealogies</p>
<ul>
<li>Description of DNA variation data</li>
<li>Wright-Fisher population model and genealogies</li>
<li>Genetic drift</li>
<li>Wright-Fisher model with mutation</li>
<li>Mutation-drift balance</li>
<li>Neutral theory</li>
<li>Selection basics</li>
</ul>
<aside class="notes">
<p>Objective: introduce the foundations of population genetics, with an emphasis on genealogies. Framing population genetic concepts in trees will facilitate the introduction of the coalescent in the afternoon.</p>
<p>The lecture presents sequence data, the Wright-Fisher model, genetic drift, mutation, and selection. The focus is to familiarise students with concepts and basic theory.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section>
<section id="dna-variation" class="title-slide slide level1 center">
<h1>DNA variation</h1>
<div>
<p><img data-src="assets/images/popgen.svg"></p>
</div>
<aside class="notes">
<p>Goal of section: look at the data that forms the foundation for population genomic analyses</p>
<p>From <span class="citation" data-cites="nei_MolecularEvolutionPhylogenetics_2000">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nei &amp; Kumar, 2000, p. 231</a>)</span>:</p>
<blockquote>
<p>The main subject of population genetics is to study the generation and maintenance of genetic polymorphism and to understand the mechanisms of evolution at the population level</p>
</blockquote>
<p><span class="citation" data-cites="casillas_MolecularPopulationGenetics_2017">(<a href="#/bibliography" role="doc-biblioref" onclick="">Casillas &amp; Barbadilla, 2017, p. 1026</a>)</span>:</p>
<blockquote>
<p>Big data samples of complete genome sequences of many individuals from natural populations of many species have transformed population genetics inferences on samples of loci to population genomics: the analysis of genome-wide patterns of DNA variation within and between species.</p>
</blockquote>
<p><span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004, p. 1</a>)</span></p>
<blockquote>
<p>Population geneticists spend most of their time doing one of two things: describing the genetic structure of populations or theorizing on the evolutionary forces acting on populations. On a good day, these two activities mesh and true insights emerge.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="dna-variation-1" class="slide level2">
<h2>DNA variation</h2>
<div class="dna small">
<table style="width:75%;">
<caption>Sequence aligmnent of four DNA sequences <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019</a>, Fig 1.1)</span>. </caption>
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
</tbody>
</table>
</div>
<p>The main data for molecular population genetics are DNA sequences. The alignment above shows a <em>sample</em> of four DNA <em>sequences</em>. Each sequence has 15 nucleotides (<em>sites</em>) “from the same <em>locus</em> (location) on a chromosome” <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(p.2 <a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019</a>)</span></p>
<div class="columns">
<div class="column" style="width:60%;">
<div class="fragment">
<p>Alternative names for <em>sequence</em>:</p>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment"><em>chromosome</em></li>
<li class="fragment"><em>gene</em></li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment"><em>allele</em> (different by origin)</li>
<li class="fragment"><em>sample</em></li>
<li class="fragment"><em>cistron</em></li>
</ul>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="fragment">
<p>We will preferentially use <em>sequence</em> or <em>chromosome</em> to refer to an entire sequence, and <em>allele</em> to refer to individual nucleotides that differ.</p>
</div>
</div>
</div>
</section>
<section id="dna-variation---monomorphic-sites" class="slide level2">
<h2>DNA variation - monomorphic sites</h2>
<div class="dna">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;">T</td>
</tr>
</tbody>
</table>
</div>
<p>The alignment has 4 DNA sequences where each sequence has length <span class="math inline">\(L=15\)</span>. A site where all nucleotides (alleles) are identical is called a <em>monomorphic</em> site (indicated with asterisks above). There are 9 monomorphic sites.</p>
</section>
<section id="dna-variation---segregating-sites" class="slide level2">
<h2>DNA variation - segregating sites</h2>
<div class="dna">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>A</strong></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
</tr>
</tbody>
</table>
</div>
<p>A site where there are different nucleotides (alleles) is called a <em>segregating</em> site (indicated with asterisks above), often denoted <em>S</em>. There are <span class="math inline">\(S=6\)</span> segregating sites.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="fragment">
<p>Alternative names for <em>segregating site</em> are:</p>
</div>
<div>
<ul>
<li class="fragment"><em>polymorphism</em></li>
<li class="fragment"><em>mutation</em></li>
<li class="fragment"><em>single nucleotide polymorphism (SNP)</em></li>
</ul>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p><em>mutation</em> here and onwards refers to the process that generates new variation and the new variants generated by this process</p>
<p>In contrast to <em>mutation</em> which corresponds to within-species variation, a <em>substitution</em> refers to DNA differences between species.</p>
</div>
</div>
</div>
</section>
<section id="dna-variation---major-and-minor-alleles" class="slide level2">
<h2>DNA variation - major and minor alleles</h2>
<div class="dna">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>A</strong></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
</tr>
</tbody>
</table>
</div>
<p>Much of the nucleotide variation we study consists of <em>bi-allelic</em> SNPs. The most common variant is called the <em>major</em> allele, and the least common the <em>minor</em> allele.</p>
<p>The set of alleles found on a single sequence is called <em>haplotype</em>.</p>
</section>
<section id="describing-dna-variation---heterozygosity" class="slide level2 smaller">
<h2>Describing DNA variation - heterozygosity</h2>
<div class="dna">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>A</strong></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
</tr>
</tbody>
</table>
</div>
<p>Once we have a sample of sequences we want to describe the observed variation. At any position the <em>i</em>th allele has sample frequency <span class="math inline">\(p_i\)</span>, where the sum of all allele frequencies is 1. For instance, at site 1, <span class="math inline">\(p_T=1\)</span> (and by extension <span class="math inline">\(p_A=p_C=p_G=0\)</span>), and at site 2 <span class="math inline">\(p_C=1/4\)</span> and <span class="math inline">\(p_T=3/4\)</span>.</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="fragment">
<h4>
Heterozygosity
</h4>
<p>The <em>heterozygosity</em> at a site <span class="math inline">\(j\)</span> is given by</p>
<p><span class="math display">\[
h_j = \frac{n}{n-1}\left(1 - \sum_i p_i^2\right)
\]</span></p>
<p>where the summation is over all alleles and <span class="math inline">\(p_i\)</span> is the frequency of the <span class="math inline">\(i\)</span>-th allele</p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<h4>
Exercise: calculate the heterozygosity at sites 1, 2 and 5
</h4>
</div>
<div class="fragment">
<!-- markdownlint-disable MD013 -->
<p><span class="math display">\[
h_1 = \frac{4}{3} \left(1 - p_T^2 \right) = 0 \\
h_2 = \frac{4}{3} \left(1 - \left(p_C^2 + p_T^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{16} + \frac{9}{16}\right)\right) = \frac{1}{2}\\
h_5 = \frac{4}{3} \left(1 - \left(p_A^2 + p_G^2\right) \right) = \frac{4}{3} \left( 1 - \left(\frac{1}{4} + \frac{1}{4}\right)\right) = \frac{2}{3}
\]</span></p>
<!-- markdownlint-enable MD013 -->
</div>
</div>
</div>
<aside class="notes">
<p>In a randomly mating population, the heterozygosity is equal to the frequency of heterozygotes. Note however that the definition of heterozygosity only relies on allele frequencies, which means it can be applied to populations that are not in Hardy-Weinberg equilibrium. It can also be applied to haploid organisms, like bacteria.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="describing-dna-variation---nucleotide-diversity" class="slide level2">
<h2>Describing DNA variation - nucleotide diversity</h2>
<div class="dna">
<table style="width:75%;">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">1</th>
<th style="text-align: center;">2</th>
<th style="text-align: center;">3</th>
<th style="text-align: center;">4</th>
<th style="text-align: center;">5</th>
<th style="text-align: center;">6</th>
<th style="text-align: center;">7</th>
<th style="text-align: center;">8</th>
<th style="text-align: center;">9</th>
<th style="text-align: center;">10</th>
<th style="text-align: center;">11</th>
<th style="text-align: center;">12</th>
<th style="text-align: center;">13</th>
<th style="text-align: center;">14</th>
<th style="text-align: center;">15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>A</strong></td>
</tr>
<tr class="even">
<td style="text-align: center;">T</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;"><strong>G</strong></td>
<td style="text-align: center;">A</td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;"><strong>C</strong></td>
<td style="text-align: center;">T</td>
<td style="text-align: center;">C</td>
<td style="text-align: center;">G</td>
<td style="text-align: center;">T</td>
</tr>
<tr class="odd">
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
<td style="text-align: center;"></td>
<td style="text-align: center;">*</td>
</tr>
</tbody>
</table>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<div class="fragment">
<h3>
Nucleotide diversity <span class="math inline">\(\pi\)</span>
</h3>
<p>The <em>nucleotide diversity</em> is the sum of site heterozygosities:</p>
<p><span class="math display">\[
\pi = \sum_{j=1}^S h_j
\]</span></p>
<p>where <span class="math inline">\(S\)</span> is the number of segregating sites</p>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<h3>
Calculate the nucleotide diversity
</h3>
</div>
<div class="fragment">
<p>Observation: <span class="math inline">\(h_i\)</span> either 1/2 or 2/3 (for sites with <span class="math inline">\(p_{major}=p_{minor}\)</span>).</p>
</div>
<div class="fragment">
<!-- markdownlint-disable MD013 -->
<p><span class="math display">\[
\pi = \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{2}{3} + \frac{1}{2} + \frac{1}{2} = 3\frac{1}{3}
\]</span></p>
<!-- markdownlint-enable MD013 -->
</div>
<div class="fragment">
<p>Often we provide <span class="math inline">\(\pi\)</span> per site:</p>
<p><span class="math display">\[
\pi = 3.33/15 = 0.222
\]</span></p>
</div>
</div>
</div>
<aside class="notes">
<p><span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">Hahn (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span> implicitly assumes we are looking at DNA polymorphism. The expression actually holds for any genetic variation at a locus, and is sometimes called the <em>gene diversity</em> <span class="citation" data-cites="nei_MolecularEvolutionPhylogenetics_2000">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nei &amp; Kumar, 2000, p. 245</a>)</span>.</p>
<p>Under the infinite sites model, <span class="math inline">\(E(\pi)=\theta=4N_e\mu\)</span>, for which reason <span class="math inline">\(\pi\)</span> sometimes is called <span class="math inline">\(\theta_\pi\)</span>. The measure gives the average number of pairwise nucleotide differences between two sequences, so an alternative expression is</p>
<p><span class="math display">\[
\pi = \frac{\sum_{i&lt;j}k_{ij}}{n(n-1)/2}
\]</span></p>
<p>The latter expression is called the <em>nucleotide diversity</em> <span class="citation" data-cites="nei_MolecularEvolutionPhylogenetics_2000">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nei &amp; Kumar, 2000, p. 251</a>)</span>.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="in-absence-of-evolutionary-forces-alleles-are-in-equilibrium" class="slide level2">
<h2>In absence of evolutionary forces alleles are in equilibrium</h2>
<h3 id="the-hardy-weinberg-equilibrium">The Hardy-Weinberg equilibrium</h3>
<p>For a locus, let <span class="math inline">\(A\)</span> and <span class="math inline">\(a\)</span> be two different alleles and let <span class="math inline">\(p\)</span> be the frequency of the <span class="math inline">\(A\)</span> allele and <span class="math inline">\(q=1-p\)</span> the frequency of the <span class="math inline">\(a\)</span> allele. In the <strong>absence</strong> of <strong>mutation</strong>, <strong>drift</strong>, <strong>migration</strong>, and other evolutionary processes, the equilibrium state is given by the <strong>Hardy-Weinberg equilibrium</strong> (<strong>HWE</strong>).</p>
<table>
<tbody>
<tr class="odd">
<td>Genotype:</td>
<td><span class="math inline">\(AA\)</span></td>
<td><span class="math inline">\(Aa\)</span></td>
<td><span class="math inline">\(aa\)</span></td>
</tr>
<tr class="even">
<td>Frequency:</td>
<td><span class="math inline">\(p^2\)</span></td>
<td><span class="math inline">\(2pq\)</span></td>
<td><span class="math inline">\(q^2\)</span></td>
</tr>
</tbody>
</table>
<p>Importantly, under the HWE assumption, we can calculate allele frequencies from genotype frequencies and vice versa:</p>
<p><span class="math display">\[
p = p^2 + pq\\
q = q^2 + pq
\]</span></p>
<aside class="notes">
<p>Segue: apart from describing the variation via e.g., diversity measures, we want to model how allele frequencies <em>change</em> in time. As <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004</a>, preface p.&nbsp;xi)</span> points out, “While genotype frequencies are easily measured, their change is not”</p>
<p>IOW: describing variation fine, but where does it come from and how does it change?</p>
<p>If assumptions of HWE hold, we have no change of variation. However, we <strong>want</strong> to look at <strong>change</strong> of variation and disentangle the forces that impose change</p>
<p>HWE assumption gives us a way to calculate allele frequencies from genotype frequencies. How well do these assumptions hold in real data? See next slide.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="natural-populations-do-mate-randomly" class="slide level2">
<h2>Natural populations do mate randomly?</h2>
<div class="cell" data-file="../../recipes/slides/foundations/hwe.R" data-hash="index_cache/revealjs/fig-hwe-1000genomes_57a5dc7d2d4ecdda7def670cd4f23dc1">
<div class="cell-output cell-output-error">
<pre><code>Error in file(file, "rt"): cannot open the connection</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in pivot_longer(x, cols = c("AA", "Aa", "aa"), names_to = "genotype"): object 'x' not found</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in by(x, x$population, function(y) {: object 'x' not found</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in ggplot(z, aes(x = p, y = value)): object 'z' not found</code></pre>
</div>
</div>
<aside class="notes">
<!-- markdownlint-disable MD013 -->
<p>See <a href="https://stackoverflow.com/questions/26587940/ggplot2-different-legend-symbols-for-points-and-lines" class="uri">https://stackoverflow.com/questions/26587940/ggplot2-different-legend-symbols-for-points-and-lines</a> for legend customization.</p>
<!-- markdownlint-enable MD013 -->
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-obsession-of-population-genetics" class="slide level2">
<h2>The obsession of population genetics</h2>
<p>Population genetics is about <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004</a>)</span></p>
<ol type="1">
<li>describing the genetic structure of populations</li>
<li>constructing theories on the forces that influence genetic variation</li>
</ol>
<div class="fragment">
<p>Questions to ponder:</p>
<ul>
<li>why does variation look the way it does?</li>
<li>how is variation maintained?</li>
<li>how does variation change over time (<span class="math inline">\(\Delta p\)</span>)?</li>
<li>what forces shape the genetic structure of <em>populations</em>?</li>
</ul>
</div>
<aside class="notes">
<p>From HWE: we want to look a the creation, maintenance and loss of variation and what forces affect it</p>
<p>Goal: describe theory behind evolving populations. This has been an obsession for a long time, even before DNA was known. We need to get back to basics.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="models-of-populations" class="title-slide slide level1 center">
<h1>Models of populations</h1>
<div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="index_files/figure-revealjs/wright-fisher-drift-models-of-population-1.svg" style="width:30.0%"></p>
</figure>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Segue: take a step back to the days before the nature of genetic inheritance was known. First focus on the population.</p>
<p>Goal of lecture is also to understand genetic relationships in the context of genealogies -&gt; coalescent.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<div>
<ul>
<li class="fragment">discrete and non-overlapping generations</li>
<li class="fragment">haploid individuals or two subpopulations (males and females)</li>
<li class="fragment">constant population size</li>
<li class="fragment">all individuals are equally fit</li>
<li class="fragment">population has no geographical or social structure</li>
<li class="fragment">no recombination</li>
</ul>
</div>
</div><div class="column" style="width:50%;">

</div>
</div>
<aside class="notes">
<p>See <span class="citation" data-cites="hein_GeneGenealogies_2004">Hein et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2004</a>)</span> for more assumptions. The second assumption means we can use 2N chromosome interchangeably for haploid (n=2N) and diploid (n=N) populations.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model-1" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-1-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-2" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-1">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-2-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-3" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-2">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3.0-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-4" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-3">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3.1-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-5" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-4">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3.1b-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-6" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-5">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3.2-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-7" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-6">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-8" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-7">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-3-untangled-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-9" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Model of populations that describes <strong>genealogical</strong> relationships of genes (chromosomes) in a population under the following assumptions <span class="citation" data-cites="hein2005gene">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hein et al., 2005</a>)</span>:</p>
<ul>
<li>discrete and non-overlapping generations</li>
<li>haploid individuals or two subpopulations (males and females)</li>
<li>constant population size</li>
<li>all individuals are equally fit</li>
<li>population has no geographical or social structure</li>
<li>no recombination</li>
</ul>
<h3 id="algorithm-8">Algorithm</h3>
<ol type="1">
<li>Setup starting population at time zero</li>
<li>Add offspring (same size) at time one</li>
<li>Select parents to offspring at random</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wf-4-1.svg" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="wright-fisher-model-10" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>

<img data-src="index_files/figure-revealjs/fig-wright-fisher-model-gganimation-.gif" width="14" class="r-stretch quarto-figure-left"><p class="caption">Figure&nbsp;1: Wright-Fisher model</p><aside class="notes">
<p>See <a href="https://mikeyharper.uk/animated-plots-with-r/" class="uri">https://mikeyharper.uk/animated-plots-with-r/</a></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="wright-fisher-model-11" class="slide level2 smaller">
<h2>Wright-Fisher model</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell" data-hash="index_cache/revealjs/fig-wright-fisher-model-gganimation-last_78fe16f7075885cb9b994f809091dd66">
<div class="cell-output-display">
<div id="fig-wright-fisher-model-gganimation-last" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;2: WF model indicating time direction from past (top) to present (bottom).</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wright-fisher-model-gganimation-last-1.svg" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<div class="fragment">
<div class="cell" data-hash="index_cache/revealjs/fig-wf-model-genealogy_559de7c785316a6ebd0c6023db219f87">
<div class="cell-output-display">
<div id="fig-wf-model-genealogy" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;3: WF model tracing the genealogies of three extant chromosomes</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-model-genealogy-1.svg" width="960"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="the-wright-fisher-sampling-model" class="slide level2">
<h2>The Wright-Fisher sampling model</h2>
<p>Let’s formalise the sampling process of the Wright-Fisher model<sup>1</sup>. We assume</p>
<ol type="1">
<li>a single locus in a haploid population of size <span class="math inline">\(2N\)</span> (or diploid of size <span class="math inline">\(N\)</span> when random mating)</li>
<li>no mutation and selection</li>
<li>discrete generations</li>
</ol>
<p>Each generation we sample <span class="math inline">\(2N\)</span> new chromosomes from the previous generation. The probability of choosing a chromosome <span class="math inline">\(v\)</span> is <span class="math inline">\(1/2N\)</span> (coin flip with probability of success <span class="math inline">\(1/2N\)</span>). Since the trials are independent, and we perform <span class="math inline">\(2N\)</span> trials, the number of offspring <span class="math inline">\(k\)</span> of a given chromosome <span class="math inline">\(v\)</span> is <a href="https://en.wikipedia.org/wiki/Binomial_distribution"><em>binomially distributed</em></a> <span class="math inline">\(\mathrm{Bin}(m, p)\)</span>, with parameters <span class="math inline">\(m=2N\)</span> and probability of success <span class="math inline">\(p=\frac{1}{2N}\)</span>:</p>
<p><span class="math display">\[
P(v=k) = {2N\choose k}\left( \frac{1}{2N} \right)^k
\left(1 - \frac{1}{2N} \right)^{(2N - k)}
\]</span></p>
<aside class="notes">
<p>We assume some familiarity with the Binomial distribution. Some mathematical formality follows.</p>
<p>Let <span class="math inline">\(p\)</span> be the probability of heads (=1) of a coin flip, and <span class="math inline">\(q=1-p\)</span> the probability of tails (=0). A coin flip can be modelled by a random variable (r.v.) <span class="math inline">\(X\)</span> that follows a <em>Bernoulli distribution</em>, where <span class="math inline">\(\mathrm{Pr(}X=1\mathrm{)}=p\)</span>, <span class="math inline">\(\mathrm{Pr(}X=0\mathrm{)}=1-p=q\)</span>. One can show (e.g., with transforms such as the <a href="https://en.wikipedia.org/wiki/Probability-generating_function">probability-generating function</a>) that a binomially distributed variable <span class="math inline">\(Y\sim Bin(n, p)\)</span> is the sum of <span class="math inline">\(n\)</span> independent Bernoulli variables <span class="math inline">\(X\)</span> (“Bernoulli trials”), or <span class="math inline">\(Y=\sum^nX\)</span>. In our case, there are <span class="math inline">\(2N\)</span> trials, each with <span class="math inline">\(p=1/2N\)</span>.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<aside><ol class="aside-footnotes"><li id="fn1"><p>For a more extensive treatment, see <span class="citation" data-cites="hermisson_MathematicalPopulationGenetics_2017">Hermisson (<a href="#/bibliography" role="doc-biblioref" onclick="">2017</a>)</span> or <span class="citation" data-cites="hein2005gene">Hein et al. (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span></p></li></ol></aside></section>
<section id="properties-of-wright-fisher-sampling" class="slide level2">
<h2>Properties of Wright-Fisher sampling</h2>
<h3 id="the-expected-number-of-offspring-is-one">The expected number of offspring is one</h3>
<p><br></p>

<img data-src="index_files/figure-revealjs/wf-expected-number-of-offspring-1.svg" width="960" class="r-stretch quarto-figure-center"><p><br></p>
<p>The expected value of a <span class="math inline">\(\mathrm{Bin}(m,p)\)</span> (binomially distributed) variable with parameters <span class="math inline">\(m=2N\)</span> and <span class="math inline">\(p=\frac{1}{2N}\)</span> is <span class="math inline">\(mp\)</span>, hence:</p>
<p><span class="math display">\[
E(v) = mp = 2N\frac{1}{2N} = 1
\]</span></p>
</section>
<section id="properties-of-wright-fisher-sampling-1" class="slide level2">
<h2>Properties of Wright-Fisher sampling</h2>
<h3 id="poisson-approximation-for-large-n">Poisson approximation for large <span class="math inline">\(N\)</span></h3>
<p><br></p>

<img data-src="index_files/figure-revealjs/wf-poisson-approximation-large-n-1.svg" width="960" class="r-stretch quarto-figure-center"><p><br></p>
<p>When <span class="math inline">\(2N\)</span> large it holds that the probability that sequence <span class="math inline">\(v\)</span> has <span class="math inline">\(k\)</span> offspring is</p>
<p><span class="math display">\[
P(v=k) \approx \frac{1}{k!}e^{-k}
\]</span></p>
<p>Since <span class="math inline">\(P(v=0) = e^{-1} \approx 0.37\)</span>, a fraction 0.37 of sequences lack descendants</p>
</section>
<section id="properties-of-wright-fisher-sampling-2" class="slide level2">
<h2>Properties of Wright-Fisher sampling</h2>
<h3 id="sec-prob-sequence-pick-same-parent">Probability that two sequences pick same parent</h3>

<img data-src="index_files/figure-revealjs/wf-prob-parents-identical-fig-1.svg" width="800" class="r-stretch quarto-figure-center"><p>Let <span class="math inline">\(u\)</span> and <span class="math inline">\(v\)</span> be two sequences. Pick a parent <span class="math inline">\(p_u\)</span> of <span class="math inline">\(u\)</span>. Then the probability that <span class="math inline">\(v\)</span> picks the same parent is (solid lines)</p>
<p><span class="math display">\[
P(p_u=p_v) = \frac{1}{2N}
\]</span></p>
<p>The probability that they pick different parents is</p>
<p><span class="math display">\[
P(p_u \neq p_v) = 1 - \frac{1}{2N}
\]</span></p>
</section>
<section id="properties-of-wright-fisher-sampling-3" class="slide level2 smaller">
<h2>Properties of Wright-Fisher sampling</h2>
<h3 id="time-for-two-sequences-to-coalesce">Time for two sequences to coalesce</h3>
<div class="columns">
<div class="column" style="width:25%;">

</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-wf-two-chrom-find-common-ancestor_82ba764843e6b721f6e4621fc4653bb9">
<div class="cell-output-display">
<div id="fig-wf-two-chrom-find-common-ancestor" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;4: Illustration of two sequences finding a common ancestor after five generations; four failures with probability <span class="math inline">\(1-1/2N\)</span>, one success with probability 1/2N</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-two-chrom-find-common-ancestor-1.svg" style="width:70.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:25%;">

</div>
</div>
<p>Time for two sequences <span class="math inline">\(u\)</span> and <span class="math inline">\(v\)</span> to find common ancestor distributed as <span class="math inline">\((1 - \frac{1}{2N})^{j-1}\frac{1}{2N}\)</span> (<span class="math inline">\(j-1\)</span> failures followed by success). This is the <a href="https://en.wikipedia.org/wiki/Geometric_distribution">geometric distribution</a> <span class="math inline">\(Ge(p)\)</span>, with parameter <span class="math inline">\(p=\frac{1}{2N}\)</span>, and expected value <span class="math inline">\(\frac{1}{p}={2N}\)</span>. That is, the <strong>expected number of generations for two sequences to find a common ancestor (i.e., coalesce) is <span class="math inline">\(2N\)</span> generations</strong>.</p>
<aside class="notes">
<p>Since a large fraction of genes lack descendants, very quickly (compared to 2N) a population will descend from a small proportion of genes.</p>
<p>Derivation of average time to having same parent (i.e., coalescence) requires knowledge of geometric distribution. The probability that the genes find common parent <span class="math inline">\(j\)</span> generations ago is</p>
<p><span class="math display">\[
(1-\frac{1}{2N})^{(j-1)}\frac{1}{2N}
\]</span></p>
<p>due to independence between generations. This is the geometric distribution <span class="math inline">\(Ge(p)\)</span> with <span class="math inline">\(p=1/2N\)</span>, which has expected value <span class="math inline">\(1/p\)</span>, i.e., 2N in this case.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="origin-and-change-of-variation" class="title-slide slide level1 center">
<h1>Origin and change of variation</h1>
<div style="display: grid; grid-template-columns: 300px 200px 200px; grid-row-gap: 0px; grid-column-gap: 100px;">
<div>
&nbsp;
</div>
<div class="circular-image">
<p>Mutation</p>
<p><img data-src="assets/images/mutation.jpg"></p>
</div>
<div class="circular-image">
<p>Selection</p>
<p><img data-src="assets/images/finches.jpg"></p>
</div>
<div>
&nbsp;
</div>
<div class="circular-image">
<p>Recombination</p>
<p><img data-src="assets/images/recombination.jpeg"></p>
</div>
<div class="circular-image">
<p>Drift</p>
<p><img data-src="assets/images/genetic-drift.png"></p>
</div>
</div>
</section>
<section id="wright-fisher-model-with-alleles" class="slide level2">
<h2>Wright-Fisher model with alleles</h2>
<div class="columns compact">
<div class="column" style="width:35%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wright-fisher-drift-0.3-1-1.svg" width="960"></p>
</div>
</div>
</div><div class="column fragment" style="width:35%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/wright-fisher-drift-0.3-2-1.svg" width="960"></p>
</div>
</div>
</div><div class="column fragment" style="width:30%;">
<p>Alleles can randomly <em>fix</em> or be <em>lost</em> through process called <strong>genetic drift</strong></p>
</div><p>Wright-Fisher model showing the evolution of population of 10 genes over 16 generations. Allele variants are shown in white and black. Starting frequency black variant is 0.3.</p>
</div>
<aside class="notes">
<p>(To avoid confusion: previously black was used to indicate parent, here black/white will refer to different alleles)</p>
<p>Enter allelic variants; we are looking at two alleles</p>
<p>Important: the fate of any allele under drift is to be <strong>lost</strong> or <strong>fixed</strong>, which ultimately means variation is <strong>lost</strong> (decays) with drift</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="sampling-model-again" class="slide level2">
<h2>Sampling model again</h2>

<img data-src="index_files/figure-revealjs/allele-sampling-model-1.svg" width="1152" class="r-stretch quarto-figure-center"><p>We assume two gene variants <span class="math inline">\(A\)</span>, <span class="math inline">\(a\)</span>, each with <span class="math inline">\(i\)</span> and <span class="math inline">\(j=2N-i\)</span> copies in generation <span class="math inline">\(t\)</span>. Let <span class="math inline">\(p_t=i/2N\)</span> be the <em>frequency</em> of <span class="math inline">\(A\)</span> in generation <span class="math inline">\(t\)</span>, and <span class="math inline">\(q_t=1-p_t\)</span> the frequency of <span class="math inline">\(a\)</span>.</p>
<p>The sampling model for next generation <span class="math inline">\(t+1\)</span>, <em>given</em> <span class="math inline">\(i\)</span> <span class="math inline">\(A\)</span> alleles in generation <span class="math inline">\(t\)</span>, is then</p>
<p><span class="math display">\[
p_{t+1} = P(v_A=k) = {2N\choose k}\left( \frac{i}{2N} \right)^k
\left(1 - \frac{i}{2N} \right)^{(2N - k)} = \mathsf{Bin}(2N, \frac{i}{2N})
\]</span></p>
<aside class="notes">
<p>NB: we here slightly change notation, with <span class="math inline">\(p_{t+1} = P(X=k)\)</span>. Also, we have omitted an index <span class="math inline">\(i\)</span> in the expressions of <span class="math inline">\(p\)</span>. The probability of observing <span class="math inline">\(k\)</span> <span class="math inline">\(A\)</span> alleles in generation <span class="math inline">\(t+1\)</span> will of course depend on the number <span class="math inline">\(i\)</span> <span class="math inline">\(A\)</span> alleles in the previous generation.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="probability-distributions-of-allele-frequencies" class="slide level2">
<h2>Probability distributions of allele frequencies</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell" data-hash="index_cache/revealjs/fig-allele-frequency-histogram-plot_6d6e16f0a5ea89795c59f75bd1d26423">
<div class="cell-output-display">
<div id="fig-allele-frequency-histogram-plot" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;5: Histogram showing the course of change of the allele frequency distribution with time <span class="citation" data-cites="kimura_1983">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kimura, 1983, fig. 3.4</a>)</span>. When N large (<span class="math inline">\(\gtrsim 100\)</span>) histogram can be approximated by continuous distribution (diffusion theory). Try recipe <a href="../../recipes/slides/index.html#sec-recipe-slides-allele-frequency-distribution"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a> for different values of N.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-allele-frequency-histogram-plot-1.svg" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-buri-table-14_5562a47ebfc60c19ee98c9fdd9ab108b">
<div class="cell-output-display">
<div id="fig-buri-table-14" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;6: Frequency distributions of the brown eye (<span class="math inline">\(bw^{75}\)</span>) allele in replicate experimental populations (<span class="math inline">\(n\sim 100\)</span>) of <em>Drosophila melanogaster</em> (8 <i class="fa-solid fa-venus" aria-label="venus"></i>, 8 <i class="fa-solid fa-mars" aria-label="mars"></i>) <span class="citation" data-cites="buri_GeneFrequencySmall_1956">(<a href="#/bibliography" role="doc-biblioref" onclick="">Buri, 1956</a>)</span></figcaption>
<p><img data-src="index_files/figure-revealjs/fig-buri-table-14-1.svg" style="width:70.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Kimura’s plot illustrates the allelic frequency distribution of replicate populations each consisting of N=10 sequences. There are two allelic types. The x axis corresponds to the proportion of populations in a given state (e.g., the y value for x=3/10 corresponds to the proportion of populations with 3 alleles of one type and 7 of the other). At <span class="math inline">\(t=0\)</span> all populations are in state 5/10.</p>
<p>Kimura’s plot is very instructive and students are encouraged to test the recipe code, increasing the number of states incrementally. For large enough N, the histograms can be approximated by continuous distributions. This observation led Kimura (even though diffusion equations were originally introduced to genetics by Fisher in 1922) to apply diffusion theory to obtain probability densities of allele frequencies, leading in turn to compact expressions of fixation probabilities, expected ages of alleles, and more. A treatment of diffusion theory is outside the scope of this course; the interested reader can consult e.g., <span class="citation" data-cites="ewens_MathematicalPopulationGenetics_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Ewens, 2004</a>)</span>.</p>
<p>The Buri experiment is an empirical demonstration of Kimura’s plot.</p>
<p>On the sampling model, <span class="citation" data-cites="charlesworth_ElementsEvolutionaryGenetics_2010">(<a href="#/bibliography" role="doc-biblioref" onclick="">Charlesworth &amp; Charlesworth, 2010, p. 231</a>)</span> says:</p>
<blockquote>
<p>It is, however, impossible to write down a simple algebraic expression for P, even without selection and mutation. [The equation] is useful for obtaining numerical results for relatively small populations, but becomes computationally demanding when N becomes very large.</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genetic-drift" class="slide level2">
<h2>Genetic drift</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>To capture dynamics, follow allele frequency trajectory as function of time.</p>
<div class="smallr">
<div class="cell" data-hash="index_cache/revealjs/wright-fisher-source_fc6a9d0852d57471ad01989470464bdb">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="do">##' Wright Fisher model - follow allele frequency distribution</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="do">##'</span></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="do">##' @param p0 Starting frequency</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="do">##' @param n Population size</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="do">##' @param generations Number of generations to simulate</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="do">##'</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>wright_fisher <span class="ot">&lt;-</span> <span class="cf">function</span>(p0, n, generations) {</span>
<span id="cb6-8"><a href="#cb6-8"></a>    x <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="at">mode =</span> <span class="st">"numeric"</span>, <span class="at">length =</span> generations)</span>
<span id="cb6-9"><a href="#cb6-9"></a>    x[<span class="dv">1</span>] <span class="ot">&lt;-</span> p0</span>
<span id="cb6-10"><a href="#cb6-10"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="dv">2</span>, <span class="fu">length</span>(x))) {</span>
<span id="cb6-11"><a href="#cb6-11"></a>        x[i] <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="at">size =</span> n, <span class="at">prob =</span> x[i <span class="sc">-</span> <span class="dv">1</span>])<span class="sc">/</span>n</span>
<span id="cb6-12"><a href="#cb6-12"></a>    }</span>
<span id="cb6-13"><a href="#cb6-13"></a>    x</span>
<span id="cb6-14"><a href="#cb6-14"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="index_cache/revealjs/wright-fisher-plot_6499baaaee1d16b3debdc278a724c026">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># Example simulation and plot</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="fu">set.seed</span>(<span class="dv">1223</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a>generations <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span>  <span class="co"># NB: haploid population size!</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span>generations, <span class="fu">wright_fisher</span>(<span class="fl">0.5</span>, n, generations), <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">ylab =</span> <span class="st">"frequency"</span>,</span>
<span id="cb7-6"><a href="#cb7-6"></a>    <span class="at">xlab =</span> <span class="st">"generation"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="cell" data-animation.hook="gifski" data-hash="index_cache/revealjs/fig-genetic-drift-N-10_92895623153bceedc7a63fb5f59e9b2f">
<div class="cell-output-display">
<div id="fig-genetic-drift-N-10" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;7: Genetic drift for different haploid(!) population sizes, starting frequency <span class="math inline">\(p_0\)</span>=0.5. Note dependency of variance on population size N.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-genetic-drift-N-10-1.gif" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Mention <em>neutral</em> here as this is how alleles behave under drift: like gas particle diffusing up and down. Allele is eventually either fixed or lost.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="genetic-drift-1" class="slide level2">
<h2>Genetic drift</h2>
<div class="columns">
<div class="column" style="width:50%;">
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/fig-genetic-drift-many-repetitions_e77d7c2b8179e5e687200e1a47a30dfd">
<div class="cell-output-display">
<div id="fig-genetic-drift-many-repetitions" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;8: Genetic drift for different combinations of starting frequency and population size for n=50 repetitions per parameter combination. Note how variation and time to fixation depends on population size and starting frequency.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-genetic-drift-many-repetitions-1.svg" width="560" height="380"></p>
</figure>
</div>
</div>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:50%;">
<div>
<ul>
<li class="fragment">fate of allele: fixation or loss <span class="math inline">\(\rightarrow\)</span> eventually loss of variation</li>
<li class="fragment"><strong>probability of fixation <span class="math inline">\(\pi(p)=p\)</span></strong>, where <span class="math inline">\(p\)</span> is the <em>current</em> frequency</li>
<li class="fragment">rate of drift (loss of variation) <span class="math inline">\(\propto \frac{1}{2N}\)</span></li>
</ul>
</div>
<div class="fragment">
<div class="center">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/gas-particle-drift_674263fbb995b2e89763594279313bd6">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="index_files/figure-revealjs/gas-particle-drift-1.svg" width="960" height="200"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<p>Mathematical treatment of drift can become complicated: easier to study dynamics of <em>heterozygosity</em></p>
</div>
<aside class="notes">
<p>Simplifying assumptions -&gt; rate of drift <span class="math inline">\(\propto \frac{1}{N}\)</span> (<span class="math inline">\(\frac{1}{2N}\)</span> in our treatment) <span class="citation" data-cites="leffler_RevisitingOldRiddle_2012">(<a href="#/bibliography" role="doc-biblioref" onclick="">Leffler et al., 2012</a>)</span></p>
<p>In reality population fluctuates and therefore Ne is substituted for N. Larger Ne -&gt; smaller fluctuation in allele frequency (as evidenced in plots) -&gt; maintain larger genetic diversity</p>
<p>Drift can be thought of as a force acting on a buoyant gas particle floating up and down at random.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="heterozygosity-dynamics" class="slide level2">
<h2>Heterozygosity dynamics</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-hash="index_cache/revealjs/fig-identity-by-state-and-descent_68793df2e3436679ae674958c36e2819">
<div class="cell-output-display">
<div id="fig-identity-by-state-and-descent" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;9: Illustration of identity by descent (IBD) and state (IBS). Alleles in generation <span class="math inline">\(n\)</span> are IBD but not IBS.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-identity-by-state-and-descent-1.svg" style="width:70.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p>Let <span class="math inline">\(\mathcal{H}_t\)</span> be the probability that two alleles are <strong>different by state</strong>. One can show that the time course evolution of <span class="math inline">\(\mathcal{H}_t\)</span> in a randomly mating population consisting of <span class="math inline">\(N\)</span> diploid hermaphroditic individuals is</p>
<p><span class="math display">\[
\mathcal{H}_t = \mathcal{H}_0 \left( 1 - \frac{1}{2N} \right)^t
\]</span></p>
<p>Important consequence: heterozygosity in WF population lost at rate <span class="math inline">\(1/2N\)</span>.</p>
</div>
</div>
<aside class="notes">
<p>Alleles in generation <span class="math inline">\(n\)</span> are IDB but not IBS</p>
<p>Motivation: mathematical description of genetic drift complicated for populations with more than one individual. Easier to study the evolution of heterozygosity.</p>
<p>Equation presented without details. Derivation in <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004, pp. 25–27</a>)</span> starts from G=probability that two alleles are IBS</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="heterozygosity-dynamics-1" class="slide level2">
<h2>Heterozygosity dynamics</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell" data-hash="index_cache/revealjs/fig-gillespie-2.4_6c1dc4ae053a5395b90d141b2393a73f">
<div class="cell-output-display">
<div id="fig-gillespie-2.4" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;10: Plot of <span class="math inline">\(\mathcal{H}_t\)</span> illustrating dependency on population size</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-gillespie-2.4-1.svg" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell" data-hash="index_cache/revealjs/fig-black-footed-ferret_4d33173619ef0c3b86a392b2b7b0f661">
<div class="cell-output-display">
<div id="fig-black-footed-ferret" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;11: Heterozygosity in black-footed ferret <span class="citation" data-cites="wisely_GeneticDiversityFitness_2002">(<a href="#/bibliography" role="doc-biblioref" onclick="">Wisely et al., 2002</a>)</span>. Example from <span class="citation" data-cites="grahamcoop_NotesPopulationGenetics_2020">Graham Coop (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span>, Fig. 4.5</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-black-footed-ferret-1.svg" style="width:90.0%"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:40%;">
<p><img data-src="assets/images/ferret.svg" style="width:40.0%"></p>
<p>Example of how rapid decline in population size can affect heterozygosity.</p>
<div class="fragment">
<p>Population size influences genetic diversity!</p>
</div>
<div class="fragment">
<p>However, <em>census</em> population size not (always) the correct measure.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Dependency on population size: for large enough populations the decline will be <strong>very</strong> slow (drift speed ~ 1/2N)</p>
<p>Practical example shows loss of heterozygosity tell-tale signature of population decline; conversely, <strong>not</strong> easy to show population decline in large populations (e.g., marine species with large <span class="math inline">\(N_e\)</span>) using heterozygosity as measure</p>
<p><span class="citation" data-cites="barton2007evolution">(<a href="#/bibliography" role="doc-biblioref" onclick="">Barton et al., 2007, p. 369</a>)</span> “The relation between genetic diversity and population size is difficult to discern, in part, because it is extremely hard to estimate the numbers of most species and because the number that matters is an average back into the distant past”(!)</p>
<p>From <span class="citation" data-cites="grahamcoop_NotesPopulationGenetics_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Graham Coop, 2020, p. 64</a>)</span></p>
<blockquote>
<p>To see how a decline in population size can affect levels of het- erozygosity, let’s consider the case of black-footed ferrets (Mustela nigripes). The black-footed ferret population has declined dramatically through the twentieth century due to destruction of their habitat and sylvatic plague. In 1979, when the last known black-footed ferret died in captivity, they were thought to be extinct. In 1981, a very small wild population was rediscovered (40 individuals), but in 1985 this population suffered a number of disease outbreaks. At that point of the 18 remaining wild individuals were brought into captivity, 7 of which reproduced. Thanks to intense captive breeding efforts and conservation work, a wild population of over 300 individuals has been established since. However, because all of these individuals are descended from those 7 individuals who survived the bottleneck, diversity levels remain low. Wisely et al.&nbsp;measured heterozygosity at a number of microsatellites in individuals from museum collections, showing the sharp drop in diversity as population sizes crashed (see Figure 4.5).</p>
</blockquote>
<p>Segue: population size important; however, <em>census</em> population size is not always the measure we want when relating to genetic diversity</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="effective-population-size" class="slide level2">
<h2>Effective population size</h2>
<p>Assumptions underlying Wright-Fisher model seldom fulfilled for natural populations. In particular</p>
<ul>
<li>non-random mating (population structure)</li>
<li>fluctuations of population census size</li>
</ul>
<p>Therefore, magnitude of drift experienced by a population different from that predicted by population size</p>
<p>Technically correct definition (but see <span class="citation" data-cites="waples_WhatNeAnyway_2022">Waples (<a href="#/bibliography" role="doc-biblioref" onclick="">2022</a>)</span>):</p>
<blockquote>
<p><span class="math inline">\(N_e\)</span> is the size of an ideal population that would experience the same rate of genetic drift as the population in question.</p>
</blockquote>
<aside class="notes">
<p>ideal population: discrete generations with random mating, no evolutionary forces besides drift, no selective advantages</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="mutation" class="slide level2">
<h2>Mutation</h2>
<div class="columns">
<div class="column" style="width:30%;">

<table>
<tbody><tr>
<td>
<div style="transform: translate(50px, 0px);">
<i class="fa-solid fa-bolt fa-2x fa-rotate-90"></i>
</div>
</td>
<td>
<div style="transform: translate(0px, 40px);">
<i class="fa-solid fa-dna fa-5x"></i>
</div>
</td>
</tr>

</tbody></table>
</div><div class="column" style="width:70%;">
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="index_files/figure-revealjs/allele-sampling-model-mutation-1.svg" width="1152"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:70%;">
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/mutation-models_ae9c567c3453d82184cb7050ec753bce">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/mutation-models-1.svg" width="960"></p>
</div>
</div>
<!-- markdownlint-enable MD013 -->
</div><div class="column" style="width:30%;">
<dl>
<dt>Two-allele</dt>
<dd>
Derive popgen stats
</dd>
<dt>Finite sites</dt>
<dd>
Recurrent mutations
</dd>
<dt>Infinite alleles</dt>
<dd>
Protein electrophoresis
</dd>
<dt>Inifinite sites</dt>
<dd>
DNA sequences
</dd>
</dl>
</div>
</div>
</div>
<aside class="notes">
<p>Mutations are generated randomly as changes in DNA. In WF model will be highlighted in black. We assume mutations are described by a Poisson process with rate <span class="math inline">\(\mu\)</span> (per generation).</p>
<p>The mutation models are:</p>
<ul>
<li>two-allele model: used to derive popgen statistics</li>
<li>finite sites model: recurrent mutations may occur, an assumption that is often omitted to facilitate calculations (e.g., coalescent)</li>
<li>inifinite alleles model: every mutation leads to new allele (e.g., protein electrophoresis)</li>
<li>infinite sites model: model commonly used for long DNA sequences</li>
</ul>
<p>Additional model not presented here is the stepwise mutation model (mainly used to model repeats):</p>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/stepwise-mutation-model_89d608c5a4dcd526167ec279c52e1b92">
<div class="cell-output-display">
<p><img data-src="index_files/figure-revealjs/stepwise-mutation-model-1.svg" width="1000"></p>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="mutation-and-drift" class="slide level2">
<h2>Mutation and drift</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p>Genetic drift “moves” frequencies to the point that variation is lost via allele fixation or loss. New variation is introduced through <strong>mutation</strong>. We typically assume mutations are described by a <a href="https://en.wikipedia.org/wiki/Poisson_point_process">Poisson process</a> with rate <span class="math inline">\(\mu\)</span> (per generation).</p>
<div class="fragment">
<p>The <strong>mutation rate</strong> is denoted <span class="math inline">\(\mu\)</span>, and the <strong>population scaled</strong> mutation rate is <span class="math inline">\(2N_e\mu\)</span> for haploid populations, <span class="math inline">\(4N_e\mu\)</span> for diploid, where <span class="math inline">\(N_e\)</span> is the <strong>effective population size</strong>.</p>
</div>
<div class="fragment">
<p>The <strong>mutation - drift balance</strong> is when the diversity lost due to drift equals the diversity gained due to mutation.</p>
</div>
</div><div class="column" style="width:40%;">
<div class="fragment">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-wf-mutation_fd89915487e293c2815fe75ec9e3cfaf">
<div class="cell-output-display">
<div id="fig-wf-mutation" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;12: Variation is introduced by mutations (black) at rate <span class="math inline">\(\mu=1e^{-4}\)</span> and is occasionally lost through genetic drift.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-wf-mutation-1.svg" width="400" height="500"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Segue: so if drift removes variation, what introduces it? Mutation!</p>
<p>Note that the mutation rate in the WF model is not meaningful; the value is simply chosen such that enough mutations show up in plot.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="tracing-the-evolution-of-mutations" class="slide level2">
<h2>Tracing the evolution of mutations</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-tracing-evolution-of-mutations_65fe57a21cd6c2bebc032a2dfb986d98">
<div class="cell-output-display">
<div id="fig-tracing-evolution-of-mutations" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;13: Different mutations suffer different fates. Most mutations are lost in a couple of generations. Mutant alleles are colored black and their genealogies are highlighted with thicker edges.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-tracing-evolution-of-mutations-1.svg" width="500" height="500"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Observation: most mutations are in fact lost</p>
</div>
<div class="fragment">
<p>Recall: fixation probability <span class="math inline">\(\pi(p)=p\)</span></p>
</div>
</div>
</div>
<aside class="notes">
<p>Difference to previous figure: here we highlight the genealogies</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="mutation-drift-balance" class="slide level2">
<h2>Mutation drift balance</h2>
<p>Drift removes variation. Mutation reintroduces it. At equilibrium the change in variation by definition is 0. In terms of <span class="math inline">\(\mathcal{H}_t\)</span> (the probability that two alleles are <em>not</em> identical by state), <span class="math inline">\(\Delta\mathcal{H}=0\)</span>.</p>
<p>One can show<sup>1</sup> the classical formula that the equilibrium <strong>heterozygosity</strong> value is</p>
<p><span class="math display">\[
\hat{\mathcal{H}} = \frac{4N_e\mu}{1 + 4N_e\mu}
\]</span></p>
<div class="fragment">
<p><span class="math inline">\(\mu\)</span> is often assumed known, and heterozygosity is easily calculated from data, which provides a way of estimating <span class="math inline">\(N_e\)</span>.</p>
</div>
<div class="fragment">
<p>The compound parameter <span class="math inline">\(4N_e\mu\)</span> is called the <strong>population scaled mutation rate</strong> and is commonly named <span class="math inline">\(\theta\)</span> such that</p>
<p><span class="math display">\[
\hat{\mathcal{H}} = \frac{\theta}{1 + \theta}
\]</span></p>
</div>
<aside class="notes">
<p><span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">Gillespie (<a href="#/bibliography" role="doc-biblioref" onclick="">2004</a>)</span>, pp.&nbsp;30–31 uses a <em>difference equation</em> approach to derive <span class="math inline">\(\mathcal{H}\)</span>. Briefly, he studies the time evolution of <span class="math inline">\(\mathcal{G}\)</span>, the probability that two alleles drawn at random without replacement from the population are identical by state. Mutations are assumed unique, i.e., the <em>infinite-alleles</em> model. It holds</p>
<p><span class="math display">\[
\mathcal{G}^\prime = (1-\mu)^2\left[ \frac{1}{2N} +
\left( 1 - \frac{1}{2N} \right) \mathcal{G} \right]
\]</span></p>
<p>where <span class="math inline">\((1-\mu)^2\)</span> is the probability that no mutation occur in either of the two sampled alleles. Since <span class="math inline">\(\mu\)</span> is small, <span class="math inline">\((1-\mu)^2 \approx 1-2\mu\)</span>, which after some manipulation gives the desired expression for <span class="math inline">\(\mathcal{H} = 1 - \mathcal{G}\)</span>.</p>
<p>On pages 46–47, he shows that the expression for can be derived in a much simpler fashion using coalescent theory. Tracing two lineages backwards in time, the probability of coalescence is 1/2N, whereas the probability of a mutation is <span class="math inline">\(1-(1-\mu)^2\approx 2\mu\)</span>; <span class="math inline">\(\mathcal{H}\)</span> is then simply the relative probability of the two events</p>
<p><span class="math display">\[
\mathcal{H} = \frac{2\mu}{2\mu + 1/2N} = \frac{4N\mu}{4N\mu + 1}
\]</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<aside><ol class="aside-footnotes"><li id="fn2"><p>see <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004, p. 29</a> – 31)</span> for a concise treatement</p></li></ol></aside></section>
<section id="the-neutral-theory-of-evolution" class="slide level2">
<h2>The neutral theory of evolution</h2>
<div class="columns">
<div class="column" style="width:40%;">
<p>Mutation drift balance, together with the observation during 50’s-60’s that polymorphism was more common than expected, is the foundation of the <strong>neutral theory</strong> of evolution: allele frequencies may change and fix due to chance alone and not selection.</p>
<div class="fragment">
<p><strong>Nearly neutral theory</strong> later developed to explain failure to predict scaling of polymorphism with population size.</p>
</div>
</div><div class="column" style="width:60%;">
<div class="cell" data-hash="index_cache/revealjs/fig-heterozygosity-function-ne_d3b39bdf13e8b5ad73f0303a5f0fb4ec">
<div class="cell-output-display">
<div id="fig-heterozygosity-function-ne" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;14: Heterozygosity predicted by the neutral theory. Shaded region shows typical heterozygosities in animals. The observed <span class="math inline">\(N_e\mu\)</span> range is higher than predicted from plot. From <span class="citation" data-cites="hurst_GeneticsUnderstandingSelection_2009">Hurst (<a href="#/bibliography" role="doc-biblioref" onclick="">2009</a>)</span>, Fig 1.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-heterozygosity-function-ne-1.svg" style="width:60.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Related to rate of substitution and molecular evolution is the work of Kimura that lead to the development of the neutral theory.</p>
<p>Motivation: if polymorphic sites deleterious, should not expect much polymorphism.</p>
<p>Low levels of polymorphism expected assuming little balancing selection <span class="citation" data-cites="hurst_GeneticsUnderstandingSelection_2009">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hurst, 2009, p. 87</a>)</span>; however electrophoretic studies showed polymorphism common. Would lead to detrimental load <span class="citation" data-cites="kimura_ProteinPolymorphismPhase_1971">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kimura &amp; Ohta, 1971</a>)</span> -&gt; therefore majority of polymorphism must evolve neutrally (dynamics). Also: rate of evolution (on protein level) too high (Haldane’s dilemma)</p>
<p>On the shaded region: the observed range of <span class="math inline">\(N_e\mu\)</span> is larger than that which is predicted by the plot, and since <span class="math inline">\(\mu\)</span> is constrained within a couple of orders of magnitude, <span class="math inline">\(N_e\)</span> must vary more than predicted by the (strictly) neutral theory.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="mutation-rate-can-be-estimated-from-substitution-rate" class="slide level2">
<h2>Mutation rate can be estimated from substitution rate</h2>
<p>Mutation enters populations and may be fixed by drift. Therefore, with time there will be fixed differences, or <strong>substitions</strong> (typically in the evolution of species) between populations, or species. In molecular evolution, the <strong>substition rate</strong>, <strong><span class="math inline">\(\rho\)</span></strong>, is the most interesting quantity.</p>
<div class="fragment">
<p>The total number of new mutations in every generation is <span class="math inline">\(2N\mu\)</span> (total number of gametes times mutation rate)</p>
</div>
<div class="fragment">
<p>New mutations fix at a rate <span class="math inline">\(1/2N\)</span></p>
</div>
<div class="fragment">
<p>Therefore, the <strong>average rate of substitution</strong>, <strong><span class="math inline">\(\rho\)</span></strong>, is <span class="math inline">\(2N\mu\times1/2N\)</span>, or</p>
<p><span class="math display">\[
\rho=\mu
\]</span></p>
<p>which is independent of population size!</p>
</div>
<div class="fragment">
<p>Practical implication: we can <em>estimate mutation rate from the substitution rate</em> at neutrally evolving sites (e.g., <span class="citation" data-cites="kumar_MutationRatesMammalian_2002">Kumar &amp; Subramanian (<a href="#/bibliography" role="doc-biblioref" onclick="">2002</a>)</span>)</p>
</div>
<aside class="notes">
<p>What happens when sufficient amount of time passes? Mutations <strong>fix</strong> and generate <strong>substitutions</strong> between species. This slide presents Kimura’s remarkable result regarding mutation rate and substitution rate!</p>
<p>On <span class="math inline">\(2N\mu\)</span>: the larger the population, the larger the number of mutations that can fix</p>
<p>If <span class="math inline">\(4N\mu&gt;&gt;0\)</span> no new mutations can fix. Issue here is there is no explicit mention of DNA sequence -&gt; introduction of infinite sites model</p>
<p>For practical note on estimation of mutation rate see e.g., <span class="citation" data-cites="hubisz_InferenceAncestralRecombination_2020">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hubisz &amp; Siepel, 2020, p. 247</a>)</span></p>
<p><span class="citation" data-cites="kumar_MutationRatesMammalian_2002">(<a href="#/bibliography" role="doc-biblioref" onclick="">Kumar &amp; Subramanian, 2002</a>, first paragraph)</span>:</p>
<blockquote>
<p>Rates of point mutation can be determined indirectly by estimating the rate at which the neutral substitutions accumulate in protein-coding genes</p>
</blockquote>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="selection" class="title-slide slide level1 center">
<h1>Selection</h1>
<div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/images/gould_finches.jpg" width="700"></p>
</figure>
</div>
</div>
<aside class="notes">
<p>For a list of methods see <a href="https://methodspopgen.com/methods-to-detect-selection/" class="uri">https://methodspopgen.com/methods-to-detect-selection/</a></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="selection-and-fitness" class="slide level2 smaller">
<h2>Selection and fitness</h2>

<img data-src="index_files/figure-revealjs/fig-selection-model-tikz-1.svg" width="700" class="r-stretch quarto-figure-center"><p class="caption">Figure&nbsp;15: The life cycle used in the fundamental model of selection <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">(<a href="#/bibliography" role="doc-biblioref" onclick="">Gillespie, 2004, fig. 3.2</a>)</span></p><div class="fragment">
<blockquote>
<p>Much confusion exists in the literature regarding how various types of selection are defined, in particular because some of the terminology is used slightly differently within different scientific communities <span class="citation" data-cites="nielsen_MolecularSignaturesNatural_2005">(<a href="#/bibliography" role="doc-biblioref" onclick="">Nielsen, 2005</a>)</span></p>
</blockquote>
</div>
<div class="columns">
<div class="column fragment" style="width:50%;">
<p><span class="math display">\[
\begin{matrix}
\mathrm{Genotype} &amp; aa &amp; aA &amp; AA \\
\mathrm{Frequency\ in\ newborns} &amp; p^2 &amp; 2pq &amp; q^2\\
\mathrm{Viability} &amp; w_{aa} &amp; w_{aA} &amp; w_{AA}\\
\mathrm{Frequency\ after\ selection} &amp; p^2w_{aa} / \bar{w} &amp;
    2pqw_{Aa} / \bar{w} &amp; q^2w_{AA} / \bar{w} \\
\mathrm{Relative\ fitness} &amp; 1 &amp; 1-hs &amp; 1-s\\
\end{matrix}
\]</span></p>
<p>where <span class="math inline">\(\bar{w} = p^2w_{aa} + 2pqw_{Aa} + q^2w_{AA}\)</span> is the <strong>mean fitness</strong>.</p>
</div><div class="column fragment" style="width:50%;">
<table>
<tbody>
<tr class="odd">
<td><span class="math inline">\(h=0\)</span></td>
<td><span class="math inline">\(a\)</span> dominant, <span class="math inline">\(A\)</span> recessive</td>
</tr>
<tr class="even">
<td><span class="math inline">\(h=1\)</span></td>
<td><span class="math inline">\(A\)</span> dominant, <span class="math inline">\(a\)</span> recessive</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(0&lt;h&lt;1\)</span></td>
<td>incomplete dominance</td>
</tr>
<tr class="even">
<td><span class="math inline">\(h&lt;0\)</span></td>
<td>overdominance (heterozygote advantage)</td>
</tr>
<tr class="odd">
<td><span class="math inline">\(h&gt;1\)</span></td>
<td>underdominance</td>
</tr>
</tbody>
</table>
<div class="flushright smallr">
<p>Notation follows <span class="citation" data-cites="gillespie_PopulationGeneticsConcise_2004">Gillespie (<a href="#/bibliography" role="doc-biblioref" onclick="">2004</a>)</span>, pp.&nbsp;61–64.</p>
</div>
</div>
</div>
<aside class="notes">
<p>The life cycle figure starts with newborns carrying the <span class="math inline">\(A\)</span> allele at frequency <span class="math inline">\(p\)</span>, like parents. To reproduce they must reach adulthood, but due to selection, the allele frequency of <span class="math inline">\(A\)</span> may shift to <span class="math inline">\(p^\prime\)</span> because different genotypes affect survival probabilities. This is typically summarized in genotype <em>viabilities</em>.</p>
<p>The Nielsen quote is a reminder that many different notations and parametrizations exist in the literature.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-most-important-equation-in-population-genetics" class="slide level2">
<h2>The most important equation in population genetics</h2>
<p><span class="math display">\[
p^\prime - p = \Delta_sp = \frac{pq[p(w_{aa} - w_{aA}) + q(w_{aA} -
w_{AA})]}{p^2w_{aa} + 2pqw_{aA} + q^2w_{AA}}
\]</span></p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-selection-fig-p_ffc6058520bca56f38360a2c5d16f5d0">
<div class="cell-output-display">
<div id="fig-selection-fig-p" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;16: Allele frequency change over time for directional, balancing, and disruptive selection, for different values of <span class="math inline">\(p_0\)</span>.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-selection-fig-p-1.svg" width="400" height="400"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-selection-fig-dp_7b2c079c909312c4bb1a686638dc3f16">
<div class="cell-output-display">
<div id="fig-selection-fig-dp" class="quarto-figure quarto-figure-center">
<figure>
<figcaption>Figure&nbsp;17: Rate of allele frequency change as a function of allele frequency for directional, balancing, and disruptive selection.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-selection-fig-dp-1.svg" width="400" height="400"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>Given the genotype fitnesses/viabilities, we can work out the difference in allele frequency <span class="math inline">\(p-p^\prime\)</span> between successive generations as a function of viabilities (or relative fitnesses). Equation relates allele frequency change to viabilities for different genotypes. It is instructive to study the equation and plot trajectories to gain an intuition for how allele frequencies evolve given the different selection regimes.</p>
<p>The left panel shows that the three modes of selection have different equilibrium points: for directional selection the favored homozygote will eventually attain 100%, balancing selection has a stable equilibrium point (attractor) for a given allelic ratio (here 25:75), and disruptive selection has an <em>unstable</em> equilibrium point (repeller) for a given allelic ratio.</p>
<p>The right panel shows the rate of <em>change</em> as a function of allele frequency. Note how the directionality changes at the equilibrium point for balancing and directional selection.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="selection-and-drift---population-size-matters" class="slide level2">
<h2>Selection and drift - population size matters</h2>
<div class="columns">
<div class="column" style="width:80%;">
<div class="cell" data-hash="index_cache/revealjs/fig-selection-and-drift_41d281d469119a489a2c966e92ca2d4d">
<div class="cell-output-display">
<div id="fig-selection-and-drift" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;18: The fixation probability relative to the neutral probability of fixation (<span class="math inline">\(p=1/2N\)</span>) under the assumption <span class="math inline">\(s&lt;0.1\)</span>. Red highlights region where <span class="math inline">\(|N_es|&lt;0.05\)</span>. Adapted from <span class="citation" data-cites="lynch_OriginsGenomeArchitecture_2007">Lynch (<a href="#/bibliography" role="doc-biblioref" onclick="">2007</a>)</span>, Fig. 4.2.</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-selection-and-drift-1.svg" width="1152"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:20%;">
<p>In red region (<span class="math inline">\(|N_es|&lt;0.005\)</span>) the probability of fixation is within 10% of neutral fixation.</p>
<p><strong>Consequence</strong>: for any population size there exists range of selection coefficients where mutant alleles <span class="math inline">\(\approx\)</span> neutral (<strong>effective neutrality</strong>).</p>
</div>
</div>
<aside class="notes">
<p>Segue: note on the strength of s and the relation to <span class="math inline">\(N_e\)</span>. For small <span class="math inline">\(N_e\)</span> the relation will hold for higher <span class="math inline">\(s\)</span>, meaning small populations will have difficulties purging deleterious mutations.</p>
<p>Even if a new mutation has, say, <span class="math inline">\(s=0.01\)</span>, it still needs to overcome effects of drift. Only has 2% change of fixation.</p>
<p>In figure, we assume <span class="math inline">\(s&lt;0.1\)</span>, which is almost always the case <span class="citation" data-cites="lynch_OriginsGenomeArchitecture_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Lynch, 2007, p. 74</a>)</span>.</p>
<p><span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 133</a>)</span></p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="substitutions-and-direct-selection" class="slide level2">
<h2>Substitutions and direct selection</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>For genes, the ratio of non-synonymous to synonymous substitutions can tell us about protein evolution.</p>
<dl>
<dt><strong><span class="math inline">\(\mathbf{d_N/d_S &lt;&lt; 1}\)</span></strong></dt>
<dd>
negative (purifying) selection
</dd>
<dt><strong><span class="math inline">\(\mathbf{d_N/d_S &lt; 1}\)</span></strong></dt>
<dd>
majority nonsynonymous deleterious, some advantageous
</dd>
<dt><strong><span class="math inline">\(\mathbf{d_N/d_S = 1}\)</span></strong></dt>
<dd>
neutral or mix neutral / advantageous / deleterious mutations
</dd>
<dt><strong><span class="math inline">\(\mathbf{d_N/d_S &gt; 1}\)</span></strong></dt>
<dd>
positive selection
</dd>
</dl>
</div><div class="column" style="width:50%;">
<div class="cell" data-file="../../recipes/slides/foundations/hsapiens-rat-dnds.R" data-hash="index_cache/revealjs/fig-ensembl-hsapiens-rat-dnds_12ca98d665704594b3f298179bcb51b6">
<div class="cell-output-display">
<div id="fig-ensembl-hsapiens-rat-dnds" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;19: <span class="math inline">\(d_n/d_s\)</span> comparisons for human-rat orthologs. For most genes, <span class="math inline">\(d_n/d_s &lt;&lt; 1\)</span> indicating purifying selection. A handful of genes (n=9) have <span class="math inline">\(d_n/d_s &gt; 1.0\)</span> which could indicate positive selection. <a href="../../recipes/slides/index.html#sec-recipe-slides-dn-ds-ratio-human-rat"><i class="fa-solid fa-rectangle-list" aria-label="rectangle-list"></i></a></figcaption>
<p><img data-src="index_files/figure-revealjs/fig-ensembl-hsapiens-rat-dnds-1.svg" width="768" height="400"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="fragment">
<div class="translatey50">
<p>Not all mutations fall in genes. Methods for detecting direct selection not applicable to studying selection on <em>single</em> mutation, or e.g., balancing. This requires looking for specific patterns of diversity surrounding locus under selection.</p>
</div>
</div>
<aside class="notes">
<p>Segue: how do we detect selection?</p>
<p>Early approach: look at protein-coding sequence where nucleotide substitutions can change the amino acid (non-synonymous, and presumably non-neutral) or be silent (synonymous, assumed neutral).</p>
<p>Not all mutations in genes. Also want to look at single advantageous mutations, or balanced polymorphisms. Must therefore look at tell-tale signs of linked neutral variants.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linked-selection-reduces-diversity-at-neighbour-loci" class="slide level2">
<h2>Linked selection reduces diversity at neighbour loci</h2>
<div class="cell" data-hash="index_cache/revealjs/fig-selective-sweep_453b993406cec290ffb65804d95cd362">
<div class="cell-output-display">
<div id="fig-selective-sweep" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;20: A selective sweep of an advantageous mutation (gray dot). Adapted from <span class="citation" data-cites="charlesworth_ElementsEvolutionaryGenetics_2010">Charlesworth &amp; Charlesworth (<a href="#/bibliography" role="doc-biblioref" onclick="">2010</a>)</span>, Fig. 8.13</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-selective-sweep-1.svg" width="960" height="300"></p>
</figure>
</div>
</div>
</div>
<p>Example of a selective sweep. If a sweep completes at a locus, it will become monomorphic, as will the <em>neighbouring</em> sites. Mutation could reintroduce variation. Recombination could increase diversity in neighbourhood, but in a manner that depends on the <em>distance</em> from the locus under selection.</p>
<aside class="notes">
<p>Alternatives to direct selection tests are those that look at linked selection. In essence, we look for characteristic patterns in the vicinity of a selected locus. A sweep leads to the selected site and neigbouring sites being monomorphic (and therefore invisible to tests based on direct selection). Without recombination, mutation could reintroduce variation. In the presence of recombination, the effect of reduced diversity will be dependent on distance from the locus under selection.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="recombination-breaks-association-between-loci" class="slide level2">
<h2>Recombination breaks association between loci</h2>
<div class="columns">
<div class="column" style="width:60%;">
<!-- markdownlint-disable MD013 -->
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="assets/images/meiosis.png" style="width:80.0%"></p>
</figure>
</div>
<!-- markdownlint-enable MD013 -->
<div class="flushright smallr translatey50">
<p><span class="citation" data-cites="miller_HumanBiology_2020">Miller (<a href="#/bibliography" role="doc-biblioref" onclick="">2020</a>)</span>, Fig. 5.12.3</p>
</div>
<p>Once per chromosome! But: rates vary between loci (hotspots), sex chromosomes vs autosomes, and in some species, recombination only occurs in one sex (e.g., <em>D.melanogaster</em>).</p>
</div><div class="column" style="width:40%;">
<div class="fragment">
<p>Main effect: association between loci breaks up.</p>
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/recombinant-chromosomes_d23f66fa110dab0f2e4522aeb3ff13eb">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="index_files/figure-revealjs/recombinant-chromosomes-1.svg" style="width:80.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<aside class="notes">
<p>The mechanism that breaks up association between loci is recombination. Cartoon to right shows the possible (haploid) outcomes of meiosis. <span class="math inline">\(r\)</span> (the recombination rate) denotes the probability that there is a recombination event between the two loci.</p>
<p>The farther apart two loci on the same chromosome, the larger the probability that a recombination event occurs between them. Note that this can be applied to loci on different chromosomes as well; they segregate independently.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linkage-disequilibrium-and-its-decay" class="slide level2">
<h2>Linkage disequilibrium and its decay</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Association between loci can be written as:</p>
<p><span class="math display">\[
D_{AB} = p_{AB} - p_Ap_B
\]</span></p>
<p>Similar expressions hold for other pairs; only need to know one <span class="math inline">\(D_{ij}\)</span> (e.g., <span class="math inline">\(D_{AB}\)</span>) so drop subscript and rewrite:</p>
<p><span class="math display">\[
p_{AB}= p_Ap_B + D
\]</span></p>
<p>If <span class="math inline">\(D\neq0\)</span> the loci are in <strong>linkage disequilibrium</strong>.</p>
<div class="fragment">
<p>Can show that decay over time is</p>
<p><span class="math display">\[
D_t = (1-r)^tD_0
\]</span></p>
<p>Recombination <em>decreases</em> D (linkage)</p>
</div>
</div><div class="column fragment" style="width:50%;">
<div class="quarto-figure quarto-figure-left">
<figure>
<figcaption>LD between pairs of autosomal SNPs for human and mouse. From <span class="citation" data-cites="laurie_LinkageDisequilibriumWild_2007">(<a href="#/bibliography" role="doc-biblioref" onclick="">Laurie et al., 2007, fig. 2</a>)</span></figcaption>
<p><img data-src="assets/images/ld_mice.png"></p>
</figure>
</div>
</div>
</div>
<aside class="notes">
<p>Handwavy introduction of LD as quantity. There are other equations, but the main point here is to point out that D measures the degree of linkage (disequilibrium); the higher D, the stronger linkage is.</p>
<p>The figure shows the difference in LD decay among wild populations compared to inbred mouse strains, where LD extends over several megabases. Therefore, LD decay informs about demographic history, and as we will see, evolutionary processes (sweeps -&gt; longer LD blocks).</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity" class="slide level2">
<h2>The effect of a selective sweep on diversity</h2>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/pgip-slim-sweep-command_51eacb4a9e001967e6e5faa0749b18cb">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">pgip-slim</span> <span class="at">--seed</span> 42 <span class="at">-n</span> 1000 <span class="at">-r</span> 1e-6 <span class="at">-m</span> 1e-7 <span class="at">--threads</span> 12 recipes/slim/selective_sweep.slim <span class="at">-l</span> 1000000 <span class="at">--outdir</span> results/slim</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="ex">pgip-tsstat</span> results/slim/slim<span class="pp">*</span>.trees <span class="at">-n</span> 10 <span class="at">--seed</span> 31 <span class="at">-s</span> pi <span class="at">-s</span> S <span class="at">-s</span> TajD <span class="at">-w</span> 500 <span class="at">--threads</span> 10 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-v</span> <span class="at">-</span> <span class="op">&gt;</span> results/slim/selective_sweep.w500.csv.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-hash="index_cache/revealjs/pgip-load-slim-data_3ac66aeeaa40c52c4eda503c1ba06518">
<div class="cell-output cell-output-error">
<pre><code>Error in file(file, "rt"): cannot open the connection</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in `$&lt;-.data.frame`(`*tmp*`, thetaW, value = numeric(0)): replacement has 0 rows, data has 401</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in `select()`:
! Can't subset columns that don't exist.
✖ Column `windows` doesn't exist.</code></pre>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<!-- markdownlint-disable MD013 -->
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-pgip-slim-sweep-plot-tajd_45bdce78bc0f2a58623012b8acdfa538">
<div class="cell-output cell-output-error">
<pre><code>Error in FUN(X[[i]], ...): object 'windows' not found</code></pre>
</div>
</div>
<!-- markdownlint-enable MD013 -->
<aside class="notes">
<p>Simulation: N=1e4, mu=5e-8, pi=4<em>N</em>mu=0.004, seek S=a*pi=0.01, so a=S/pi which gives n.&nbsp;One needs to ramp up recombination rate to see the dip in TajD clearly.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity-1" class="slide level2">
<h2>The effect of a selective sweep on diversity</h2>
<!-- markdownlint-disable MD013 -->
<div class="cell" data-hash="index_cache/revealjs/pgip-slim-sweep-command-2_74273fb0acada19733235fc43559f319">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="ex">pgip-slim</span> <span class="at">--seed</span> 42 <span class="at">-n</span> 1000 <span class="at">-r</span> 1e-6 <span class="at">-m</span> 1e-7 <span class="at">--threads</span> 12 recipes/slim/selective_sweep.slim <span class="at">-l</span> 1000000 <span class="at">--outdir</span> results/slim</span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="ex">pgip-tsstat</span> results/slim/slim<span class="pp">*</span>.trees <span class="at">-n</span> 10 <span class="at">--seed</span> 31 <span class="at">-s</span> pi <span class="at">-s</span> S <span class="at">-s</span> TajD <span class="at">-w</span> 500 <span class="at">--threads</span> 10 <span class="kw">|</span> <span class="fu">gzip</span> <span class="at">-v</span> <span class="at">-</span> <span class="op">&gt;</span> results/slim/selective_sweep.w500.csv.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<!-- markdownlint-enable MD013 -->
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-pgip-slim-sweep-plot-tajd-2_27c4821634a05a67f6a8904a2345b981">
<div class="cell-output cell-output-error">
<pre><code>Error in fortify(data): object 'data_mean' not found</code></pre>
</div>
</div>
<aside class="notes">
<p>Important: individual runs show a lot of stochasticity, but the signal is clear for the <em>mean</em>.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-effect-of-a-selective-sweep-on-diversity-2" class="slide level2">
<h2>The effect of a selective sweep on diversity</h2>
<div class="cell" data-hash="index_cache/revealjs/pgip-load-slim-data2_d3d5b29b7f3d29b2a54ab05fe2bc46b9">
<div class="cell-output cell-output-error">
<pre><code>Error in file(file, "rt"): cannot open the connection</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in `$&lt;-.data.frame`(`*tmp*`, thetaW, value = numeric(0)): replacement has 0 rows, data has 401</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in `dplyr::select()`:
! Can't subset columns that don't exist.
✖ Column `windows` doesn't exist.</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="index_cache/revealjs/fig-pgip-slim-sweep-plot-pi-S_ffe777f64e408c17cbe5c6cfc5c74287">
<div class="cell-output cell-output-error">
<pre><code>Error in pivot_longer(., cols = c("pi", "thetaW", "pi_minus_thetaW"), : object 'data_mean' not found</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in `geom_line()`:
! Problem while computing aesthetics.
ℹ Error occurred in the 2nd layer.
Caused by error in `FUN()`:
! object 'windows' not found</code></pre>
</div>
</div>
<aside class="notes">
<p>cf <span class="citation" data-cites="nielsen_MolecularSignaturesNatural_2005">Nielsen (<a href="#/bibliography" role="doc-biblioref" onclick="">2005</a>)</span>. Not clear how scaling of statistics to have mean=1 is done.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>

<section id="the-phases-of-a-selective-sweep" class="slide level2">
<h2>The phases of a selective sweep</h2>
<div class="columns">
<div class="column" style="width:70%;">
<div class="cell" data-hash="index_cache/revealjs/fig-selective-sweep-phases_1c334fc387646d680bcda7651aa0d29c">
<div class="cell-output-display">
<div id="fig-selective-sweep-phases" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;21: Time goes from left to right. As sweep progresses, tree topology changes. Adapted from <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">Hahn (<a href="#/bibliography" role="doc-biblioref" onclick="">2019</a>)</span>, Figure 8.1</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-selective-sweep-phases-1.svg" width="960" height="500"></p>
</figure>
</div>
</div>
</div>
</div><div class="column fragment" style="width:30%;">
<p>Amount of diversity depends on fixation time. A neutral locus fixes in <span class="math inline">\(4N_e\)</span> generations; for <span class="math inline">\(s=0.0001\)</span>, it takes approximately <span class="math inline">\(0.29N_e\)</span> generations.</p>
<p>Selections changes the genealogy (different topology, shorter branches), an aspect used in many linkage-based tests for selection.</p>
</div>
</div>
<aside class="notes">
<p>A selective sweep changes the genealogy (shorter branches, different <em>topology</em>) compared to the neutral case, a fact that can be utilised in linkage-based tests for selection.</p>
<p>Note that the amount of variation surrounding a sweep depends heavily on the <em>rate</em> of the sweep. <span class="citation" data-cites="hahn_MolecularPopulationGenetics_2019">(<a href="#/bibliography" role="doc-biblioref" onclick="">Hahn, 2019, p. 167</a>)</span> points out that an advantageous allele with <span class="math inline">\(s=0.0001\)</span> will take approximately <span class="math inline">\(0.29N_e\)</span> generations to fix (compared to <span class="math inline">\(4N_e\)</span> generations for a neutral locus)</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="linked-selection-may-constrain-levels-of-diversity" class="slide level2">
<h2>Linked selection may constrain levels of diversity</h2>
<div>
<div class="cell" data-hash="index_cache/revealjs/fig-hitchhiking-vs-background-selection_4263c63ea2c9c0c7a63ed47196e7fc71">
<div class="cell-output-display">
<div id="fig-hitchhiking-vs-background-selection" class="quarto-figure quarto-figure-left">
<figure>
<figcaption>Figure&nbsp;22: Hitchhiking (left) versus background selection (right).</figcaption>
<p><img data-src="index_files/figure-revealjs/fig-hitchhiking-vs-background-selection-1.svg" width="1000"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="hitchhiking">Hitchhiking</h3>
<ul>
<li>alleles linked to locus under selection “hitchhike” to high frequencies <span class="citation" data-cites="smith_HitchhikingEffectFavourable_1974">(<a href="#/bibliography" role="doc-biblioref" onclick="">Smith &amp; Haigh, 1974</a>)</span></li>
<li>evidence: positive correlation between putative neutral diversity and recombination <span class="citation" data-cites="corbett-detig_NaturalSelectionConstrains_2015">(<a href="#/bibliography" role="doc-biblioref" onclick="">Corbett-Detig et al., 2015</a>)</span></li>
</ul>
</div><div class="column" style="width:50%;">
<h3 id="background-selection">Background selection</h3>
<ul>
<li>loci linked to a deleterious locus will be purged from population and thus reduce diversity <span class="citation" data-cites="charlesworth_EffectDeleteriousMutations_1993">(<a href="#/bibliography" role="doc-biblioref" onclick="">Charlesworth et al., 1993</a>)</span></li>
<li>similar patterns to hitchhiking</li>
</ul>
</div>
</div>
<aside class="notes">
<p>Linked selection has been proposed as solution to the low range of diversity seen between species with orders of magnitude ranges in census population sizes.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="summary" class="slide level2">
<h2>Summary</h2>
<p>We have looked at the <strong>Wright-Fisher model</strong> as a model of populations and <strong>genealogies</strong>*</p>
<p><strong>Genetic drift</strong> moves allele frequencies up and down at random and removes variation at rate <span class="math inline">\(\propto 1/2N\)</span></p>
<p><strong>Mutation</strong> reintroduces variation. The <strong>Neutral theory</strong> posits most mutations are neutral and dynamics follow <strong>mutation drift</strong> equilibrium.</p>
<p>Methods to detect <strong>selection</strong> are based on <strong>direct selection</strong> or studying patterns of variation caused by <strong>linked selection</strong>.</p>
</section>
<section id="bibliography" class="slide level2 unnumbered unlisted smaller scrollable">
<h2>Bibliography</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-barton2007evolution" class="csl-entry" role="listitem">
Barton, N. H., Briggs, D. E. G., Eisen, J. A., Goldstein, D. B., &amp; Patel, N. H. (2007). <em>Evolution</em>. <span>Cold Spring Harbor Laboratory Press</span>.
</div>
<div id="ref-buri_GeneFrequencySmall_1956" class="csl-entry" role="listitem">
Buri, P. (1956). Gene <span>Frequency</span> in <span>Small Populations</span> of <span>Mutant Drosophila</span>. <em>Evolution</em>, <em>10</em>(4), 367–402. <a href="https://doi.org/10.1111/j.1558-5646.1956.tb02864.x">https://doi.org/10.1111/j.1558-5646.1956.tb02864.x</a>
</div>
<div id="ref-casillas_MolecularPopulationGenetics_2017" class="csl-entry" role="listitem">
Casillas, S., &amp; Barbadilla, A. (2017). Molecular <span>Population Genetics</span>. <em>Genetics</em>, <em>205</em>(3), 1003–1035. <a href="https://doi.org/10.1534/genetics.116.196493">https://doi.org/10.1534/genetics.116.196493</a>
</div>
<div id="ref-charlesworth_ElementsEvolutionaryGenetics_2010" class="csl-entry" role="listitem">
Charlesworth, B., &amp; Charlesworth, D. (2010). <em>Elements of <span>Evolutionary Genetics</span></em>. <span>Roberts and Company Publishers</span>.
</div>
<div id="ref-charlesworth_EffectDeleteriousMutations_1993" class="csl-entry" role="listitem">
Charlesworth, B., Morgan, M. T., &amp; Charlesworth, D. (1993). <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1205596">The <span>Effect</span> of <span>Deleterious Mutations</span> on <span>Neutral Molecular Variation</span></a>. <em>Genetics</em>, <em>134</em>(4), 1289–1303.
</div>
<div id="ref-corbett-detig_NaturalSelectionConstrains_2015" class="csl-entry" role="listitem">
Corbett-Detig, R. B., Hartl, D. L., &amp; Sackton, T. B. (2015). Natural <span>Selection Constrains Neutral Diversity</span> across <span>A Wide Range</span> of <span>Species</span>. <em>PLOS Biology</em>, <em>13</em>(4), e1002112. <a href="https://doi.org/10.1371/journal.pbio.1002112">https://doi.org/10.1371/journal.pbio.1002112</a>
</div>
<div id="ref-ewens_MathematicalPopulationGenetics_2004" class="csl-entry" role="listitem">
Ewens, W. J. (2004). <em>Mathematical <span>Population Genetics</span></em> (S. S. Antman, J. E. Marsden, L. Sirovich, &amp; S. Wiggins, Eds.; Vol. 27). <span>Springer</span>. <a href="https://doi.org/10.1007/978-0-387-21822-9">https://doi.org/10.1007/978-0-387-21822-9</a>
</div>
<div id="ref-gillespie_PopulationGeneticsConcise_2004" class="csl-entry" role="listitem">
Gillespie, J. H. (2004). <em>Population <span>Genetics</span>: <span>A Concise Guide</span></em> (2nd edition). <span>Johns Hopkins University Press</span>.
</div>
<div id="ref-grahamcoop_NotesPopulationGenetics_2020" class="csl-entry" role="listitem">
Graham Coop. (2020). <em>Notes on <span>Population Genetics</span></em>. <a href="https://github.com/cooplab/popgen-notes">https://github.com/cooplab/popgen-notes</a>
</div>
<div id="ref-hahn_MolecularPopulationGenetics_2019" class="csl-entry" role="listitem">
Hahn, M. (2019). <em>Molecular <span>Population Genetics</span></em> (First). <span>Oxford University Press</span>.
</div>
<div id="ref-hein2005gene" class="csl-entry" role="listitem">
Hein, J., Schierup, M. H., &amp; Wiuf, C. (2005). <em>Gene genealogies, variation and evolution: <span>A</span> primer in coalescent theory</em>. <span>Oxford University Press</span>. <a href="https://books.google.se/books?id=CCmLNAEACAAJ">https://books.google.se/books?id=CCmLNAEACAAJ</a>
</div>
<div id="ref-hein_GeneGenealogies_2004" class="csl-entry" role="listitem">
Hein, J., Schierup, M., &amp; Wiuf, C. (2004). Gene genealogies, variation and evolution. A primer in coalescent theory. In <em>Systematic Biology - SYST BIOL</em> (Vol. 54).
</div>
<div id="ref-hermisson_MathematicalPopulationGenetics_2017" class="csl-entry" role="listitem">
Hermisson, J. (2017). <em>Mathematical population genetics</em>. <a href="https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017">https://www.mabs.at/fileadmin/user_upload/p_mabs/Lecture_Notes_2017</a>
</div>
<div id="ref-hubisz_InferenceAncestralRecombination_2020" class="csl-entry" role="listitem">
Hubisz, M., &amp; Siepel, A. (2020). Inference of <span>Ancestral Recombination Graphs Using ARGweaver</span>. In J. Y. Dutheil (Ed.), <em>Statistical <span>Population Genomics</span></em> (pp. 231–266). <span>Springer US</span>. <a href="https://doi.org/10.1007/978-1-0716-0199-0_10">https://doi.org/10.1007/978-1-0716-0199-0_10</a>
</div>
<div id="ref-hurst_GeneticsUnderstandingSelection_2009" class="csl-entry" role="listitem">
Hurst, L. D. (2009). Genetics and the understanding of selection. <em>Nature Reviews Genetics</em>, <em>10</em>(2), 83–93. <a href="https://doi.org/10.1038/nrg2506">https://doi.org/10.1038/nrg2506</a>
</div>
<div id="ref-kimura_1983" class="csl-entry" role="listitem">
Kimura, M. (1983). <em>The neutral theory of molecular evolution</em>. <span>Cambridge University Press</span>. <a href="https://doi.org/10.1017/CBO9780511623486">https://doi.org/10.1017/CBO9780511623486</a>
</div>
<div id="ref-kimura_ProteinPolymorphismPhase_1971" class="csl-entry" role="listitem">
Kimura, M., &amp; Ohta, T. (1971). Protein <span>Polymorphism</span> as a <span>Phase</span> of <span>Molecular Evolution</span>. <em>Nature</em>, <em>229</em>(5285), 467–469. <a href="https://doi.org/10.1038/229467a0">https://doi.org/10.1038/229467a0</a>
</div>
<div id="ref-kumar_MutationRatesMammalian_2002" class="csl-entry" role="listitem">
Kumar, S., &amp; Subramanian, S. (2002). Mutation rates in mammalian genomes. <em>Proceedings of the National Academy of Sciences</em>, <em>99</em>(2), 803–808. <a href="https://doi.org/10.1073/pnas.022629899">https://doi.org/10.1073/pnas.022629899</a>
</div>
<div id="ref-laurie_LinkageDisequilibriumWild_2007" class="csl-entry" role="listitem">
Laurie, C. C., Nickerson, D. A., Anderson, A. D., Weir, B. S., Livingston, R. J., Dean, M. D., Smith, K. L., Schadt, E. E., &amp; Nachman, M. W. (2007). Linkage <span>Disequilibrium</span> in <span>Wild Mice</span>. <em>PLOS Genetics</em>, <em>3</em>(8), e144. <a href="https://doi.org/10.1371/journal.pgen.0030144">https://doi.org/10.1371/journal.pgen.0030144</a>
</div>
<div id="ref-leffler_RevisitingOldRiddle_2012" class="csl-entry" role="listitem">
Leffler, E. M., Bullaughey, K., Matute, D. R., Meyer, W. K., Ségurel, L., Venkat, A., Andolfatto, P., &amp; Przeworski, M. (2012). Revisiting an <span>Old Riddle</span>: <span>What Determines Genetic Diversity Levels</span> within <span>Species</span>? <em>PLOS Biology</em>, <em>10</em>(9), e1001388. <a href="https://doi.org/10.1371/journal.pbio.1001388">https://doi.org/10.1371/journal.pbio.1001388</a>
</div>
<div id="ref-lynch_OriginsGenomeArchitecture_2007" class="csl-entry" role="listitem">
Lynch, M. (2007). <em>The origins of genome architecture</em>. <span>Sinauer Associates</span>.
</div>
<div id="ref-miller_HumanBiology_2020" class="csl-entry" role="listitem">
Miller, C. (2020). <em>Human <span>Biology</span></em>. <span>Thompson Rivers University</span>.
</div>
<div id="ref-nei_MolecularEvolutionPhylogenetics_2000" class="csl-entry" role="listitem">
Nei, M., &amp; Kumar, S. (2000). <em>Molecular <span>Evolution</span> and <span>Phylogenetics</span></em>. <span>Oxford University Press</span>.
</div>
<div id="ref-nielsen_MolecularSignaturesNatural_2005" class="csl-entry" role="listitem">
Nielsen, R. (2005). Molecular <span>Signatures</span> of <span>Natural Selection</span>. <em>Annual Review of Genetics</em>, <em>39</em>(1), 197–218. <a href="https://doi.org/10.1146/annurev.genet.39.073003.112420">https://doi.org/10.1146/annurev.genet.39.073003.112420</a>
</div>
<div id="ref-smith_HitchhikingEffectFavourable_1974" class="csl-entry" role="listitem">
Smith, J. M., &amp; Haigh, J. (1974). The hitch-hiking effect of a favourable gene. <em>Genetics Research</em>, <em>23</em>(1), 23–35. <a href="https://doi.org/10.1017/S0016672300014634">https://doi.org/10.1017/S0016672300014634</a>
</div>
<div id="ref-waples_WhatNeAnyway_2022" class="csl-entry" role="listitem">
Waples, R. S. (2022). What <span>Is Ne</span>, <span>Anyway</span>? <em>Journal of Heredity</em>, <em>113</em>(4), 371–379. <a href="https://doi.org/10.1093/jhered/esac023">https://doi.org/10.1093/jhered/esac023</a>
</div>
<div id="ref-wisely_GeneticDiversityFitness_2002" class="csl-entry" role="listitem">
Wisely, S. M., Buskirk, S. W., Fleming, M. A., McDonald, D. B., &amp; Ostrander, E. A. (2002). Genetic <span>Diversity</span> and <span>Fitness</span> in <span>Black-Footed Ferrets Before</span> and <span>During</span> a <span>Bottleneck</span>. <em>Journal of Heredity</em>, <em>93</em>(4), 231–237. <a href="https://doi.org/10.1093/jhered/93.4.231">https://doi.org/10.1093/jhered/93.4.231</a>
</div>
</div>

<div class="footer footer-default">
<p>Population genetics foundations</p>
</div>
</section></section>

    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-pointer/pointer.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'pointer': {"key":"q","color":"red","pointerSize":8,"alwaysVisible":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'h.v',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'default',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: true,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1280,

        height: 720,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, RevealPointer, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script type="text/javascript">
      function hyperlink_logo() {
        var button = document.createElement( 'div' );
        button.className = "logo-button";
        button.id = "logo-button";
        button.style.visibility = "visible";
        button.style.position = "absolute";
        button.style.zIndex = 30;
        button.style.fontSize = "24px";
        button.style.left = "130px";
        button.style.bottom = "14px";
        button.style.top = "auto";
        button.style.right = "auto";
        if (window.location.host.startsWith("localhost")) {
          button.innerHTML = `<a href="http://${window.location.host}/slides" title="Slides listing"><i class="fa fa-home"></i></a>`
        } else {
          button.innerHTML = '<a href="https://nbisweden.github.io/workshop-pgip/slides" title="Slides listing"><i class="fa fa-home"></i></a>';
        }
        document.querySelector( ".reveal" ).appendChild( button );
      }
    window.document.addEventListener("DOMContentLoaded", function (event) {
      hyperlink_logo();
    });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>